{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "authorship_tag": "ABX9TyMrsj8xfgVfdhKlEPKX/pGZ",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Satish-Kumar-1/Personal-Cancer-Diagnosis/blob/main/personalized_cancer_diagonsis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 90
        },
        "id": "YB_1Miz4AEP6",
        "outputId": "a80913b3-299f-4473-e46a-b2d6ae4c83e8",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-476af8a1-eb4b-4438-ad8e-153e80a8fb06\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-476af8a1-eb4b-4438-ad8e-153e80a8fb06\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script>// Copyright 2017 Google LLC\n",
              "//\n",
              "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
              "// you may not use this file except in compliance with the License.\n",
              "// You may obtain a copy of the License at\n",
              "//\n",
              "//      http://www.apache.org/licenses/LICENSE-2.0\n",
              "//\n",
              "// Unless required by applicable law or agreed to in writing, software\n",
              "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
              "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
              "// See the License for the specific language governing permissions and\n",
              "// limitations under the License.\n",
              "\n",
              "/**\n",
              " * @fileoverview Helpers for google.colab Python module.\n",
              " */\n",
              "(function(scope) {\n",
              "function span(text, styleAttributes = {}) {\n",
              "  const element = document.createElement('span');\n",
              "  element.textContent = text;\n",
              "  for (const key of Object.keys(styleAttributes)) {\n",
              "    element.style[key] = styleAttributes[key];\n",
              "  }\n",
              "  return element;\n",
              "}\n",
              "\n",
              "// Max number of bytes which will be uploaded at a time.\n",
              "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
              "\n",
              "function _uploadFiles(inputId, outputId) {\n",
              "  const steps = uploadFilesStep(inputId, outputId);\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  // Cache steps on the outputElement to make it available for the next call\n",
              "  // to uploadFilesContinue from Python.\n",
              "  outputElement.steps = steps;\n",
              "\n",
              "  return _uploadFilesContinue(outputId);\n",
              "}\n",
              "\n",
              "// This is roughly an async generator (not supported in the browser yet),\n",
              "// where there are multiple asynchronous steps and the Python side is going\n",
              "// to poll for completion of each step.\n",
              "// This uses a Promise to block the python side on completion of each step,\n",
              "// then passes the result of the previous step as the input to the next step.\n",
              "function _uploadFilesContinue(outputId) {\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  const steps = outputElement.steps;\n",
              "\n",
              "  const next = steps.next(outputElement.lastPromiseValue);\n",
              "  return Promise.resolve(next.value.promise).then((value) => {\n",
              "    // Cache the last promise value to make it available to the next\n",
              "    // step of the generator.\n",
              "    outputElement.lastPromiseValue = value;\n",
              "    return next.value.response;\n",
              "  });\n",
              "}\n",
              "\n",
              "/**\n",
              " * Generator function which is called between each async step of the upload\n",
              " * process.\n",
              " * @param {string} inputId Element ID of the input file picker element.\n",
              " * @param {string} outputId Element ID of the output display.\n",
              " * @return {!Iterable<!Object>} Iterable of next steps.\n",
              " */\n",
              "function* uploadFilesStep(inputId, outputId) {\n",
              "  const inputElement = document.getElementById(inputId);\n",
              "  inputElement.disabled = false;\n",
              "\n",
              "  const outputElement = document.getElementById(outputId);\n",
              "  outputElement.innerHTML = '';\n",
              "\n",
              "  const pickedPromise = new Promise((resolve) => {\n",
              "    inputElement.addEventListener('change', (e) => {\n",
              "      resolve(e.target.files);\n",
              "    });\n",
              "  });\n",
              "\n",
              "  const cancel = document.createElement('button');\n",
              "  inputElement.parentElement.appendChild(cancel);\n",
              "  cancel.textContent = 'Cancel upload';\n",
              "  const cancelPromise = new Promise((resolve) => {\n",
              "    cancel.onclick = () => {\n",
              "      resolve(null);\n",
              "    };\n",
              "  });\n",
              "\n",
              "  // Wait for the user to pick the files.\n",
              "  const files = yield {\n",
              "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
              "    response: {\n",
              "      action: 'starting',\n",
              "    }\n",
              "  };\n",
              "\n",
              "  cancel.remove();\n",
              "\n",
              "  // Disable the input element since further picks are not allowed.\n",
              "  inputElement.disabled = true;\n",
              "\n",
              "  if (!files) {\n",
              "    return {\n",
              "      response: {\n",
              "        action: 'complete',\n",
              "      }\n",
              "    };\n",
              "  }\n",
              "\n",
              "  for (const file of files) {\n",
              "    const li = document.createElement('li');\n",
              "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
              "    li.append(span(\n",
              "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
              "        `last modified: ${\n",
              "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
              "                                    'n/a'} - `));\n",
              "    const percent = span('0% done');\n",
              "    li.appendChild(percent);\n",
              "\n",
              "    outputElement.appendChild(li);\n",
              "\n",
              "    const fileDataPromise = new Promise((resolve) => {\n",
              "      const reader = new FileReader();\n",
              "      reader.onload = (e) => {\n",
              "        resolve(e.target.result);\n",
              "      };\n",
              "      reader.readAsArrayBuffer(file);\n",
              "    });\n",
              "    // Wait for the data to be ready.\n",
              "    let fileData = yield {\n",
              "      promise: fileDataPromise,\n",
              "      response: {\n",
              "        action: 'continue',\n",
              "      }\n",
              "    };\n",
              "\n",
              "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
              "    let position = 0;\n",
              "    do {\n",
              "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
              "      const chunk = new Uint8Array(fileData, position, length);\n",
              "      position += length;\n",
              "\n",
              "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
              "      yield {\n",
              "        response: {\n",
              "          action: 'append',\n",
              "          file: file.name,\n",
              "          data: base64,\n",
              "        },\n",
              "      };\n",
              "\n",
              "      let percentDone = fileData.byteLength === 0 ?\n",
              "          100 :\n",
              "          Math.round((position / fileData.byteLength) * 100);\n",
              "      percent.textContent = `${percentDone}% done`;\n",
              "\n",
              "    } while (position < fileData.byteLength);\n",
              "  }\n",
              "\n",
              "  // All done.\n",
              "  yield {\n",
              "    response: {\n",
              "      action: 'complete',\n",
              "    }\n",
              "  };\n",
              "}\n",
              "\n",
              "scope.google = scope.google || {};\n",
              "scope.google.colab = scope.google.colab || {};\n",
              "scope.google.colab._files = {\n",
              "  _uploadFiles,\n",
              "  _uploadFilesContinue,\n",
              "};\n",
              "})(self);\n",
              "</script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving kaggle.json to kaggle.json\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'kaggle.json': b'{\"username\":\"satishkumar0912\",\"key\":\"dc71fffb70f2b880d42ad5a523a03bfc\"}'}"
            ]
          },
          "metadata": {},
          "execution_count": 1
        }
      ],
      "source": [
        "from google.colab import files\n",
        "files.upload()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir ~/.kaggle\n",
        "!cp kaggle.json ~/.kaggle/\n",
        "!chmod 600 ~/.kaggle/kaggle.json"
      ],
      "metadata": {
        "id": "bgfLbngZA8UC"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!kaggle competitions download -c msk-redefining-cancer-treatment"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "orKmi5tWBS84",
        "outputId": "06344201-1706-4f80-ebac-c38b4397d2c3"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Downloading msk-redefining-cancer-treatment.zip to /content\n",
            " 99% 167M/169M [00:09<00:00, 22.9MB/s]\n",
            "100% 169M/169M [00:09<00:00, 18.8MB/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!unzip msk-redefining-cancer-treatment.zip"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8gUJUh4qBef4",
        "outputId": "093ce6ed-2b86-4ed5-e5c7-a7c591eec325"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  msk-redefining-cancer-treatment.zip\n",
            "  inflating: stage1_solution_filtered.csv.7z  \n",
            "  inflating: stage2_sample_submission.csv.7z  \n",
            "  inflating: stage2_test_text.csv.7z  \n",
            "  inflating: stage2_test_variants.csv.7z  \n",
            "  inflating: stage_2_private_solution.csv.7z  \n",
            "  inflating: test_text.zip           \n",
            "  inflating: test_variants.zip       \n",
            "  inflating: training_text.zip       \n",
            "  inflating: training_variants.zip   \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!unzip training_variants.zip"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cMweildZE59v",
        "outputId": "7ff0f414-67f1-43c3-86c4-b4777c38e8be"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  training_variants.zip\n",
            "  inflating: training_variants       \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Import important libraries"
      ],
      "metadata": {
        "id": "XGCMjsWnEOp7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import time\n",
        "import warnings\n",
        "from nltk.corpus import stopwords\n",
        "from sklearn.decomposition import TruncatedSVD\n",
        "\n",
        "import re\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "\n",
        "from sklearn.metrics import accuracy_score, log_loss, confusion_matrix\n",
        "from sklearn.feature_extraction.text import TfidfVectorizer\n",
        "from sklearn.feature_extraction.text import CountVectorizer\n",
        "\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.calibration import CalibratedClassifierCV\n",
        "\n",
        "import seaborn as sns"
      ],
      "metadata": {
        "id": "5DQowQHkDn8Q"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Load the dataset"
      ],
      "metadata": {
        "id": "qnxuzR8sTd7Z"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data = pd.read_csv(\"/content/training_variants\")\n",
        "print('Number of data points: ', data.shape[0])\n",
        "print('Number of features: ', data.shape[1])\n",
        "print('Features: ', data.columns.values)\n",
        "data.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 258
        },
        "id": "w2AP_uCrEqOl",
        "outputId": "9c767130-123f-4b89-8ab0-18b23501a0a1"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of data points:  3321\n",
            "Number of features:  4\n",
            "Features:  ['ID' 'Gene' 'Variation' 'Class']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   ID    Gene             Variation  Class\n",
              "0   0  FAM58A  Truncating Mutations      1\n",
              "1   1     CBL                 W802*      2\n",
              "2   2     CBL                 Q249E      2\n",
              "3   3     CBL                 N454D      3\n",
              "4   4     CBL                 L399V      4"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-1858b85d-07bf-440a-be22-63891a455b53\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>FAM58A</td>\n",
              "      <td>Truncating Mutations</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>CBL</td>\n",
              "      <td>W802*</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>CBL</td>\n",
              "      <td>Q249E</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>CBL</td>\n",
              "      <td>N454D</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>CBL</td>\n",
              "      <td>L399V</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-1858b85d-07bf-440a-be22-63891a455b53')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-1858b85d-07bf-440a-be22-63891a455b53 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-1858b85d-07bf-440a-be22-63891a455b53');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-88d1fe78-bb79-4ddc-98da-56d571dbee87\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-88d1fe78-bb79-4ddc-98da-56d571dbee87')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-88d1fe78-bb79-4ddc-98da-56d571dbee87 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data",
              "summary": "{\n  \"name\": \"data\",\n  \"rows\": 3321,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 958,\n        \"min\": 0,\n        \"max\": 3320,\n        \"num_unique_values\": 3321,\n        \"samples\": [\n          1057,\n          812,\n          2658\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 264,\n        \"samples\": [\n          \"PDGFRA\",\n          \"KEAP1\",\n          \"ARID1A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Variation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2996,\n        \"samples\": [\n          \"E1384K\",\n          \"H597Y\",\n          \"V851A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Class\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2,\n        \"min\": 1,\n        \"max\": 9,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          8,\n          2,\n          6\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data.Class.unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yNF8FybEFQbl",
        "outputId": "220062bb-0e6e-4061-cda3-4af470543c2b"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([1, 2, 3, 4, 5, 6, 7, 8, 9])"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!unzip training_text.zip"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ToZefkbrG0a8",
        "outputId": "26c113d3-a4f5-46e9-f0e8-dfc29d5fce42"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  training_text.zip\n",
            "  inflating: training_text           \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## Reading the text data\n",
        "\n",
        "data_text = pd.read_csv(\"/content/training_text\", sep = \"\\|\\|\", engine = \"python\", names = [\"ID\", \"Text\"], skiprows = 1 )\n",
        "print(\"Number of data points: \", data_text.shape[0])\n",
        "print(\"Number of features: \", data_text.shape[1])\n",
        "print(\"Features: \", data_text.columns.values)\n",
        "data_text.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 258
        },
        "id": "PWU00j0-F1lw",
        "outputId": "332eb4af-2002-4c34-dab8-2d29600c0fa1"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of data points:  3321\n",
            "Number of features:  2\n",
            "Features:  ['ID' 'Text']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   ID                                               Text\n",
              "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
              "1   1   Abstract Background  Non-small cell lung canc...\n",
              "2   2   Abstract Background  Non-small cell lung canc...\n",
              "3   3  Recent evidence has demonstrated that acquired...\n",
              "4   4  Oncogenic mutations in the monomeric Casitas B..."
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-912acb67-c5e2-4bf3-90e1-5ff502833a02\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>Recent evidence has demonstrated that acquired...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-912acb67-c5e2-4bf3-90e1-5ff502833a02')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-912acb67-c5e2-4bf3-90e1-5ff502833a02 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-912acb67-c5e2-4bf3-90e1-5ff502833a02');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-e57246fb-366f-44ed-83fc-385cdab76ea4\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-e57246fb-366f-44ed-83fc-385cdab76ea4')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-e57246fb-366f-44ed-83fc-385cdab76ea4 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data_text",
              "summary": "{\n  \"name\": \"data_text\",\n  \"rows\": 3321,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 958,\n        \"min\": 0,\n        \"max\": 3320,\n        \"num_unique_values\": 3321,\n        \"samples\": [\n          1057,\n          812,\n          2658\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1920,\n        \"samples\": [\n          \"The FLT3 receptor tyrosine kinase is constitutively activated by internal tandem duplication (ITD) in its juxtamembrane domain or tyrosine kinase domain in 30% of acute myeloid leukemia (AML) cases.1-3 Alternatively, FLT3 can also be activated by mutations in the kinase domain (such as the D835Y mutation, which are observed in 7% of AML cases.1,2 Both in vitro and in vivo data have demonstrated that FLT3-ITD and FLT3-D835Y encode constitutively activated kinases that drive the proliferation and survival of hematopoietic cells. Furthermore, FLT3-ITD mutations confer a bad prognosis in AML,1,4,5 and FLT3 inhibitors are expected to improve the outcome of AML patients in this subgroup. These data provide the rationale for the use of FLT3 kinase inhibitors for the treatment of AML.2  While pre-clinical experiments with cell-based assays and mouse models documented a potent effect of various FLT3 kinase inhibitors, first clinical trials showed only moderate efficacy.2 One complicating factor for the treatment of AML is the heterogeneity of the disease, with FLT3 mutations being only one of the many mutations present in human AML samples. Molecular data also suggest that FLT3 mutations tend to occur late in the multi-step development of AML, limiting the effect of FLT3 inhibitors to the subset of AML cells within the leukemia that harbor FLT3 mutations.  In an ongoing multinational phase II trial of the FLT3 inhibitor AC220 as monotherapy, a complete remission rate of 45% was reported at an interim report at the EHA meeting in 2011.6,7 Despite these promising results, patients treated with AC220 develop resistance to this inhibitor and Smith et al. have demonstrated that this is due to the acquisition of FLT3 kinase mutations.8 In that study, mutations at 3 different positions in the kinase domain of FLT3 were identified that confer resistance to AC220: F691, D835 and Y842.8 These mutations were identified by an in vitro resistance screen, and in AML samples obtained from patients who developed resistance to AC220 treatment. Although these mutations conferred resistance to AC220, defined by a clear shift in the IC50 values compared to FLT3-ITD and by their identification as acquired mutations in relapsed patients during AC220 treatment, they could still be inhibited by higher concentrations of the drug.  Eight years ago, we predicted resistance mutations in the FLT3 kinase domain using an in vitro mutagenesis technology in FLT3-ITD transformed Ba/F3 cells.9 In this way, we identified 3 positions in FLT3 (N676, F691 and G697) that upon mutation conferred high level resistance to PKC412. One of these mutations (N676D) was later also identified as a resistance mutation in an AML patient treated with PKC412, indicating the validity of the in vitro predictions.10 The F691L mutation in FLT3 was recently identified as a resistance mutation in AML patients treated with AC220.8  Here we show that also the G697R and N676D mutations confer resistance to AC220. Ba/F3 cells dependent on the expression of FLT3-ITD or FLT3-ITD with additional N676D, F691I, F691L or G697R mutation were treated with increasing concentrations of AC220. We measured the proliferation of these cells over a period of 24 h and we determined the level of autophosphorylation of FLT3 as a measurement of its activity after 90 min of inhibitor treatment. All mutations conferred resistance to AC220, with G697R and F691L/I mutations being the most resistant, and N676D conferring a lower level of resistance (Figure 1). In addition, we also used the Ba/F3 cells expressing FLT3-D835Y, and we observed that also these cells were highly resistant to AC220 (Figure 1).  Figure 1. Figure 1. Sensitivity of FLT3-ITD mutants to the AC220 inhibitor. Ba/F3 cells expressing FLT3-ITD (W51 mutation as described by Kelly et al.11), FLT3-D835Y and different FLT3-ITD mutants were treated with increasing concentrations of AC220 and their proliferation ... Our data show that there are additional mutations that confer resistance to AC220; in particular, the G697R mutation that confers resistance similar to the recently described F691L mutation. All these mutations can be easily acquired by a single nucleotide change in the FLT3 gene sequence. Notably, the F691L, G697R and N676D mutations that confer resistance to AC220 remain sensitive to ponatinib or sorafenib, other known FLT3 inhibitors currently being tested in clinical trials. These results point out that combinations of AC220 and ponatinib or sorafenib may be useful to overcome resistance to single agents.12,13\",\n          \"Bruton\\u2019s tyrosine kinase (BTK), a member of the TEC family of kinases, plays a crucial role in B-cell maturation and mast cell activation. Although the structures of the unphosphorylated mouse BTK kinase domain and the unphosphorylated and phosphorylated kinase domains of human ITK are known, understanding the kinase selectivity profiles of BTK inhibitors has been hampered by the lack of availability of a high resolution, ligand-bound BTK structure. Here, we report the crystal structures of the human BTK kinase domain bound to either Dasatinib (BMS-354825) at 1.9 A\\u02da resolution or to 4-amino-5-(4-phenoxyphenyl)-7H-pyrrolospyrimidin- 7-yl-cyclopentane at 1.6 A\\u02da resolution. This data provides information relevant to the development of small molecule inhibitors targeting BTK and the TEC family of nonreceptor tyrosine kinases. Analysis of the structural differences between the TEC and Src families of kinases near the Trp-Glu-Ile motif in the N-terminal region of the kinase domain suggests a mechanism of regulation of the TEC family members. Keywords: Bruton\\u2019s tyrosine kinase; BTK; Dasatinib; Celera compound; TEC-family; crystal structure Introduction Bruton\\u2019s tyrosine kinase (BTK) is a member of the TEC kinase family, nonreceptor tyrosine kinases that play important roles in T-cell receptor-, B-cell receptor-, and Fcc-receptor-mediated signaling. BTK participates in signal transduction from B-cell antigen receptors resulting in phospholipase C-c2-mediated calcium mobilization1 which, in turn, affects pre-B-cell functional maturation and expansion. Since BTK is required for B-cell function, it is an important target for the potential treatment of inflammatory diseases that involve B-cell activation. Mutations in the human BTK gene are responsible for Abbreviations: BTK, Bruton\\u2019s tyrosine kinase; BTK-KD, BTK kinase domain; ITK, interleukin-2 inducible T-cell kinase; PH, pleckstrin homology; SH, Src homology; TEC, tyrosine kinase expressed in hepatocellular carcinoma; XLA, X-linked agammaglobulinemia. Disclosure: All authors (with the exception of M. J. Romanowski) are employees of Biogen Idec and own company stock. Michael J. Romanowski\\u2019s current address is Novartis Institutes for Biomedical Research, Cambridge, MA 02139. *Correspondence to: Laura F. Silvian, Biogen Idec, Inc., 12 Cambridge Center, Cambridge, MA 02142. E-mail: laura.silvian@biogenidec.com Published by Wiley-Blackwell. VC 2010 The Protein Society PROTEIN SCIENCE 2010 VOL 19:429\\u2014439 429 X-linked agammaglobulinemia (XLA), a male immunodeficiency that results in a deficit of mature B cells and serum immunoglobulin.2,3 Several compounds that inhibit BTK kinase activity in biochemical assays have been described in the literature and differ in their kinase selectivity profiles. One weak compound, LFM-A13 (a-cyano-bhydroxy-b-methyl-N-(2,5-dibromophenyl)-propenamide) is a BTK inhibitor with an IC50 of 2.5 lM in a biochemical assay, but also inhibits PLK3 and JAK2.4\\u20136 However, it was found to be somewhat specific for BTK, exhibiting 100-fold higher IC50 values for related tyrosine kinases such as JAK1, HCK, EGFR, and insulin-receptor kinase (IRK).7 Another compound, Dasatinib ([N-(2-chloro-6-methylphenyl)- 2-(6-(4-(2-hydroxyethyl)piperazin-1-yl)-2-methylpyrimidin-4-ylamino)thiazole-5-carboxamide] or BMS- 354825) [Fig. 1(A)], originally used to target BCRAbl, has been shown to bind to BTK with an IC50 of 5 nM8 but also binds to other kinases such as SRC family members (HCK, SRC, and CSK), and ephrin receptors, FGR, PDGFRa, and YES.9 BTK was identified as a target of Dasatinib through pull-down experiments in the CML cell line K562.8 The reversible Celera compound, 3-cyclopentyl-1-(4-phenoxyphenyl)-1H-pyrazolo pyrimidin-7-amine,10 was recently described by Pan et al. 11 as a potent inhibitor of unphosphorylated BTK (8.2 nM IC50 in a biochemical assay). However, it also inhibits Lck and Src with IC50 values of 2 and 70 nM, respectively.10 It is chemically similar to the commercially available 4-amino-5-(4- phenoxyphenyl)-7H-pyrrolo[2,3-d]pyrimidin-7-yl-cyclopentane [B43; Fig. 2(A)] described as a potent inhibitor of Lck.12 Finally, an irreversible inhibitor from Pharmacyclics11 is currently in Phase I for B-cell lymphomas. It is expected to bind irreversibly to Cys481 in the BTK kinase domain active site (0.72 nM IC50 in a biochemical assay) and its selectivity profile is better than the reversible binder because it exhibits greater selectivity against Lck, which lacks this cysteine (>1000-fold selectivity in a biochemical assay). Future design of potent, specific BTK inhibitors would be facilitated by the structures of these compounds bound to BTK, to discern whether there are regions surrounding the ligand that are unique to this kinase. BTK is composed of several domains: an N-terminal pleckstrin homology (PH) domain, a prolinerich TEC homology domain, two SRC homology domains (SH3 followed by SH2), and a C-terminal kinase domain (BTK-KD). Mutations in all domains of human BTK have been found to lead to XLA and missense mutations have been found in all domains except for the SH3 domain.13 Structures have been Figure 1. BTK-KD Y551E/Dasatinib crystal structure. A: Chemical structure of Dasatinib. B: Electron density (2Fo-Fc map at 1 sigma) for Dasatinib within a surface representation of the BTK protein in the human BTK-KD-Y551E/Dasatinib complex. C: Overall view of the BTK kinase domain bound to Dasatinib. Inhibitor (magenta), amino-terminal lobe (green), carboxy-terminal lobe (gray), C-helix (red), hinge (blue), and activation loop (yellow). D: Close-up of the active site and residues within 5 A\\u02da of the bound Dasatinib. The surface is colored according to the convention of base (blue) and acid (red). 430 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain solved for the kinase domains of apo-murine BTK7 and human ITK,14 but a high-resolution structure of a full-length protein with regulatory domains is not available. Low-resolution structures of BTK solved by small angle X-ray scattering have revealed an extended, linear arrangement of the SH3, SH2, and kinase domains, which contrasts with structures of autoinhibited full-length Src and Abl kinases in which a more compact arrangement of the SH2 and SH3 domains allows for the SH2 domain to bind near the C-terminal tail of the kinase domain.15 Structural studies of the Src family of tyrosine kinases have revealed that these proteins can adapt two conformations: an autoinhibitory state of the protein, referred to as an \\u2018\\u2018assembled regulatory domain\\u2019\\u2019 conformation, and an active, more open, structure, where the SH2 domain does not interact with the unphosphorylated C-terminal tail.16 Here, we describe the 1.94 A\\u02da resolution crystal structure of the human BTK-KD Y551E mutant bound to Dasatinib and a 1.6 A\\u02da resolution crystal structure of the unphosphorylated human BTK-KD bound to B43. We observe that the two structures differ in the orientation of the C-helix, similar to conformational changes observed in Src kinase family members that are locked into active or inactive states. Both BTK-KD structures reveal ordered density for the WEX motif at the N-terminus of the kinase domain, where X is a hydrophobic residue. The location of the tryptophan side chain at the base of the C-helix provides an explanation for how the WEX motif acts as an important regulatory element for the TEC family of kinases, similar to its role in regulation of the Src family of kinases, and suggests that the two families have a similar mechanism of regulation. Results Protein purification and characterization BTK-KD and BTK-KD Y551E were purified to  95% purity using a simple, three-step process utilizing two successive glutathione-Sepharose chromatography steps followed by size exclusion chromatography. Mass spectrometry indicated that the majority of the wild-type BTK-KD and BTK-KD Y551E was intact and unphosphorylated (wild-type calculated mass \\u00bc 32,641.6 Da, measured mass \\u00bc 32,643 Da; mutant calculated mass \\u00bc 32,607.5 Da, measured mass \\u00bc 32,609 Da), although 2 and 8%, respectively, were missing the first 4 N-terminal residues (wildtype calculated mass \\u00bc 32,317.2 Da, measured mass Figure 2. BTK-KD/B43 crystal structure. A: Chemical structure of B43. B: Electron density (2Fo-Fc map at 1 sigma) for B43 within a surface representation of the BTK protein in the human BTK-KD-B43 complex. C: Overall view of the BTK kinase domain bound to B43. Inhibitor (green), amino-terminal lobe (green), carboxy-terminal lobe (gray), C-helix (red), hinge (blue), and activation loop (yellow). D: Close-up of the active site and residues within 5 A\\u02da of the bound B43. The surface is colored according to the convention of base (blue) and acid (red). The active site cysteine is shown with a yellow surface. Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 431 \\u00bc 32,319 Da and mutant calculated mass \\u00bc 32,283.2 Da, measured mass \\u00bc 32,285 Da). Crystal structures The crystal structures of BTK-KD Y551E/Dasatinib and BTK-KD/B43 complexes were determined using 1.94 A\\u02da diffraction data (Rfree \\u00bc 25.8%) and 1.6 A\\u02da diffraction data (Rfree \\u00bc 23.1%), respectively (Table I). The electron density maps clearly revealed the positions of the ligands. Figure 1(B) depicts the electron density for Dasatinib and Figure 2(B) reveals the electron density for B43. The human BTK-KD-Y551E/Dasatinib complex structure exhibits a \\u2018\\u2018C-helix in,\\u2019\\u2019 active conformation in which the catalytic lysine forms H-bonds to Glu445 of the C-helix [Fig. 1(C,D)]. The bilobal fold of the BTK KD is similar to that reported for other tyrosine kinase structures: the N-terminal lobe (residues 392\\u2013473) consists of five antiparallel b-sheets and two a-helices and contains the glycine-rich loop that covers the active site. The C-terminal lobe (residues 482\\u2013659) consists of nine a-helices and two b-strands and the activation loop. Finally, the hinge region (residues 474\\u2013481) connects the two lobes of the kinase and contains the active site cysteine (Cys481). The ordered region of the Dasatinib cocrystal structure encompasses residues 392\\u2013658 but electron density for parts of the activation loop (residues 542\\u2013558) and residues 435\\u2013441 is not visible. However, the DFG-portion of the activation loop is visible and is in the so-called \\u2018\\u2018DFG-in\\u2019\\u2019 conformation. The unphosphorylated BTK-KD/B43 cocrystal structure, in contrast, displays a \\u2018\\u2018C-helix-out\\u2019\\u2019 inactive conformation [Figs. 2(C) and 3(B)]. Most of the kinase is ordered except the tip of the glycine rich loop (residues 410\\u2013414). The activation loop displays a \\u2018\\u2018DFG-in\\u2019\\u2019 conformation and is completely ordered. It is composed of two alpha helices, in which Arg544 within the first helix of the activation loop forms a salt bridge to Glu445 of the C-helix. Dasatinib interactions Dasatinib makes several hydrogen bonds to the hinge and occupies a hydrophobic pocket behind the Thr474 gatekeeper residue [Fig. 1(B,D)]; similar to the previously reported structures of Dasatinibbound Abl (PDB ID: 2GQG),17 Lyn (PDB ID: 2ZVA),18 and cSrc (PDB ID: 3G5D).19 Its H-bond interactions to the hinge are described in Figure 1(D), including interactions with the backbone carbonyl and amide of 477, water-mediated interactions with the Tyr476 side chain, and an interaction between the Thr474 side chain and the compound amide nitrogen. The glycine-rich loop of BTK-KD curls toward Dasatinib to form a hydrophobic interaction with Phe413 [Fig. 1(D)]. The ortho-chloro, methyl phenyl substituent of Dasatinib is twisted to enter a hydrophobic pocket, composed of both hydrophobic and hydrophilic residues Met449, Val458, Leu460, Ile472, Lys430, Glu445, and Ser538. Finally, Glu445 of the C-helix forms a salt bridge with the catalytic Lys430; the epsilon amino group of this lysine is poised to make a pi-cation interaction with the Dasatinib ortho-chloro, methyl phenyl ring [Fig. 1(D)]. B43 interactions The B43 compound makes several hydrogen bonds to the hinge and occupies a hydrophobic pocket behind the Thr474 gatekeeper residue [Fig. 2(D)]. The 4-amino pyrrolopyrimidine of B43 occupies the position of the adenine ring of ATP and the cyclopentyl ring occupies the space generally occupied by the ATP ribose in typical protein kinase structures. The 4-amino pyrrolopyrimidine makes many interactions with the hinge; the exocyclic amine directly interacts Table I. Data Collection and Refinement Statistics BTK-KD/B43 complex BTK-KD Y551E/Dasatinib complex Data collection Space group P21212 P21212 Cell dimensions a (A\\u02da ) 72.5 73.4 b (A\\u02da ) 104.3 104.6 c (A\\u02da ) 38.0 38.1 Wavelength (A\\u02da ) 1.54 0.98 Resolution (A\\u02da ) 50\\u20131.6 50\\u20131.94 Rsym (%)a 8.0 (36.3)a 8.3 (61.9) I/r(I) a 18.5 (4.6)a 32.1 (3.3) Multiplicity 3.8 (4.1)a 6.8 (6.6) Total no. reflections/no. unique reflections 405,732/39,020 490,862/22,496 Completeness (%)a 98.1 (98.5)a 99.2 (99.2) Refinement Resolution (A\\u02da ) 24.5\\u20131.6 28.4\\u20131.94 No. reflections 36,293 21,141 Rwork (Rfree) b (%) 19.2 (23.1) 19.9 (25.8) No. molecules per asymmetric unit 1 1 No. atoms Protein 2161 2016 Ligand 28 33 Water 495 164 Average B-factors (A\\u02da c ) Protein atoms 25.8 23.4 Ligand 30.5 41.7 Water 53.5 35.7 R.m.s.d Bond lengths (A\\u02da ) 0.007 0.024 Bond angles () 1.14 2.05 Ramachandron plot % Allowed 98.5 98.2 % Generously allowed 1.5 1.4 % Not allowed 0 0.4 Rsym \\u00bc R|Ihkl  <Ihkl>|/RIhkl. a The value in parentheses is for the highest resolution bin (approximate interval, 0.1 A\\u02da ). b Rwork \\u00bc RIhkl| |Fo  |Fc| |/RIhkl|Fo| for all data except 5%, which is used for the Rfree calculation. 432 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain with the gatekeeper Thr474 hydroxyl and the backbone carbonyl of 475, the N-3 of the pyrimidine accepts a hydrogen bond from the backbone amine of Met477, and the N-1 of the pyrimidine forms a water-mediated hydrogen bond network to the hydroxyl of Tyr476 and the backbone carbonyl of Ala 478 (Fig. 2). The distal phenyl group of the phenoxyphenyl is twisted 38 out of plane of the phenylether, such that it enters a hydrophobic pocket composed of only hydrophobic residues, Phe442, Met449, Leu460, Ile472, Phe540, and Leu542, and is in proximity to Asp539 of the \\u2018\\u2018DFG\\u2019\\u2019 motif. Phe540 of this motif forms a face-to-edge pi-stacking interaction with the phenoxyphenyl group of B43. Asp539 of the DFG-motif forms a salt bridge with the catalytic Lys430 but does not form direct hydrogen bond interactions with the compound. Discussion Relevance of structures for drug discovery The structures of the human BTK-KD Y551E/Dasatinib and BTK-KD/B43 complexes we report here differ from the publicly available structure of apo-murine BTK-KD (PDB ID: 1K2P) and are arguably more relevant for drug discovery for diseases in which inhibition of BTK may be desired. When the apo-mouse BTK structure is superimposed on the human BTK-KD/B43 structure (RMSD of 1.19 A\\u02da over 230 aligned a-carbons), the biggest differences are observed in the activation loop and in the glycine-rich loop. The activation loop of the mouse apoBTK-KD structure adapts an extended configuration with Tyr551 pointed toward solvent. In the mouse apo-BTK structure, the glycine loop also caves into the active site and occludes the ATP binding pocket. Because the mouse and human BTK-KD\\u2019s are 98.3% identical, and only four amino acids are replaced in the mouse sequence (Q379K, K433R, K625R, and T653S), it is likely that the kinase domain flexibility observed in the apo-murine BTK-KD structure is due to a lack of occupancy of a compound in the active site, rather than due to an intrinsic structural difference between the mouse and human species. Selectivity of compound inhibition For both Dasatinib and the reversible Celera compound, the size and hydrogen bonding nature of the gatekeeper residue of a given kinase generally correlates with its degree of biochemical inhibition9 (Table II). Most of the kinases that are inhibited by 10 lM Dasatinib with a Kd < 1 nM, or that are inhibited by 10 lM Celera compound with less than 5% residual activity, have a threonine gatekeeper. A valine residue in this gatekeeper position is tolerated for the Celera compound binding, but is not as well tolerated for Dasatinib binding to the Ret and KDR kinases (Table II). Because the threonine gatekeeper forms H-bond interactions with both compounds, it is possible that the H-bonding binding energy plays a greater role in binding Dasatinib compared to the Celera compound. An alternative explanation for the poor binding of Dasatinib to valine gatekeeper containing kinases KDR and Ret is that there are differences in side chains within 5 A\\u02da of the compound (Table III). In particular, one residue in the back pocket that forms close hydrophobic interactions with Dasatinib in BTK is Met449, which is replaced by a leucine in KDR and Ret. Because the back pocket in the Dasatinib cocrystal structure Figure 3. Differences between the \\u2018\\u2018inactive\\u2019\\u2019 B43 and \\u2018\\u2018active\\u2019\\u2019 Dasatinib-bound conformation of BTK. A: Overlay of BTK-KD bound to B43 (green), ITK bound to staurosporin (magenta, PDB ID: 1SNU) and P38a (yellow, PDB ID: 1P38). This indicates that the larger phenylalanine gatekeeper of ITK compared to the Thr474 gatekeeper of BTK prevents B43 from binding. It also indicates that the shorter hinge of P38a blocks the B43 binding site. B: Overlay of B43 structure (green) and Dasatinib structure (magenta) indicates that they superimpose well near the hinge but differ near the C-helix, resulting in a shift of >10 A\\u02da in the conformation of the Glu445 side chain and the residues that surround the hydrophobic back pocket. Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 433 is composed of mixed hydrophobic and hydrophilic residues, Dasatinib may have a greater reliance on Met449 compared to B43, whose back pocket is completely surrounded by hydrophobic residues. Either explanation (i.e., different energy considerations for H-bonding to a threonine gatekeeper for different compounds or differences in the makeup of side chains that line Dasatinib\\u2019s unique hydrophobic pocket), could explain why Dasatinib does not bind as well to Ret and KDR. The exception to the rule of requiring a small gatekeeper for compound binding is p38a, EGFR, and NIMA (never in mitosis gene A)-related kinase 11 (NEK11) kinases, which have threonine gatekeepers, but are only moderately inhibited by both small molecules. P38a kinase has a shorter hinge, and thus its reduced affinity can be ascribed to a smaller binding site [Fig. 3(A)]. Similarly, there are differences in the other residues within 5 A\\u02da of the two small molecules (Table III), which could account for the differences in affinity for NEK11 and EGFR. Both differ in the residues that make up the hydrophobic pocket behind the threonine gatekeeper, with EGFR substituting a cysteine for Val458 and NEK11 substituting a larger phenylalanine for Leu460 and a leucine for Met449. Furthermore, NEK11 contains a glycine substitution for Ser538, a residue which is likely to enhance the flexibility of the DFG motif which follows (Table III). In contrast, Abl, Lyn, and Src, kinases which are inhibited well by both small molecules, show no significant variation in the residues that make up the hydrophobic pocket. In these kinases Ser538 is replaced with alanine, a residue with similar conformational flexibility, and is not likely to effect the flexibility of the activation loop (Tables II and III). As the size of the gatekeeper residue increases from threonine to phenylalanine [as in ITK, Fig. 3(A)], methionine (as in Igf1R, JAK1, JNK3, and SYK) (Table II), or leucine (as in Aurora A) (Table II) neither inhibitor reduces kinase activity to the full extent because both binding modes require insertion of the compound past the bulky gatekeeper into the hydrophobic back pocket. Table II. Gatekeeper Residue in Various Kinases, % Activity for Celera Compound and Kd Values for Dasatinib Kinase Gatekeeper % Activity in the presence of 10 lM Celeraa Dasatinib Kd (nM) b Ack1 Ser 3 6 Abl Thr 3 0.5 Bmx Thr 1 0.0 BTK Thr 1 1.0 CSK Thr 0 0.2 EGFR Thr 20 100 FGR Thr 3 0.5 FYN Thr 3 0.7 Hck Thr 5 0.3 Kit Thr 4 0.6 Lck Thr 13 0.2 Lyn Thr 4 0.6 P38a Thr 57 30 Src Thr 1 0.2 Yes Thr 1 0.3 SIK Thr 17 ND RIPK2 Thr 4 30 cRAF Thr 11 600 EphB1 Thr 65 0.4 NEK11 Thr 64 7000 RET Val 0 200 KDR/VEGFR2 Val 6 3000 RSK1 Leu 80 ND RSK2 Leu 83 ND AurA Leu 101 2000 IGF1R Met 87 ND JAK2 Met 88 ND JNK3 Met 76 ND Syk Met 50 3000 Itk Phe 18 ND ND, not determined. a Upstate kinase profiler. b Taken from Ref. 9. Table III. Residues Within 5 A\\u02da of Compounds in the BTK-KD/B43 and BTK-KD Y551E/Dasatinib Structures NTD GRL UPH LWH HYP AVL 428b 430c 408b 409d 413d 416b 472d 474a,c 475b 476b 477b 479d 480b 481c 458c 460b 449b 538b 539b 540b 542b BTK A K L G F V I T E Y M N G C V L M S D F L NEK11 Ve K L G F V I T E YCe Ge Re De V Fe Le Ge DFVe EGRF A K L G F V Le T Qe Le M Fe G CCe LMTe DFL ABL A K L G Ye V I T E FMYe G Ne V L MAe DFL LYN A K L G F V I T E Y M Ke G Se V L MAe DFL SRC A K L G F V I T E Y M Ke G Se V L MAe DFL RET A K L G F V Le Ve E YAe Ye G Se I e L Le SDFL KDR A K L G F V Ve Ve E Fe Ce Fe G Ne VLLe SDFL NTD, N-terminal domain; GRL, glycine rich loop; UPH, upper hinge; LWH, lower hinge; HYP, hydrophobic pocket; AVL, activation loop. a Gatekeeper residue. b Residue interacts with both B43 and Dasatinib in BTK. c Residue only interacts with Dasatinib in BTK. d Residue only interacts with B43 compound in BTK. e Differences in sequence from BTK sequence. 434 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain Overall conformational changes In the BTK-KD Y551E/Dasatinib complex structure, the Gly-rich loop of BTK partially shields Dasatinib from solvent, as found in the Abl kinase structure (data not shown). Because of the curled-in glycinerich loop in the BTK structure, a hydrophobic interaction is formed between Phe413 and the exocyclic methyl of the Dasatinib pyridine ring [Fig. 1(D)], and between Gln412 and Asp539 of the \\u2018\\u2018DFG\\u2019\\u2019 motif. In every other respect, however, the overall conformation of the BTK-KD/Dasatinib structure is similar to the known Dasatinib-bound conformations of Lyn,18 cSrc,19 and Abl,17 including the compound\\u2019s H-bond interactions with the hinge and the position of its C-helix. The nonvisible residues differ between the two BTK structures, which can be correlated with the different interactions made with different compounds or by the mutation in the activation loop. The tip of the glycine-rich loop is disordered in the B43 structure but is ordered in the Dasatinib structure, while the activation loop is disordered in the Dasatinib structure but is well ordered in the B43 structure. The curled glycine-rich loop forms van der Waals contacts with Dasatinib, between a methyl substituent and the Phe413 side chain, whereas no similar interactions take place between B43 and the BTK-KD protein residues and the glycine-rich loop is disordered. One might expect that the Y551E mutation in the activation loop of the BTK-KD Y551E/Dasatinib structure is responsible for the activation loop disorder; the mutated residue is electrostatically incompatible with the conformation of the activation loop seen in the BTK-KD/B43 structure. In the BTK-KD/B43 structure, the Y551 residue is in close proximity to an Asp521 side chain; this is likely to be electrostatically repelled by mutation of the tyrosine to a glutamate. While it is often hard to pinpoint why flexible regions of crystal structures are disordered, it appears that formation of critical molecular interactions produces ordered electron density for the more flexible regions of BTK. Comparison of the structures of the human BTK-KD-Y551E/Dasatinib complex and the BTKKD/B43 complex reveals a change of conformation from catalytically \\u2018\\u2018active\\u2019\\u2019 to \\u2018\\u2018inactive.\\u2019\\u2019 The Dasatinib complex is more similar to the ATP-bound conformation of most kinases, in which a conserved glutamate from the C-helix forms a salt bridge to the catalytic lysine. In fact, no crystals could be formed with the unphosphorylated, wild-type BTK kinase construct, prompting us to make the Y551E mutant as a mimic of the phosphorylated wild-type protein. In contrast, the BTK-KD/B43 complex shows an outward shift of the C-helix [Fig. 3(B)] relative to its position in the Dasatinib structure, the conserved salt bridge from the glutamate to the catalytic lysine breaks, and a large hydrophobic pocket opens behind the gatekeeper residue. The ability of different kinases to adapt a C-helix out conformation might enable the design of specific inhibitors that targets this larger hydrophobic pocket. Furthermore, Cys481 in the active site of BTK-KD could also be exploited to gain kinase selectivity in which a small molecule may be irreversibly bound to this cysteine via a covalent bond.11 To determine the overall similarity of the BTKKD/B43 structure to other kinases, the B43 complex structure was submitted to the Dali-lite server for structure alignment and scoring20 (http://ekhidna. biocenter.helsinki.fi/dali_server/). The top hits, inactive Hck (PDB ID: 1QCF), inactive SRC (PDB ID: 2SRC), inactive ABL (PDB ID: 2G1T), ITK (PDB ID: 1SM2), and mouse BTK (PDB ID: 1K2P), could be aligned with the human BTK over more than 260 a-carbons and with an rmsd of 2.0 A\\u02da or better. The highest scoring hits, excluding the TEC family of kinases, were all inactive conformations of tyrosine kinases from the Src and Abl families, consistent with their overall sequence similarities to human BTK. The conformation of the activation loop and C-helix in the human BTK-KD/B43 structure is very similar to the inactive Src structure (PDB ID: 2SRC) with an rmsd 1.64 A\\u02da over 257 a-carbons; in Src the activation loop forms two alpha helices and occludes access of the substrate peptide. The overall conformation of the BTK-KD Y551E/Dasatinib structure is similar to the active c-Src (CSK) structure (PDB ID: 1Y57) where the activation loop is swung out and the C-helix moves toward the active site (Fig. 4). Comparison of Src and TEC family kinases The phosphorylation-triggered regulation of BTK and Src differ. Unlike the Src family, the TEC family of nonreceptor tyrosine kinases lacks a conserved tyrosine in the C-terminus that could be phosphorylated to then bind to the SH2 domain.16 BTK is regulated by the phosphorylation of two tyrosine residues, Tyr223 in the SH3 domain and Tyr551 in the activation loop of the kinase domain, both of which participate in kinase activation.21 In a recent study of BTK autophosphorylation, the Y551F mutant was shown to have a 5 to 10-fold lower enzymatic activity than the wild-type protein, indicating that this tyrosine plays an important role in BTK activation.22 Furthermore, mutation of a conserved tryptophan in the N-terminal W-E-X motif, in which X is a hydrophobic residue, also appears to effect the activities of the two kinase families differently. In Src, mutation of the Trp to Ala increases kinase activity while in BTK, mutation of the Trp to Ala reduces kinase activity.23 The human BTK structures described here include ordered density for the WEI motif (residues 392\\u2013395), an area which was disordered in the Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 435 available murine BTK structure (PDB ID: 1K2P) and the human ITK structure (PDB ID: 1SM2). This enables a structural comparison of the TEC family and the Src family kinases in this conserved region. The Trp side chain shifts from being solvent exposed in the \\u2018\\u2018inactive\\u2019\\u2019 BTK-KD/B43 complex structure to being wedged into a pocket behind the inward C-helix in the \\u2018\\u2018active\\u2019\\u2019 BTK-KD Y551E/Dasatinib complex structure [Fig. 4(A,B)]. A structural superposition of the two BTK structures with the inactive SRC (PDB ID: 2SRC) 24 and an active CSK structure (PDB ID: 1Y57) 16 show that the side chain of Trp395 superimposes in the active structures of both kinase families. In the \\u2018\\u2018inactive\\u2019\\u2019 conformations, a lysine or methionine side chain from the rotated C-helix sterically occludes the tryptophan side chain, and the Trp side chains are not superimposable [Fig. 4(B)]. The similar structural shift observed suggests that the Src and Tec families utilize a similar means of activating the conformation of the kinase domain. However, the inactive conformations of the Trp side chain differ and the differential energy required in switching from inactive to active conformations may differ; this could explain why mutation of this residue to alanine produces different regulatory effects on the two proteins. This study reports the structures of human BTKKD with Dasatinib or 4-amino-5-(4-phenoxyphenyl)- 7H-pyrrolo[2,3-d]pyrimidin-7-yl-cyclopentane bound at the active site. Analysis of the binding modes reveals that the size and H-bonding potential of the gatekeeper residue in different kinases and the nature of the residues that make up the hydrophobic pocket behind the gatekeeper residue are critical in dictating if either compound binds. We propose that the presence of a distinct hydrophobic pocket in the B43 structure suggests that one could design compounds with limited selectivity for kinases that can adapt the \\u2018\\u2018C-helix-out\\u2019\\u2019 conformation. Movements in the C-helix, presumably induced by phosphorylation of the activation loop residue Tyr551, can shift BTK into an active conformation and could generate a second metal binding site containing Glu445 that Lin et al. have shown is important for optimal catalytic activity of the Tyr551-phosphorylated BTK.25 Finally, we show that, like the Src-family of kinases, BTK can adapt a similar conformational rotamer of Trp395 in its \\u2018\\u2018active\\u2019\\u2019 conformation, which is correlated with a similar movement of the C-helix. With a growing recognition that BTK plays a key role in many B cell lymphomas and autoimmune diseases, these structures will assist with selective drug design.  Bruton\\u2019s tyrosine kinase (Btk) is required for B cell development and B cell antigen receptor (BCR) function. Cross-linking of BCR induces phosphorylation of Btk at Tyr551 and Tyr223. However, the functional requirement of these phosphorylation for BCR signaling remains unclear. We demonstrate here that mutation of Tyr551, not Tyr223, abrogates the BCR-induced calcium mobilization. Not only Lyn, but also Syk was required for tyrosine phosphorylation of Btk in BCR signaling. These results suggest that transphosphorylation of Btk on Tyr551 is essential for BCR function and that this phosphorylation is mediated through the concerted actions of Lyn and Syk.  The B cell antigen receptor (BCR)1 is composed of surface immunoglobulin noncovalently associated with a pair of Ig\\u03b1/Ig\\u03b2 disulfide-linked heterodimers, which are essential for signal transduction. Stimulation of the BCR induces the enzymatic activation and tyrosine phosphorylation of three distinct families of nonreceptor cytoplasmic protein tyrosine kinases (PTKs), the Src family, Syk, and Btk. The Src family kinases are rapidly activated after BCR engagement, and their activation correlates with the initial tyrosine phosphorylation of the immunoreceptor tyrosine-based activation motif on the BCR Ig\\u03b1 and Ig\\u03b2 subunits (reviewed in Refs.1-4).  Temporally, activation of Src family kinases is followed by Btk and Syk (5). This sequential activation potentially places Btk and Syk downstream of Src family kinases. Utilizing co-overexpression system in fibroblasts and COS cells, it has been demonstrated that Lyn transphosphorylates Btk on Tyr551 in the catalytic domain, a site homologous to the Src family kinase consensus autophosphorylation site (6, 7). This results in a 5\\u201310-fold increase in Btk enzymatic activity (7). The increase in activity also leads to increased autophosphorylation at Tyr223 in the SH3 domain of Btk (8). The identical phosphopeptides were generated after cross-linking of the BCR, indicating that these sites are also tyrosine phosphorylated in B cells (7). Although the importance of phosphorylation of Tyr551 and Tyr223 for fibroblast transformation has been examined (8, 9), functional significance of phosphorylation of Tyr551 and Tyr223 of Btk in BCR signaling remains elusive.  To genetically define the functional relationship among Lyn, Syk, and Btk in BCR signaling, we established each PTK-deficient DT40 B cells (10, 11). Our previous results have shown that the BCR-induced calcium mobilization is abrogated in Btk-deficient DT40 cells and that the loss of Btk does not significantly affect the activation of Lyn and Syk in BCR signaling (11). Here we show that BCR-induced tyrosine phosphorylation of Btk is abolished in Lyn/Syk double-deficient DT40 cells, suggesting that Btk acts downstream of Lyn and/or Syk in BCR signaling. Moreover, this phosphorylation is partially inhibited in Lyn- or Syk-deficient cells, indicating contribution of both Lyn and Syk to Btk phosphorylation. The Btk Y223F mutant was able to restore the BCR-induced calcium mobilization, whereas the Y551F mutant could not. Thus, these results suggest that phosphorylation of Tyr551 of Btk through Lyn and Syk is essential for BCR signaling.  Previous Section Next Section EXPERIMENTAL PROCEDURES  Cells, Antisera, and DNA Transfection DT40 cells were cultured in RPMI 1640 supplemented with 10% fetal calf serum, penicillin, streptomycin, and glutamine. Anti-chicken IgM mAb M4 and anti-phospholipase C (PLC)-\\u03b32 Ab were described previously (10). The anti-phosphotyrosine mAb (4G10) and anti-T7 mAb were obtained from Upstate Biotechnology, Inc. and Novagen, respectively. T7-tagged and mutant Btk cDNAs were created by polymerase chain reaction, and the resulting constructs were confirmed by DNA sequencing. These cDNAs were cloned into pApuro expression vector (10). For DNA transfection into DT40 cells, DNA was linearized, electroporated, and selected in the presence of puromycin (0.5 \\u03bcg/ml). The expression of Btk was analyzed by Western blotting.  Immunoprecipitation and Immunoblot Analysis DT40 cells were stimulated by mAb M4 for indicated time. Cells were solubilized in Nonidet P-40 lysis buffer (1% Nonidet P-40, 150 mM NaCl, 20 mM Tris, pH 7.5, 1 mM EDTA) containing 50 mM NaF, 10 \\u03bcM molybdate, and 0.2 mM sodium vanadate supplemented with protease inhibitors described previously (10). Cell lysates were sequentially incubated (1 h at 4\\u2009\\u00b0C for each incubation with Ab and protein A-Sepharose). For immunoblotting, samples were separated on SDS-PAGE and transferred to nitrocellulose membrane (Amersham Corp.). Filters were incubated with mAb 4G10 or anti-T7 mAb. After washing, filters were developed using a sheep anti-mouse IgG Ab conjugated to horseradish peroxidase and enhanced chemiluminescence (ECL).  For in vitro kinase assay, the immunoprecipitates were washed with 20 mM Hepes, pH 8, and 150 mM NaCl after washing with lysis buffer. Added to each sample was 50 \\u03bcl of kinase buffer (20 mM Hepes, pH 8, 10 mMmagnesium acetate, 10 mM MnCl2) in the absence or presence of ATP (1 \\u03bcM). Recombinant glutathioneS-transferase fusion protein containing a cytoplasmic domain of mouse Ig\\u03b1 (glutathione S-transferase/Ig\\u03b1) was made and used as an exogenous substrate (6). The reactions were allowed at 30\\u2009\\u00b0C for 10 min and terminated by the addition of sample buffer.  Calcium and Phosphoinositide Analysis For calcium analysis, cells (5 \\u00d7 106) were resuspended in phosphate-buffered saline containing 20 mM Hepes, pH 7.2, 5 mM glucose, 0.025% bovine serum albumin, and 1 mM CaCl2 and loaded with 3 \\u03bcMFura-2/AM at 37\\u2009\\u00b0C for 45 min. Cells were washed twice and adjusted to 106 cells/ml with continuous monitoring of fluorescence spectrophotometer (model F-2000; Hitachi) at an excitation wavelength of 340 nm and an emission wavelength of 510 nm. Calibration and calculation of calcium level were done as described (12).  For phosphoinositide analysis, cells (106/ml) were labeled with myo-[3H]inositol (10 \\u03bcCi/ml, 105 Ci/mmol) for 4\\u20135 h in inositol-free RPMI 1640 supplemented with 10% dialyzed fetal calf serum, then stimulated in the presence of 10 mM LiCl with mAb M4. The soluble inositol phosphates were extracted with trichloroacetic acid at indicated time points, and applied to AG1-X8 (formate form) ion exchange columns (Bio-Rad) preequilibrated with 0.1 M formic acid. The columns were washed with 10 ml of H2O and 10 ml of 60 mMammonium formate, 5 mM sodium tetraborate. Elution was performed with increasing concentrations of ammonium formate (0.1\\u20130.7M).  Previous Section Next Section RESULTS  We have used a genetic approach to determine the requirements for Btk tyrosine phosphorylation following BCR engagement. We utilized three mutants of the chicken B cell line DT40, generated by inactivation of either lyn, syk, or both genes by homologous recombination (10, 11). Since the Ab raised against chicken Btk does not immunoprecipitate efficiently, we expressed an epitope-tagged version of Btk (designated T7-Btk) into wild type and these mutant DT40 cells. Clones expressing similar levels of T7-Btk in these deficient DT40 cells were selected. These clones were lysed prior to and following BCR ligation, and T7-Btk was immunoprecipitated with anti-T7 mAb. As shown in Fig. 1, Btk was inducibly tyrosine-phosphorylated following BCR stimulation in wild type DT40 cells, consistent with previous reports (5, 13-15). In contrast to wild type cells, Lyn/Syk double-deficient DT40 cells failed to exhibit any Btk tyrosine phosphorylation following BCR ligation, indicating requirement of Lyn and/or Syk for the BCR-induced tyrosine phosphorylation of Btk. In Lyn-deficient DT40 cells, tyrosine phosphorylation of Btk at 1 and 3 min after BCR ligation was significantly reduced, whereas this phosphorylation at 10 min reached almost the same level as that in wild type cells. Compared with Lyn-deficient DT40 cells, Syk-deficient cells showed a complementary time course of the BCR-induced phosphorylation of Btk; phosphorylation of Btk was only observed at 1 and 3 min after BCR stimulation. These data suggest that either Lyn or Syk alone is capable of phosphorylating Btk at least to some extent and that the concerted actions of Lyn and Syk are required for full phosphorylation of Btk in BCR signaling.  Figure 1 View larger version: In this page In a new window Download as PowerPoint Slide Figure 1 BCR-induced tyrosine phosphorylation of Btk in various DT40 mutant cells. At the indicated time points after M4 stimulation (4 \\u03bcg/ml), cells were lysed in 1% Nonidet P-40 lysis buffer and immunoprecipitated with anti-T7 mAb. Immunoprecipitates were loaded onto 8% SDS-PAGE, and the blotted membrane was incubated with mAb 4G10. After the filter was stripped, the same blot was reprobed with anti-T7 mAb. pY, phosphotyrosine.  It has been demonstrated recently that transphosphorylation of Btk at Tyr551 results in increased its enzymatic activity, leading to autophosphorylation of a second tyrosine Tyr223 in a fibroblast system (8). To determine that this sequential phosphorylation of Btk occurs also after BCR stimulation, we transfected Y551F and Y223F mutants of Btk into Btk-deficient DT40 cells (Fig. 2 A). Stimulation of BCR did not enhance tyrosine phosphorylation of Btk(Y551F) (Fig. 3). This result suggests the possibility that Tyr551 of Btk is only target of Lyn/Syk-dependent phosphorylation after BCR stimulation in DT40 cells. Alternatively, phosphorylation at Tyr551 is a critical step for subsequent phosphorylation of Btk in BCR signaling context. In contrast to Btk(Y551F), Btk(Y223F) showed increased tyrosine phosphorylation upon BCR stimulation, although this phosphorylation was only observed at 1 min after receptor cross-linking (Fig. 3). Thus, these results implicate that a primary target of tyrosine through concerted actions of Lyn and Syk is Tyr551 of Btk in BCR signaling, leading to phosphorylation of Tyr223. Since Btk(Y223F) did not show significant tyrosine phosphorylation at 3 and 10 min after receptor stimulation, our finding also suggests that phosphorylation of Tyr223 is required for sustained Btk phosphorylation in BCR signaling.  Figure 2 View larger version: In this page In a new window Download as PowerPoint Slide Figure 2 Schematic representation of Btk cDNA constructs (A) and expression of mutant Btk in Btk-deficient DT40 cells (B). DT40 cells expressing indicated constructions were lysed with Nonidet P-40 lysis buffer (2 \\u00d7 106 cells/lane) and immunoprecipitated with anti-T7 mAb. Immunoprecipitates were divided, and half of them were used for Western blotting with anti-T7 mAb. The remaining half were used for in vitro kinase assay as described under \\u201cExperimental Procedures\\u201d and blotted with mAb 4G10. In the absence of ATP, in vitro kinase samples of these immunoprecipitates showed no significant tyrosine phosphorylation. pY, phosphotyrosine.  Figure 3 View larger version: In this page In a new window Download as PowerPoint Slide Figure 3 BCR-induced tyrosine phosphorylation of Btk(Y223F) and Btk(Y551F). At the indicated time points after M4 stimulation (4 \\u03bcg/ml), Btk-deficient cells expressing these constructs were lysed in 1% Nonidet P-40 lysis buffer and immunoprecipitated with anti-T7 mAb. Immunoprecipitates were loaded onto 8% SDS-PAGE, and the blotted membrane was incubated with mAb 4G10. After the filter was stripped, the same blot was reprobed with anti-T7 mAb. pY, phosphotyrosine.  Utilizing Btk-deficient DT40 cells expressing Btk(Y551F) and Btk(Y223F), we analyzed the effects of these mutations on BCR signaling. Wild type and kinase-negative T7-Btk were also transfected into Btk-deficient DT40 cells as a positive and a negative control, respectively (Fig. 2 A). To determine whether these mutations affect tyrosine kinase activity, Btk immunoprecipitates were used forin vitro kinase assays with glutathioneS-transferase/Ig\\u03b1 as an exogenous substrate. Btk(R525Q) exhibited no kinase activity, indicating that the immunoprecipitates are largely free of contaminating tyrosine kinases. When the extent of tyrosine phosphorylation is normalized to the amount of protein present in each kinase assay, both Btk(Y551F) and Btk(Y223F) had similar transphosphorylation activity compared with wild type Btk (Fig.2 B). We showed previously that the BCR-induced PLC-\\u03b32 activation is abrogated in Btk-deficient DT40 cells, leading to loss of calcium mobilization (11). Thus, we examined whether these mutations were able to restore these defects or not. As shown in Fig.4 A, DT40 cells expressing Btk(Y223F) exhibited normal calcium mobilization, whereas Btk(Y551F) was able to mobilize only small amount of calcium upon receptor cross-linking. Consistent with these data, cross-linking of BCR on DT40 cells expressing Btk(Y223F) stimulated inositol 1,4,5-trisphosphate (IP3) production and tyrosine phosphorylation of PLC-\\u03b32, although these parameters were lower than wild type Btk (Fig. 4,B and C). This might reflect the lower expression level of Btk(Y223F) than wild type Btk (Fig. 2 B). The BCR-induced IP3 production and tyrosine phosphorylation of PLC-\\u03b32 in DT40 cells expressing Btk(Y551F) was essentially the same as those in DT40 cells expressing Btk(R525Q) (Fig. 4 and data not shown). These results demonstrate that phosphorylation of Tyr551, not Tyr223, is essential for BCR signaling.  Figure 4 View larger version: In this page In a new window Download as PowerPoint Slide Figure 4 Functional restoration of BCR signaling by Btk(Y223F) and Btk(Y551F). Calcium mobilization (A), IP3 generation (B), and tyrosine phosphorylation of PLC-\\u03b32 (C). C, Btk-deficient DT40 cells expressing these mutants were stimulated for 3 min with M4 (4 \\u03bcg/ml). Cells were lysed in 1% Nonidet P-40 lysis buffer and immunoprecipitated with anti-PLC-\\u03b32 Ab. Immunoprecipitates were loaded on 6% SDS-PAGE, and the blotted membrane was incubated with mAb 4G10. After the filter was stripped, the same blot was reprobed with anti-PLC-\\u03b32 Ab. pY, phosphotyrosine.  Previous Section Next Section DISCUSSION  In Lyn-deficient DT40 cells, BCR-induced tyrosine phosphorylation of Btk was significantly inhibited at 1 and 3 min after stimulation (Fig. 1), indicating that Btk phosphorylation is mediated by Lyn in BCR signaling in these early time points. However, at 10 min after stimulation, this phosphorylation reached almost the same level in wild type cells, suggesting that tyrosine phosphorylation of Btk at 10 min after BCR stimulation is independent of Lyn. In contrast to Lyn-deficient DT40 cells, Syk-deficient cells show the profound inhibition of the BCR-induced tyrosine phosphorylation of Btk at 10 min after receptor cross-linking, implicating that this Btk phosphorylation is mediated by Syk. Taken together, these data suggest that the initial Btk phosphorylation and sustained phosphorylation are mediated by coordinated actions of Lyn and Syk after BCR cross-linking.  Our conclusion is somewhat inconsistent with the previous reports using COS cell and fibroblast expression systems (6, 7). In these systems, Lyn is able to phosphorylate Btk, whereas Syk is not. One of the possibilities about this disparity between the heterologous systems and DT40 B cell system is that Syk is maximally activated through transphosphorylation at Tyr518 and/or Tyr519 by Lyn in DT40 cells (16). In contrast, overexpressed Syk itself may not be fully activated in heterologous systems. To test this possibility, we transfected T7-Btk into Syk-deficient cells expressing Syk mutant in which Tyr518/Tyr519 is changed to Phe518/Phe519. In this mutant DT40 cell, the BCR-induced tyrosine phosphorylation of Btk was similar to that in Syk-deficient DT40 cells (data not shown), implicating that the requirement of Syk is due to up-regulated Syk through transphosphorylation of Tyr518/Tyr519. It is also possible that Btk may not be a direct substrate of Syk. Assuming that another PTK acts downstream of Syk in BCR signaling, our data might be accounted for by the involvement of this Syk-regulated PTK in tyrosine phosphorylation of Btk.  It has been reported that two tyrosines, Tyr551 and Tyr223, are phosphorylated after cross-linking of BCR. Btk(Y551F) exhibited no tyrosine phosphorylation upon BCR cross-linking, whereas Btk(Y223F) showed an increase of tyrosine phosphorylation at 1 min after receptor stimulation, suggesting that Tyr551 phosphorylation is a prerequisite for subsequent phosphorylation of Btk in BCR signaling events. Thus, these results support the previous contention that phosphorylation of Btk at Tyr551 is followed by its autophosphorylation at Tyr223 (7, 8). Since Btk(Y223F) did not show significant tyrosine phosphorylation at 3 and 10 min after receptor stimulation, phosphorylation of Tyr223 appears to be required for sustained Btk phosphorylation. Recent crystallographic analysis of Itk (Btk/Tec family PTK expressed in T cells) may provide insights into the function of phosphorylation of Tyr223 of Btk (17). Based on this analysis, the proline-rich domain adjacent to the SH3 domain of Btk/Tec family kinases contains an SH3 ligand (Fig. 2 A), allowing intramolecular interaction. Interestingly, Tyr223of Btk is located within the interface of this interaction. Thus, phosphorylation of Tyr223 may disrupt this intramolecular interaction, thereby changing the conformation of Btk. This conformational change might be required for sustained phosphorylation of Btk in BCR signaling.  Our functional data clearly indicate that Tyr551 is essential for BCR signaling, whereas Tyr223 is dispensable for BCR-induced PLC-\\u03b32 activation. Since phosphorylation of Tyr551 was already reported to increase the kinase activity of Btk with 5\\u201310-fold (7), one of the consequence of phosphorylation of Tyr551 is increased kinase activity upon BCR cross-linking. Although we carried out in vitro kinase assay on Btk immunoprecipitates in wild type DT40 cells, the BCR-induced activation of its in vitro kinase activity could not be reproducibly observed. Previous reports suggest that the magnitude of the BCR-induced activation of Btk is significantly smaller than that of maximum Btk phosphorylation by Lyn in heterologous systems (6, 7). Indeed, in our hand, this 5\\u201310-fold activation of Btk in COS cells could be reproducibly detected. Thus, the most likely explanation is that the increase of Btk enzymatic activity through phosphorylation of Tyr551 in DT40 cells is too small for our detection system.  In the case of Syk, it has been demonstrated that in addition to recruitment of Syk to phosphorylated Ig\\u03b1/Ig\\u03b2, phosphorylation at tyrosine (Tyr519 and/or Tyr518) within its activation loop is critical for BCR signal transduction (16). In this report, we show that phosphorylation of Tyr551 on Btk activation loop is also an obligatory mechanism for its participation in BCR signaling. Thus, cytoplasmic PTK cascade through phosphorylation of tyrosine located in the activation loop may be one of the general mechanisms for cytoplasmic signal transduction.  The Tec family kinases are tyrosine kinases that function primarily in hematopoietic cells. The catalytic activity of the Tec kinases is positively influenced by the regulatory domains outside of the kinase domain. The current lack of a full-length Tec kinase structure leaves a void in our understanding of how these positive regulatory signals are transmitted to the kinase domain. Recently, a conserved structure within kinases, the \\u2018regulatory spine\\u2019, has been identified that assembles and disassembles as a kinase switches between its active and inactive states. Here we define the residues that comprise the regulatory spine within Tec kinases. Compared to previously characterized systems, the Tec kinases contain an extended regulatory spine that includes a conserved methionine within the C-helix and a conserved tryptophan within the SH2-kinase linker of Tec kinases. This extended regulatory spine forms a conduit for transmitting the presence of the regulatory domains of Tec kinases to the catalytic domain. We further show that mutation of the gatekeeper residue at the edge of the regulatory spine stabilizes the regulatory spine resulting in a constitutively active kinase domain. Importantly, the regulatory spine is preassembled in this gatekeeper mutant rendering phosphorylation on the activation loop unnecessary for its activity. Moreover, we show that the disruption of the conserved electrostatic interaction between Btk R544 on the activation loop and Btk E445 on the C-helix also aids in the assembly of the regulatory spine. Thus, the extended regulatory spine is a key structure that is critical for maintaining the activity of Tec kinases.  Keywords: Itk and Btk, regulatory spine, kinase activation, gatekeeper residue, phosphorylation Go to: INTRODUCTION Protein kinases catalyze the transfer of a phosphate group from ATP to a hydroxyl containing amino acid side chain; either Ser/Thr for serine-threonine kinases or Tyr for tyrosine kinases[1]. The activity of protein kinases is exquisitely regulated within the cell[1; 2; 3; 4]. The switch from the catalytically inactive state of a kinase to the active state is often accompanied by the phosphorylation of a key residue within the activation loop, a large flexible loop that lies between the two lobes of the kinase domain structure[1]. Phosphorylation of the activation loop residue can trigger concerted movements in other mobile elements within the kinase domain such as the C-helix and the DFG motif that brings about the assembly of the catalytically critical residues in the kinase active state[1]. The active state conformations associated with different kinases are nearly identical across the many kinases for which high resolution structures have been solved[2; 3; 5; 6; 7]. It is not surprising that the active states of distinct kinases are very similar since the phospho-transfer chemistry carried out by different kinases is the same. In contrast, it is becoming clear that the structural features associated with the inactive state of various kinases differ widely.  The search for features that are conserved within the structures of active kinases has recently led to the identification of a structure termed the \\u2018regulatory spine\\u2019[8; 9]. The regulatory spine was first identified by Local Spatial Patterns alignment analysis using a set of serine-threonine and tyrosine kinase structures that included PKA as a model kinase[8; 9]. The regulatory spine defines a stretch of amino acid residues that is assembled only in the active state of kinases. Unlike consensus sequences that consist of a continuous stretch of amino acids in the primary structure, the regulatory spine consists of disparate residues that span the N- and C-terminal lobes of the kinase domain. Assembly of the regulatory spine has been proposed to be a crucial step in the activation of protein kinases and this structure is disrupted in structures of kinases in the inactive state. Moreover, the regulatory spine has been proposed as the mechanism by which allosteric effects can be propagated to the kinase active site[8; 9].  The Tec kinases are immunologically related tyrosine kinases which consist of five mammalian members: Itk, Btk, Tec, Txk and Bmx[10; 11]. Tec kinases share similar domain architecture as Src, Abl and Csk family of kinases, in that they have the SH3-SH2-kinase domain cassette[10]. Despite domain similarities, we and others have shown that the regulation of Tec kinases is distinct from Src and Abl kinases[10; 12; 13; 14]. Unlike the Src and Abl kinases, the N-terminal regulatory domains of Tec kinases positively regulate the activity of the kinase domain[14]. While the isolated kinase domains of Src and Abl are active, the isolated kinase domains of the Tec family exhibit poor catalytic activity[13; 14; 15]. Regulation of the Tec kinases is in fact more similar to the Csk family of enzymes, whose N-terminal regulatory domains are essential for the catalytic activity of the kinase domain[16]. Moreover, while the structures of full-length Src and Csk kinases are available, the structure of a full-length Tec kinase remains elusive[2; 17; 18]. This leaves a significant gap in our mechanistic understanding of the regulation of the Tec family kinases.  In this manuscript, we define the regulatory spine within the Tec family kinases and show that mutation of the spine residues leads to different effects in the context of the isolated kinase domain verses full-length Itk. Moreover, we show that the assembly of a stable regulatory spine within members of the Tec kinase family is critically dependent on the presence of the SH2-kinase linker region. The residues that comprise the regulatory spine within PKA are insufficient to promote the assembly of a stable regulatory spine within the Tec kinases. We extend the Tec regulatory spine to include a conserved methionine within the C-helix and a conserved tryptophan residue within the SH2-kinase linker of Tec kinases. Together with the spine residues originally identified within the PKA kinase domain, the conserved methionine and tryptophan residues form a continuous structure that links the kinase active site to the N-terminal regulatory domains of Tec kinases. Furthermore, we show that the stabilization of the regulatory spine by the gatekeeper threonine to methionine mutation eliminates the need for phosphorylation on the activation loop for Btk kinase activity. Thus, the extended regulatory spine is a structure that is critical for the regulation of Tec kinases and if stabilized appropriately, is sufficient for activation in the absence of activation loop tyrosine phosphorylation and in the absence of the non-catalytic Tec regulatory domains.  Go to: RESULTS The regulatory spine controls Tec kinase activity  The regulatory spine within PKA consists of five residues: L95, L106, Y164, F185, and D220[9]. These residues are spread throughout the primary sequence of the PKA kinase domain: L95 is located on the C-helix, L106 is on the N-terminus of the \\u03b24 strand, Y164 is part of the \\u2018HRD\\u2019 motif, F185 is from the \\u2018DFG\\u2019 motif and D220 is on the F-helix within the C-terminal lobe[9]. Alignment of the structure of the Btk and Itk kinase domains with that of PKA shows that the corresponding regulatory spine residues within the Btk and Itk kinase domains should consist of: Btk M449, L460, H519, F540 and D579 and Itk M409, L420, H479, F500 and D539 respectively (Fig. 1a).  Figure 1 Figure 1 Identification of the regulatory spine in Tec kinases To test the role of the predicted spine residues in the Tec kinases, we wished to take advantage of the rapid bacterial expression and purification system that has been developed for the isolated kinase domains of Itk and Btk[19]. One issue that arises, however, is the fact that the isolated kinase domains of the Tec kinases exhibit poor catalytic activity and so the expected loss of function mutations in the regulatory spine would be difficult to characterize given the already low activity of the wild type kinase domains. In order to take advantage of the ease of the bacterial expression system, we needed an isolated Btk and Itk kinase domain with higher catalytic activity. It has been demonstrated previously that introduction of a bulky hydrophobic residue such as isoleucine or methionine at the gatekeeper position activates multiple kinases[20]. Indeed, we find that the Itk F434M and Btk T474M isolated kinase domains are more active when compared to wild-type isolated kinase domain (manuscript in preparation). We therefore probed the importance of the predicted regulatory spine residues in Itk and Btk by mutating them individually to alanine in the context of the Itk F434M and Btk T474M isolated kinase domain mutants. Activity measurements are carried out by monitoring the phosphorylation levels of a peptide substrate in a radioactive assay or by detecting the level of autophosphorylation on the activation loop tyrosine (Y511 in Itk and Y551 in Btk) by western immunoblotting. While phosphorylation of the activation loop tyrosine is achieved by the activity of the Src family kinases Lck and Lyn respectively in vivo, under in vitro conditions Tec kinases autophosphorylate on the activation loop tyrosine[21; 22; 23].  Disruption of the regulatory spine residues by mutation to alanine is predicted to disrupt kinase activity. Indeed, mutation of Btk M449, H519, F540 and D579 and Itk M409, H479, F500 and D539 to alanine within the isolated kinase domain of Itk F434M and Btk T474M leads to decreased phosphorylation on the activation loop tyrosine and drastically reduces the catalytic activity (Fig. 1b, c, d and e). Itk kinase residues such as M409, F500 and H479 are highly conserved with other kinases with well-defined roles in catalysis[8]. Itk M409 is involved in binding the substrate ATP[12]. Itk F500 is part of the \\u2018DFG\\u2019 motif at the start of the activation loop segment and is responsible for stabilizing the conformation of the preceding aspartate, and the C-helix[8]. Itk H479 is part of the \\u2018HRD\\u2019 motif where it serves as a scaffold for D499 and F500[8]. It is therefore not surprising that mutation at these sites leads to a loss in activity. However, Itk D539 is not part of any previously characterized regulatory motif and therefore the drastic reduction in Itk activity upon mutation of this residue to alanine highlights the importance of the regulatory spine in regulating Itk activity. Moreover, since the regulatory spine residues are conserved within the Tec kinase family, this structure would be predicted to be critical for regulating the activity of all Tec kinase family members.  Unexpectedly, mutation of Itk L420 and Btk L460 to alanine, which based on the previous PKA work is predicted to be part of the regulatory spine, failed to inactivate the isolated kinase domains of Itk F434M or Btk T474M (Fig. 1b, c, d and e). In fact, the isolated kinase domains of both Itk L420A/F434M and Btk L460A/T474M double mutants showed an increase in activity when compared to the isolated kinase domain of Itk F434M and Btk T474M, respectively. To ensure that the activating effect of the Leu to Ala mutation was not an artifact of working with the activated (gatekeeper mutant) isolated kinase domains of Itk and Btk, the regulatory spine residues were mutated to alanine in the context of wild-type isolated kinase domains of Itk and Btk. Since the catalytic activity of the isolated kinase domain of Tec kinases is generally poor, the detection of activation loop phosphorylation levels by western immunoblotting is difficult and for Itk, in particular, phosphorylation Y511 is below the detection limit by western immunoblotting. Hence, the activity of the wild-type and regulatory spine mutants of the isolated kinase domains of Itk and Btk are monitored by their ability to phosphorylate a peptide substrate in the radioactive assay.  The results of the regulatory spine mutations in the context of the wild type kinase domains mirror the results for the activating gatekeeper mutants described above. Mutation of Btk M449, H519, F540 and D579 to alanine within the isolated wild type kinase domain of Btk leads to decreased catalytic activity (Fig. 1f). Circular dichroism spectra of the Btk mutants overlay well with that of wild-type Btk showing that the decreased activity of the Btk mutants are not due to unfolding of the kinase (Supp. Fig. 1). Since the activity of the isolated Itk wild-type kinase domain is not significantly above background levels, the resulting activity of the Itk spine mutants: Itk M409A, H479A, F500A and D539A can not be interpreted except to say activity is no greater than wild type (Fig. 1g). However, consistent with our earlier results with the gatekeeper mutant, the Itk L420A and Btk L460A mutants are both more active than wild-type isolated kinase domain of Itk and Btk, respectively (Fig. 1f and g). Interestingly, previous studies on the regulatory spine within the Abl kinase have shown that mutation of Abl L320 to glycine (corresponding to Itk L420 and Btk L460), leads to only a slight decrease in kinase activity of full length Abl[20]. This is also consistent with modeling studies of Abl that have shown that Abl L320 has a modest impact on regulatory spine formation[20]. Thus, Itk L420 and Btk L460 do not play a major role in the assembly of the regulatory spine within the context of the isolated kinase domain of Itk and Btk.  The regulatory spine is not assembled in the structure of phosphorylated Itk kinase domain  Activation of Tec kinases requires the phosphorylation of a conserved tyrosine within the activation loop[12; 21; 23]. Separate high-resolution structures of the Btk kinase domain with Y551 either phosphorylated (active) or unphosphorylated (inactive) have recently been reported[24]. A comparison of the inactive and active states of Btk kinase domain shows clear differences in the region of the regulatory spine (Fig. 2a). The regulatory spine residues assemble into a linear arrangement in the structure of the active Btk kinase, while interactions between the regulatory spine residues (in particular L460, M449 and F540) are disrupted in the structure of inactive Btk kinase. The structural differences in this region between active and inactive Btk are consistent with the role of the regulatory spine as defined previously for PKA (Fig. 2a).  Figure 2 Figure 2 The regulatory spine is assembled only in the active state of kinases Crystal structures of the isolated kinase domain of Itk that is either unphosphorylated or phosphorylated on Y511 in the activation loop are also available[12]. Surprisingly, the available Itk kinase domain structures overlay quite well, with little or no conformational differences between them regardless of the phosphorylation state of the activation loop. Inspection of the Itk kinase domain structures shows that the regulatory spine residues in the Itk kinase domain structures adopt a disrupted configuration similar to the inactive Btk kinase domain (Fig. 2b). An additional hallmark of an active kinase domain structure is the formation of a crucial ion pair between a conserved lysine positioned within the \\u03b23 strand of the kinase domain and a conserved glutamate on the C-helix[1]. The distance between the corresponding residues (Itk K390 and E405) within the structures of both phosphorylated and unphosphorylated Itk kinase domain is on the order of 8.0 \\u00c5 which is significantly greater than the 3 to 4 \\u00c5 distance that is observed in the structures of other active kinases (Fig. 2b)[12]. Indeed, the distance between the same conserved ion pair within Btk, (Btk K430 and E445) changes from 14.6 to 3.8 \\u00c5 upon activation of the kinase (Fig. 2a). These observations suggest that both the phosphorylated and unphosphorylated Itk kinase domains adopt a conformation consistent with the \\u2018inactive\\u2019 enzyme. The inability of the Itk kinase domain to fully assemble into an active conformation, despite being phosphorylated on Y511 in the Itk activation loop, points to additional requirements for activation.  A notable difference between the Itk and Btk crystal structures is that the construct used for the crystallization of the Btk kinase domain included the SH2-kinase linker, while that of the Itk kinase did not (Fig. 2a & b)[12; 24]. We hypothesize that the absence of the SH2-kinase linker within the Itk construct used for crystallography might prevent the assembly of a stable regulatory spine in the kinase domain despite phosphorylation on the activation loop Y511.  Extension of the Tec kinase regulatory spine  We have previously shown that the SH2-kinase linker, the 17 amino acids between the SH2 domain and the kinase domain, is critical for the activity of both Btk and Itk, exerting a positive effect on the catalytic function of the kinase domain[14]. More specifically, a conserved tryprophan (Itk W355, Btk W395) in the SH2-kinase linker, as well as a methionine residue in the C helix (Itk M410, Btk M450) are crucial residues in the Tec family kinase regulatory apparatus (Fig. 3a)[14]. Like the regulatory spine mutants described in Figure 1, point mutations of Itk W355, Btk W395 or Itk M410 to alanine all result in a significant drop in catalytic activity (Fig. 3b). Examination of the crystal structures of Btk shows that Btk W395 and Btk M450 are located at the \\u2018top\\u2019 of the regulatory spine and, in the active Btk structure, serve to extend the hydrophobic packing of the spine residues well into the N-terminus of the kinase domain (Fig. 3c & d). In the inactive Btk structure, repositioning of the W395 and M450 side chains disrupts the extended regulatory spine (Fig. 3c & d) in much the same manner that repositioning of M449 in the inactive Btk structure disrupts the core of the spine (Fig. 2a). The active and inactive Btk structures consist of the same amino acids (residues 382\\u2013659) and differ with respect to the activation loop tyrosine (Y551). In the active structure the activation loop tyrosine is mutated to glutamate (Y551E) to mimic phosphorylated Btk, and for the inactive structure wild type Btk in the unphosphorylated form was used for crystallization[24].  Figure 3 Figure 3 Extension of the regulatory spine within Tec kinases to include the SH2-kinase linker Compared to the regulatory spine that has been defined for PKA[9], the Tec kinase family requires at least two additional residues (Itk W355 & M410) to fully assemble the regulatory spine. The tryptophan residue is located in the SH2-kinase linker region providing an explanation for the positive regulatory role of this non-catalytic region[14]. The available Itk kinase domain structures are entirely consistent with this finding; without the contribution of the SH2-linker region, the regulatory spine of Itk does not assemble into the active conformation. As a result, the structures of the phosphorylated and unphosphorlated Itk kinase domain fragments are very similar to each other and neither resembles an active kinase. Future structures of Tec kinases with and without the SH2-kinase linker in various activation (phosphorylation) states will be required to fully probe this hypothesis.  Itk L420 and Btk L460 are essential for the formation of the extended regulatory spine within full-length Tec kinases  The conserved tryptophan in the SH2-kinase linker is critical for the activity of full-length Tec kinases[14]. Upon activation, the tryptophan side-chain forms part of a hydrophobic pocket that is lined by the conserved methionine on the C-helix and leucine on the \\u03b24 strand (Itk M410, L420 and Btk M450, L460) (Fig. 3d). Although Itk L420 and Btk L460 were shown to not be a part of the regulatory spine in the context of the isolated kinase domain (Fig. 1), the structure of the Btk kinase domain (which contains the SH2-kinase linker region) suggests that this leucine would be critical for the formation of the extended regulatory spine in the context of the full-length kinase. We therefore tested the role of this leucine in the context of full-length Itk.  As shown in Figure 3e, mutation of Itk L420 to alanine within the context of full-length Itk disrupted the activity of Itk. Consistent with our earlier results, mutation of Itk M409, F500 and D539 to alanine within full length Itk also inactivated the kinase. Problems with cloning prevented us from testing the activity of the full-length Itk H479A mutant. Together, Itk M409, L420, H479, F500, D539, M410 and W355 form a conserved allosteric signaling network that is absolutely required for the activity of Itk. Based on sequence conservation, the extended regulatory spine defined here likely controls the activity of the entire Tec kinase family.  Phosphorylation on the activation loop tyrosine is not required for the activity of the Btk T474M mutant  It is well established that activation loop phosphorylation is required for activation of tyrosine kinases[1; 8] and specifically for the Tec kinases, phosphorylation on the activation loop tyrosine results in at least a ten-fold increase in activity[21; 25]. Structures of active and inactive kinase domains from a number of different kinases demonstrate a conserved electrostatic network that switches between the active and inactive conformations (Fig. 4a)[1; 8]. In the inactive state of Btk, the conserved glutamate (Btk E445) on the C-helix (of the Lys-Glu ion pair) is associated with a conserved arginine (Btk R544) on the activation loop[24; 26]. Activation of Btk by phosphorylation on Btk Y551 (the activation loop tyrosine) leads to a specific interaction between pY551 and R544, and a concomitant loss of the association between the R544 and E445[24; 26]. Btk E445 on the C-helix then swings toward the kinase active site and associates with the conserved lysine side chain (Btk K430) to bring about the assembly of the active state of the kinase[24; 26]. The conformational adjustment of the C-helix brings M449 and M450 in line with the other residues of the regulatory spine (Fig. 4a). Thus, phosphorylation on the activation loop tyrosine is an initiating step in the process of regulatory spine assembly.  Figure 4 Figure 4 The regulatory spine is preassembled in the Btk T474M gatekeeper mutant Since mutation of the gatekeeper residue to methioine appears to stabilize the regulatory spine within Btk, we next probed the requirement of activation loop phosphorylation in the context of the T474M gatekeeper mutation. The activity of the BtkT474M/Y551F double mutant was compared to that of the Btk T474M mutant by monitoring phosphorylation of an exogenous substrate. We find that mutation of the activation loop Btk Y551 to phenylalanine has no effect on the activity of the Btk T474M mutant (Fig. 4b & c). The Btk T474M/Y551F double mutant and the Btk T474M single mutant are both active, whereas the wild type Btk kinase requires activation loop phosphorylation; the Btk Y551F single mutant exhibits poor activity (Fig. 4b). These data are consistent with the idea that phosphorylation on the activation loop tyrosine initiates the assembly of the regulatory spine by disrupting the association between Btk E445 and Btk R544. Once pY551 competes with E445 for association with R544, the E445/K430 salt bridge is formed, bringing the C-helix (and importantly M449) into the regulatory spine structure. Under conditions that pre-organize the regulatory spine structure (as in the Btk T474M mutant), phosphorylation on the activation loop Y551 is no longer required to trigger the conformational changes that accompany Btk activation.  We further tested the importance of this switched electrostatic network by mutating Btk R544. We reasoned that mutation of Btk R544 will activate the kinase as inward movement of the C-helix is restrained by the E445:R544 interaction and loss of this electrostatic interaction should facilitate movement of the C-helix. Indeed, mutation of Btk R544 to serine in the context of wild-type isolated kinase domain of Btk leads to a two-fold increase in activity (Fig. 4d & e). In contrast, mutation of Btk R544 to serine in the context of Btk T474M isolated kinase domain does not further activate the kinase (Fig. 4d & e). These results again suggest that pre-assembly of the kinase regulatory spine (by mutation) can overcome the regulatory interactions that normally control kinase activity. Thus, for the wild-type kinase under physiological conditions, phosphorylation on the activation loop sets into motion a cascade of events: (1) formation of the pY551:R544 interaction with concomitant disruption of the Btk E445:R544 interaction, (2) inward movement of the C-helix and formation of the E445:K430 interaction and (3) assembly of the regulatory spine, that culminates in the formation of an active kinase (Fig. 4a).  Go to: DISCUSSION The structures of the isolated kinase domain of Btk in both the active and inactive states have provided significant insight into Tec kinase activation[24; 26]. However, the lack of a full-length structure of any Tec kinase leaves numerous unanswered questions regarding the regulation of Tec kinases. While the SH2 domain and the SH2-kinase linker region of Tec kinases have been shown to be required for the activity of Tec kinases[12; 13; 14], the mechanism by which these regulatory domains positively influence catalytic activity has not been clear. Here we identify an intramolecular allosteric signaling network that extends from the SH2-kinase linker into the kinase domain of Tec kinases. This intramolecular connectivity defines how the non-catalytic Tec regulatory domains, through the SH2-kinase linker region, impinge on the kinase domain and stabilize the regulatory spine. In the absence of the Tec regulatory domains, whether achieved by deletion or conformational changes within the full-length molecule, the regulatory spine is disrupted and kinase activity is inhibited.  Comparing our results to those previously published for PKA shows that the Tec kinases require an extended spine structure and it is this longer regulatory spine that couples kinase activity to regions outside of the kinase domain. Specifically, a conserved methionine in the C-helix and a conserved tryptophan in the SH2-linker swing in to \\u2018cap\\u2019 the spine in the active Btk structure and are critical for the activity of the Tec kinases. Mutation of these extended spine residues disables kinase activity even if the \\u2018core\\u2019 residues defined for PKA are intact.  Further support for the importance of the entire regulatory spine comes from examination of the genetic mutations in Btk that are associated with X-linked agammaglobulinemia (XLA) in humans, a disease that is characterized by impaired B cell development[27]. Several of the extended regulatory spine residues identified in this study are mutated in XLA patients[27]. In addition to nonsense mutations at Btk M449, M450 and W395, which cause premature truncation of the protein, there are also missense mutations at Btk M450, F450 and D579 (M450I, F450S and D579V). These mutations likely destabilize the regulatory spine and prevent Btk kinase activation.  Activation loop tyrosine phosphorylation has long been known to influence kinase activity. In addition to priming the kinase domain for the assembly of the regulatory spine, phosphorylation on the conserved Tyr of the activation loop is also thought to stabilize the substrate-binding site of the activation loop[1]. However, our work has shown that in the context of the activating gatekeeper Btk T474M mutation, loss of Y551 phosphorylation (by mutation to phenylalanine) has no affect on substrate phosphorylation. Pre-assembly of the regulatory spine (mutation of T474M) removes any requirement for activation loop phosphorylation. These results are consistent with earlier studies that focused on phosphorylation of Y551 in Btk[25]. That earlier work did show that phosphorylation on Btk Y551 significantly alters catalytic activity of Btk; kcat of the Btk Y551F mutant is greatly diminished compared to wild-type Btk presumably due to loss of the trigger for regulatory spine assembly. However, the affinity for a peptide substrate is unaffected by Y551 phosphorylation; Km of a peptide substrate for Btk Y551F mutant is the same as wild-type Btk. As well, an alternative direct substrate docking mechanism has been identified for the Tec kinases, in which an SH2 domain within the substrate docks onto the kinase domain outside of the active site and facilitates substrate phosphorylation[28; 29; 30]. It is therefore possible that the peptide substrate-binding region on the activation loop may only play a minor role in substrate recognition in Tec kinases and that activation loop phosphorylation plays one significant role: initiation of spine assembly by altering electrostatic interactions in and around the active site.  We have previously shown that the regulation of Tec family kinases is similar to that of Csk (C-terminal Src kinase)[14]. Unlike the Src and Abl family of kinases, the SH3 and SH2 regulatory domains of Csk and Tec kinases positively influence the catalytic activity of the kinase domain[14; 16]. Like Itk and Btk, the isolated kinase domain of Csk has poor catalytic activity[16; 31; 32]. In order to define the elements within the Csk kinase domain that are responsible for the poor catalytic activity, a mutagenesis study was carried out in which Src:Csk kinase domain chimeras were created[33]. A Src:Csk chimeric kinase domain where the N-terminal lobe of Csk was replaced with the N-terminal lobe of Src was shown to exhibit activity comparable to that of full-length Csk[33]. Further mutagenesis of the N-terminal lobe, identified the C-helix and the \\u03b2 turn between the \\u03b24 and \\u03b25 strands (the \\u2018top\\u2019 of the regulatory spine) as key structural elements that are required for the activity of the isolated kinase domain of Csk[33].  There is remarkable correlation between the elements that are identified in the Csk mutagenesis study and the extended regulatory spine of Tec kinases identified in this study. Since the Csk and Tec family kinases are both positively regulated by domains outside of the kinase domain, it is not surprising that they would share the key determinants of catalytic activity. Indeed, as suggested in the original identification of the kinase regulatory spine[9], this feature likely plays a critical role in all active kinases. Our work confirms this notion for the Tec family kinases, and illustrates that extension of the regulatory spine by two residues couples this regulatory feature to allosteric events occurring outside of the kinase domain. Moreover, our data suggest that despite a high degree of sequence conservation, the five core regulatory spine residues do not assemble into the active conformation in the context of the isolated Itk or Btk kinase domains. The energetics of spine assembly seem to be tailored for each specific regulatory environment; the Tec kinase domains are maintained in their inactive state (spine assembly does not occur even when the activation loop is phosphorylated) until additional regions of the protein (in this case the SH2-linker) are present to drive complete spine assembly forward and fully activate the kinase.\",\n          \"More than 50% of transitional cell carcinomas of the bladder show loss of heterozygosity of a region spanning the TSC1 locus at 9q34 and mutations of TSC1 have been identified in 14.5% of tumours. These comprise nonsense mutations, splicing mutations, small deletions and missense mutations. Missense mutations are only rarely found in the germline in TSC disease. Therefore, we have examined six somatic missense mutations found in bladder cancer to determine whether these result in loss of function. We describe loss of function via distinct mechanisms. Five mutations caused mutually exclusive defects at mRNA and protein levels. Of these, two mutations caused pre-mRNA splicing errors that were predicted to result in premature protein truncation and three resulted in markedly reduced stability of exogenous TSC1 protein. Primary tumours with aberrant TSC1 pre-mRNA splicing were confirmed as negative for TSC1 expression by immunohistochemistry. Expression was also significantly reduced in a tumour with a TSC1 missense mutation resulting in diminished protein half-life. A single TSC1 missense mutation identified in a tumour with retained heterozygosity of the TSC1 region on chromosome 9 caused an apparently TSC2- and mTOR-independent localization defect of the mutant protein. We conclude that although TSC1 missense mutations do not play a major role in causation of TSC disease, they represent a significant proportion of somatic loss of function mutations in bladder cancer.  Go to: INTRODUCTION Tuberous Sclerosis Complex (TSC) is an autosomal dominant tumour suppressor gene syndrome with an incidence of 1 in 6000\\u201310 000 births. TSC is characterized by the development of benign growths, called hamartomas, in the kidneys, heart, brain and skin, and patients present clinically with a variety of developmental disorders (1). TSC is caused by mutations affecting either of the tumour suppressor genes TSC1 or TSC2. TSC1 on chromosome 9q34 encodes hamartin (2) and TSC2 on chromosome 16p13.3 encodes tuberin (3). Approximately half of large TSC families show linkage to 9q34 and half to 16p13.3 (4\\u20136). Tumour development in TSC patients is thought to occur as the result of a somatic \\u2018second-hit\\u2019 in either TSC1 or TSC2, according to Knudson\\u2019s tumour suppressor model (7). Loss of heterozygosity (LOH) of TSC1 or TSC2 has been reported in some TSC hamartomas, such as renal angiomyolipomas. However, loss of the wild-type allele in brain lesions is rare, suggesting the possibility of tissue-specific haploinsufficiency of TSC genes (8\\u201310).  Co-localization and co-immunoprecipitation of TSC1 and TSC2 in mammalian cells (11,12) and direct binding in yeast two-hybrid assays provide a tentative explanation for the similar disease phenotype in TSC patients with mutations in either TSC1 or TSC2 genes (2,13). Functionally, the TSC1/TSC2 complex is positioned at the centre of multiple growth signalling pathways and is a key integrator of signals controlling protein translation and cell growth (14). Activation of the TSC1/TSC2 complex in growth-limiting conditions attenuates signalling through mTOR via specific GTPase activating protein (GAP) activity of TSC2 towards RHEB (15,16).  While epithelial malignancy is not a common feature of TSC, studies in this laboratory and others have implicated loss of function of TSC1 in bladder tumorigenesis (17\\u201319). Loss of heterozygosity (LOH) for markers on chromosome 9 is observed in more than 50% of bladder tumours of all grades and stages (20) and sub-chromosomal LOH analyses have identified the TSC1 locus at 9q34 as a common critical region of deletion between markers D9S149 and D9S66 (19,21\\u201323). To date, we have screened 154 bladder tumours by fluorescent single strand conformation polymorphism (F-SSCP) analysis and direct sequencing, and found an overall mutation frequency of 14.5%. The mutation spectrum comprises nonsense (35%), missense (26%), frameshift (26%), in-frame deletions (3%) and splicing (10%) mutations (24) (Platt et al., in preparation). In all cases but one, TSC1 missense mutations were tumour-specific somatic events. TSC1 is the only gene on 9q that has been found to be mutated in bladder tumours, and may therefore be the critical gene on this chromosome arm implicated in >50% of all bladder tumours.  Missense mutations of TSC1 have not routinely been confirmed as functionally inactivating in TSC disease, though two recent reports provide evidence that in a few cases these are likely to be disease-causing (25,26). Here we sought to determine whether the TSC1 missense mutations identified in bladder tumours constituted inactivating mutations. We anticipated that discrete amino acid changes of mutant proteins might allow the identification of functionally important residues. Wild-type and mutant TSC1 constructs were retrovirally delivered into TSC1-null bladder tumour cell lines and functionally characterized. All somatic TSC1 missense mutations perturbed TSC1 function by causing aberrant splicing, protein instability or protein mislocalization. Defects were confirmed in primary tumours by RT\\u2013PCR analysis of mutant transcripts and immunohistochemical analysis.  Go to: RESULTS Missense mutations of TSC1 identified in bladder tumours  Previously, we identified 8 mutations including 2 missense mutations in a series of 62 bladder tumours (24). Screening of an additional 92 tumours (Platt et al., in preparation) identified an additional 15 mutations of which 4 were missense mutations. In total, therefore, 6 missense mutations have been identified in 154 tumours, representing 26% of all mutations found (Table 1). Mutations were determined as tumour-specific by genotyping of paired tumour and blood samples. 1250C>T (Thr417Ile) was previously described in the germline in TSC disease in Japanese patients, but was not confirmed as causative of TSC (27,28). 1250C>T was also identified in our laboratory in the patient\\u2019s constitutional DNA, and the tumour sample retained heterozygosity for microsatellite markers at the TSC1 gene locus (24). Threonine 417 was previously identified as a site of CDK1-dependent phosphorylation (29). To determine the biological significance of this variant, 1250C>T (Thr417Ile) was characterized here alongside tumour-specific missense mutations. Five of six other tumours with TSC1 missense mutations showed LOH for markers at the TSC1 locus (Table 1). There was no relationship between mutation and tumour stage or grade. Missense mutations were predominantly N-terminal but did not localize to a common functional domain (Fig. 1).  Figure 1. Figure 1. Positions of amino acid substitutions in relation to described functional domains of hamartin. Table 1. Table 1. TSC1 missense variants identified in bladder tumours To assess possible functional implications of amino acid substitutions, conservation of missense mutant residues was determined in TSC1 orthologs. His68, Phe158 and Phe216 were conserved among Rattus, Mus, Drosophila, Fugu and Gallus orthologs. His105, His206 and Thr417 differed only in Drosophila, which shares 31% identity with human TSC1. Ser35 differed in Drosophila, Fugu and Gallus orthologs. According to the BLOSUM62 scoring matrix (30), His68Arg and His105Arg substitutions are considered conservative. Ser35Cys, His206Asp and Thr417Ile are less conservative and Phe158Cys and Phe216Asp are least conservative (Table 1).  Re-expression of missense mutant TSC1 proteins in TSC1-null urothelial cells  We anticipated that missense mutations might lead to loss of function at the amino acid level. Thus, our initial approach was to express wild-type and missense mutant TSC1 cDNAs in TSC1-null bladder cell lines and investigate function of the mutant proteins. Missense mutant proteins were C-terminally FLAG- or GFP-tagged, expressed in 97-1 and HCV29 cells and characterized for TSC2 binding activity, and mTOR suppressive activity in nutrient-starved conditions.  Expression of mutant proteins was investigated by immunoblotting following retroviral transduction and selection of neomycin resistant mass cell populations. Expression levels of mutant proteins were markedly different; Ser35Cys, His105Arg, Phe216Asp and Thr417Ile were expressed at high levels similar to wild-type protein, whereas His68Arg and Phe158Cys were much reduced and His206Asp was not detected (Fig. 2A). Repeated independent infections of these two cell lines and of telomerase-immortalized normal human urothelial cells (TERT-NHUC) and assessment of mass populations of cells following selection resulted in entirely reproducible expression levels of all mutant proteins (data not shown). The transcription of missense RNA was confirmed by real time RT\\u2013PCR and uniform levels of wild-type and missense transcripts were detected (Fig. 2B). The consistently reduced or absent protein expression of His68Arg and Phe158Cys and His206Asp missense mutant forms despite the presence of RNA expression suggested possible effects of these missense changes on protein stability.  Figure 2. Figure 2. (A) Immunoblot showing levels of wild-type and mutant TSC1-FLAG proteins and endogenous TSC2 in 97-1 cell lines. (B) Measurement of TSC1 RNA levels by real time RT\\u2013PCR analysis of wild-type and mutant TSC1-FLAG mRNA transcript levels in 97-1 cell ... Mutant TSC1 proteins retain interaction with TSC2  The interaction between TSC1 and TSC2 appears to be important in maintaining the stability of each of the proteins. TSC1/TSC2 binding specifically augments TSC2 expression by limiting its ubiquitination (31,32). TSC2 GAP activity towards RHEB defines TSC1/2 mediated control of mTOR signalling, and phosphorylation of downstream effectors of mTOR is constitutive and refractory to amino acid withdrawal in cells lacking TSC1 or TSC2 (33\\u201335). Where expressed, missense mutant TSC1 proteins stabilized TSC2 levels in 97-1 cells (Fig. 2A) and co-immunoprecipitated with it (Fig. 3A). However, our data cannot exclude minor effects on interaction or TSC2 stability.  Figure 3. Figure 3. (A) Immunoblot showing TSC1 and TSC2 in 97-1 Neo control and FLAG-tagged wild-type and mutant TSC1 cell line lysates immunoprecipitated with anti-TSC1 and non-specific mouse IgM antibodies. (B) Immunoblot showing expression levels of GFP-tagged wild-type ... While His206Asp-FLAG was undetected in transduced 97-1 cells, we were able to achieve low-level expression of His206Asp-GFP in HCV29 cells, suggesting some stabilizing effect of the GFP tag. TSC2 co-immunoprecipitation and S6 phosphorylation assays were not accurately quantitative, but expression of all missense mutant proteins in TSC1-null cells reduced S6 phosphorylation in amino acid starved conditions, relative to vector alone (Fig. 3B), indicating that none of these mutant forms of TSC1 abolish interaction with TSC2.  Aberrant RNA splicing caused by TSC1 missense mutations  High-level expression and TSC2 binding activity of TSC1 Ser35Cys, His105Arg and Phe216Ala proteins suggested no functional effect of amino acid substitutions caused by 104C>G, 314A>G and 648T>A missense mutations. However, the expression of proteins from exogenous missense cDNAs did not allow the assessment of possible effects of mutations at the pre-mRNA level. We speculated that TSC1 missense mutations may cause loss of function through introduction of splicing errors in mutant transcripts in vivo. Therefore, wild-type and missense mutant TSC1 pre-mRNA sequences were screened for effects of mutations on splice site definition using the neural network algorithm (http://www.fruitfly.org/seq_tools/splice.html).  Interestingly, significant differences were seen between splice site scores of wild-type and 104C>G and 314A>G TSC1 transcripts. The 104C>G mutation, positioned 3 bp upstream of the TSC1 exon 3/4 junction, resulted in a reduced exon 3/4 splice motif score. The 314A>G mutation resulted in the introduction of a high-scoring 5\\u2032 donor splice site, immediately upstream of the A/G transversion, by generation of a novel consensus splice motif. No differences in splice site scores were observed between wild-type and 648T>A or 1250C>T transcripts or between wild-type and mutant transcripts that generated low exogenous protein expression.  To examine splicing of missense mutant TSC1 transcripts, fragments spanning the mutation site and flanking intron/exon junctions were amplified from respective tumour cDNAs by RT\\u2013PCR and sequenced. An RT\\u2013PCR product reproducibly amplified from cDNA of the tumour containing 104C>G showed increased electrophoretic mobility relative to a control fragment amplified from TERT-NHUC cDNA (Fig. 4A). Sequencing of the tumour product revealed use of a cryptic 5\\u2032 donor site in exon 4, the adjacent downstream exon to the 104C>G mutant exon (Supplementary Material, Fig. S1). The reduced splice motif score at the exon 3/4 junction was shown to have no effect on exon 3/4 splicing. If translated, this 104C>G mutant transcript would generate 26 new amino acids from an alternate reading frame and truncate prematurely at residue 79.  Figure 4. Figure 4. Semi quantitative RT\\u2013PCR analysis of TSC1 RNA in normal urothelial cells and TSC1 mutant tumour samples, showing increased electrophoretic mobility of RT\\u2013PCR products amplified from cDNA from tumours containing TSC1 104C>G (A) ... A faster migrating band was also reproducibly amplified from cDNA of the tumour containing 314A>G (Fig. 4B), compared to control TERT-NHUC cDNA. Sequence analysis of PCR products revealed a 50-nucleotide deletion in the transcript generated from the missense allele. The 314A>G mutation created a de novo 5\\u2032 donor splice site immediately upstream of 314A>G and resulted in the splicing of 50 nucleotides from the 3\\u2032 end of exon 5. The effect seen at the mRNA level was entirely consistent with the prediction made by in silico analysis. If translated, the 314A>G mutant transcript is expected to generate two new C-terminal amino acids and truncate prematurely at residue 107.  Exon skipping through introduction of a de novo 5\\u2032 splice site by point mutation is a relatively well-described mechanism of aberrant pre-mRNA splicing (36\\u201338). No normal transcript was amplified from cDNA from tumours containing TSC1 104C>G or 314A>G mutations, suggesting that all splicing occurs via the novel sites. The abundance of missense transcripts appeared lower than normal transcript by semi-quantitative RT\\u2013PCR, when standardized to HPRT. However, by real time RT\\u2013PCR, TSC1 transcript abundance was relatively higher in tumours with TSC1 missense mutations causing transcript splicing or protein stability defects, than in uncultured or cultured TERT-NHUC, when standardized to SDHA (data not shown).  Missense mutant TSC1 proteins exhibit reduced protein stability  The demonstration of consistently low-level exogenous protein expression, despite uniform mRNA transcript abundance, suggested diminished stability of TSC1 His68Arg, Phe158Cys and His206Asp mutant proteins. By inhibiting protein synthesis with cycloheximide, we determined that His68Arg, Phe158Cys and His206Asp mutant proteins were turned-over relatively faster than wild-type TSC1, in a proteasome-dependent manner (Fig. 5 A and B and data not shown). Moreover, by 35S labelling, we were able to show that His68Arg, Phe158Cys and His206Asp proteins had markedly shorter half-lives than wild-type TSC1 (Fig. 5C).  Figure 5. Figure 5. (A) Immunoblot showing turnover of wild-type and S35C and H68R missense TSC1 proteins in cycloheximide (CHX)-treated cells. Cells were cultured in full growth medium supplemented with 100 \\u00b5g/ml CHX or DMSO vehicle alone, and lysed at time-points ... Phe216Ala TSC1 protein shows altered localization  Previous studies have described a granular, cytoplasmic localization of endogenous TSC1 in vitro and in vivo and also of overexpressed TSC1 in COS-7 cells (39\\u201341). Localization of monomeric TSC1 and of TSC1 complexed with TSC2 is likely influenced by culture conditions given that components of the Akt-mTOR signalling cascade are membrane localized when activated. In addition, TSC2 is reported to shuttle into the nucleus in a cell cycle and phosphorylation-dependent manner (42\\u201344).  It was anticipated that the characterization of discrete missense amino acid changes may offer insight into potentially mTOR-independent or bladder-specific functions of TSC1. Our results indicated that TSC1 104C>G, 203A>G, 314A>G, 473T>G and 616C>G missense mutations cause loss of function by generic mechanisms of altered message or reduced protein stability. However, sequencing of RT\\u2013PCR products amplified from cDNA from the tumour containing the TSC1 648T>A mutation showed no altered splicing of the mutant transcript. Also, at the protein level, Phe216Ala was expressed at high levels similar to wild-type TSC1. The substitution of a phenylalanine residue for an alanine residue constitutes the loss of a high-molecular weight, hydrophobic benzyl group and is assigned a non-conservative \\u20132 BLOSUM62 score. The lack of effect on splicing or protein stability raised the possibility that this mutation may cause a defect in TSC1 protein function per se.  We determined the localization of GFP-tagged mutant proteins in nutrient replete and deficient conditions. Wild-type, Phe216Ala, Ser35Cys, His68Arg and Thr417Ile proteins were compared. In complete growth medium, wild-type TSC1 and Ser35Cys, His68Arg and Thr417Ile mutant proteins showed diffuse punctate cytoplasmic distribution with distinct cytoplasmic foci. In amino acid deficient medium, these proteins became localized almost exclusively to large cytoplasmic bodies. Intriguingly, the Phe216Ala substitution markedly altered the localization of TSC1; Phe216Ala was exclusively cytoplasmic and diffuse in full growth medium and did not redistribute to discrete bodies in amino acid deficient conditions (Fig. 6). However, it was shown to stabilize and to co-immunoprecipitate with endogenous TSC2, and to attenuate growth signalling through mTOR in starved conditions (Fig. 3B). These observations were reproduced in TSC1-transduced 97-1 and HCV29 cell lines and TERT-NHUC and suggest that localization of TSC1 to cytoplasmic foci is not a requirement for TSC2 binding or negative regulation of mTOR. Also, defective localization of Phe216Ala is likely TSC2- and mTOR-independent, possibly suggesting an independent function of TSC1 at cytoplasmic foci.  Figure 6. Figure 6. Localization of wild-type TSC1 (A) and TSC1 Ser35Cys (B), Phe216Ala (C) and Thr417Ile (D) mutant proteins in amino acid-starved 97-1 cells. Cells were cultured in full growth medium on highly optically clear microscopy dishes to sub-confluence and amino ... Unlike the other five tumours with missense mutations, the tumour containing the TSC1 648T>A mutation did not show LOH of 9q, as confirmed by analysis of pure microdissected tumour cell populations (data not shown). Both normal and mutant alleles were detected by sequencing of tumour cDNA.  TSC1 protein expression in bladder tumour tissues  To examine the effects of pre-mRNA splicing defects and protein stability defects caused by missense mutations at the tumour level, TSC1 mutant and wild-type tumours were screened for expression of TSC1 protein by immunohistochemistry. Sensitivity and specificity of a rabbit monoclonal anti-TSC1 antibody was confirmed by staining of paraffin-embedded pellets of TSC1-null HCV29 cells and HCV29 in which wild-type TSC1 had been re-expressed (Fig. 7A and D). Normal ureter showed strong cytoplasmic TSC1 expression in the urothelium (Fig. 7E). Tumours with wild-type TSC1 showed strong cytoplasmic staining (Fig. 7F), and a tumour with a TSC1 73\\u201377\\u0394 5 small deletion causing premature truncation at residue 27, showed no TSC1 expression (Fig. 7C). Both tumours with homozygous TSC1 missense mutations causing pre-mRNA splicing defects were negative for TSC1 expression (Fig. 7G and H). Figure 7H shows normal urothelium with strong TSC1 staining adjacent to immunonegative TSC1 314A>G mutant tumour cells. The 648T>A (Phe216Ala) mutant tumour showed moderate TSC1 expression consistent with normal stability of the Phe216Ala protein and retention of chromosome 9q heterozygosity in the tumour (Fig. 7I). Of the three missense mutations causing reduced protein stability (TSC1 203A>G, 473T>G and 616C>G), tumour material was available only for the 616C>G (His206Asp) mutant sample. The 616C>G mutation was identified from tumour material resected in 2003, and TSC1 expression was shown to be low in three tumour resections from this patient in successive years (Fig. 7J\\u2013L).  Figure 7. Figure 7. TSC1 immunostaining of HCV29 Neo (A) and TSC1 (D) cell pellets, normal ureter (negative (B) and positive (E) antibody controls) and TSC1 73\\u201377\\u0394 5 (C) and TSC1 wild-type (F) bladder tumours. TSC1 staining of TSC1 missense mutant bladder ... Go to: DISCUSSION We have demonstrated that bladder tumour-derived TSC1 missense mutations result in loss of TSC1 function and that this occurs via distinct mechanisms. An overall TSC1 mutation frequency of \\u223c14.5% is found in bladder cancer (24) (Platt et al., in preparation) and missense mutations comprise 26% of mutations found to date. The identification of deleterious missense mutations in bladder tumours indicates a causative role of loss of TSC1 function via this mechanism in bladder tumorigenesis. The vast majority of TSC1 mutations in TSC disease are predicted to be protein truncating in nature, and no significant genotype/phenotype correlations have been observed (2,5,45\\u201348). Non-chain terminating TSC1 mutations (missense or in-frame deletions) are rare in TSC disease (27,47,49,50) (www.LOVD.nl/TSC1). Although missense mutations have been reported previously, most have not been confirmed as disease-causing. Several have been revealed as rare polymorphisms or associated with other nonsense mutations in the same patient (46,50). Others have been disregarded as potentially disease-causing on the basis of conservative amino acid changes, or have not been characterized further (27,51). However, a recent publication by Jansen et al. (26) that identified three missense mutations (L916R, M224R and E412V) in affected individuals reported functional analyses on two of these. It was reported that E412V affected RNA splicing. Interestingly, in transfections of constructs of wild-type TSC1 and the M224R variant, the latter showed lower levels of protein expression that were unable to completely suppress S6 phosphorylation, results similar to those described here for missense variants with reduced protein half-life.  We have shown that two of six bladder tumour-derived missense mutations result in pre-mRNA splicing defects, three lead to reduced stability of mutant proteins and intriguingly, one mutation causes a TSC2- and mTOR-independent localization defect of the mutant protein. Wild-type TSC1 showed a granular cytoplasmic distribution in cells in full growth medium and became localized almost exclusively to large cytoplasmic bodies in amino acid starved cells. In contrast, TSC1 was cytoplasmic and diffuse in serum-starved cells. Preliminary results indicated that TSC1 bodies were dynamic, non-aggresomal structures (data not shown). Interestingly, the Phe216Ala substitution abolished the localization of TSC1 to these bodies. As expected from its position outside the recognized TSC2 binding domain, TSC1 Phe216Ala retained TSC2 binding activity and mTOR suppressive activity in amino acid starved conditions, suggesting that localization of TSC1 to cytoplasmic foci was not a requirement for TSC2 binding or for suppression of mTOR signalling. This may also suggest that TSC1 has a separable function at cytoplasmic foci that is independent of TSC2 and independent of negative regulation of mTOR by TSC1/TSC2. However, it is not clear whether the formation of intensely staining bodies in conditions of over-expression of TSC1 and/or TSC2 (39,52) is relevant to the normal physiological condition. Further investigation of the function of Phe216Ala is now required, ideally at levels of expression that are closer to normal and in both TSC1-null cells and those expressing endogenous wild-type TSC1.  Although the missense variant 1250C>T (T417I) was found in the patient\\u2019s germline and this patient showed no symptoms of TSC, we included this variant in our analyses as threonine 417 has previously been identified as a site of CDK1-dependent phosphorylation (29). We found no evidence for a functional defect and conclude that this represents a rare polymorphism. As this patient was Japanese, as were both TSC patients in whom this variant was reported previously (27,28), this rare variant may be confined to the Japanese population.  Evidence suggests that cis-acting mutations affecting splicing of some tumour suppressor genes can have causal roles in tumour initiation and progression (53). Predicted and confirmed effects of TSC-related and bladder tumour-derived TSC1 missense changes on pre-mRNA splicing reported here underscore the importance of using RNA-based techniques, together with conventional mutation detection methods, to effectively identify disease-causing mutations. One of the mutations that deleteriously affected splicing (104C>G) had an unexpected effect. The use of a cryptic 5\\u2032 donor site in an adjacent downstream exon is a non-conventional splicing event; disease-associated mutations at splice junctions typically result in skipping of the mutant exon (38,54,55). The 104C>G mutation was shown to reduce the exon 3 5\\u2032 splice site motif score and to disrupt an enhancer sequence spanning exon 3/4. 104C>G also disrupts a U1 snRNA-binding motif, spanning the 5\\u2032 splice site from position \\u20133 to +8 (Supplementary Material, Fig. S1). In combination with a normally low-scoring exon 4 3\\u2032 acceptor site, these factors may result in reduced splicing efficiency of introns 3\\u20134. Spliceosome assembly is directed by juxtaposed splicing elements, and were introns 4\\u20135 to splice first, splicing machinery at the 3\\u2032 acceptor site of exon 4 may stimulate recruitment of U1 snRNA to possible binding sites in exon 4, and result in use of a cryptic 5\\u2032 donor site.  In bladder tumours as in TSC disease, nonsense, deletion and frameshift TSC1 mutations result in premature protein truncation and loss of protein function. We have now shown that missense mutations cause loss of function by aberrant splicing or reduced protein stability. The identification of tumour-specific TSC1 mutations in the context of chromosome 9 LOH argues strongly for a direct role of loss of TSC1 function in the aetiology of these tumours. These data are consistent with TSC1 acting as a tumour suppressor gene in bladder cancer in accordance with Knudsen\\u2019s two-hit hypothesis. Biallelic inactivation of TSC1 or TSC2 may not be necessarily required in some TSC-related tumours (9). The discrepancy between frequency of TSC1 mutation and frequency of LOH in the TSC1 gene region in bladder tumours suggests that haploinsufficiency of TSC1 may contribute to tumour development in some cases. LOH of chromosome 9 is a particularly frequent event in bladder cancer and to date, TSC1 is the only gene on 9q found to be mutated in bladder tumours. The contribution of partial loss of TSC1 to clonal expansion of tumour cells with 9q LOH is unknown, and haploinsufficiency remains a possibility in those bladder tumours with 9q LOH and no TSC1 mutation. LOH at the TSC1 locus may accompany another event driving loss of chromosome 9. Deletion of the CDKN2A locus at 9p21, which occurs in up to 50% of bladder tumours (56,57), or of an as yet unidentified chromosome 9 tumour suppressor gene, may constitute such a driving force. Loss of one copy of TSC1 may therefore be an advantageous gratuitous hit or \\u2018passenger event\\u2019 (58).  TSC is not a cancer prone syndrome and TSC lesions very rarely progress to malignancy. That TSC patients do not have an increased risk of developing bladder cancer, or other sporadic cancers, may be explained by the requirement of cumulative genetic insults and the typically late onset of malignant disease. The timing and order of initiating and subsequent genetic events in hamartoma and bladder tumour development is likely to be critical in determining malignant potential. Why TSC1 appears to be involved in bladder cancer and not other epithelial cancers is an unresolved question.  Frequent LOH of 9q is found in other tumour types, including ovarian carcinoma, gallbladder carcinoma, nasopharyngeal carcinoma and non-small cell lung cancer (59\\u201365). However, reports of TSC1 and TSC2 mutation status in sporadic tumours other than bladder are very few. No mutations were observed in sporadic glial and glioneuronal tumours or renal cell carcinomas (66,67). Fifty-three and 39% of lung adenocarcinomas and precursor lesions, respectively, were found to have LOH of 9q (68). Subsequently, a screen of 47 lung adenocarcinomas identified three confirmed TSC1 mutations. However, LOH and mutations were not detected simultaneously (69).  In ovarian carcinoma, gallbladder carcinoma, nasopharyngeal carcinoma and lung adenocarcinoma, LOH or deletion is described in both TSC gene regions (63,68,70,71). If loss of TSC1/2 complex function is the pathogenic effect of loss of TSC1, it may be expected that bladder tumours would also show loss of TSC2 function. In an LOH screen of 16p, we have detected LOH for markers at the TSC2 locus in 16% of bladder tumours (Platt et al., in preparation). It will be important to screen for mutations in the retained copy of TSC2 in cases with LOH to determine whether there is also an involvement of TSC2 or whether there is an important independent role of TSC1 in urothelial cells.  Abstract We have surveyed the mutations of TSC1 and TSC2 from 38 (25 sporadic, 11 familial, and 2 unknown) Japanese patients with tuberous sclerosis complex. In 23 of 38 subjects, we detected 18 new mutations in addition to 4 mutations that had been previously reported. We also found 3 new polymorphisms. The mutations were not clustered on a particular exon in either of the genes. Seven TSC1 mutations found in 3 familial and 4 sporadic cases were on the exons (3 missense, 2 nonsense point mutations, a 1-base insertion, and a 2-bp deletion). Fifteen TSC2 mutations were found in 5 familial cases, 10 sporadic cases, and 1 unknown case. The 12 mutations were on the exons (8 missense, 1 nonsense point mutations, a 1-bp insertion, a 5- bp deletion, and a 4-bp replacement) and 3 point mutations were on the exon\\u2013intron junctions. Although the patients with TSC2 mutations tend to exhibit relatively severe mental retardation in comparison to those with TSC1 mutations, a genotype\\u2013phenotype correlation could not yet be established. The widespread distribution of TSC1/TSC2 mutations hinders the development of a simple diagnostic test, and the identification of individual mutations does not provide the prediction of prognosis. Key words Tuberous sclerosis complex \\u2022 TSC1 gene \\u2022 TSC2 gene \\u2022 Hamartin \\u2022 Tuberin \\u2022 Mutation Introduction Tuberous sclerosis complex (TSC) is an autosomal dominant disease with an estimated prevalence of 1 in 10,000 in the Caucasian population and 1 in 31,000 in the Japanese population (Ohno et al. 1981; Gomez 1988). Multiple organs are affected in TSC, including the central nervous system (mental retardation, epilepsy), skin (facial angiofibroma, shagreen patches, hypopigmented macules, periungual fibromas), heart (cardiac rhabdomyomas), and kidney (cysts, angiomyolipoma). There are two causative genes of TSC. The TSC1 (hamartin) is located on 9q34 and consists of 23 exons. The TSC2 (tuberin) is located on 16p13.3 and consists of 41 exons (European Chromosome 16 Tuberous Sclerosis Consortium 1993; Van Slegtenhorst et al. 1997). Loss of heterozygosity (LOH) of either TSC1 or TSC2 in the affected tissues indicates that both act as a tumor suppressor (Green et al. 1994; Henske et al. 1995, 1996; Sepp et al. 1996). Multiple isotypes of tuberin are generated by alternative splicing at two different exons (exon 25 and 31) of TSC2 (Xiao et al. 1997). Since the identification of the TSC1/ TSC2 genes, the mutations in Caucasian TSC patients have been extensively surveyed (Kumar et al. 1995 a,b; Vrtel et al. 1996; Wilson et al. 1996; Bajek et al. 1997; Jones et al. 1997; Maheshwar et al. 1997; Van Slegtenhorst et al. 1997; Au et al. 1997, 1998). The mutations in Japanese TSC patients, however, have remained unknown. In the present study, we surveyed the mutations in Japanese patients and assessed the relationship between the mutations and clinical symptoms. Materials and methods Patients and sample collection Thirty-eight (25 sporadic, 11 familial, and 2 unknown) Japanese TSC patients were clinically diagnosed according to the criteria (Gomez 1988). In some cases, ultrasound renogra392 phy, MRI, and CT brain scan gave supportive evidence for the diagnosis. Genomic DNAs were prepared from the blood samples as described (Nanba et al. 1995). Healthy Japanese volunteer DNA as normal controls were extracted by the Nucleic Acid Purification System (MagExtractor MFX-2000; Toyobo, Osaka, Japan) from 100\\u00edl blood. Polymerase chain reaction-single strand conformation polymorphism (PCR-SSCP) analysis All coding exons of TSC1 and TSC2 were analyzed for every patient. Sets of PCR primers for the 41 exons of TSC2 were designed based on the TSC2 sequence, except for the primers for exon 13 and 25, which were kindly obtained by Dr. J.R. Sampson at the University of Wales, Cardiff, UK (Maheshwar et al. 1996) (Table 1). The primer information for the 21 coding exons of the TSC1 gene was obtained from Dr. O Hino in the Department of Experimental Pathology, Cancer Institute, Tokyo, Japan, and Dr. D.J. Kwiatkowski in Brigham and Womens\\u2019 Hospital, Boston, USA, except for the primers for exon 9 (CACTGAGTTGACACTCTGAAG and CTGAACTAAGTCTTACTCCAG), which were originally designed. PCR contained 25ng of the genomic DNA as a template, 10pmol of each primer, 0.25\\u00edM of each dNTP, 0.1 unit of Ampli Taq Gold (Perkin-Elmer, Norwalk, CT, USA), and a buffer in a total volume of 5\\u00edl. Cycling conditions were as follows: an initial denaturation at 95\\u00b0C for 10 min followed by 30 cycles of denaturation at 95\\u00b0C for 1 min, annealing at 65\\u00b0C for 1min and elongation at 72\\u00b0C for 1min, with a final extension at 72\\u00b0C for 5min. All the PCR products were analyzed on 1.5% agarose gel. SSCP analysis was performed as described (Orita et al. 1989; Yuasa et al. 1997). Using a minigel (10 3 10 cm), the samples were analyzed under four different electrophoresis conditions, in a combination of two sets of gel mixtures (12% polyacrylamide gel with or without 5% glycerol) and two temperature conditions (4\\u00b0C or 22\\u00b0C). Table 1. Primers for TSC2 gene analysis Exon Product size (bp) 1 F: AAGGTTATGCCCACCAGAGA; R: AGACACAGGTAGCTCACTCA 296 2 F: CACTGGCCCCTTTTTCTTCT; R: CTCTAGTAGCTCAACTGGAT 150 3 F: CTTGGAGAGCACATCCTCAC; R: CAGCCTGACGTCACCCATCC 195 4 F: CTGATCCTGTGGCTTTTGTC; R: AGAAACCTCCAACCCAAGGT 199 5 F: TCGCAAACTGCGCGACTTCT; R: CAGAACAGGATCTCAATTCT 173 6 F: GACTGAGCTCGGTGCTCCCT; R: TGAAATGGGCCGGCCCTGGG 109 7 F: GAAGGAGGTGGGAAGGAAG; R: CATCCATGTGCTCTCAGGAA 278 8 F: CTTTGGGAGGAGATGGTGG; R: ATGTGTGCGTTGGTAAGGGA 286 9 F: CGTCTCTCTGGGGAACACTT; R: TCACTGCACACAGAAACCGC 176 10 F: TGCTGGCCGGGCTCGTGTTC; R: TGCTCCCCGGAGCTCCTGCC 198 11 F: GCCTGTGTCATCGTGCCTGG; R: AAAGGCCCCACCCAGGAGGG 194 12 F: CAACACCGGCTCTTCTTTTG; R: AGGCAAGGCTACAGAGGACC 153 13a F: GGGCTGTGGCGGGCACTC; R: AGAGGTCCGGCAGACCGAAGTC 245 14 F: TTGGCCTCCCTTGTGCCTGT; R: GCCCCGGGTACAACGGCTCT 203 15 F: GACTCAGAACCATGAGCCTG; R: AACAGACTCCAACACAACGC 251 16 F: CGTGGTGAGCTGCGTCCTCT; R: ACTGCGTGCGCAACCCCAGC 178 17 F: GGCTTTCACCATCCTCTTCC; R: TCATGGGCCGTGGGCAGGTC 160 18 F: GCCTCAGCTGCTTCTCTTGC; R: AACGTGGGACGGATGGTCCC 216 19 F: CCTGACGCCTCCTCTCCTCG; R: CCCAACCCCAGGCCGGCATG 185 20 F: TCATGCCTGGATTTGGTCAT; R: TCAAGCTCGCCTGCTCTGAC 187 21 F: GGGCCTGAGGTGTCCTGTCT; R: AGGTGCAGTGCTGCAGGGCG 246 22 F: GGCTCCCTGACCACCTCTCC; R: ACAGGGCCTGGGGCGACCAC 148 23 F: GTGCCAGCCCCCTTCTCATC; R: ACAGCCTTCACCGCCCTGAG 152 24 F: TCACTGTCTGGGTGTGCTC; R: TGCCCTGCCTCACATACTC 260 25a F: CCCTCCACTGGCTTGTTCTCC; R: CGGGCAAGACGATGAGGTCAT \\u2014b 26 F: CCTGGTCACGGCCTCTCCCT; R: CCAGCCCTGTAGTGCCGCCT 213 27 F: GGTTTCACGCTCCCTGTCTT; R: GGAGGAAAGGAAGGTGCAGT 200 28 F: ACCCTGTGCGTGGGATTCTC; R: AGGTGGCTGGGGCCAAGGCT 163 29 F: CCTCATGCCCGTCTTCTTC; R: AGATGGCGCTCAGGCCAC 267 30 F: AAGGTAGTCTGCCGCCTCCG; R: CCAAGGCCCGCCATGCCACT 251 31 F: CGTCGACACGGCCTTCCCTT; R: ACCGCTTCCCTCTGAGAGGC 118 32 F: AGCAGCCCCGTCTGTGTCCT; R: GTCCAGGCTCTGAGCCACAC 171 33 1F: CTCTTTGGGATGGTCCTTTC; 1R: CTGTGACCGGGCCTTAACCT 275 33 2F: CCTGGGGACAAGGCCGACGT; 2R: CCCGCCCGGAAGCAAGAGGC 317 34 F: CCACCATCCCCTCCCTGT; R: TAGCCAGAGCCCTCTGCAA 206 35 F: CTCTGGCTAAGCTCCCTGT; R: TGAGCACTTCATGCTGTAGG 326 36 F: GGGAGTGATGCCACCCTG; R: ATGGGAGCCGTGCCCTGA \\u2014b 37 F: CATCCGGCCCTGCTCACCCT; R: GCCTCACTGACGGCCCTCAC 187 38 F: TGACCCTTTCTCTTGTCCGG; R: TCTGCACTGCAGCCCCACCC 131 39,40 F: TGGCGTGACCACCAAGTCTC; R: GCTGAGGAGCCCCATATTCC 325 41 F: CAGACTTACTGCCCAAGCCGCCT; R: CCCCGCACCAAGCAGACAAA 342 a Data were kindly provided by Dr. J.R. Sampson (University of Wales, UK) b The exact sizes of the PCR products were not known 393 Sequencing Variant bands were reamplified and subcloned into pGEMT Vector (Promega, Madison, WI, USA). The DNA sequences were determined using the Thermo Sequenase Fluorescent Labeled Primer Cycle Sequencing Kit (Amersham, Buckinghamshire, England) with Cy-5-labeled M13 universal or reversal primers and ALFred Automatic Sequencer (Pharmacia Biotech, Tokyo, Japan). At least ten independent clones of each PCR product were sequenced in both directions. The sequence data were analyzed by Genetyx computer software (Software Development, Tokyo, Japan). Results SSCP analysis A total of 25 variants were detected by SSCP analysis (Fig. 1) from 23 patients (60% of the total 38 patients). Seven variants of TSC1 gene were detected in 3 familial cases and 4 sporadic ones. Fifteen variants of TSC2 gene were detected in 5 familial cases, 10 sporadic cases, and one unknown case. L574L of TSC1 and I531I of TSC2 genes were considered to be polymorphisms because the point mutations did not cause amino acid substitution. M322T of TSC1 was detected in the normal controls (6 of 80) and was supposed to represent a polymorphism (Table 2A). All other missense mutations were not detected in 60 or 100 normal controls. Any DNA sample from the parents was not available except for P28. Mutations in TSC1 gene Two nonsense, two frame shift, and three missense mutations were located in various exons of the gene. Two mutations (R509X, Q654E) in three patients were located in exon 15, and the R509X was found both in a sporadic patient (P26) and a familial patient (P3). Two mutations (W676X and T417I) were found in one patient (P31). Mutations in TSC2 gene One nonsense, 3 frame shift, and 8 missense mutations were identified in 16 patients (Table 2B). The two single nucleotide substitutions (2760 1 1 G to A, 5179 2 1 G to T) of an exon\\u2013intron junction were predicted to cause a splicing error. Although a substitution (5178 1 5 G to A) in the intron was not in the consensus sequence of the splicing, the mutation was located 5bp distant from the exon and was not detected in 100 normal controls. The mutation might cause a splicing error. L717R in a patient with pulmonary tuberous sclerosis was reported previously (Zhan et al. 1999). Both the V769E mutation (P2 and P21) and R1459X mutation (P6 and P8) were detected in 2 unrelated sporadic patients. Two missense mutations (H137R and D647N) Fig. 1. PCR-SSCP analysis of the TSC1/TSC2 genes. Arrows indicate the aberrant bands found in the patients\\u2019 samples. C, control samples from healthy subjects; P, patients\\u2019 samples 394 were found in the same patient (P32). P1657L in P28 was confirmed to be a de novo mutation by examination of the parents\\u2019 samples. Clinical symptoms and mutation types The clinical symptoms and the gene mutations are summarized in Table 3. The number of the patients with mental retardation (1 and 11) in TSC2 was significantly larger than that in TSC1 (\\u00f82 5 4.8, P 5 0.0285). The severity of mental retardation of the patients with TSC1 mutations appeared to be milder than that of the patients with TSC2 mutations. Complete information on renal and cardiac involvements was not available and could not be assessed. None of the mutation types (missense, nonsense, frame shift, and splicing) appeared to be correlated with a specific clinical symptom. In particular, the R509X mutation in TSC1 was detected in two patients, and the severity of mental retardation in these patients were quite different. Discussion Mutations in a given gene may be screened by various methods such as Southern blot analysis, SSCP, heteroduplex analysis, protein truncation test, and RT-PCR followed by SSCP (Kumar et al. 1995a,b; Vrtel et al. 1996; Wilson et al. 1996; Bajek et al. 1997; Jones et al. 1997; Maheshwar et al. 1997; Au et al. 1997, 1998). In the present study, we employed SSCP to survey the mutations of TSC1 and TSC2 genes because the method is simple enough to cover the multiple numbers of exons. To increase the mutation detection rate of SSCP, we used four electrophoresis conditions by a combination of different sets of temperature and glycerol content of the gel. Our mutation detection rate was 60% (23 of 38), which is comparable with those in other reports (Jones et al. 1997; Van Slegtenhorst et al. 1997; Au et al. 1998). There are several possible explanations for the failure to detect of TSC1/TSC2 mutations in the 15 patients. First, it can be false negative. One report showed that the SSCP Table 2A. Mutations in TSC1 Type of Nucleotide Patient Amino acid change Location mutation alteration S/F* 46 216PheFS-216X(a) exon 7 Frame shift 866delTT F 11 311SerFS-340X exon 10 Frame shift 1210insT F 31 T417l exon 12 Missense 1471 CtoT S 26,3 R509X(b) exon 15 Nonsense 1746 CtoT F,S 23 Q654E(b) exon 15 Missense 2181 CtoG S 31 W676X exon 16 Nonsense 2249 GtoA S 12 T899S exon 21 Missense 2918 CtoG S 46,27,3 L574L exon 15 Polymorphism 1947 TtoC 17,34,36 M322T exon 10 Polymorphism 1186 TtoC Table 2B. Mutations in TSC2 Type of Nucleotide Patient Mutation Location mutation alteration S/F* 32 H137R Exon 4 Missense 428 AtoG S 48 238LeuFS-337X Exon 7 Frame shift 725 insT S 30 F320L Exon 9 Missense 978 TtoG 43 337GluFS-351X Exon 10 Frame shift 1029GATCdel, CATGGCAT ins F 22 R611Q(c) Exon 16 Missense 1850 GtoA F 47 597HisFS-614X Exon 16 Frame shift 1808delACTAC S 32 D647N Exon 17 Missense 1957 AtoG S 1 L717R Exon 19 Missense 2168 TtoG S 21,2 V769E Exon 20 Missense 2324 TtoA F,S 33 \\u2014 Intron 24 Splicing 2760 1 1 GtoA S 27 V963M Exon 25 Missense 2905 GtoA S 6,8 R1459X Exon 33 Nonsense 4393 CtoT S,S 28 P1657L(d) Exon 38 Missense 5042 CtoT S 24 \\u2014 Intron 39 Splicing 5178 1 5 GtoA F 41 \\u2014 Intron 39 Splicing 5179 2 1 GtoT F 33 1531l Exon 14 Polymorphism 1611 CtoT *S, sporadic; F, familial The initiator ATG begins at base 222 of TSC1 and at base 19 for TSC2 The original amino acid and its position in the protein (numbered from the initiator Met as 1) are followed by a new amino acid for missense mutations, and by X for nonsense mutation a,b The identical change was reported by Van Slegtenhorst et al. (1997) c The identical change was reported by Wilson et al. (1996) d The identical change was reported by Maheshwar et al. (1997) 395 method to detect a point mutation on a 250-bp DNA fragment was 63% effective (Sheffield et al. 1993). Second, it may be true negative in those cases in which the mutation lies outside of the coding sequence, either in the 59-regulatory elements, within the intronic sequences, or in the 39- UTR. In addition, a large-scale rearrangement of genomic DNA may yield true negative results in the SSCP of the coding exons. Although there is no report of a large-scale mutation in TSC1, there are indeed some in TSC2 (Jones et al. 1997; The European Chromosome 16 Tuberous Sclerosis Consortium 1993; Au et al. 1997; Van Slegtenhorst et al. 1997). The mutation surveys of TSC1/TSC2 genes in Caucasian TSC patients revealed no clustering of the mutation in either gene (Wilson et al. 1996; Jones et al. 1997; Van Slegtenhorst et al. 1997; Au et al. 1998). In accordance with these previous findings, the 22 mutations (18 new and 4 reported) found in this study were distributed on various exons and there was no clustering of the mutations. Again in accordance with the previous findings, there was no obvious relationship between the type and location of the mutation and the clinical symptoms. However, the patients with TSC2 mutations appeared to have more severe mental retardation compared with the patients with TSC1 mutations. The same tendency has been noted (Jones et al. 1997). It is therefore at least clear that the clinical phenotype cannot be predicted from the genotype. Because of the heterogeneity of the mutations, it also appears impossible at present to develop a simple diagnostic test for TSC. Acknowledgements We thank Dr. Satoko Kumada, Dr. Akira Uchiyama (Tokyo Metropolitan Fuchu Ryoiku Center), Dr. Youich Sakakihara, Dr. Hidekazu Takeshita (Departement of Pediatrics, Graduate School of Medicine, University of Tokyo), Dr. Khotaro Table 3. Clinical data of mutation detected patients Patient Mental White Subependymal number Age Sex retardation AF macules nodules Seizures S/F Mutations Location of mutation 46 9 M 1 2 1 1 1 F Frame shift TSC1 exon 7 11 7 F 6 1 1 1 1 F Frame shift TSC1 exon 10 31 8 F 1 n.d. 1 1 1 S Missense/nonsense TSC1 exon 12; TSC1 exon 16 3 32 F 11 n.d. n.d. n.d. n.d. S Nonsense TSC1 exon 15 26 21 M 2 2 1 1 2 F Nonsense TSC1 exon 15 23 21 M 1 2 1 1 1 S Missense TSC1 exon 15 12 32 F 2 1 1 1 1 S Missense TSC1 exon 21 32 2 M n.d. n.d. 1 1 1 S Missenses TSC2 exon 4; TSC2 exon 17 48 24 M 11 1 1 1 1 S Frame shift TSC2 exon 7 30 nd nd n.d. n.d. n.d. n.d. n.d. n.d. Missense TSC2 exon 9 43 8 M 11 1 1 1 1 F Frame shift TSC2 exon 10 22 6 F 2 2 1 1 1 F Missense TSC2 exon 16 47 6 F 1 1 1 1 1 S Frame shift TSC2 exon 16 1 17 F 2 1 1 1 1 S Missense TSC2 exon 19 21 44 F 11 1 1 n.d. 1 F Missense TSC2 exon 20 2 32 M 11 1 1 1 1 S Missense TSC2 exon 20 33 48 M 11 n.d. n.d. n.d. 1 S Splicing TSC2 intron 24 27 13 M 2 2 1 1 1 F Missense TSC2 exon 25 6 8 M 11 n.d. n.d. n.d. n.d. F Nonsense TSC2 exon 33 8 9 F 11 1 1 1 1 S Nonsense TSC2 exon 33 28 3 F 1 n.d. 1 1 1 S Missense TSC2 exon 38 24 12 M 2 2 1 1 1 F Splicing TSC2 intron 39 41 7 M 2 1 1 1 1 F Splicing TSC2 intron 39 n.d., information not available; AF, angiofibroma; S, sporadic; F, familial Izumi, Dr. Masahiro Nakagawa (Third Department of Internal Medicine, Faculty of Medicine, Kagoshima University), Dr. Ryutaro Kira (Department of Pediatrics, Kyushu University), Dr. Hitoshi Sejima (Department of Pediatrics, Shimane Medical University), Dr. Masahito Miyazaki (Department of Pediatrics, School of Medicine, Tokushima University), Dr. Akira Kuramochi (Department of Dermatology, Saitama Medical School), and Dr. Sadahisa Ohta (Department of Neurosurgery, Tokyo Medical and Dental University) for collecting the samples of the TSC patients. This work was supported by the Ministry of Health and Welfare of Japan by a special grant for neurocutaneous disease, and by the Ministry of Education, Science, Sport and Culture of Japan by a grant-in-aid for scientific research.  Twenty-seven Japanese patients with the tuberous sclerosis complex (TSC), consisting of 23 sporadic and 4 familial cases, were tested for mutations in the TSC1 and TSC2 genes, using single-strand conformational polymorphism analysis and direct sequencing. Four possible pathogenic mutations were found in the TSC1 gene, including three frame shifts and a nonsense mutation in a familial case. All mutations were expected to result in a truncated hamartin gene product. The TSC2 gene analysis identified six possible pathogenic mutations only in the sporadic cases, including two frame shifts, one in-frame deletion, and three missense mutations. Two of the TSC2 mutations were expected to result in a truncated tuberin gene product. These results of the Japanese TSC patients were compatible with the reports from Europe and the United States, i.e., (1) TSC1 mutations are rarer in sporadic cases than in familial cases, (2) substantial numbers of sporadic cases arise from mutations in the TSC2 gene, and (3) mutations of the TSC1 gene may cause premature truncation of hamartin.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "## Preprocessing the text\n",
        "\n",
        "import nltk\n",
        "nltk.download('stopwords')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nsotR_rcHslw",
        "outputId": "9e6b86e9-c15c-41be-d9f3-ac7a335fcadb"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "[nltk_data] Downloading package stopwords to /root/nltk_data...\n",
            "[nltk_data]   Unzipping corpora/stopwords.zip.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "hVVdKzelVUPE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# stop_words = set(stopwords.words('english'))\n",
        "\n",
        "# def nlp_preprocessing(total_text, index, columns):\n",
        "#     string = \"\"\n",
        "\n",
        "#     total_text = re.sub('[^a-zA-Z0-9\\n]', ' ', total_text)\n",
        "\n",
        "#     total_text = re.sub('\\s+', ' ', total_text)\n",
        "\n",
        "#     total_text = total_text.lower()\n",
        "\n",
        "#     for word in total_text.split():\n",
        "#         if not word in stop_words:\n",
        "#             string += word + \" \"\n",
        "\n",
        "#     data_text[columns][index] = string"
      ],
      "metadata": {
        "id": "uWqLfVJjIL4i"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "7iK9fThjWE46"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(len(stopwords.words('english')))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IxKDK08BWGVl",
        "outputId": "f5fa7244-eaf9-4cfd-a43d-25927efcfa3f"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "179\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "8469TQGQeAGM"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "stop_words = set(stopwords.words('english'))\n",
        "\n",
        "def nlp_preprocessing(total_text, index, column):\n",
        "    if type(total_text) is not int:\n",
        "        string = \"\"\n",
        "\n",
        "        total_text = re.sub('[^a-zA-Z0-9\\n]', ' ', total_text)\n",
        "        total_text = re.sub('\\s+',' ', total_text)\n",
        "        total_text = total_text.lower()\n",
        "\n",
        "        for word in total_text.split():\n",
        "            if not word in stop_words:\n",
        "                string += word + \" \"\n",
        "\n",
        "        data_text.loc[column, index] = string\n",
        "\n",
        ""
      ],
      "metadata": {
        "id": "syosKBf6VZ4P"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#text processing stage.\n",
        "# start_time = time.clock()\n",
        "for index, row in data_text.iterrows():\n",
        "    if type(row['Text']) is str:\n",
        "        nlp_preprocessing(row['Text'], index, 'Text')\n",
        "    else:\n",
        "        print(\"there is no text description for id:\",index)\n",
        "# print('Time took for preprocessing the text :',time.clock() - start_time, \"seconds\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "w3loX7zhdoYt",
        "outputId": "f4dcf01d-e694-4b99-96a6-c6c594edddde"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "there is no text description for id: 1109\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "there is no text description for id: 1277\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "there is no text description for id: 1407\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "there is no text description for id: 1639\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "there is no text description for id: 2755\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n",
            "<ipython-input-14-2828ca736759>:15: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
            "  data_text.loc[column, index] = string\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# start_time = time.time\n",
        "\n",
        "for index, row in data_text.iterrows():\n",
        "    if type(row['Text']) is str:\n",
        "        nlp_preprocessing(row['Text'], index, 'Text')\n",
        "\n",
        "    else:\n",
        "        print(\"There is no text description for id: \", index)\n",
        "\n",
        "# print(\"Time took for preprocessing the text: \", time.time-start_time, \"seconds\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "dhBZ5pQpJ6cZ",
        "outputId": "b9e657ac-1e60-46a6-fb17-4b15d6866552"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "There is no text description for id:  1109\n",
            "There is no text description for id:  1277\n",
            "There is no text description for id:  1407\n",
            "There is no text description for id:  1639\n",
            "There is no text description for id:  2755\n",
            "There is no text description for id:  Text\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_text.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 288
        },
        "id": "homEeUHrLDPe",
        "outputId": "71563a8b-b73d-44df-d9df-4c3773bd62f9"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    ID                                               Text    0    1    2    3  \\\n",
              "0  0.0  Cyclin-dependent kinases (CDKs) regulate a var...  NaN  NaN  NaN  NaN   \n",
              "1  1.0   Abstract Background  Non-small cell lung canc...  NaN  NaN  NaN  NaN   \n",
              "2  2.0   Abstract Background  Non-small cell lung canc...  NaN  NaN  NaN  NaN   \n",
              "3  3.0  Recent evidence has demonstrated that acquired...  NaN  NaN  NaN  NaN   \n",
              "4  4.0  Oncogenic mutations in the monomeric Casitas B...  NaN  NaN  NaN  NaN   \n",
              "\n",
              "     4    5    6    7  ... 3311 3312 3313 3314 3315 3316 3317 3318 3319 3320  \n",
              "0  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "1  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "2  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "3  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "4  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "\n",
              "[5 rows x 3318 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-31ea201f-4e89-4109-97e3-62f1b9ebcf0a\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Text</th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "      <th>4</th>\n",
              "      <th>5</th>\n",
              "      <th>6</th>\n",
              "      <th>7</th>\n",
              "      <th>...</th>\n",
              "      <th>3311</th>\n",
              "      <th>3312</th>\n",
              "      <th>3313</th>\n",
              "      <th>3314</th>\n",
              "      <th>3315</th>\n",
              "      <th>3316</th>\n",
              "      <th>3317</th>\n",
              "      <th>3318</th>\n",
              "      <th>3319</th>\n",
              "      <th>3320</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1.0</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2.0</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3.0</td>\n",
              "      <td>Recent evidence has demonstrated that acquired...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4.0</td>\n",
              "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 3318 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-31ea201f-4e89-4109-97e3-62f1b9ebcf0a')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-31ea201f-4e89-4109-97e3-62f1b9ebcf0a button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-31ea201f-4e89-4109-97e3-62f1b9ebcf0a');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-21baeacb-0866-481f-8825-1a23b692a6e2\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-21baeacb-0866-481f-8825-1a23b692a6e2')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-21baeacb-0866-481f-8825-1a23b692a6e2 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "data_text"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_text.Text[1]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 140
        },
        "collapsed": true,
        "id": "BogXqIp1N1lf",
        "outputId": "2ba6f057-ec2f-494b-8fe6-dabeff775f7b"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "\" Abstract Background  Non-small cell lung cancer (NSCLC) is a heterogeneous group of disorders with a number of genetic and proteomic alterations. c-CBL is an E3 ubiquitin ligase and adaptor molecule important in normal homeostasis and cancer. We determined the genetic variations of c-CBL, relationship to receptor tyrosine kinases (EGFR and MET), and functionality in NSCLC.  Methods and Findings  Using archival formalin-fixed paraffin embedded (FFPE) extracted genomic DNA, we show that c-CBL mutations occur in somatic fashion for lung cancers. c-CBL mutations were not mutually exclusive of MET or EGFR mutations; however they were independent of p53 and KRAS mutations. In normal/tumor pairwise analysis, there was significant loss of heterozygosity (LOH) for the c-CBL locus (22%, n\\u200a=\\u200a8/37) and none of these samples revealed any mutation in the remaining copy of c-CBL. The c-CBL LOH also positively correlated with EGFR and MET mutations observed in the same samples. Using select c-CBL somatic mutations such as S80N/H94Y, Q249E and W802* (obtained from Caucasian, Taiwanese and African-American samples, respectively) transfected in NSCLC cell lines, there was increased cell viability and cell motility.  Conclusions  Taking the overall mutation rate of c-CBL to be a combination as somatic missense mutation and LOH, it is clear that c-CBL is highly mutated in lung cancers and may play an essential role in lung tumorigenesis and metastasis.  Go to: Introduction In the US alone, each year approximately 219,400 people are diagnosed with lung cancers, out of which more than 145,000 of them succumb to the disease [1]. This number is roughly equivalent to the combined mortality rates of cancers of the breast, prostate, colon, liver, kidney and melanoma [1]. In addition the prognosis is usually poor and the five-year survival rate is less than 15%. There are also significant ethnic differences for lung cancer, and the outcome is worse for blacks compared to whites. Gender differences are also striking with women having significantly better prognosis as compared to men. There are a number of genetic alterations that can occur in lung cancer. As an example, in NSCLC, mutations in KRAS, p53, EGFR and MET have been identified. Many of these pathways, especially Receptor Tyrosine Kinases (RTKs) are controlled by c-CBL.  CBL (Casitas B-lineage lymphoma) is a mammalian gene located on human chromosome 11q23.3 [2] and is involved in cell signaling and protein ubiquitination [3]. CBL proteins belong to the RING finger class of ubiquitin ligases (E3) and there are three homologues c-CBL, CBL-b, CBL-3 [4]. The c-CBL and CBL-b genes are ubiquitously expressed with the highest levels in hematopoietic tissues [5]. c-CBL consists of four regions encoding for functionally distinct protein domains: the N-terminal tyrosine kinase binding (TKB) domain, the linker region, the catalytic RING finger domain, the proline-rich region and the c-terminal ubiquitin-associated (UBA) domain that also overlaps with a leucine-zipper (LZ) domain [3]. Both TKB and RING finger domains are essential for ligand-induced ubiquitination of RTKs [6], [7], [8], [9]. The RING finger domain is required for the recruitment of E2 ubiquitin-conjugating enzymes. The TKB domain includes four-helix bundle (4H), a calcium-biding EF hand, and a modified SH2 domain, which binds to phosphotyrosine residues [3], [10], [11], [12]. In addition, the proline-rich region of c-CBL can associate with the SH3 domain of Grb2, which can indirectly recruit c-CBL to RTKs via the GRB2 adaptor protein [7], [13], [14].  c-CBL also binds to EGFR and acts as the E3 that targets EGFR for ubiquitination and degradation. Furthermore, CBL desensitizes EGF signaling and opposes cellular proliferation induced by EGF [15]. EGF activation also appears to activate the tyrosine kinase SRC, which phosphorylates c-CBL and in turn activates the ubiquitination and degradation of EGFR [16], [17], [18]. A recent study shows that defective endocytosis of EGFR is characterized by a deletion mutant and the point mutation L858R, whereby its association with c-CBL and subsequent ubiquitination are impaired [19]. Recently, the first human c-CBL mutations were reported in acute myeloid leukemia (AML) patients [20]. The mutation R420Q inhibits FMS-like tyrosine kinase 3 (FLT3) internalization and ubiquitination [20].  Not only can E3 activity be important in oncogenesis, c-CBL has a dual but separate function as a signal transduction molecule. We have previously shown that c-CBL is important in binding CRKL and BCR/ABL in hematopoietic cells. Also, it can bind and modulate functions of cytoskeleton by binding to proteins like talin and paxillin. The TKB domain is important in binding to a number of molecules, and they then function in signal transduction.  Given the critical role of CBL in normal homeostasis and cancer, we hypothesized that it might be mutated in lung cancers. In this study, we report novel c-CBL somatic mutations S80N/H94Y, Q249E and W802* in Caucasian, Taiwanese and African-American lung cancer patients, respectively. Expressing these mutations in NSCLC cell lines lead to increased proliferation and cell motility. We show that c-CBL mutations occur with or without MET or EGFR mutations but are mutually exclusive of a LOH at the c-CBL locus. Additionally, c-CBL LOH is associated with either MET or EGFR mutations. We thus hypothesize that c-CBL mutations might contribute to the oncogenic potential of MET and EGFR in lung cancer.  Go to: Methods Ethics Statement  Written consent on all research on human subjects has been obtained from the Institutional Review Board, University of Chicago and covers all research performed in the laboratory. The following is their contact information: Institutional Review Board, The University of Chicago, McGiffert Hall, 5751 S. Woodlawn Ave., 2nd floor, Chicago, IL 60637. Written informed consents were received from all patients whose tissue samples were used for this study.  Tissue Samples  Lung cancer tissue and paired adjacent normal lung tissues were obtained from 50 Caucasian, 29 African-Americans and 40 Taiwanese NSCLC patients who were recruited at the University of Chicago Hospital (Chicago, USA) (Caucasian and African-American patients) and Taipei Veterans General Hospital of Taiwan (Taiwanese patients) after obtaining appropriate Institutional Review Board permission and informed consent from the patients. Out of 119 samples, 77 were men, 38 were women and 4 were unknown with age at diagnosis ranging from 47 to 90 years. In terms of tumor types, 53 were adenocarcinoma, 32 were squamous cell carcinoma and 34 were large cell carcinoma. 49 were stage I, 14 were stage II, 34 were stage III, and 13 were stage IV (Table S1).  Cell Culture  Human non-small cell lung carcinoma cells A549 and H358 were maintained in DMEM and RPMI-1640, respectively. Human embryonic kidney 293T cells were cultured in DMEM. Media were supplemented with 10% fetal bovine serum, 100 units/ml of penicillin, and 100 µg/ml of streptomycin (Invitrogen, Carlsbad, CA). Cells were cultured at 37°C in a humidified incubator containing 5% CO2.  c-CBL Gene Mutational Analysis  Exons 2 to 16 of c-CBL gene were individually amplified by polymerase chain reaction (PCR). Primers are listed in Table S2. PCR conditions were 1 cycle of 95°C for 5 minutes; 35 cycles of 94°C for 30 seconds, 58°C for 30 seconds and 72°C for 2 minutes; and one cycle of 72°C for 10 minutes. PCR products were treated with ExoSAP-IT (USB Corporation, Cleveland, OH) and sequenced by Big-Dye Terminator Chemistry (Applied Biosystems, Foster City, CA). Sequencing was performed on the forward coding strand with confirmation of c-CBL alterations performed by sequencing the reverse strand as well. Chromatograms were analyzed for mutations using Mutation Surveyor v2.61 (Softgenetics, State College, PA).  Plasmid Constructs and Site-Directed Mutagenesis  The wild-type c-CBL cDNA insert was subcloned into the pAlterMax expression vector using XhoI and SalI restriction enzyme sites (Promega, Madison, WI). Using this parental plasmid pAlterMax-c-CBL, the TKB domain double mutation (S80N/H94Y), the point mutation (Q249E), and the C-terminal point mutation W802* of c-CBL were created using the following primers: 5′-GCTGGCGCTAAAGAATAACCCACCTTATATCTTAGAC-3′ and 5′-CTACCAGATACCTACCAGTATCTCCGTACTATCTTGTC-3′ for the double mutation S80N/H94Y; 5′-CTTTACCCGACTCTTTGAGCCCTGGTCCTCTTTGC-3′ for Q249E, and 5′-CAGCTCCTCCTTTGGCTGATTGTCTCTGGATGGTGATC-3′ for W802* along with their complementary primers using the QuickChange Site-Directed Mutagenesis XL kit (Stratagene, La Jolla, CA) according to the manufacturer's instructions. The constructs were confirmed for the point mutations by standard DNA sequencing of both strands.  Loss of Heterozygosity (LOH) Analysis  Five microsatellites on chromosome 11 (3 on 11q at or within 200 kb up or downstream of the c-CBL gene and 2 control markers on 11p) were selected for analysis (Table S3). Established microsatellite markers and respective primer sequences were selected from the GeneLoc database (http://genecards.weizmann.ac.il/geneloc/index.shtml, Weizmann Institute of Science, Rehovot, Israel). Primers were custom designed and each forward primer was fluorescently labeled at the 5′ end with FAM, PET, NED, or VIC (Applied Biosystems). Primer annealing temperatures and duplex scores were evaluated with NIST Primer Tools (http://yellow.nist.gov:8444/dnaAnalysis/primerToolsPage.do; National Institute of Standards and Technology, Gaithersburg, MD). Primers were verified by performing PCR with control DNA (isolated from TK6 cells) and resolving the products on agarose gels. Bands were visualized with an UV transilluminator. Genomic DNA was extracted from tumor samples and paired normal lung tissue. Primers were grouped into multiplex combinations shown in Table S4. Marker D11S929 served as an internal control to check for consistency in PCRs and of peaks from capillary electrophoresis. Multiplex PCRs were carried out in a volume of 10 µL that contained 1 µL genomic DNA (20–50 ng), 0.5 µM of each primer (1.0 µM total for each primer pair), 400 µM dNTPs, 1X PCR buffer containing MgCl2, and 0.2 U Taq DNA polymerase. PCR was performed on the ABI GeneAmp 9700 PCR System under the following conditions: 5 min at 94°C; 30 cycles of 30 sec at 94°C, 1 min at 60°C, 1 min at 72°C; and 5 min at 72°C. The PCR products were separated by capillary electrophoresis on an ABI 3130XL DNA Analyzer. Chromatograms were analyzed with Peak Scanner 1.0 and GeneMapper 3.7 software (Applied Biosystems) for allelic alterations. The area of the peaks produced by the DNA PCR products was quantified for each allele. The ratio of the allelic areas was calculated for each tumor and paired normal DNA sample. When the qLOH (allelic ratio for the tumor peaks divided by the allelic ratio of paired normal sample) was ≤0.5 or ≥2.0 for c-CBL and at least one other 11q marker in at least two separate experiments, the sample was considered as having an allelic imbalance and interpreted as LOH. Samples were evaluated in at least two separate experiments and samples showing prospective LOH at c-CBL repeated a third time which included a new control marker at the BAX locus (data not shown) on chromosome 19 to verify integrity of sample DNA.  Transfection of c-Cbl Constructs  The A549 cell line was transfected using the Fugene HD (Roche, Nutley, NJ) reagent according to the manufacturer's instructions. Eight µg of plasmid DNA, containing either no insert (empty vector), wild-type c-CBL, S80N/H94Y c-CBL, Q249E c-CBL or W802* CBL was used for transfection in a 6-well culture plate. Cells were harvested 48 h after transfection and analyzed for expression.  c-CBL Knockdown  c-CBL knockdown was performed using lentiviral transduction using MISSION lentiviral transduction particles (Sigma-Aldrich, St. Louis, MO) as per manufacturer's instructions. Briefly, 1×105 H358 cells/well were seeded in 6-well plates and infected the following day with c-CBL lentiviral shRNA constructs. To generate stable c-CBL knockdown cell lines, cells were selected for 2 days with 1 µg/ml puromycin. c-CBL levels were determined using whole cell lysates by immunoblotting with anti-CBL antibody (Santa Cruz Biotechnologies, Santa Cruz, CA).  Cell Viability Assay  Cells were transfected as described above in the transfection assay. Forty-eight hours after transfection, viability of cells was assessed using Trypan Blue exclusion.  Wound Healing Assay  A549 cells were seeded in 6-well plates and cultured for 48 h until 100% confluent. The medium was then changed and the cells were transfected as described in the transfection assay. Twelve hours after transfection, a straight scratch was made across the cell layer using a 1 ml pipette tip. The cells were then gently washed with 1× PBS to remove cellular debris and the media was replaced. Photographs were taken of the wound region every 12 h until 48 h.  Western Blot Analysis  Forty eight hours after transfection, cells were collected and washed twice in 1X PBS, then lysed in ice-cold lysis buffer (0.5M Tris-HCl with pH 7.4, 1.5 M NaCl, 2.5% deoxycholic acid, 10 mM EDTA, 10% NP-40, 0.5 mM DTT, 1 mM phenylmethylsulfonyl fluoride, 5 µg/mL leupeptin, and 10 µg/mL aprotinin) for 5 minutes. The lysate was centrifuged at 13,000 rpm for 20 minutes at 4°C, and protein content of the supernatant was measured. Total cell lysates (50 µg/well) were separated by SDS-PAGE electrophoresis and the gels transferred onto nitrocellulose membranes (Whatman, Piscataway, NJ). Membranes were blocked with 5% non-fat dry milk in phosphate-buffered saline containing Tween-20 (PBST) (1X PBS, 0.1% Tween-20) for 1 h at room temperature and incubated with the appropriate primary antibody at 4°C overnight. Membranes then were washed three times with PBST and probed with appropriate horseradish peroxidase (HRP)-conjugated secondary antibody for 1 h at room temperature. The membranes were again washed three times in PBST and bands were visualized using Western blot chemiluminescence reagent (BioRad, Valencia, CA) on a Chemidoc Gel documentation system (BioRad, Valencia, CA). Antibodies were obtained from Santa Cruz Biotechnologies and used at the following dilutions (c-CBL, 1∶5000; c-MET, 1∶5000; EGFR, 1∶5000; ubiquitin, 1∶1000; HA, 1∶5000 and β-actin, 1∶10,000).  Flow Cytometry  Cell cycle analysis was carried out by flow cytometry. Approximately 2×106 cells were grown in media containing 10% FBS. Cells were harvested by trypsin/EDTA treatment, washed with 1X PBS three times and fixed with ice-cold 70% ethanol for 2 h. Cells were washed again with cold PBS and stained with a solution containing 25 µg/mL propidium iodide, 200 µg/mL RNase A, and 0.1% Triton X-100 for 30 minutes in the dark. Cell cycle analysis was performed using a Guava PCA-96 flow cytometer (Guava Technologies, Millipore, Billerica, MA).  Ubiquitin Ligase Activity  293T cells were maintained in culture in DMEM supplemented with 10% FBS and 1% penicillin (100 units/mL) and streptomycin (100 µg/mL) were transfected with 0.2 µg EGFR-pcDNA3 and 2 µg HA-tagged c-CBL constructs as indicated using calcium phosphate according to manufacturer's protocol (Profection, Promega, Madison, WI). Twenty-four hours post-transfection, cells were starved overnight in DMEM supplemented with 0.5% FBS, and then treated with or without EGF (100 ng/ml) for 15 min. The cells were collected and washed two times in ice-cold PBS containing 0.2 mM sodium orthovanadate then lysed in ice-cold lysis buffer (10 mM Tris HCl, pH 7.5, 150 mM NaCl, 5 mM EDTA, 1% Triton X100, 10% Glycerol, 2 mM sodium orthovanadate and protease inhibitors). Lysates were cleared of debris by centrifugation at 16,000 g for 10 min at 4°C. EGFR immunoprecipitations were performed on 200 µg of cleared lysate using 250 ng of rabbit-anti-EGFR and Protein A/G Plus Sepharose overnight at 4°C. Precipitations were washed 5 times in lysis buffer before boiling in Laemmli buffer. Elutions were immunoblotted with anti-ubiquitin and EGFR. Twenty micrograms of cleared lysate were immunoblotted for each of the c-CBL constructs using anti-HA.  Statistical Analysis  Mutation rates between different groups were compared using Fisher's exact test. For continuous variables, group comparisons were performed using analysis of variance (ANOVA) followed by Sidak's adjustment for multiple comparisons. Experiments involving repeated measurements over time were analyzed using repeated measures ANOVA with the Greenhouse-Geisser adjustment to the degrees of freedom. Analyses were conducted using STATA (v10.1) software (Stata Corporation, College Station, TX).  Go to: Results c-CBL Gene Mutations in Lung Cancer  To investigate the role of c-CBL in lung cancer, we analyzed its genomic DNA in tumor and paired normal samples drawn from multiple ethnicities. The lung tumor samples represented Caucasians (n\\u200a=\\u200a50), African-Americans (n\\u200a=\\u200a29), and Taiwanese (n\\u200a=\\u200a40) lung cancer patients. We designed 12 pairs of primers to sequence the coding region of c-CBL gene that spans exons 2 to 16 (Table S2). We identified 8 unique somatic mutations in c-CBL exons among 8 different patients. A variation L620F, a known SNP (rs2227988) in exon 11 was also detected. Importantly, the eight novel non-synonymous mutations were confirmed by sequencing both strands of c-CBL genomic DNA obtained from lung tumor samples (Table 1). Moreover, none of the 8 mutations were detected in the corresponding normal tissue, indicating that these were somatic mutations. Four synonymous single nucleotide variations (SNVs) were also identified but were not used further in this study.  Table 1 Table 1 c-CBL mutation analysis in 119 lung cancer patient tumor tissues. Three of the 8 novel non-synonymous mutations were located in the TKB (tyrosine kinase binding) domain (S80N, H94Y, and Q249E), one in the RING finger domain (V391I), one in the proline-rich region (72515_72517 del ATG), and three in the C-terminal region (W802*, R830K, and A848T) of the c-CBL protein (Figure 1A and Figure S1). In Figure 1B, we show model chromatograms of representative samples.  Figure 1 Figure 1 c-CBL mutations and LOH in non-small cell lung cancer. 11q LOH of c-CBL Gene  Paired lung tumor and normal lung tissue samples from Taiwanese patients (n\\u200a=\\u200a37) were investigated for LOH. Eight (21.6%) showed LOH at the c-CBL locus on chromosome 11 while 29 samples (78.4%) revealed normal allelic contribution at the microsatellite markers (Figures 1C and D).  c-CBL Mutations in Different Ethnic Groups  The c-CBL double mutant S80N/H94Y was found in the same patient and the overall mutation rate for c-CBL in lung tumors was 6.7% (8/119). The frequency of c-CBL mutation was highest in large cell carcinoma (14.7%; 5 of 34 patients) followed by squamous carcinoma (6.3%; 2 of 32 patients) and the least was observed in adenocarcinoma (AD) (1.8%; 1 of 53 patients), although these rates were not statistically significant (p\\u200a=\\u200a0.292). Mutation rates were 6.0% among Caucasians (0 of 20 in AD; 0 of 10 in SQ; and 3 of 20 in LC), 13.8% in African-Americans (1 of 10 in AD; 1 of 10 in SQ; and 2 of 9 in LC), and 2.5% (0 of 23 in AD; 1 of 12 in SQ; and 0 of 5 in LC) in the Taiwanese population. Additionally two Taiwanese patients with lung cancer (one squamous and one adenocarcinoma) had the known SNP L620F. Ethnic differences were not statistically significant, however the power to detect differences was low.  Mutations in MET and EGFR Can Be Co-Associated with c-CBL Alterations  Since East Asians with lung cancer have a higher frequency of EGFR and MET mutations in lung tumors [21], [22], we also determined mutations in EGFR and MET in the same Taiwanese cohort samples and compared the results with the observed c-CBL alterations (LOH and/or mutations). In the 37 samples tested, we did not find any overlap between c-CBL mutations and c-CBL LOH (Figure 2). Of the three c-CBL mutants (including the known L620F SNP, rs2227988), one of the samples had a MET mutation (N375S) and the other had an EGFR mutation (L858R). Among the 8 samples that had a LOH at the c-CBL locus 5 had an additional mutation in MET (N375S) and 2 had an EGFR exon 19 deletion. Twenty-six samples had neither c-CBL mutation nor c-CBL LOH (3 patients had a c-CBL mutation but no c-CBL LOH). Among these 26 samples 9 had a MET mutation (8 N375S, 1 L211W), 13 had an EGFR mutation (7 exon 9 deletion, 6 L858R) and 4 had no other MET or EGFR mutation. Thus the rate of MET or EGFR mutations among patients with LOH at the c-CBL locus (7 of 8) was similar to that seen in patients without c-CBL mutation or LOH (22 of 26 patients) (p\\u200a=\\u200a0.99). These 4 patients with no identifiable mutation in c-CBL, MET or EGFR represented 10.8% of the 37 patients analyzed in the Taiwanese patient cohort. Conversely, 89.2% Taiwanese lung cancer patients have an identifiable mutation in either c-CBL, MET or EGFR or a combination of the three genes (Figure 2). Additionally, we determined p53 and KRAS mutations in these Taiwanese cohorts. Two p53 and 1 KRAS mutation were detected. The single KRAS mutation overlapped with one p53 mutation. This patient also had the EGFR exon 19 deletion but had no c-CBL mutation. The other p53 mutation sample had a c-CBL LOH with concurrent MET N375S mutation. Thus, in the Taiwanese samples analyzed, p53/KRAS mutations and c-CBL mutations were mutually exclusive (data not shown).  Figure 2 Figure 2 c-CBL mutations and relationship to MET and EGFR mutations in lung cancer. Cellular Functions of c-CBL Alterations in the Context of Lung Tumorigenesis  A. E3 activity is intact in the mutant c-CBL proteins To investigate whether the different c-CBL mutations affect the E3 activity, EGFR was chosen as a model substrate for c-CBL E3 function. All of the c-CBL mutants tested enhanced ubiquitination of the activated EGFR similar to the wild-type c-CBL protein. This result demonstrates that the catalytic activity of the c-CBL mutants is not impaired when EGFR was the substrate. (Figure 3A).  Figure 3 Figure 3 Ubiquitination, viability, expression and cell cycle analysis of various c-CBL mutants. B. Effect on lung cancer cell viability The effect of a representative c-CBL mutant from each of the three ethnic backgrounds on lung cancer cell viability in cell lines was determined. S80N/H94Y double mutation, Q249E, and W802* were identified in lung tumor samples obtained from a Caucasian, a Taiwanese and an African-American, respectively. As described in methods, the c-CBL wild-type (WT) and the above three mutants were expressed after cloning them into pAlterMax vector in A549 cells. These cells express relatively low basal levels of endogenous c-CBL (data not shown). Transfection efficiency was comparable between different groups and the number of cells transfected with c-CBL wild-type construct was about 70% compared to control cells that were transfected with the empty vector. Cells transfected with S80N/H94Y, Q249E and W802* c-CBL mutant constructs resulted in increased number of viable cells that was 132.3%, 120.8% and 147.9% higher respectively, relative to the empty vector control transfected cells and significantly different from the wild-type construct (p\\u200a=\\u200a0.022, p\\u200a=\\u200a0.049, and p\\u200a=\\u200a0.008, respectively) (Figure 3B). Relative levels of c-CBL protein in whole cell lysates prepared from samples obtained from a parallel experiment was determined. The c-CBL protein levels in samples representing untransfected and empty vector transfected cells were comparable and those representing the c-CBL WT and the three c-CBL mutants were comparable (Figure 3C).  C. Effect on cell cycle To investigate if the increases in cell viability in different c-CBL mutants are due to increased cellular proliferation, a cell cycle analysis was performed. A549 cells were transfected with the c-CBL WT or the three different mutants: S80N/H94Y, Q249E and W802*. The empty vector transfectant was used as a control. Forty-eight hours after transfection cell cycle analysis was performed as described in materials and methods. There was no significant change in the subG1, G1 or the S phase of the cell cycle among the different mutants compared to the WT construct (p\\u200a=\\u200a0.64, p\\u200a=\\u200a0.40, and p\\u200a=\\u200a0.28, respectively). The G2/M phase of the cell cycle showed an increase in cell numbers for the three mutants, S80N/H94Y, Q249E and W802*, when compared to the WT but again the difference was not statistically significant (p\\u200a=\\u200a0.25) (Figure 3D).  D. Effect on cell motility To investigate the effect of the expression of the above three c-CBL mutants on cell migration, we carried out wound healing assay as described in materials and methods. The closing of the scratch or the wound was monitored at 0, 12, 24, 36, and 48 h. (Figure 4A). In all the samples, that represented cells transfected with mutants, the wound gap was much smaller than that seen in the sample that represented cells transfected with c-CBL WT (p<0.001). We also determined the rate of wound closure for all the five groups. At 48 h, wild-type c-CBL transfectants showed 61.1% open wound while the S80N/H94Y, Q249E and W802* mutants showed 18.7%, 23.9% and 34.3% open wound respectively (p<0.001) (Figure 4B).  Figure 4 Figure 4 c-CBL mutations affect wound healing in A549 cells. E. c-CBL knockdown increases cell viability It is hypothesized that the LOH seen in our samples could lead to decreased expression of c-CBL. Thus we tested the effect of c-CBL knockdown in lung cancer cells. Compared to A549, H358 lung cancer cells express relatively high levels of endogenous c-CBL (data not shown). c-CBL expression was knocked down using lentiviral construct that expressed c-CBL specific shRNA and compared the results with those that were transduced with scrambled shRNA and the results are shown in Figure 5. We identified several clones that revealed varying degrees of c-CBL knockdown showing different sets of c-CBL lentiviral shRNA knockdown efficiency (Figure 5A). Of all the clones tested, Clone 27 was chosen for further experiments. Equal amount of cells were seeded in a 6-well plate and the cell proliferation was measured at various times and the results are depicted in Figure 5B. As expected, number of cells increased in a time dependent fashion from 100 to 190% relative to scrambled shRNA as control in a span of 48 h (p\\u200a=\\u200a0.0002) (Figure 5B). The cell cycle phases in H358 cells that were knocked down with c-CBL shRNA were looked at and compared with the scrambled shRNA. There were no discernable differences between these two constructs in the different phases of the cell cycle (data not shown).  Figure 5 Figure 5 Knockdown of c-CBL using an shRNA increases cell proliferation. Go to: Discussion Our results demonstrate that c-CBL is somatically mutated (or has LOH) in lung cancers, and can significantly contribute to enhanced cell viability and motility. There was also a high prevalence of LOH with respect to c-CBL in lung tumors that harbored MET or EGFR mutation.  In the present study, we have demonstrated the occurrence of c-CBL mutations in lung cancer patients, especially with different ancestral variations. Mutations in c-CBL have been recently reported in juvenile myelomonocytic leukemia and myeloid malignancies. In the AML study, the mutation R420Q located in the junction of the RING finger and the linker region inhibited FMS-like tyrosine kinase 3 (FLT3) internalization and ubiquitination [20], thus contributing to the gain-in-function for the RTK. In addition, mutations such as H398Y, C384R, and L380P mapped to the RING finger domain and the linker region of c-CBL that is required for its E3 activity [23], [24], [25], [26], [27]. Additionally, homozygous mutations in the RING finger domain of the c-CBL gene were described as a result of acquired Uniparental Disomy (UPD) [26]. It is important to note that our results indicate LOH at 11q23 locus and these are mutually exclusive from missense mutations of c-CBL. The somatic mutations were all heterozygous. The mutations in AML led to abrogation of the E3 activity leading to prolonged RTK activation. In addition mutants located on the linker region surrounding the RING finger domain exhibited enhanced AKT signaling in response to cytokine stimulation [26]. In addition, it was shown in NH3T3 cells, that neither mutations in the RING finger nor the linker region causes transformation, however while certain mutations perturbs the ubiquitination, others affect receptor recycling and prolong kinase activity [28].  We report here c-CBL mutations that mapped not only to the RING finger domain, but also to the TKB domain, proline-rich domain and the C-terminal region, but none mapped to the linker region as reported in the AML studies described above [23], [24], [25], [26], [29]. In addition, 8 mutants that we detected were found in different ethnic backgrounds. For example, S80N/H94Y, Q249E, W802* were detected in Caucasians, Taiwanese and African-Americans, respectively. The results point out not only the difference between lung cancer and other cancers, but also genetic polymorphism among different races in the same cancer. Interestingly, there is a large disparity between African-American and other ethnic populations with lung cancer [30]. We have previously shown that there was a low frequency of EGFR and MET mutation in African-Americans as compared with Taiwanese and Caucasians [31]. In this study, the number of African-American samples analyzed was relatively fewer and we found 3 mutations that are unique to this ethnicity. It would now behoove us to further study the genetic alterations that can occur and determine the targeted therapeutics for African-Americans. Our results provide evidence of the importance of c-CBL in tumorigenesis and potential signaling. Our prediction, based on the AML data, would be that the V391I RING finger domain mutation would affect the E3 activity. Also, it will be important to determine the binding partners for c-CBL in the TKB domain and proline rich domain mutations. It has previously been shown that the TKB domain can bind to growth factor receptors and it will be important to determine the cross-binding of these mutants to MET and or EGFR. It would also be important in the future to look at fluorescence in situ hybridization/copy number changes in c-CBL in lung cancer.  c-CBL plays an important role in down regulating RTK-mediated signaling through K63 poly-ubiquitination and subsequent downregulation of RTKs followed by lysosomal degredation [3]. Mono-ubiquitination or ubiquitinated with K63-linked chains of substrates by c-CBL may lead to enhancement of biological and biochemical functions (reviewed in Hermann et al, 2007 [32]). The mutations that we analyzed in our studies all point out to the fact that the E3 activity of c-CBL on EGFR is intact; the EGFR levels in the various mutants remain same (Figure S2). Multiple kinases, both RTKs and non-RTKs could be acted upon by c-CBL, including ERBs, PDGFR, FMS, MET, c-Kit, VEGFR, FLT-1, RON, FGFR, IR, as well as SYK, FYN, LCK, FGR, LYN and c-ABL [3]. In lung cancers the relevant substrates of c-CBL in terms of degradation or signal transduction are yet to be identified.  The observation that c-CBL somatic mutations, especially S80N/H94Y, Q249E and W802* showed increased cell viability and cell motility that is in agreement with the physiological role for D-cbl in the regulation of apoptosis and differentiation identified in Drosophila is very significant [33]. It has been previously shown that activating c-CBL mutation downregulates EGFR signaling and decreases cellular proliferation and migration in breast cancer cell lines [34]. Although the role of c-CBL in the negative regulation of RTKs is well substantiated, thereby suggesting that it is a natural tumor suppressor, studies in cancer cells have revealed both tumor suppressor and tumor promoting activities depending on the type of c-CBL mutation and the number of alleles at the c-CBL locus [24]. In agreement with the above, the three c-CBL mutants described here appear to have tumor growth and metastasis promoting properties. Although these mutants are outside of the RING finger or the linker region of c-CBL, their downstream effects are significant so as to cause increased proliferation and migration, but the substrate affected by these mutations are not known yet. This raises the possibility that some of the cellular functions of c-CBL are independent of its ubiquitin-ligase activity, an area that we are currently investigating. The oncogenic nature of RTKs, addiction of cancers to growth signals and given the clustering of c-CBL, EGFR and MET mutations, it is possible that the transforming effect of c-CBL mutations is most likely a combinatorial effect of the three. We also show that LOH for c-CBL was found in a significant number of samples that harbored MET or EGFR mutations. The fact that about 7% of lung tumor samples are likely to have c-CBL mutations and an additional 22% are likely to harbor c-CBL-related LOH, makes c-CBL a highly mutated molecule in lung cancer. Since LOH alone is not enough to cause a transforming event [35], [36], [37], associated mutation in the MET or EGFR locus or yet another RTK discussed above may play a role in carcinogenesis. We predict that this LOH in c-CBL results in haploinsufficiency that downplays RTK ubiquitination leading to hyperactivity of the RTKs. However, whether this is sufficient cause for tumorigenesis remains to be determined. Consistent with our hypothesis is the fact that c-cbl-/- mice have increased kinase activity in lymphocytes, but is not sufficient for tumor formation [35], [36], [37]. c-CBL LOH could also lead to increased expression of c-CBL from the other allele to compensate for a loss of an allele. Alternately, there could be some form of synergy working with reduced c-CBL levels and mutated receptors that exacerbate the phenotype of each alone.  Previous studies from our lab and others have shown that East Asians with lung cancers have relatively high frequencies of gain-of-function of mutations in RTKs such as EGFR and MET [31]. In a cohort of Japanese patients an activating MET mutation was identified in the splice region that deletes the juxtamembrane domain that is involved E3 activity of c-CBL [38]. This study also found that activation of MET is mutually exclusive of EGFR, KRAS and HER2 gene mutations [38]. We failed to detect such mutations in significant numbers in lung tumor samples obtained from African-Americans (n\\u200a=\\u200a29) and Caucasian (n\\u200a=\\u200a50) patients. One MET mutation was identified in each of the groups whereas 1 and 3 EGFR mutations were identified in the African-American and Caucasian cohorts respectively. EGFR mutations have earlier been identified as one of the key mutations affecting lung adenocarcinoma patients in a comprehensive study of 188 patients [39]. Our study encompasses different histologies of NSCLC. However, the published series did not find any mutations in c-CBL or MET unlike our study that encompassed different subtypes of NSCLC. It is important to note, we have recently shown that MET mutations in lung cancer are in majority germline [31]. We have reported earlier c-CBL mutations in a small cohort of Taiwanese lung cancer samples [40]. In our efforts to understand the ethnic differences in the lung oncogenome, we also looked at PAX transcription factors such as PAX5 and PAX8 that are highly expressed in lung cancers; however there was no preferential expression or mutations of the above genes in lung tumor samples of African-Americans. In this study, we show relatively high frequency of c-CBL mutations in lung cancers, especially in the large cell type among Caucasians and particularly among African-Americans. We therefore propose c-CBL as an efficacious target for lung cancers in African-Americans that needs to be further substantiated. This is all the more important because the prognosis for African-Americans with lung cancer, especially for men is much poorer compared to their Caucasian counterparts [41].  In conclusion, the results presented in this study demonstrate that c-CBL is frequently mutated or even lost in lung cancers. Our results support a role for c-CBL mutants that are independent of its ubiquitination activity. Given the relatively high mutation rates in c-CBL as well as RTKs such as MET and EGFR, it is likely that their combined effect could be synergistic in promoting tumorigenesis.\""
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "result = pd.merge(data, data_text, on = 'ID', how = 'left')"
      ],
      "metadata": {
        "id": "CBiwzBFIN7z-"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "result.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 322
        },
        "id": "0nV8-gRZOPAf",
        "outputId": "ef52a283-0031-4676-ec6c-97f3298e43b8"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   ID    Gene             Variation  Class  \\\n",
              "0   0  FAM58A  Truncating Mutations      1   \n",
              "1   1     CBL                 W802*      2   \n",
              "2   2     CBL                 Q249E      2   \n",
              "3   3     CBL                 N454D      3   \n",
              "4   4     CBL                 L399V      4   \n",
              "\n",
              "                                                Text    0    1    2    3    4  \\\n",
              "0  Cyclin-dependent kinases (CDKs) regulate a var...  NaN  NaN  NaN  NaN  NaN   \n",
              "1   Abstract Background  Non-small cell lung canc...  NaN  NaN  NaN  NaN  NaN   \n",
              "2   Abstract Background  Non-small cell lung canc...  NaN  NaN  NaN  NaN  NaN   \n",
              "3  Recent evidence has demonstrated that acquired...  NaN  NaN  NaN  NaN  NaN   \n",
              "4  Oncogenic mutations in the monomeric Casitas B...  NaN  NaN  NaN  NaN  NaN   \n",
              "\n",
              "   ... 3311 3312 3313 3314 3315 3316 3317 3318 3319 3320  \n",
              "0  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "1  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "2  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "3  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "4  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  \n",
              "\n",
              "[5 rows x 3321 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-08f1feff-0cfc-4f92-926b-61b7ae7e659c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "      <th>Text</th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "      <th>3</th>\n",
              "      <th>4</th>\n",
              "      <th>...</th>\n",
              "      <th>3311</th>\n",
              "      <th>3312</th>\n",
              "      <th>3313</th>\n",
              "      <th>3314</th>\n",
              "      <th>3315</th>\n",
              "      <th>3316</th>\n",
              "      <th>3317</th>\n",
              "      <th>3318</th>\n",
              "      <th>3319</th>\n",
              "      <th>3320</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>FAM58A</td>\n",
              "      <td>Truncating Mutations</td>\n",
              "      <td>1</td>\n",
              "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>CBL</td>\n",
              "      <td>W802*</td>\n",
              "      <td>2</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>CBL</td>\n",
              "      <td>Q249E</td>\n",
              "      <td>2</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>CBL</td>\n",
              "      <td>N454D</td>\n",
              "      <td>3</td>\n",
              "      <td>Recent evidence has demonstrated that acquired...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>CBL</td>\n",
              "      <td>L399V</td>\n",
              "      <td>4</td>\n",
              "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 3321 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-08f1feff-0cfc-4f92-926b-61b7ae7e659c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-08f1feff-0cfc-4f92-926b-61b7ae7e659c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-08f1feff-0cfc-4f92-926b-61b7ae7e659c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-bd6006f3-043e-4301-baf5-b678091c3346\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-bd6006f3-043e-4301-baf5-b678091c3346')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-bd6006f3-043e-4301-baf5-b678091c3346 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "result"
            }
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "result = result.drop(result.iloc[:, 5:], axis = 1)"
      ],
      "metadata": {
        "id": "iYBPsJnefxom"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "result[result.isnull().any(axis = 1)]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "MqCfiS_UOTA-",
        "outputId": "59bfcb1d-14a0-423f-aaa7-a8a6a8a09a91"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        ID    Gene             Variation  Class Text\n",
              "1109  1109   FANCA                S1088F      1  NaN\n",
              "1277  1277  ARID5B  Truncating Mutations      1  NaN\n",
              "1407  1407   FGFR3                 K508M      6  NaN\n",
              "1639  1639    FLT1         Amplification      6  NaN\n",
              "2755  2755    BRAF                 G596C      7  NaN"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-46ecd1c4-af33-4ae5-ac20-f80559dd3f58\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1109</th>\n",
              "      <td>1109</td>\n",
              "      <td>FANCA</td>\n",
              "      <td>S1088F</td>\n",
              "      <td>1</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1277</th>\n",
              "      <td>1277</td>\n",
              "      <td>ARID5B</td>\n",
              "      <td>Truncating Mutations</td>\n",
              "      <td>1</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1407</th>\n",
              "      <td>1407</td>\n",
              "      <td>FGFR3</td>\n",
              "      <td>K508M</td>\n",
              "      <td>6</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1639</th>\n",
              "      <td>1639</td>\n",
              "      <td>FLT1</td>\n",
              "      <td>Amplification</td>\n",
              "      <td>6</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2755</th>\n",
              "      <td>2755</td>\n",
              "      <td>BRAF</td>\n",
              "      <td>G596C</td>\n",
              "      <td>7</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-46ecd1c4-af33-4ae5-ac20-f80559dd3f58')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-46ecd1c4-af33-4ae5-ac20-f80559dd3f58 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-46ecd1c4-af33-4ae5-ac20-f80559dd3f58');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-f9d5588b-a16b-4b6b-b85b-f5c9f15ee71b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f9d5588b-a16b-4b6b-b85b-f5c9f15ee71b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-f9d5588b-a16b-4b6b-b85b-f5c9f15ee71b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "repr_error": "0"
            }
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "result.loc[result['Text'].isnull(), 'Text'] = result['Gene'] + ' ' + result['Variation']"
      ],
      "metadata": {
        "id": "OOr5GLn3Owe_"
      },
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "result[result['ID'] == 1109]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 81
        },
        "id": "AE_AHro0PKhf",
        "outputId": "448141b4-3ed4-417c-d223-f961e2e6a800"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        ID   Gene Variation  Class          Text\n",
              "1109  1109  FANCA    S1088F      1  FANCA S1088F"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5162b1a8-cb62-49f6-a811-18ee3d59c02a\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1109</th>\n",
              "      <td>1109</td>\n",
              "      <td>FANCA</td>\n",
              "      <td>S1088F</td>\n",
              "      <td>1</td>\n",
              "      <td>FANCA S1088F</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5162b1a8-cb62-49f6-a811-18ee3d59c02a')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5162b1a8-cb62-49f6-a811-18ee3d59c02a button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5162b1a8-cb62-49f6-a811-18ee3d59c02a');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"result[result['ID'] == 1109]\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1109,\n        \"max\": 1109,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1109\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"FANCA\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Variation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"S1088F\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Class\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"FANCA S1088F\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "result.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "hOoi7SsmioGh",
        "outputId": "e0373a10-69cd-4784-a3d5-02db17e610b8"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   ID    Gene             Variation  Class  \\\n",
              "0   0  FAM58A  Truncating Mutations      1   \n",
              "1   1     CBL                 W802*      2   \n",
              "2   2     CBL                 Q249E      2   \n",
              "3   3     CBL                 N454D      3   \n",
              "4   4     CBL                 L399V      4   \n",
              "\n",
              "                                                Text  \n",
              "0  Cyclin-dependent kinases (CDKs) regulate a var...  \n",
              "1   Abstract Background  Non-small cell lung canc...  \n",
              "2   Abstract Background  Non-small cell lung canc...  \n",
              "3  Recent evidence has demonstrated that acquired...  \n",
              "4  Oncogenic mutations in the monomeric Casitas B...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-a510b3ec-ac1f-4f5c-a8ad-861b3297d1ee\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>FAM58A</td>\n",
              "      <td>Truncating Mutations</td>\n",
              "      <td>1</td>\n",
              "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>CBL</td>\n",
              "      <td>W802*</td>\n",
              "      <td>2</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>CBL</td>\n",
              "      <td>Q249E</td>\n",
              "      <td>2</td>\n",
              "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>CBL</td>\n",
              "      <td>N454D</td>\n",
              "      <td>3</td>\n",
              "      <td>Recent evidence has demonstrated that acquired...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>CBL</td>\n",
              "      <td>L399V</td>\n",
              "      <td>4</td>\n",
              "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a510b3ec-ac1f-4f5c-a8ad-861b3297d1ee')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-a510b3ec-ac1f-4f5c-a8ad-861b3297d1ee button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-a510b3ec-ac1f-4f5c-a8ad-861b3297d1ee');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-dcd46f64-5868-4c7a-a2cc-cfdc52c865e8\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-dcd46f64-5868-4c7a-a2cc-cfdc52c865e8')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-dcd46f64-5868-4c7a-a2cc-cfdc52c865e8 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "result",
              "summary": "{\n  \"name\": \"result\",\n  \"rows\": 3321,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 958,\n        \"min\": 0,\n        \"max\": 3320,\n        \"num_unique_values\": 3321,\n        \"samples\": [\n          1057,\n          812,\n          2658\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 264,\n        \"samples\": [\n          \"PDGFRA\",\n          \"KEAP1\",\n          \"ARID1A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Variation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2996,\n        \"samples\": [\n          \"E1384K\",\n          \"H597Y\",\n          \"V851A\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Class\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2,\n        \"min\": 1,\n        \"max\": 9,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          8,\n          2,\n          6\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1925,\n        \"samples\": [\n          \"Quantitative multiplex PCR and genomic real-time PCR were used to complete an RB1 mutation analysis in 57 of 433 and 72 of 262 patients with hereditary and isolated unilateral retinoblastoma, respectively. These patients were selected because in previous analyses, which focused mainly on the identification of point mutations, no RB1 mutation was found. We identified gross deletions and insertions in peripheral blood DNA from 26 of 57 patients (46%) with hereditary retinoblastoma, and in six of 72 patients (8.3%) with isolated unilateral disease. In addition, we identified 32 somatic mutations in tumor DNA from 31 of 72 patients (43%) with isolated unilateral retinoblastoma. Together with our previous results, we found that gross RB1 alterations were present in the peripheral blood DNA from 65 of 433 (15%) and 17 of 262 (6.5%) patients with bilateral or familial and isolated unilateral retinoblastoma, respectively. Including reported gross deletions, an analysis of the frequency of breakpoints per intron length shows higher densities in introns 13, 16, 23, and 24. Genotype\\u2013phenotype analyses showed that on the whole, carriers of gross deletions develop fewer retinoblastomas compared to patients who are heterozygous for other types of RB1 null mutations. Specifically, carriers of cytogenetic and submicroscopic whole gene deletions often have unilateral tumors only. By contrast, almost all patients with gross deletions with one breakpoint in RB1 have bilateral retinoblastoma. Hum Mutat 26(5), 437\\u2013445, 2005. r 2005 Wiley-Liss, Inc. KEY WORDS: retinoblastoma; RB1; expression; genomic rearrangement; tumor INTRODUCTION Retinoblastoma (MIM] 180200), a malignant childhood tumor of the eye, is caused by a two-step inactivation [Cavenee et al., 1983; Knudson, 1971] of the RB1 gene (Entrez GeneID 5925, www.ncbi.nlm.nih.gov/entrez, GenBank L11910.1). Individuals who are heterozygous for an oncogenic germline mutation in RB1 usually develop multiple tumors in both eyes (bilateral retinoblastoma). In most families that segregate an oncogenic RB1 gene mutation, retinoblastoma is transmitted as an autosomaldominant trait with almost complete penetrance (hereditary retinoblastoma). About 50% of children with retinoblastoma have tumors in one eye only and no family history of retinoblastoma (sporadic unilateral retinoblastoma). In most of these patients the two RB1 gene mutations that initiated tumor development occurred in the somatic cells of the patient and are not detected in DNA from peripheral blood. However, about 12% of children with sporadic unilateral retinoblastoma have the hereditary form of retinoblastoma [Klutz et al., 1999; Richter et al., 2003; Schu\\u00a8ler et al., 2005]. Optimal genetic counseling and planning of ophthalmologic surveillance in relatives depends on the identification of the oncogenic RB1 gene mutation in the index patient. To meet this diagnostic requirement, mutation analysis has become available as a routine clinical service. It is important for the success of such a service that no mutation be missed [Noorani et al., 1996; Richter et al., 2003]. The vast majority of RB1 mutations in patients with retinoblastoma reported so far are point mutations (single-base substitutions, short-length alterations involving less than 20 bp, and complex mutations). The spectrum of this kind of oncogenic RB1 alterations has been delineated (LSDB of the RB1 gene, http://rb1-lsdb.d-lohmann.de) [Lohmann, 1999] and consistent associations between the nature of the oncogenic point mutation and the phenotypic expression have emerged [Lohmann and Gallie, 2004]. Specifically, carriers of RB1 alleles with nonsense or frameshift mutations usually develop bilateral retinoblastoma, whereas many missense and in-frame mutations are associated with unilateral retinoblastoma or incomplete penetrance. Until recently there have been only a relatively few reports on gross Published online 26 August 2005 in Wiley InterScience (www. interscience.wiley.com). DOI 10.1002/humu.20234 Received 5 April 2005; accepted revised manuscript 13 June 2005. Grant sponsor: Deutsche Forschungsgemeinschaft (Klinische Forschergruppe Ophthalmologische Onkologie und Genetik; Grant numbers: DFG KFO 109/1-1; TP II-1; Grant sponsor: IFORES Doktorandenstipendium (Medizinische Fakulta\\u00bat der Universita\\u00bat Duisburg Essen).  Correspondence to: Dietmar Lohmann, Institut fu\\u00bar Humangenetik, Universita\\u00batsklinikum Essen, Hufelandstrasse 55, IG1, 45122 Essen, Germany. E-mail: retinoblastoma@mac.com r 2005 WILEY-LISS, INC. deletions (i.e., deletions involving Z20 bp) in patients with retinoblastoma. In early studies that used Southern blot analysis for mutation detection, gross rearrangements were identified in peripheral blood DNA of about 12.5\\u201316% of patients with bilateral or familial retinoblastoma [Blanquet et al., 1991; Kloss et al., 1991]. Recently, PCR-based methods have facilitated the identification of gross RB1 alterations, and this has increased the number of reported mutations considerably [Richter et al., 2003; Houdayer et al., 2004]. Using quantitative multiplex PCR [Richter et al., 2003] and genomic real-time PCR, we were able to complete RB1 mutation detection in samples from 61 of 129 unrelated patients with retinoblastoma. We had previously analyzed these patients using a combination of different mutation detection techniques, but with negative results. We also searched the relevant literature for mutation reports. We used this data set to analyze the spectrum of gross alterations and to determine associations with phenotypic expression. MATERIALS AND METHODS Patient Samples DNA from peripheral blood of 433 index patients with hereditary retinoblastoma (bilateral or familial retinoblastoma) was available for genetic testing. In addition, DNA from both peripheral blood and fresh frozen tumor was available from 262 patients with isolated unilateral retinoblastoma. Informed consent was obtained for each case. Genetic counseling was provided. All DNA samples had previously been screened for the presence of point mutations in RB1 (GenBank reference sequence L11910.1) using SSCP and HDA [Lohmann et al., 1996, 1997; Klutz et al., 1999; Yilmaz et al., 1998] or DHPLC (D.R. Lohmann, unpublished results). All small mutations were confirmed by direct sequencing. DNA from tumor samples was also tested for the presence of hypermethylation in the promoter region of RB1 using published protocols [Zeschnigk et al., 1999, 2004]. Samples that were sent in before 1993 were also analyzed by Southern blot hybridization as described previously [Greger et al., 1989; Kloss et al., 1991]. Moreover, a cytogenetic analysis of peripheral blood lymphocytes was performed in patients who showed developmental problems indicative of a 13q- deletion syndrome. Samples from 129 patients without known mutations were used in the present study. Genotyping of Polymorphic STR Loci in the RB1 Gene A microsatellite analysis of two polymorphic short tandem repeat (STR) loci of the RB1 gene (RB1.20 [Yandell and Dryja, 1989] and RBi2 [Toguchida et al., 1993]) was performed on all samples using primers and protocols as described previously [Lohmann et al., 1995]. Quantitative Multiplex PCR The multiplex PCR protocol described by Richter et al. [2003] was used for quantitative multiplex PCR of exons 1\\u201327 and the 30 end of the promoter of the RB1. We modified the protocol by using 1) an additional internal comparative control target (SHGC-78184, chromosome 10, www.ncbi.nlm.nih.gov/genome/sts/sts.cgi?uid 5 96074), and 2) other primers for PCR amplification of exon 15/16 (forward Rbg76819se: ATT CAA TGC TGA CAC AAA TAA GGT TTC, product length 325 bp) and exon 4 (reverse RBg42075as: AGT GTA ACC CTA ATA AAA TG, product length 267 bp). Custom-synthesized primers labeled with 6-FAM fluorescent dye were obtained from Applera (Weiterstadt, Germany). For amplification reactions we used the Qiagen Multiplex PCR Kit (Qiagen, Hilden, Germany). Since we used a different PCR instrumentation and reaction chemistry, we had to make minor changes to the original PCR protocol [Richter et al., 2003] to obtain comparable amounts of product for each exon when we analyzed the normal DNA. Electrophoresis and detection of fluorescence-labeled products were performed using an ABI 3100 Genetic Analyzer (Abi, Weiterstadt, Germany). Peak area integrals were determined using GeneScan software (Abi, Weiterstadt, Germany). Each signal peak integral specific for a given exon was normalized against the signals of two internal controls. As external controls we used DNA samples without copy number changes, and samples with known alterations. All calculations were carried out using Excel spreadsheet software (Microsoft, Redmond, WA). Genomic Real-Time PCR Real-time PCR was performed on genomic DNA using Taqman technology with minor groove binder probes on an ABI Prism 7000 Sequence Detection System (Abi, Weiterstadt, Germany). We used a protocol similar to the one described by Hu et al. [2004]. Real-time PCR reactions were carried out at a final concentration of 300 nM for forward and reverse primers, and 250 nM for the Taqman probes (the primers and probes are listed in Table 1). One rationale for the choice of regions for real-time PCR was to identify deletions extending over the border of large introns (introns 2 and 17). Each reaction contained 60 ng of genomic DNA and 12.5 ml of the TaqMan Universal PCR Master Mix (Abi, Weiterstadt, Germany) in an individual reaction volume of 25 ml. The thermocycling conditions were 501C for 2 min and 951C for 10 min, followed by 40 cycles of 951C for 15 sec and 601C for 1 min. Each of the 96-well, real-time PCR optical plates included reactions with serial dilutions of genomic DNA (200, 100, 50, 25, and 12.5 ng of DNA in 25 ml reaction volume), which was used to generate standard curves for each Taqman assay. With these standard curves the real-time PCR cycle numbers (ct) could be used to determine the amount of PCR product. We compared the amplification kinetics of the PCR targets located in specific exons of the RB1 gene to the amplification of a control target located in the imprinting center of chromosome 10 (Karin Buiting, unpublished data) (Table 2). We developed Taqman assays for exons 2, 4, 7, 10, 17, 18, and 24. The high GC content of the region of exon 1 was extremely hard to amplify in a constant manner, so we used a Taqman assay spanning a region between 534 and 496 bp upstream of the RB1 (GenBank accession L11910.1) to detect deletions of the 50 end of the gene. Statistical Analysis All analyses were performed with JMP software (www.jmpin. com). A contingency analysis of association between genotypes and phenotypic expression (bilateral retinoblastoma, unilateral retinoblastoma, or unaffected mutation carrier) was performed using likelihood ratio Chi-square statistics. A bivariate linear fit model was used to analyze the breakpoint frequency per intron length. RESULTS QMPX-PCR identified RB1 mutations in peripheral blood DNA from 26 of 57 patients (46%) with hereditary retinoblastoma, and in six of 72 patients (8.3%) with isolated unilateral disease. In addition, we identified 32 somatic mutations in tumor DNA from 31 of 72 patients (43%) with isolated unilateral retinoblastoma. 438 HUMAN MUTATION 26(5), 437^445, 2005 TABLE 1. Sequences of Primers and Hybridization Oligos for Realtime PCR Target region Oligo name Position in L11910 Sequence (50 to 30 ) Promoter Rb1up-RT-MGB ^534 CCTGCGTTAGTTTGCT RB1up-RT-se ^563 TGAGAACATGCAGTGTTTGGTTT RB1up-RT-as ^496 TGGAGCCAGAAGCCTTTATCC Exon 2 RB2-RT-MGB 5458 ACTGCATTATGTCAGAAATTAAAGATACCAGATCATGTCA RB2-RT-se 5412 TTCATTTGGTAGGCTTGAGTTTGA RB2-RT-as 5522 TCCCAAGTTAACCAAGCTCTCTCT Exon 4 RB4-RT-MGB 41981 AATGCTATGTCAAGACTG RB4-RT-se 41951 AAAGAAATTGATACCAGTACCAAAGTTG RB4-RT-as 42027 AGTGCAAACAATACATCATACTTCTTCA Exon 7 RB7-RT-MGB 56971 TTTAATTTATGCCCCTTTTACTTTCTCATTCAGCAGT RB7-RT-se 56941 CATGTTGCTCAAAGAACCATATAGGAAGT RB7-RT-as 57034 TGGTACCCACTAGACATTCAATAAGC Exon 10 RB10-RT-MGB 64283 TGTAATGACATGTAAAGGAT RB10-RT-se 64260 CTGTGCCTCTGTGTGCTGAGA RB10-RT-as 64332 AACCTTGAAAGAAAAAAGTCACTGACA Exon 17 RB17-RT-MGB 78156 AGCCTTTGATTTTTACAAAGTGATCG RB17-RT-se 77951 CCATGATTCTGAATGTGCTTAAT RB17-RT-as 78207 GTTGCCTTCTGCTTTGATAAAACTT Exon 18 RB18-RT-MGB 150047 CAACTGATCACCTTGAATC RB18-RT-se 150024 ACAATCAAAGGACCGAGAAGGA RB18-RT-as 150092 TGGAGAGGAAGATTAAAGAGGACAAG Exon 24 RB24-RT-MGB 170273 TGGTATTCCTAATAGTTCAGAATG RB24-RT-se 170244 TGGTTCTAGGGTAGAGGTAACCTTTAA RB24-RT-as 170325 CAATTTTGCAGAGATGAGCATAAATAC TABLE 2. Summary of Mutations  50 Breakpoint 30 Breakpoint Change Systematic description Sample numbers 50 of Promoter 30 of stop codon Deletion c.-138^27841?del M5926}, M6958, M6989}, M7395}, M8001, M10134}, M10291}, M11144}, M11667, M13912, M15801, M18067}, M18968}, M20054, M20472, M20537, M20744}, M21644, M21801, M22138}, M23239, M23403, M25277, M26029}, M27372 50 of Promoter 30 of stop codon Insertion c.-138^27841?ins M14070}, M24003} 50 of Promoter Intron 17 Deletion c.-138^16951?del M24862}, M24796 50 of Promoter Intron 13 Deletion c.-138^13321?del M23292} 50 of Promoter Intron 2 Deletion c.-138^2641?del M12769} 50 of Promoter Intron 2 Insertion c.-138^2641?ins M21850 Intron 1 Intron 6 Deletion c.138-?^6071?del M23720 Intron 1 Intron 3 Deletion c.138-?^3801?del M23632 Intron 2 30 of stop codon Deletion c.265-?^27841?del M14381} Intron 2 Intron 7 Deletion c.265-?^27841?del M21203} Intron 4 30 of stop codon Deletion c.501-?^27841?del M7998} Intron 4 Intron 13 Deletion c.501-?^13321?del M7608} Intron 4 Intron 9 Deletion c.501-?^9391?del M13913} Intron 6 Intron 23 Deletion c.608-?^24891?del M11037, M23477 Intron 7 30 of stop codon Deletion c.719-?^27841?del M7858} Intron 8 30 of stop codon Deletion c.862-?^27841?del M24580 Intron 9 30 of stop codon Deletion c.940-?^27841?del M13539} Intron 12 30 of stop codon Deletion c.1216-?^27841?del M14710} Intron 12 Intron 17 Deletion c.1216-?^16951?del G65 Intron 13 Intron 17 Deletion c.1333-?^16951?del M11802}, M12835 Intron 16 30 of stop codon Deletion c.1499-?^27841?del M8265} Intron 17 30 of stop codon Deletion c.1696-?^27841?del M2004}, M10080}, M14765}, M23292} Intron 17 Intron 26 Insertion c.1696-?^27131?ins M25404} Intron 17 Intron 23 Deletion c.1696-?^24891?del M8504, M23066 Intron 17 Intron 18 Deletion c.1696-?^18141?del M18916} Intron 20 Intron 24 Deletion c.2107-?^25201?del M26422 Intron 21 Intron 23 Deletion c.2212-?^24891?del M20451 Intron 23 30 of stop codon Deletion c.2490-?^27841?del M5392, M12814 Intron 23 Intron 24 Deletion c.2490-?^25201?del M3623, M4716, M14372  Systematic description according to HGVS recommendations (www.hgvs.org). Mutation numbering is based on cDNA sequence (GenBank reference sequence NM^000321.1). }, somatic mutations only identi\\u00a2ed in DNA from tumor. HUMAN MUTATION 26(5), 437^445, 2005 439 Whole Gene Deletions The results of the QMPX-PCR of peripheral blood DNA from 14 patients showed half of the normal copy numbers for all products specific for regions within the RB1 gene, thus indicating hemizygous loss of one RB1 allele. In all patients deletions were confirmed by real-time PCR of targets within RB1 and, in informative families, by genotyping of polymorphic STR loci within this gene. In one patient, it was possible to detect the deletion by conventional cytogenetic analysis of peripheral blood lymphocytes. QMPX-PCR analysis of tumor DNA from patients with isolated unilateral retinoblastoma showed that 13 of 43 tumors with LOH at intragenic polymorphic loci had a deletion of one RB1 allele (i.e., a hemizygous RB1 deletion). It must be noted that this figure (13/43, 30%) cannot be used to estimate somatic deletions that cause hemizygosity because these 43 tumors are only a subset of our whole sample set, which comprises 262 tumors (an estimate of the proportion of hemizygous RB1 deletions was provided by Hagstrom and Dryja [1999]). Because of this bias, hemizygous deletions of the whole RB1 gene in tumors with LOH were not included in our mutation count. In 11 of 22 tumors that had not shown LOH in PCR-based genotyping of intragenic STR loci, we identified whole gene deletions of both RB1 alleles (i.e., a homozygous RB1 deletion). Real-time PCR on tumor DNA with probes within the RB1 gene showed complete or almost complete loss of the target sequences, thus confirming the presence of homozygous deletions. It is reasonable to assume that the finding of no allele loss by PCR-based genotyping of STR loci within the RB1 gene in tumors with homozygous RB1 deletions is due to products amplified from residual constitutional DNA that was present in the tumor samples. In all 11 patients with isolated unilateral retinoblastoma and homozygous RB1 deletions in the tumor, no deletion was detected in peripheral blood DNA. DeletionsWith One BreakpointWithin the RB1 Gene In peripheral blood DNA from five patients (M5392, M12814, M24580, and M24796) we identified gross deletions with one breakpoint localized to an intron of the RB1 gene (Fig. 1). Three of these deletions involved the 30 -end of RB1 (M5392, M12814, and M24580), including one mutation that involved parts of the regions coding for the pocket domains (M24580). Analysis of tumor DNA from patients with isolated unilateral retinoblastoma showed deletions involving the 50 and 30 ends of RB1 in three (M24682, M12769, and M23292) and 10 (M13539, M14710, M8265, M2004, M10080, M14765, M23292, M14381, M7998, and M7858) cases, respectively. These deletions were in agreement with the pattern of allele loss that was obtained by STR analysis, which showed LOH for the STR loci outside the deleted regions and no allele loss at marker loci involved in homozygous deletions. DeletionsWith Both BreakpointsWithin the RB1 Gene We detected deletions with both breakpoints within the RB1 gene in peripheral blood DNA from 13 patients (G65, M12835, M8504, M23066, M23632, M23720, M26422, M20451, M3625, M4716, M14400, M11037, and M23477) (see Fig. 1). This number includes three patients (M3625, M4716, and M14400) with a recurrent deletion of exon 24 who we reported in a previous paper [Albrecht et al., 2004]. The results of real-time-PCR and STR analysis indicated that two patients (M11037 and M23477) are mutational mosaics rather than heterozygous mutation carriers. We did not develop a real-time-PCR assay to confirm the deletion of exons 22 and 23 identified in patient M20451. However, this patient had familial retinoblastoma, and identical results were obtained by QMPX-PCR of blood DNA from two affected relatives. QMPX-PCR of tumor DNA from five patients (M11802, M18916, M21203, M7608, and M13913) showed intragenic deletions that were not present in the peripheral blood DNA of these patients. Duplications of All or Parts of the RB1 Gene In two patients with isolated unilateral retinoblastoma (M14070 and M24003), all probes specific for the RB1 gene showed triplication in tumor DNA but had normal copy numbers in constitutional DNA. It is possible that these tumors have three copies of the long arm of chromosome 13, which is a genetic alteration that has occasionally been observed in retinoblastoma [Herzog et al., 2001] but is unlikely to result in functional inactivation of the RB1 gene. Tumor DNA from two other patients with isolated unilateral retinoblastoma (M21850 and M24504) showed patterns of dosage changes compatible with duplications of contiguous parts of the RB1 gene. The mutation in tumor M21850 was also detected in the peripheral blood DNA of the patient (M21850). The pattern of copy number changes in tumor DNA from another patient (M21950) with isolated unilateral retinoblastoma does not allow for a simple interpretation because discontinuous regions along the gene showed gains and losses. The complex pattern of copy number changes was confirmed by realtime PCR, but this assay gave no further clue as to how to interpret the structure of the mutant allele. Therefore, this mutation was not included in the mutation count. Genotype^Phenotype Correlations Phenotypic expression in patients with cytogenetic and submicroscopic whole gene deletions. To calculate the frequency of gross alterations and to analyze associations between mutant genotypes and phenotypic expression, we included patients in whom we had previously identified a constitutional mutation by Southern blot hybridization or cytogenetic analysis [Horsthemke et al., 1987; Kloss et al., 1991; Lohmann et al., unpublished results]. In all, gross deletions were identified in the peripheral blood DNA of 15% (65/433) of index patients with hereditary unilateral retinoblastoma, and 6.5% (17/262) of patients with isolated unilateral retinoblastoma. Of these 82 mutations, 38 (46%) were whole gene deletions of which 27 (71%) were also detectable by conventional cytogenetic analysis. We cannot exclude that the presence of a 13qdeletion syndrome leads to preferential ascertainment of patients with cytogenetic deletions. If this is true, our figures are biased in favor of identification of cytogenetic deletions. Familial retinoblastoma was less frequent in patients with cytogenetic deletions compared to patients with submicroscopic whole gene deletions (2/27 vs. 4/11, P 5 0.0034 Fisher\\u2019s exact test, two-tailed; Fig. 2A). However, the distribution of phenotypic expression of retinoblastoma (bilateral retinoblastoma, unilateral retinoblastoma, or unaffected mutation carriers) was almost identical between carriers of cytogenetic and submicroscopic whole gene deletions (Fig. 2B). Phenotypic expression in patients with gross deletions with de\\u00a2ned breakpoints. To widen our analysis of the spectrum of phenotypic expression associated with gross RB1 deletions, we included patients reported in the literature [Bremner et al., 1997; Canning and Dryja, 1989; Dryja et al., 1993; Hashimoto et al., 1991; Houdayer et al., 2004; Kato et al., 1994; Kloss et al., 1991; Richter et al., 2003; Zajaczek et al., 1999]. In cases of familial retinoblastoma we also included nonpenetrant carriers. In all, we analyzed genotype\\u2013phenotype correlations in 440 HUMAN MUTATION 26(5), 437^445, 2005 186 patients and nonpenetrant heterozygous carriers of RB1 deletions with well defined breakpoints. As shown in Figure 3, almost all carriers of deletions with one breakpoint in the RB1 gene have bilateral retinoblastoma. Carriers of whole gene deletions with both breakpoints outside of RB1, and carriers of gross deletions with both breakpoints in RB1 develop fewer tumors (likelihood ratio test, Po0.006 and Po0.001, respectively). It has been noted that Q3 point mutations located in the regions that code for the A and B pocket domains of the pRb protein are associated with more severe phenotypic expression (for review see [Lohmann and Gallie, 2004]). Gross deletions with one or with both breakpoints in the RB1 gene may leave these regions intact. Individuals who carry mutant alleles with intact pocket coding regions have significantly fewer tumors than patients with mutations FIGURE 1. Graphical representation ofRB1and the extent of gross deletions and insertions identi\\u00a2ed in patients with retinoblastoma. Vertical bars show the relative location ofRB1exons, some of which are identi\\u00a2ed by numbers. Horizontal black bars below represent the extent of the deletion. Gray bars indicate possible locations of the deletion breakpoints. }, somatic mutations identi\\u00a2ed only in DNA from tumor. Positions marked by an asterisk ( ) were checked by genomic real-time PCR. [Color \\u00a2gure can be viewed in the online issue, which is available at www.interscience.wiley.com.] HUMAN MUTATION 26(5), 437^445, 2005 441 that deleted all or parts of these regions (likelihood ratio test, Po0.001; Fig. 4A). Also, individuals who carry alleles with gross deletions that do not alter the reading frame show a higher proportion of unaffected carriers and patients with unilateral retinoblastoma compared to patients with frameshift deletions (likelihood ratio test, Po0.001; Fig. 4B). When the comparison of phenotypic expression is restricted to individuals with mutations that leave the pocket domain regions intact, it appears that most carriers with in-frame mutations have unilateral retinoblastoma or remain unaffected, whereas most patients with frameshift changes have bilateral retinoblastoma (likelihood ratio test, Po0.001; Fig. 4C). Distribution of breakpoints of gross deletions within RB1. Of the 82 mutations identified in our patients, 44 had at least one deletion breakpoint located within RB1. Reviewing the reported mutations and entries in the human gene mutation database, we found a further 24 gross deletions with at least one breakpoint within the RB1 (www.hgmd.org) [Richter et al., 2003, Houdayer et al., 2004; Zajaczek et al., 1998, 1999; Ata-UrRasheed et al., 2002; Bremner et al., 1997; Dryja et al., 1989, 1993; Kato et al., 1994]. To find out if the frequency of deletion breakpoints per length varies between the 26 intronic regions of the RB1, we plotted the number of observed breakpoints for a given intron against its length (Fig. 5). The result of the linear fit analysis suggests that the number of observed breakpoints per intron length is about the same for most intron sequences of the RB1 (linear fit curve in Fig. 5 with 95% confidence limits) and is higher in introns 13, 16, 23, and 24, and, to a lesser extent, also elevated in introns 1, 2, 7, and 12. Since we have not determined the exact location of breakpoints within introns, we cannot exclude the presence of breakpoint cluster regions in some introns. No breakpoints of gross deletions were observed in introns 5, 10, 15, and 26 (intron 15, with a length of 79 bp, is out of scale). DISCUSSION Large deletions involving the RB1 gene were the first mutations known to cause hereditary predisposition to retinoblastoma [Francke and Kung, 1976; Friend et al., 1986]. However, until recently, few gross deletions have been reported in patients with retinoblastoma. This changed only after quantitative multiplex PCR facilitated the identification of gross alterations [Richter et al., 2003]. We used this method to complete mutation identification in clinical samples from patients with retinoblastoma. Including the results obtained with previous methods [Horsthemke et al., 1987; Kloss et al., 1991], we identified gross alterations in peripheral blood cells of 65 of 433 unrelated patients with bilateral or familial retinoblastoma (15%), and in 17 of 262 patients with isolated unilateral retinoblastoma (6.1%). If we correct for multiple testing, the relative proportions of the gross alterations compare well with recent reports by Richter et al. [2003] (29/224, 13%, and 4/134, 3%, respectively) and Houdayer et al. [2004] (17/102, 17%, and 1/90, 1.1%, respectively). Of the 82 gross alterations that we detected in constitutional cells of unrelated patients, 27 (33%) were also detectable by conventional cytogenetic analysis. FIGURE 2. Cell plots. A: Proportions of patients with isolated and familial retinoblastoma among patients with cytogenetic and submicroscopic whole RB1 gene deletions. B: Proportions of una\\u00a1ected mutation carriers and patients with unilateral and bilateral retinoblastoma among patients with cytogenetic and submicroscopic whole RB1 gene deletions. [Color \\u00a2gure can be viewed in the online issue, which is available at www.interscience.wiley.com.] FIGURE 3. Cell plot with proportions of una\\u00a1ected mutation carriers and patients with unilateral and bilateral retinoblastoma among patients with whole RB1 gene deletions (both breakpoints outside of RB1), RB1deletions with one breakpoint inside of RB1, and with both breakpoints inside of RB1. [Color \\u00a2gure can be viewed in the online issue, which is available at www. interscience.wiley.com.] 442 HUMAN MUTATION 26(5), 437^445, 2005 Cytogenetic deletions involving RB1 in 13q14 are known to be related to a contiguous gene syndrome that is characterized by developmental abnormalities (13q- deletion syndrome [Allderdice et al., 1969; Brown et al., 1995]). To identify such patients, a cytogenetic analysis should be performed in carriers of whole RB1 gene deletions. The reduced reproductive fitness of patients with a 13q- deletion syndrome is likely to account for the low proportion of familial disease in this group of patients (2/27, 7.4%). Matsunaga [1980] was the first to recognize that the proportion of unilateral disease is higher among carriers of a 13q- deletion compared to other patients with hereditary retinoblastoma. To explain this genotype\\u2013phenotype association, Bunin et al. [1989] suggested that the lower number of tumors in carriers of a 13qdeletion may be due to a reduced spectrum of effective second mutations. In fact, in more than 60% of retinoblastomas chromosomal mechanisms such as mitotic recombination or nondisjunction cause loss of the second RB1 allele [Cavenee et al., 1983; Hagstrom and Dryja, 1999; Zhu et al., 1992]. In a carrier of a 13q- deletion these alterations would result in homozygous loss of all genes contained in the deleted region. This could make the cells nonviable if the genes that are essential for basic cellular functions are lost. Consequently, only a local inactivation of the second RB1 allele will result in tumor formation. Our data confirm that patients with cytogenetic deletions involving the RB1 locus develop fewer retinoblastoma foci compared to patients who carry other types of RB1 null mutations [Lohmann et al., 1996]. Moreover, we found that unusual mild retinoblastoma predisposition is not restricted to patients with cytogenetic deletions but is also observed in carriers of submicroscopic whole RB1 gene deletions. This suggests that the putative genes whose homozygous loss is incompatible with cell viability are located in the neighborhood of the RB1. It is FIGURE 4. Cell plots with proportions of una\\u00a1ected mutation carriers and patients with unilateral and bilateral retinoblastoma among patients with grossRB1gene deletions. A:Comparison of the number of individuals with mutations that leave intact or include all or parts of the regions coding for the pocket domains. B: Comparison of the number of individuals with mutations that leave the reading frame intact or result in a frameshift. C: Comparison of the number of carriers with frameshift and in-frame mutations in individuals with mutations that leave intact the regions coding for the pocket domains. [Color \\u00a2gure can be viewed in the online issue, which is available at www.interscience.wiley.com.] FIGURE 5. Graph showing the number of observed RB1gross deletions with breakpoints located to introns of the RB1plotted against the length of the intron (logarithmic scale). Curves represent the results of the linear \\u00a2t analysis and 95% con\\u00a2dence limits. [Color \\u00a2gure can be viewed in the online issue, which is available at www.interscience.wiley.com.] HUMAN MUTATION 26(5), 437^445, 2005 443 notable in this respect that almost all carriers of gross deletions that have one breakpoint within the RB1 develop bilateral retinoblastoma (25 of 26 mutation carriers, including reported cases). In other words, mutations that delete either parts of the 50 boundary or parts of the 30 boundary of the RB1 gene are not associated with an unusually mild predisposition. If interpreted within the framework of the hypothesis by Bunin et al. [1989], this suggests that some combination of essential genes with members on both sides of the RB1 are required for cell viability. As noted above, carriers of gross deletions with one breakpoint within the RB1 show a distribution of phenotypic expression similar to that observed in patients with frameshift or nonsense changes [Lohmann et al., 1996], which are the predominant types of mutations in patients with retinoblastoma [Lohmann, 1999; Richter et al., 2003]. By comparison, patients with gross deletions with both breakpoints within the RB1 develop significantly fewer retinoblastoma foci (Fig. 3). However, a more detailed analysis reveals that lower tumor numbers and incomplete penetrance are associated with gross deletions that leave the regions that code for the pocket domains intact (Fig. 4A) or do not result in a premature termination codon (Fig. 4B). These effects seem to be additive, as most patients with in-frame gross deletions that do not affect the pocket domains have only unilateral tumors (17/35, 49%) and 13 of 35 carriers (37%) do not develop retinoblastoma (Fig. 4C). Possibly, mutant proteins that might be expressed from alleles with in-frame deletions and intact pocket domains have sufficient residual tumor-suppressive function. The local distribution of gross RB1 gene deletions does not show repeated occurrence of specific deletions with intragenic breakpoints. There may be no homologous low copy repeats within the sequence of this gene that can act as targets of non-allelic homologous recombination, which is the predominant mechanism for recurrent gross deletions [Shaw and Lupski, 2004]. A more detailed analysis of the distribution of breakpoints of gross deletions within the RB1 gene shows an excess of rate of breaks per length of sequence in introns 23, 16, 13, and, to a lesser extent, 24. Recently we reported that the vicinity of intron 23 shows a strong scaffold/ matrix attached region (S/MAR), and that this structure is possibly linked to occurrence of genomic breakage in this region [Albrecht et al., 2004]. Following this line of evidence, in the current study we used the software MARwiz (www.futuresoft.org) to identify potential S/MAR structures in other introns of the RB1 gene. This tool showed that regions with strong S/MAR potential are located adjacent to introns 16 and 13. This finding suggests that the presence of S/MAR structures is the reason for the higher number of breakpoints in these introns. However, to confirm the presence of S/MARs, further tests (e.g., stress-induced duplex destabilization (SIDD) [Benham et al., 1997]) must be performed. Genetic testing has become a desirable component of routine care for patients with retinoblastoma and their families, and is also beneficial from an economic perspective [Noorani et al., 1996]. However, to realize these advantages the causative mutations must be identified. On the whole, we detected oncogenic RB1 mutations in 402 of 433 patients (93%) with hereditary retinoblastoma, and 227 of 262 patients (87%) with isolated unilateral retinoblastoma. It will be interesting to determine the genetic cause of retinoblastoma in the remaining 7% and 13% of patients, respectively  Biochemical analysis of the retinoblastoma protein's function has received considerable attention since it was cloned just over 20 years ago. During this time pRB has emerged as a key regulator of the cell division cycle and its ability to block proliferation is disrupted in the vast majority of human cancers. Much has been learned about the regulation of E2F transcription factors by pRB in the cell cycle. However, many questions remain unresolved and researchers continue to explore this multifunctional protein. In particular, understanding how its biochemical functions contribute to its role as a tumor suppressor remains to be determined. Since pRB has been shown to function as an adaptor molecule that links different proteins together, or to particular promoters, analyzing pRB by disrupting individual protein interactions holds tremendous promise in unraveling the intricacies of its function. Recently, crystal structures have reported how pRB interacts with some of its molecular partners. This information has created the possibility of rationally separating pRB functions by studying mutants that disrupt individual binding sites. This review will focus on literature that investigates pRB by isolating functions based on binding sites within the pocket domain. This article will also discuss the prospects for using this approach to further explore the unknown functions of pRB. Background The retinoblastoma susceptibility gene (RB-1) was the first tumor suppressor gene to be cloned [1]. Since that time its encoded protein (pRB) has emerged as a key regulator of cell cycle entry and appears to be one of the most frequent targets for inactivation in human cancer [2-5]. The retinoblastoma protein is most frequently inactivated in cancer by the negative regulatory activity of cyclin dependent kinases [6]. Only in small cell lung carcinoma [7] and retinoblastoma [8] is the RB-1 gene a frequent target for direct mutation. Loss of heterozygosity at the RB-1 locus has been reported in many different sporadic cancers, suggesting that it is directly mutated outside of the lung and retina, but on a less frequent basis [9]. Based on pRB's prominent and ubiquitous role in cancer many investigators have focused their efforts on trying to determine its biochemical function. A fundamental component to this type of investigation is the evaluation of mutant alleles to determine which aspects of cell physiology require pRB. Ideally the analysis of very specific mutants will determine which protein interactions account for how pRB functions. This review will focus on the efforts that have been made to rationally separate different aspects of pRB's function in proliferative control and cancer. Given that the focus of this review is on the dissection of pRB function, some reports will inevitably be omitted because they lack a structural component. By no means is Published: 13 September 2007 Cell Division 2007, 2:26 doi:10.1186/1747-1028-2-26 Received: 30 August 2007 Accepted: 13 September 2007 This article is available from: http://www.celldiv.com/content/2/1/26 \\u00a9 2007 Dick; licensee BioMed Central Ltd. This is an Open Access article distributed under the terms of the Creative Commons Attribution License (http://creativecommons.org/licenses/by/2.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original work is properly cited. Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 2 of 15 (page number not for citation purposes) this is meant to diminish their validity, it is hoped that all aspects of pRB function will eventually fit into a framework of defined protein interactions, unfortunately not all are at this stage. In general terms, pRB has an established role in mediating a G1 arrest in development and in response to many growth regulatory signals [5]. Some examples are DNA damage [10], or growth inhibiting cytokines such as TGF-\\u03b2 [11]. In addition, pRB plays a key role in the permanent cell cycle exit of differentiating cells and this has been demonstrated both in cell culture and in vivo using gene targeted mice [12,13]. The Retinoblastoma protein plays an essential function in permanent cell cycle arrest and differentiation of adipocytes [14], myotubes [15], osteoblasts [16], and neurons [17]. In mechanistic terms, much of pRB's ability to control the cell cycle has been linked to its ability to regulate transcription [4,18]. The RB protein binds to E2F transcription factors and blocks their ability to induce transcription of genes that are needed to advance the cell cycle. In turn, once pRB is brought to a promoter by an E2F it can recruit chromatin remodeling proteins such as histone methyltransferases [19], DNA methyltransferases [20], histone deacetylases [21,22], and helicases [23] among others to further repress transcription [24]. Stimulation of cell cycle entry results in phosphorylation of pRB by cyclin/Cdk complexes and causes the release of E2Fs and chromatin regulators at the G1 to S-phase transition (Figure 1). Beyond the simple model elaborated in Figure 1 where pRB represses transcription in G1 to control S-phase entry, it should be noted that recent work on pRB to validate this mechanism is not without its controversies. However, one common theme to pRB function is that it works as an adaptor protein that can nucleate complexes containing multiple interacting partners. For this reason, understanding the consequence of the loss of one type of pRB interaction in isolation holds enormous promise to characterizing how pRB works in controlling proliferation, or other functions, that make it a tumor suppressor. Nature's bounty; using tumor-derived mutations in RB-1 to guide functional analysis Before describing what has been learned from studying pRB through a structure-function approach it is worth understanding what mutant resources are available. For example, oncogenic mutations in Ras are clustered at codons 12 and 61 and these mutants have revealed the importance of the GTP bound state [25]. This insight was made possible by the fact that oncogenic mutations discretely cripple intrinsic GTP hydrolysis while leaving interactions with downstream effectors intact. Generation of mutations such as the S17N substitution has created a dominant negative form of Ras whose over expression blocks endogenous Ras signaling [26]. At this point there are no known cancer causing mutations in RB-1 that offer discrete separation of activities. Much of the effort to dissect pRB function using cancer causing alleles has come from mutations found in retinoblastoma families with partially-penetrant inheritance [27-30], particularly the R661W substitution (Figure 2). However, these alleles encode proteins that contain multiple interaction defects. Furthermore, there has been relatively limited use of gainof-function and dominant negative alleles of RB-1 to help clarify its function despite efforts to develop these reagents [31-35]. Instead, a number of tumor derived RB-1 mutants have actually emerged as excellent negative controls! In addition to Ras, another relevant comparison of mutations is with p53. The TP53 gene is the most frequently mutated gene in cancer [36]. Cataloguing hundreds of tumor derived missense changes has been invaluable, as it has revealed the importance of the DNA binding domain in p53 tumor suppressor function. Some of these alleles have also turned out to have dominant negative activity. At present the vast majority of mutations found in retinoblastoma patients are deletions or nonsense changes, suggesting that these are null alleles [8,37]. There are 35 known cancer causing missense mutations in RB-1 from retinoblastoma patients and most of these alleles have only been reported once [37]. Taken together, this suggests that the paucity of partially defective RB-1 mutants has prevented the type of analysis of pRB that has played a crucial role in understanding other oncogenes and tumor suppressors. In short, nature has not been very generous in providing mutations to aid in analyzing the mechanism of pRB function. Transcriptional control of Figure 1 the cell cycle by pRB Transcriptional control of the cell cycle by pRB. In G1 pRB is bound to E2Fs and masks their ability to activate transcription. In turn, pRB can recruit a number of chromatin remodeling factors that can further inhibit the initiation of transcription. Mitogen signaling then activates cyclin/Cdk complexes that phosphorylate pRB in late G1 and early Sphase leading to the disassembly of this complex and the activation of transcription of genes needed for progression through S-phase. E2F DP pRB Chromatin remodeling factor G1 E2F DP pRB S P P P P P Cdk2 E Cdk4 D + Chromatin remodeling factor Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 3 of 15 (page number not for citation purposes) Cut and paste, elucidating the domain structure of pRB The retinoblastoma protein does not possess any commonly recognized DNA binding or protein interaction domains, although crystallographic data demonstrates that pRB contains two cyclin-folds and it has similarity to other proteins with this structure [38,39]. Our present knowledge of the domain structure of pRB has been determined by deletion mapping. These efforts have defined the large and small 'pocket' domains (Figure 2). The 'small pocket' was initially identified as the minimal fragment of pRB that can interact with viral proteins like SV40 TAg, HPV E7, and adenovirus E1A [40-42]. Crystal structure data indicates that the A and B halves of the small pocket each represent a single cyclin fold and they interact to form a dumbbell shaped globular domain [38]. Experiments designed to define the minimal growth-suppressing domain identified a slightly bigger piece of pRB that is called the 'large pocket' [43]. The large pocket is also capable of complementing pRB's tumor suppressor activity in vivo when expressed in place of the full-length protein [44]. In addition, the cancer-causing missense mutations in RB-1 that have been described are found mostly within the large pocket and this has focused efforts most intensively on this domain. Puzzlingly, there are two other proteins in the RB-family called p107 and p130 that have extensive similarities to pRB, particularly within the small pocket [45,46]. Despite this similarity, there is little experimental evidence to suggest these proteins have tumor suppressive functions on their own as p107-/-;p130+/- and p107+/-;p130-/- mice are not tumor prone [45]. This has further focused the question on what does the pRB pocket domain do that p107 and p130 can't do? Dissection of interactions and separating functions within the large pocket domain has been one of the rate limiting steps to advancing our understanding of pRB by a structure-function approach. Initially, efforts to mutagenize pRB were guided by sequence conservation and these have lead to a number of insightful reports that separate different aspects of pRB function [47-49]. However, crystal structures of the pRB small pocket have revealed that many of the most conserved amino acids are buried and this suggests why some of these mutants have partial loss of function for multiple interaction types [38]. Generation of synthetic mutant alleles of RB-1 that seek to create discrete defects have been greatly aided by the crystal structure of the small pocket. This has provided a detailed description of which amino acids are on the surface of this domain. Furthermore, recent co-crystal structures containing pRB and pieces of E2Fs have further indicated the sites of important protein-protein interactions within the large pocket region [50-52]. The use of this information to generate mutations and guide experimentation into the mechanism of pRB function will hopefully pave the way for a more rational understanding of how pRB controls cell proliferation and suppresses cancer development. The perspective that is presented above is meant to emphasize that our mechanistic understanding of pRB is still in its infancy because we are only now developing the tools to discover how pRB works, or to definitively test many of the long-standing models of its function. To further emphasize the importance of mapping and disrupting individual protein-protein interaction sites in mechanistic studies of pRB, the ensuing sections of this review will be organized around individual protein-protein interaction sites or activities mediated by the large pocket domain. To truly focus this review on structurefunction analysis of pRB, the work that is discussed will highlight reports that attempt to dissect function within the large pocket domain, and not merely demonstrate its involvement. In this way the goal is not only to discuss recent work, but also to present it from a viewpoint that will reveal important gaps in our mechanistic knowledge of pRB by attempting to describe its function as a sum of a few discrete parts. Regulation of E2F transcription factors, is it the key to pRB function or just another in a long list? E2F1 was one of the first pRB interacting proteins to be identified [53,54]. It is now known that pRB interacts with the first four members of the E2F family [55]. Each of these E2Fs needs to partner with a DP subunit to stably interact with pRB and to bind to E2F promoter elements [4]. Each of E2Fs 1, 2, and 3 are potent stimulators of cell Commonly used RB-1 mutations and pRB domain structure Figure 2 Commonly used RB-1 mutations and pRB domain structure. The open reading frame of pRB is shown along with the locations of the A and B parts of the pocket. Arrows indicate the positions and alterations found in commonly studied low penetrance and high penetrance mutant alleles. Below the open reading frame, structural regions of pRB are identified along with their approximate locations. A B Small Pocket Large Pocket 1 928 N-terminus C-terminus Partially functional low penetrance alleles: Null alleles commonly used as negative controls: C706F del738-75(exon 22) delN480 R661W C712R del829-88(exons 24 & 25) Spacer del703-37(exon 21) del127-66(exon 4) A B Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 4 of 15 (page number not for citation purposes) cycle progression from G1 to S-phase and are referred to as 'activator' E2Fs [18]. Direct interaction with pRB blocks this activating function because pRB masks its activation domain [56-60]. Over expression of activator E2Fs can bypass pRB's growth suppressing activity [61]. Likewise, ectopic proliferation in Rb deficient mouse embryos can be suppressed by reducing E2F activity by crossing to E2F1, 2, or 3 knock out strains [62-64]. For these reasons this interaction is viewed as central to how pRB functions in proliferative control and as a tumor suppressor gene. From a structure-function perspective, investigating the properties of pRB mutants that are deficient for E2F binding has tested the importance of this interaction. The data that has emerged from these studies will be discussed below and it shows that E2F regulation is necessary for a maximal cell cycle arrest by pRB, but it also indicates that pRB has other growth regulating functions. Initial reports of the disruption of E2F regulation by pRB used the low penetrance pRB allele, R661W. This mutant protein is defective for binding to E2F transcription factors in in vitro binding assays [28,47], and endogenous interactions have been evaluated in knock-in mice where this protein shows a dramatic reduction in binding to all activator E2Fs. In cell culture assays, this mutant retains some activity in proliferative control even though transcriptional regulation of E2Fs has been lost [47,65]. In addition, Saos-2 cells that are derived from an osteosarcoma are able to undergo aspects of bone differentiation and senescence in response to R661W [47]. The functional effects of the R661W mutation in a knock in mouse model are similar [66]. These animals develop to a slightly later stage of gestation than the knock out, despite showing complete deregulation of E2F transcription and control of fibroblast proliferation. The authors offer a number of carefully examined paradigms for differentiation in erythrocyte and macrophage lineages that show R661W retains at least partial activity. Thus, the R661W mutation may have more defects than loss of E2F regulation, but it still retains some of pRB's growth regulating activity. Two reports describe the structure of the small pocket bound to the pRB binding domain of an E2F [50,52]. These have provided molecular insight into this important interaction. Substitution of amino acids that mediate E2F and pRB interactions has revealed that clean removal of this E2F binding site can eliminate transcriptional repression of E2F reporters [67,68]. Similar to what has been found with R661W, discrete loss of E2F binding to pRB does not abrogate all proliferative control activity as most models of pRB function (such as that pictured in Figure 1) have predicted [67,68]. If removing E2F binding does not eliminate G1-S regulation by pRB, but only cripples it, what functions remain to control the cell cycle? One candidate that has been proposed to mediate pRB arrest that is distinct from E2F regulation is p27. Using mutants of pRB that have impaired E2F binding, p27 levels were shown to be elevated during senescence induction in Saos-2 cells [49]. More recently, an insightful report by Ji et al. showed that cells induced to express wild type pRB are capable of arresting in G1 before repression of E2F targets takes place [69]. These authors further demonstrate that the immediate activity in cell cycle arrest comes from the pRB C-terminus' ability to interfere with Skp2 targeting p27 for degradation. The R661W mutant retains this function, reconciling this activity with previous observations of cell cycle regulation by this mutant. These experiments establish a direct link between pRB and the control of cyclin/Cdk activity that is separate from transcriptional regulation of cyclin and Cdk genes. A more detailed evaluation of the mechanism of Skp2 regulation by pRB reveals that APCCdh1 and Skp2 simultaneously bind to pRB [70]. This interaction targets Skp2 for degradation. Analysis of the interaction sites for these molecules with pRB reveals that Skp2 contacts the Cterminal region of pRB [69] and Cdh1 makes use of the LXCXE binding cleft in combination with E2F contacts in the large pocket [70]. Based on the similarity of contact sites used by APC and Skp2 in degradation, and those by E2F/pRB/chromatin regulating complexes, it appears that pRB has two discrete cell cycle control mechanisms that are mediated by similar regions of the large pocket (Figure 3). This implies that in G1 two separate populations of pRB are engaged in distinct growth regulating mechanisms. In the process of removing E2F binding from pRB by mutagenesis it was also determined that pRB retains the Binding sites for E2Fs and othe ties Figure 3 r proliferative control activiBinding sites for E2Fs and other proliferative control activities. The open reading frame of pRB is shown with locations of the A and B parts of the pocket. Regions of contact that have been mapped for E2F transcription factors, chromatin regulators, and protein degradation machinery that are involved in controlling proliferation are diagramed. A B E2Fs 1 928 Chromatin Regulators (LXCXE binding) Transcriptional Control Interactions: Protein Degradation Interactions: Cdh1 Skp2 E2F1 Apoptotic Regulation: Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 5 of 15 (page number not for citation purposes) ability to bind to E2F1 using a separate interaction site [67,68]. This E2F1 only, or 'Specific' binding site, is located exclusively in the C-terminal portion of pRB [67]. Furthermore, apoptotic induction by E2F1 can be inhibited by a pRB mutant that is deficient for binding to all other E2Fs [67]. Mutation of the E2F1 'Specific' binding site inhibits this activity even when the other E2F binding site is present [71]. Chau et al. have shown that a pRB mutant that is incapable of regulating cell cycle advancement and is only capable of binding to E2F1, can inhibit apoptosis induced by DNA damage [68]. This suggests that E2F1 regulation may have a role in DNA damage signaling independent of cell cycle regulation by pRB. Recently, the C-terminus of pRB was co-crystallized with fragments of E2F1 and DP1 [51]. This structure reveals the molecular contacts between a well-conserved portion of pRB that is not found in p107 or p130 and a region of E2F1 that is different from the other E2F family members. In addition, swap experiments exchanging the contact site on E2F1 with the analogous region of E2F3 demonstrate that this interaction is truly unique to E2F1 [71]. Mechanistically, it remains to be determined how pRB uses this interaction with E2F1, however, the identification of a novel pRB-E2F interaction through structure-function analysis reveals that there are new regulatory pathways that have yet to be explored. Much emphasis has been placed on the central role of pRB-E2F interactions in cell cycle control. Remarkably, experiments that investigate the requirement for these interactions reveal that they are important, but alone can not explain how pRB controls proliferation. Future work will need to quantitatively define the contribution of these non-E2F mechanisms of cell cycle control. The R661W mutation suggests that non-E2F and E2F dependent cell cycle control mechanisms can be separated, however the pleiotropic nature of this mutation leaves us unsure how much each contributes. Alleles that discretely remove each arrest mechanism alone, or both together, will help answer how much each contributes to pRB's overall function. More importantly, do E2F and Cdh1/ Skp2 interactions account for all activity in cell cycle control by pRB or are their even more mechanisms? It will also be necessary to use model organisms to investigate how much E2F regulation contributes to pRB's role as a tumor suppressor compared with Cdh1/Skp2 or others. This type of analysis will ultimately determine how central E2F regulation really is to pRB's function or whether our models of pRB-E2F regulation in cancer need to be reconsidered. From viral transformation to the histone code, the LXCXE binding cleft has something for everyone As described above, the small pocket region of pRB was initially mapped as the minimal domain capable of binding to Adenovirus E1A, SV40 TAg, and HPV E7 [40]. Evaluation of the transforming properties of these viral oncogenes revealed that a peptide motif called LXCXE was necessary for pRB binding and transformation [72-77]. The convergent evolution of these very different viruses highlights the importance of RB family inactivation in viral transformation. Molecular insight into this interaction was revealed when the LXCXE motif was co-crystallized with the small pocket of pRB [38,78]. Beyond revealing the intermolecular contacts between the LXCXE motif and the pRB small pocket, this work also revealed that the contact site of the LXCXE motif on pRB is one of the most well conserved structural features of pRB family proteins [38] (Figure 4). This region of the pocket is often called the LXCXE binding cleft. The conservation of amino acids in the LXCXE binding cleft implies that propRB and the LXCXE motif Figure 4 pRB and the LXCXE motif. (A) A repressor complex composed of an E2F transcription factor, pRB, and a chromatin regulator is shown in G1. The interaction between pRB and the chromatin regulator is mediated by the peptide sequence LXCXE on the chromatin remodeling factor and the LXCXE binding cleft on pRB. Viral proteins like TAg, E7, and E1A can disrupt this complex because they contain an LXCXE motif. Disruption of this pRB-containing complex can advance the cell cycle into S-phase. (B) The LXCXE binding cleft on pRB is a shallow groove formed between two parallel helices. The amino acid sequence of these helices in human pRB is aligned to the analogous sequences from pRBfamily proteins from humans, corn, fruit flies, worms, and algae. Shaded amino acids are identical between all protein comparisons. The asterisk denotes amino acids whose side chains make direct contact with LXCXE in the crystal structure reported by Lee et al. [38]. E2F DP pRB Chromatin remodeling factor G1 LXCXE A) B) 752 764 pRB p107 p130 RRB1 RBF LIN-35 Mat3p E2F DP pRB Chromatin remodeling factor S-phase LXCXE Viral Oncoprotein LXCXE Viral Infection 700 715 * * * ** H. sapiens H. sapiens H. sapiens Z. mays D. melanogaster C. elegans C. reinhardtii LDQIMMCSMYGICKVK LDQLLLCAFYIMAKVT LDQLLMCAIYVMAKVT IDQLILCCLYGVAKVC LDQNIMCAIYIYIRVK LDQILLCCVFVIMKIN IDQIMLSTLYGYCKVH IIVFYNSVFMQRL LIKFYNTIYVGRV LIQFYNNIYIKQI IITFYNEVFVPAA IIHFYNHTYVPLM IIKYYNIEFRDRI IIGFYNAVFVPAM Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 6 of 15 (page number not for citation purposes) tein interactions mediated by it are a key component of pRB's function. Amino acid sequence conservation provides a straightforward rationale for studying the LXCXE binding cleft on pRB. Unfortunately, understanding how pRB uses LXCXE type interactions in proliferative control remains ill defined. One reason is that there are 28 proteins that have been reported to interact with pRB using an LXCXE motif (Table 1), raising the question of whether all of these interactions are necessary for proliferative control events mediated by pRB, or whether some are needed for more specialized functions. It is beyond the scope of this review to discuss how each of these molecules is proposed to partner with pRB, however, most cellular proteins that possess an LXCXE motif are believed to work with pRB in transcriptional regulation. Some of these proteins have an enzymatic activity or other property that can modify chromatin structure, these include BRG1, Brm, DNMT1, HDAC1 and 2, HP1, RBP2, RIZ1, and Suv39h1 [19-23,79- 83]. Others have no clear biochemical activities, but contribute to repression of transcription in cell culture assays, or are themselves known transcription factors. These include CtIP, EID1, ELF1, HBP1, RBP1, and UBF [82-90]. It is difficult to generalize the other LXCXE motif proteins, but they participate in functions as varied as DNA replication and apoptosis while many have no clearly defined functions. Examination of Brm, RBP1, RBP2, and HBP1 irrespective of pRB has revealed roles for them in regulating proliferation or maintaining terminal differentiation [91-94]. Given the broad possibilities for how pRB can use LXCXE type interactions (because of the vast number of interacting proteins that contain this motif), recent efforts have focused on mutating the LXCXE binding site on pRB and examining the defects caused by loss of all LXCXE interactions with pRB. Surprisingly, the ability to control cell proliferation remained largely intact in Saos-2 cell cycle arrest assays [95-98]. However, in some assays of permanent cell cycle arrest, mutations in this region of pRB were shown to allow cell cycle re-entry [96,98]. Recently, the generation of a knock-in mutant mouse strain has been described in which the LXCXE binding site on pRB has been disrupted [99]. These mice are viable, implying that cell cycle exit and differentiation occurs normally during development. It remains to be seen if there are proliferative control defects in these mice or if they are cancer prone because of them. However, the defect that was highlighted in this report was genomic instability. Disruption of pRB-LXCXE interactions caused defects in heterochromatin condensation due to diminished histone H4-K20 trimethylation. Chromosomal regions normally characterized by this modification become fused causing errors in mitosis. Thus, one in vivo manifestation of defective LXCXE interactions is abnormal chromatin structure. Another approach to identify proteins that cooperate with pRB by contacting the LXCXE binding site has been to purify pRB containing protein complexes. Isolation of the fruit fly RBF proteins with a complex of other transcriptional regulators has offered insight into the assembly of a transcriptional repressor module called dREAM [100,101]. Sequence comparisons have revealed that this is an evolutionarily conserved complex whose components are found in distantly related organisms including humans, flies, and C. elegans. This complex of proteins contains subunits that are dependent on the LXCXE binding cleft for interaction with pRB in in vitro binding experiments [100]. The use of ChIP assays to identify target promoters and RNAi to deplete individual members of this complex has revealed that it functions to block transcription of a vast number of developmentally regulated genes and is also involved in controlling DNA synthesis in Drosophila oocytes [100-103]. Mutations in the LXCXE binding cleft in knock-in mice would be expected to disrupt assembly of this complex, so how can dREAM be essential for regulating so many transcriptional targets while the knock-in mice remain viable? A recent clue has Table 1: Proteins that use an LXCXE interaction motif to bind to pRB Protein Biochemical Functions References AhR Aryl hydrocarbon receptor [158] ASK1 Ser/Thr kinase [159] Bog Unknown [160] BRG1 ATP dependent helicase [161] Brm ATP dependent helicase [110] CtIP Transcriptional repressor [84] Cyclin D1 kinase regulatory subunit [162] Cyclin D2 kinase regulatory subunit [163] Cyclin D3 kinase regulatory subunit [164] DNA Pol delta DNA polymerase [165] DNMT1 DNA methyltransferase [20] EID-1 Unknown [85] [86] ELF1 Transcription factor [87] HBP1 Transcriptional repressor [88] [89] HP1 trimethyl histone H3 binding protein [19] HDAC1 histone demethylase [21] [22] [80] HDAC2 histone deacetylase [81] Hec1 Unknown [166] Hsp75 Heat-shock protein [167] p202 Unknown [168] p204 DNA replication [169] RFC p145 DNA replication [170] RBP1 Chromatin remodeling/ transcriptional repression [82] RBP2 histone demethylase [82] RIM Unknown [171] RIZ methyltransferase [83] [172] Suv39h1 methyltransferase [19] UBF PolI transcription factor [90] Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 7 of 15 (page number not for citation purposes) come from the isolation of endogenous protein complexes containing the p130 protein [104]. This work has revealed that endogenous human dREAM components only assemble with p130 and not other pocket proteins, suggesting that human pRB differs from pocket proteins found in lower eukaryotes because it does not participate in the dREAM complex [104]. In summary, mice with an LXCXE binding cleft mutation have defective chromatin structure and this is in agreement with chromatin regulating enzymes using an LXCXE motif to bind to pRB [99]. However, the lack of obvious cell cycle phenotypes caused by disruption of LXCXE interactions in cell culture and knock-in mice reveals that there is not a universal requirement for them in all aspects of pRB's proliferative control activity [95-99]. The investigation of transcriptional repression mechanisms using model organisms like C. elegans and Drosophila has provided tremendous insight into the RB family of proteins. However, their RB gene families are slightly different in that they do not possess an exact orthologue of human pRB. Because of these differences, the dREAM complex seems unlikely to be a key cellular partner of pRB in mammals [104], leaving in question how pRB uses the LXCXE binding cleft? The fact that mice with LXCXE cleft mutations are viable creates a perplexing contradiction since its maintenance creates susceptibility to the pathogenic effects of DNA tumor viruses [105]. Since the LXCXE binding cleft on pRB is one of the most well conserved features of RB family proteins [38] it must have an important function. Determining the cellular context where LXCXE interactions with pRB are essential is the first step to understanding this aspect of pRB function. Further exploration of phenotypes in LXCXE binding cleft mutant mice should guide this analysis. Identifying the exact LXCXE dependent partners that cooperate with pRB remains a daunting task. A true understanding of how pRB uses LXCXE interactions will likely require some rationalization of the many possible binding partners. Induction of differentiation and transcriptional activation; is pRB a switch hitter? In the preceding sections little effort has been made to discriminate between cell cycle arrest and terminal differentiation. Both involve a block in cell proliferation; however, terminal differentiation also includes a significant change in gene expression patterns as these cells take on a specialized function. In addition to participating in cell cycle exit, pRB has been proposed to augment expression of cell type specific genes during differentiation [12,13]. In this way pRB can aid in transcriptional repression as well as activation, depending on the cellular context. This section of the review will focus on the mechanisms that pRB participates in to facilitate transcription in differentiation. The activity of transcription factors that induce differentiation can be augmented by pRB. These include MyoD, CBFA1/RUNX2, C/EBPs, NF-IL6, Pax transcription factors, and nuclear hormone receptors [14,16,106-112]. In addition, pRB dependent inhibition of proteins like EID- 1, Id2, and RBP2 has also been shown to activate transcription during differentiation [85,86,93,113]. So how does pRB induce differentiation? From a structure-function perspective the best evidence for pRB's role in differentiation has come from a synthetic mutant pRB called \\u2206663 that substitutes five amino acids in the pRB pocket starting at codon 663 to the peptide sequence NAIRS. This protein is defective for E2F binding, transcriptional repression, and induction of a G1 arrest. However, expression of this mutant induces markers of both muscle and bone differentiation [47]. It has also been reported that mutations in the LXCXE cleft of pRB prevent it from fully activating MyoD dependent transcription during myogenesis [114]. Because of the structural component underlying the data on pRB's role in muscle and bone differentiation, these paradigms will be discussed further. In the case of muscle, pRB is necessary for MyoD to maximally induce late markers of differentiation [15,106]. Precisely how pRB augments MyoD is not known but it seems to involve influencing phosphorylation of MEF2C that is necessary for maximal transcription of late markers [115]. In the case of bone differentiation, direct interaction between pRB and HES1 with RUNX2/CBFA1 induces DNA binding and enhanced transcriptional activation of bone specific transcriptional targets [116]. Conversely, induction of muscle or bone differentiation can be aided by pRB's ability to antagonize HDAC1, EID-1, and RBP2 in transcriptional regulation. HDAC1 is able to inhibit MyoD in cycling cells and upon induction of differentiation pRB titrates HDAC1 away from MyoD allowing it to activate transcription. Similarly, EID-1 can inhibit p300/ CBP acetylase activity that is necessary for MyoD to activate transcription. At the onset of differentiation, pRB binds to EID-1 and targets it for degradation thereby allowing MyoD to recruit p300 to stimulate transcription of muscle specific genes [85,86]. In a similar manner, pRB has been proposed to interact with RBP2 and remove it from promoters where it inhibits transcriptional activation through its H3-K4 demethylase activity [93,117]. In assays of CBFA1/RUNX2 dependent activation, pRB's stimulatory effect is abrogated by siRNA depletion of RBP2 [93]. Likewise, MyoD's inability to stimulate differentiation in pRB deficient fibroblasts can be rescued by simultaneous depletion of RBP2 [93]. These data strongly suggests that RBP2, HDAC1, and EID-1 along with pRB play critical roles in differentiation through antagonism. These mechanisms offer insight into how pRB can function in aspects of differentiation beyond a mere G1 arrest. Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 8 of 15 (page number not for citation purposes) However, many questions remain regarding pRB's role in differentiation. Since the work highlighted above does not reveal much common ground between these different mechanisms, but a single mutant (\\u2206663) is capable of inducing many of them, does pRB have one type of protein-protein interaction just for differentiation? Or are there many, subtle activation mechanisms where no one mechanism is essential? If the second possibility is true, then does this imply that pRB is really just a cell cycle regulator and its role in differentiation is indirect? When considering this possibility it is important to remember that CKI proteins p16 and p21 can block proliferation, but can not enhance transcription by MyoD [15]. Only pRB has this dual ability, suggesting that it must have additional activities in differentiation beyond G1 arrest. In order to move forward and come to a firm conclusion about how pRB works in differentiation, new structurefunction data is imperative to define which interaction surfaces are essential. The \\u2206663 mutant is defective for E2F binding as well as contact with LXCXE proteins suggesting that neither of these is likely to be necessary for transcriptional activation [47]. Curiously, each of HDAC1, EID-1, and RBP2 are proposed to contact pRB using an LXCXE motif [22,82,85,86]. This apparent contradiction raises the question of whether LXCXE interactions are really necessary in differentiation? Rb deficient embryos that are rescued by a normal placenta can develop to birth where they die due to defects in myogenesis [118], revealing a place in development where pRB plays an essential function in this differentiation paradigm. Isaac et al. report that homozygous mutation of the LXCXE binding cleft results in a viable mouse [99], indicating that LXCXE interactions with pRB are not be essential for muscle development. However, Benevolenskaya et al. have shown that RBP2 can interact with the R661W mutant of pRB and mutation of the LXCXE motif on RBP2 has only a minimal effect on the interaction with pRB [93]. This suggests that possession of an LXCXE motif does not mean that it is essential for mediating an interaction with pRB. In short, structure-function analysis of pRB's role in differentiation has wet our appetite about what it may do, but has left us unsure of how it actually participates. Future work in this area would be greatly aided by a class of pRB mutants that can block proliferation, but not induce differentiation, rather than the reciprocal properties that have already been reported. This type of mutant will focus our attention on which contacts are essential for pRB dependent transcriptional activation. Regulation by post-translational modifications, just an on-off switch or are there different settings? The retinoblastoma protein is a target for many posttranslational modifications. The most extensively studied aspect of pRB regulation is phosphorylation [4], however, it can be acetylated [119] and sumoylated [120] as well. In addition, apoptotic signaling can modify pRB function by proteolytic cleavage [121]. Regulation of pRB by phosphorylation was first observed as a mobility shift in SDS-PAGE where pRB is faster migrating in G1 phase cells and slower migrating in S, G2, and M-phase [122-125]. This behavior gave rise to the terms hypophosphorylated (pRB) and hyperphosphorylated (ppRB) when referring to these bands in western blots. Binding assays to study pRB interactions with E2Fs, chromatin regulators, and other binding partners have revealed an almost uniform preference for binding to the hypophosphorylated form of pRB [4,24]. This indicates that pRB binds to its interacting partners in G1 when it is underphosphorylated implicating this as the active form. Phosphorylation at the beginning of S-phase then prevents pRB from interacting with other proteins until the end of mitosis when it is dephosphorylated [126]. Based on this data, phosphorylation offers a relatively clear onoff mechanism for pRB regulation and this has been discussed in great depth in many previous reviews [3-7,18]. However, a considerable number of experiments raise the question of how closely is this golden rule actually followed? The earliest studies of pRB phosphorylation have used cell synchronization procedures to enrich for cells containing active pRB in G1 and inactive ppRB from the other phases [55,125]. The presence of pRB/E2F complexes in other phases of the cell cycle has often been described as a technical artifact of the synchronization procedure, or the lag between new synthesis of pRB and its phosphorylation. A number of recent reports have demonstrated by ChIP that pRB/E2F complexes can be detected on DNA in S-phase cells [127,128]. Furthermore, analysis of RBF/dE2F complexes from fruit flies indicates that some are resistant to inactivation by cyclin/Cdks and these repress transcription at all stages of the cell cycle as revealed by synchronous cell divisions in the developing eye disc [129,130]. Taken together this indicates there are circumstances where pRB family proteins are resistant to inactivation by cyclin/Cdks despite being exposed to high levels of their activity. Phospho-peptide maps have suggested as many as 21 phosphorylation sites on pRB [131,132]. Through sequence analysis, human pRB contains as many as 16 potential cyclin/Cdk phosphorylation sites containing the Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 9 of 15 (page number not for citation purposes) S/T-P consensus [31]. These phosphorylation sites are located in three general areas of the pRB open reading frame, the N-terminus near the beginning of the pocket domain, the spacer region, and the C-terminal domain just after the small pocket (Figure 5). In G1, cyclin D/ Cdk4 complexes make use of a Cdk4 binding site in the extreme C-terminus of pRB that is required for efficient phosphorylation [133]. Ablation of all three D-cyclin genes reduces phosphorylation on residues 249, 252, 807, 811, and 826 of pRB suggesting that they are primarily the targets of D associated kinases [134]. Cyclin E/Cdk2 becomes active near the G1-S boundary and uses an interaction motif called RxL or Cy to recognize pRB and phosphorylate the remaining sites [135-137]. There are a number of potential RxL motifs in the C-terminus of pRB, but mutagenesis indicates that two adjacent motifs that begin with lysine residues 870 and 873 are the most important [135]. These experiments reveal a concerted mechanism in which D and E cyclin/Cdk complexes bind to pRB at discrete locations in the C-terminus and phosphorylate it at other locations. A number of reports have suggested that the C-terminus of pRB is largely disordered and flexible [38]. This allows substrate recognition through binding while flexibility ensures access to phospho-acceptor sites. Using mutagenesis 11 of the possible 16 Cdk phosphorylation sites have been substituted to alanine to prevent inactivation of pRB [31-33]. Experiments using these constructs have revealed that this non-phosphorylatable form of pRB can block cell proliferation in fibroblasts and tumor derived cell lines [32,33,138]. Examination of pRB mutants that have only a subset of these substitutions reveals that there is no single phosphorylation site that plays a critical role in the inactivation of E2F binding to pRB [31,32,139]. However, interactions at the LXCXE binding cleft can be regulated by a subset of phosphorylation sites in the C-terminal region of pRB [140,141]. Curiously, arrest of Saos-2 cells with non-phosphorylatable pRB can be bypassed by cyclin/Cdks without phosphorylating pRB [138]. Thus, phosphorylation of pRB to control advancement of the cell cycle may be dispensable. In addition, expression of non-phosphorylatable pRB in mammary epithelia of transgenic mice initially causes slowed proliferation during development, but eventually ductal infiltration of the fat pad catches up and exhibits hyperplasia and in some cases carcinogenesis [142]. The authors of this study argue that non-phosphorylatable pRB would also remain active for the inhibition of E2F induced apoptosis and this could explain its oncogenic properties. Regardless of the interpretation, it is clear from these transgenic studies that phosphorylation of pRB is not ubiquitously required for cell division to occur. In addition to post-translational modifications that occur in response to mitogenic signals, pRB is also modified in response to DNA damage. Acetylation of lysines 873 and 874, as well as phosphorylation of serine 612 occurs rapidly in response to double strand DNA breaks [143,144]. The immediate implication of these findings is that posttranslational modifications on pRB can occur under circumstances other than proliferative control. In fact, DNA damage activates G1 and S-phase arrest mechanisms that dephosphorylate ppRB at cyclin/Cdk sites [128,145,146]. This implies that modifications added to pRB following DNA damage result in a gain in function. The exact mechanism of action that these modifications initiate is not clear. Phosphorylation of Ser612 has been reported to enhance binding to E2Fs as an anti-apoptotic mechanism [144] while acetylation of 873, 874 has been suggested to release E2Fs to induce cell death [143]. To further add to the confusion of how acetylation of pRB regulates its function, Nguyen and co-workers have shown that pRB becomes acetylated during muscle differentiation [147]. Using a non-acetylatable mutant where lysines 873 and 874 are changed to arginines, they demonstrated that this mutation leaves E2F transcriptional regulation intact. Interestingly, the resulting muscle cells are susceptible to re-entry into S-phase when stimulated by serum suggesting that acetylation is necessary for pRB to enforce permanent cell cycle exit. Clearly more work is needed to define the role of pRB 'activation' in differentiation or following DNA damage, but these studies illustrate that there is Post-translational mo Figure 5 difications of pRB Post-translational modifications of pRB. The open reading frame of pRB is depicted with the A and B regions of the pocket indicated. Cyclin/Cdk phosphorylation sites that have been investigated by mutagenesis are shown as are sites of phosphorylation following DNA damage. The location of acetylated and sumoylated lysines are indicated. Arrows indicate the approximate location of Cdk4 and cyclin E binding sites that are needed for optimal phosphorylation of pRB. Caspase cleavage of pRB occurs on the C-terminal side of aspartate 886, removing amino acids 887 to 928. 1 928 A B PhosphorylationDNA Damage: Cyclin/Cdk: Sumoylation: Acetylation: Ser 780, Ser 788, Ser 795, Ser 807, Ser 811, Thr 821, Thr 826 Ser 612 Thr252 Thr356 Ser 608 Lys 873, Lys 874 Lys 720 Caspase Cleavage: Cyclin/Cdk Binding: Cyclin E/A Cdk4 Asp 886 A B Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 10 of 15 (page number not for citation purposes) more to the regulation of pRB than a simple on-off mechanism at the G1-S boundary. One implication of acetylation of lysines 873 and 874 is that there may be a hierarchy of pRB regulation where acetylation blocks phosphorylation because the Cy motifs at this same position are no longer recognized by cyclin E/Cdk2. This could be advantageous in differentiation and DNA damage because it would prevent promiscuous phosphorylation of pRB and ensure cell cycle arrest. The most recently discovered modification of pRB is sumoylation. Sumoylation is a ubiquitin like modification that has recently been implicated in numerous biological functions, including the induction of senescence [148]. The retinoblastoma protein has been reported to be sumoylated in the small pocket near the LXCXE binding cleft at lysine 720 [120]. The exact mechanistic consequence of modifying pRB with sumo is not fully understood, however, it has been proposed to compete with proteins that bind to the LXCXE cleft. In other studies of sumoylation, the overall effect of this modification on p53 and RBP1 is an induction of senescence [149,150]. Whether sumoylation of pRB plays a role in the induction of senescence is not known, but the pRB LXCXE binding cleft has been proposed to interact with chromatin regulators to establish repressive heterochromatin as part of a permanent arrest [151]. This raises the question of whether HP1 and other chromatin regulators can contact pRB if sumo is interfering with access to this surface. Tumor necrosis factor alpha (TNF-\\u03b1) stimulates inflammation and apoptosis [152]. It signals to activate so called 'initiator' caspases that cleave substrates as part of a signaling cascade that ultimately leads to the activation of 'effector' caspases that ultimately kill the cell. Caspase cleavage of pRB occurs at an intermediate step of TNF-\\u03b1 signaling and results in the irreversible removal of the Cterminal 42 amino acids [153,154]. This is defined as an intermediate step in cell death because expression of a caspase resistant mutant of pRB inhibits progression of signaling and the effector caspases are never activated [153,154]. Using a knock-in mouse model where the DEADG recognition sequence in pRB is changed to DEAAE (called pRB-MI), Chau et al. showed that intestinal epithelia are uniquely resistant to TNF-\\u03b1 mediated cell death [155]. Furthermore, crossing the pRB-MI mice to p53-/- mice results in frequent colonic adenomas that are rare in p53-/- mice. This work suggests that pRB is a critical target in apoptotic induction by TNF-\\u03b1 and disruption of this pathway in the intestine allows damaged cells to survive and begin to form tumors [156]. Unfortunately, the mechanistic details of how pRB is regulated by proteolysis are less well understood. Two possibilities are that cleavage eliminates binding to a pro-apoptotic factor that pRB normally sequesters and its release allows cell death can to ensue [154,155], alternatively removal of the last 42 amino acids destabilizes pRB so that it's expression is eventually lost [153]. In this way cells would effectively behave as if they were pRB deficient and E2F1 would be liberated to kill cells. Given that TNF-\\u03b1 can induce cell death without the need for new protein expression, the former possibility seems more likely although there remains much to be done to understand this mechanism [155]. To summarize our present knowledge of post-translational regulation of pRB, most attention has been focused on phosphorylation and this modification has been shown to block pRB's growth inhibitory properties. Despite the preponderance of evidence for negative regulation of pRB by phosphorylation, it is clear that exceptions exist to this regulatory mechanism. Perhaps other regulatory modifications can 'mark' pRB to retain function in the face of cyclin/Cdks as described above, or inactivate it without phosphorylation. The answer to how pRB regulation is impacted by multiple signaling pathways will need to await further investigation. Never the less, these different types of modifications are revealing further complexity to an already multi-functional protein. Conclusion A common theme in this review is that a large number of proteins contact a relatively small region of pRB. Mutations that affect the pocket domain can simultaneously disrupt many aspects of pRB and have made dissecting interactions within this domain very difficult. Perhaps the reason why a vast majority of tumor-derived alleles of RB- 1 eliminate the pocket domain completely [8] is because they represent the easiest way to disrupt everything that pRB does. Never the less, with a multi-functional protein like pRB where so much is mediated by just the large pocket domain, it is even more important to define how different interacting proteins can act together with it to control cell growth. Without this level of understanding we will never be able to say how pRB really works. Despite the challenges of a structure-function based approach to pRB, discussing its function based on different protein interaction sites, or structural domains, as described above readily covers most aspects of pRB's known role in proliferative control and cancer. In nearly each case the use of mutants with defined defects reveals that the prevailing view of pRB function is part of how it works but unlikely to be the whole explanation. These structure-function approaches to study pRB have revealed many unexplained activities and highlight areas for further investigation. Just how important are these unexplored areas? A recent paper by Binne et al. shows that without Cdh1, pRB has almost no ability to induce a G1 cell cycle arrest [70]. Likewise, Sellers et al. have shown a Cell Division 2007, 2:26 http://www.celldiv.com/content/2/1/26 Page 11 of 15 (page number not for citation purposes) dramatic separation of proliferative and differentiation activities [47]. Thus, a number of these emerging areas are likely quite central to pRB function but are simply less well understood. Structure-function approaches offer a reliable means to examine how proteins interact with one another to carry out various cellular processes. Combined with genetic analysis that defines sufficiency and necessity, the purpose of individual protein-protein interactions can be very effectively defined. Despite using this approach, we still struggle to draw firm conclusions about what makes pRB a tumor suppressor. The properties of the low-penetrance allele R661W suggest that E2F interactions are an essential component of this function. However, since this allele is only partially penetrant, and some carriers of this mutation can live cancer free [157], is it telling us that E2F binding is only a small part of tumor suppression by pRB? Does this implicate Cdh1/Skp2 binding in tumor suppression because it is part of the proliferative control mechanism that functions in the absence of E2F control in the R661W mutant? Or should we exclude Cdh1/Skp2 interactions because they are not deficient in R661W? Clearly more data on tumor predisposition caused by partially defective pRB molecules is essential to draw firm conclusions about this crucial question. Since there are a limited number of tumor derived mutations to choose from, generating more synthetic alleles with discrete defects is likely the only way forward. For example, if gene-targeted mice carrying the LXCXE cleft mutation turn out to be tumor prone, then we would conclude that LXCXE interactions are part of pRB's tumor suppressor function. This would undoubtedly motivate more effort to identify precisely how pRB uses this type of interaction compared with another binding site that is excluded by a similar tumor incidence study. This type of systematic approach to investigating pRB tumor suppression could be brought to bear on many pRB interactors and binding sites. However, as lists like that in Table 1 reveal, pRB interacting proteins are numerous [24]. A potential opportunity to better focus our efforts on what makes pRB a tumor suppressor may also come from comparisons with the related pocket proteins p107 and p130. As stated earlier, these molecules share extensive similarity with pRB in the pocket domain, yet on their own neither appears to meet the genetic definition of a tumor suppressor gene. Focusing on proteins that interact with pRB and not p107 and p130 could help to guide our selection of interactors that are most relevant in cancer. By this way of thinking, a complex like APCcdh1 would hold considerable appeal because it selectively binds pRB over p107 and p130 even in in vitro binding assays [70]. Alternatively, the dREAM complex that we now know only contacts p130 in vivo [104] would be a less appealing candidate to cooperate with pRB in tumor suppression. Advancing our understanding of pRB function in this most crucial of questions will clearly require multiple approaches. The investigation of endogenous pRB protein complex composition, combined with structure-function analysis of discrete mutant alleles of the RB-1 gene, offers a logical means to separate pRB function into distinct parts to define its role in cancer.  To find correlation between the type of mutations observed and the severity of the disease using multiple techniques like polymerase chain reactions (PCR), quantitative multiplex PCR, sequencing and RNA analysis.  Methods  Prospective, observational study. Patients who had been screened for mutations in the RB1 gene were included in the study. Patient details including demographic data; age and sex, laterality, international classification of intraocular retinoblastoma (ICIOR) staging, modality of management, histopathology high risk factors if the eyes were enucleated and metastasis rate were assessed.  Results  Seventy four patients were studied. Fifty three patients had bilateral and 21 unilateral disease. Complete genetic data was analyzed for 74 patients and complete clinical correlation was established for all the 49 patients with mutations. Of the total mutations identified, 11/49 (22.4%) of patients had large deletions, 12/49 (24.5%) had small deletions or insertions, 14/49 (28.6%) had nonsense mutations, 7/49 (14.3%) had splice mutations and 5/49 (10.2%) of patients had missense mutations. Four cases were familial. Group E ICIOR stage at presentation was noted in 40% of patients with large deletions, 33% with small deletions whereas 38.5% with splice mutations and 44.4% of patients with missense mutations presented with Group B ICIOR. Twenty five percentages of eyes with large deletions had high risk features on histopathology and one patient among these developed metastasis.  Conclusion  Current laboratory testing of RB1 mutations may be feasible in determining the severity of the disease and patient counseling. The study provides a starting point for looking at correlations.  Keywords: Retinoblastoma, RB1 mutation, ICIOR groups, Deletions, Multiplex-PCR Go to: 1.\\u2003Introduction Retinoblastoma (Rb) is the most common intraocular malignancy in children, with a reported incidence ranging from 1 in 15,000 to 1 in 18,000 live births (Bishop and Madsen, 1975). It is second only to uveal melanoma in the frequency of occurrence of malignant intraocular tumors. There is no racial or gender predisposition in the incidence of retinoblastoma. Retinoblastoma is bilateral in about 25\\u201335% of cases (Shields and Shields, 1992). The average age at diagnosis is 18 months, unilateral cases being diagnosed at around 24 months and bilateral cases before 12 months (Shields and Shields, 1992).  Retinoblastoma (Rb) is brought about by biallelic inactivation of the human retinoblastoma susceptibility gene, RB1 (GenBank accession number L11910), located on chromosome 13q14 that codes for the RB protein. The cytogenetic deletions examined in retinoblastomas have assigned the genetic locus of the disease to q14 of chromosome 13 linked with the polymorphic marker gene enzyme esterase D (Friend et al., 1986). A successful labeling of normal and tumor cells with homozygous RB1 gene deletion in the human tumors, showed that most RB1 deficient tumor cells resemble cone photoreceptors, suggesting that cone photoreceptors is the cell of origin (Xu et al., 2009).  Retinoblastoma arises due to two genetic events involving both the alleles of RB1 and occurs in two forms, the hereditary and non-hereditary. Mutation of both the alleles of RB1 is required for tumor initiation. In case of hereditary transmission of the disease, one allele is mutated in the germline and the other mutation occurs somatically in the developing retina. In non-hereditary disease, mutations of both the alleles somatic (Knudson, 1971; Mairal et al., 2000; Corson and Gallie, 2007). The RB1 gene shows a high degree of mutational heterogeneity in retinoblastoma with over 900 mutations reported till date (Valverde et al., 2005). Approaches using multiple techniques including quantitative multiplex PCR and sequencing (Richter et al., 2003), or DHPLC (denaturing high-performance liquid chromatography) along with quantitative multiplex PCR for short fluorescent fragments (QMPSF) Houdayer et al., 2004 could achieve detection rate of 80\\u201389% (Richter et al., 2003; Houdayer et al., 2004). Precise identification of the RB1 gene mutation could help in enhancing the clinical management of the relatives at risk (Gallie et al., 1995). Genetic testing of retinoblastoma is being employed to screen for and detect carriers of RB1 mutations among relatives of affected individuals and for prenatal testing (Gallie et al., 1999). This in turn can facilitate prompt management of the disease and better visual outcome in affected children (Gallie et al., 1999).  Our group first developed and published a combinatorial approach for detection of RB1 mutations (Parsam et al., 2009) and in the present study the authors attempted to correlate clinical features of the disease with RB1 mutations in 49 patients with known mutations.  Go to: 2.\\u2003Materials and methods 2.1. Study: prospective, observational  2.1.1. Technique Mutational analysis of RB1 was carried out to detect different types of mutations as previously described (Parsam et al., 2009). Once our group had the genetic data, authors then correlated the different types of mutations with patient details like demographic data; age and sex, laterality, international classification of intraocular retinoblastoma (ICIOR) staging, modality of management, high risk factors if the eyes were enucleated and metastasis rates. The following histopathological entities were taken as high risk features: (Honavar et al., 2002) anterior chamber seeding, iris infiltration, ciliary body infiltration, massive choroidal infiltration, invasion of optic nerve lamina cribrosa, retrolaminar optic nerve invasion, invasion of optic nerve transection, scleral infiltration and extrascleral extension.  Go to: 3.\\u2003Results Seventy four patients were studied. Fifty three patients had bilateral and 21 unilateral retinoblastoma. By combining the three different approaches as elucidated in the techniques above, mutations were detected in 49 patients (44 bilateral and five unilateral) (Parsam et al., 2009). Complete genetic data was analyzed for 74 patients and complete clinical correlation was established for all the 49 patients with mutations. Since the study is about the clinical correlation among patients of retinoblastoma where mutations were identified, all the results would restrict to these 49 patients. There were 28 males (57%) and 21 females (43%) among the study subjects. The mean age was 17.2 months (range 3\\u201338 months). Of the total mutations identified, 11/49 (22.4%) of patients had large deletions, 12/49 (24.5%) had small deletions or insertions, 14/49 (28.6%) had nonsense mutations, 7/49 (14.3%) had splice mutations and 5/49 (10.2%) of patients had missense mutations. Four cases were familial. 23/49 (46.94%) of the mutations identified were not reported earlier to the best of our knowledge and these were novel mutations identified in our study (Parsam et al., 2009). The clinical details among each class are as follows.  3.1. Large deletions  Eleven of the 49 patients had large deletions accounting for 22.4% of all the mutations. There were six male and five female patients. 9/11 (81.8%) patients had bilateral retinoblastoma whereas 2/11 (18.2%) had unilateral retinoblastoma at presentation. Of the 20 eyes that were studied in this group, eight (40%) presented with Group E ICIOR followed by groups B, D and C ICIOR. Two of the 20 (10%) presented with extra ocular extension and one eye had presented after enucleation elsewhere. Seven (35%) eyes were enucleated. The other modalities of management in this group included chemoreduction and focal therapy. Focal therapy here included cryotherapy and transpupillary thermotherapy. Plaque brachytherapy was evaluated separately. External beam radiotherapy (EBRT) was given as a part of multimodal approach for the extra ocular retinoblastomas. Five of the 20 eyes (25%) had high risk histopathological features and these patients underwent adjuvant chemotherapy. One patient with extra ocular extension presented with metastasis and was subjected for palliative treatment (Table 1).  Table 1 Table 1 Clinical profile of retinoblastoma patients with large RB1 deletions. 3.2. Small deletions  Twelve of the 49 patients had small deletions accounting for 24.5% of all the mutations. There were eight male and four female patients. All these 12 patients had bilateral retinoblastoma at presentation. Of the 24 eyes that were studied in this group, eight (33.3%) presented with Group E ICIOR followed by groups B, D and C ICIOR. Two (8.3%) presented with extra ocular extension and one eye had presented after enucleation elsewhere. Eight (34%) eyes were enucleated. The other modalities of management in this group included chemoreduction, focal therapy and plaque brachytherapy. External beam radiotherapy (EBRT) was given as a part of multimodal approach for the extra ocular retinoblastoma. Two (8.3%) of the eyes had high risk histopathological features and these patients underwent adjuvant chemotherapy (Table 2).  Table 2 Table 2 Clinical profile of retinoblastoma patients with small RB1 deletions. 3.3. Nonsense mutations  Fourteen of the 49 patients had nonsense mutations accounting for 28.6% of all the mutations. There were six male and eight female patients. Thirteen of the 14 patients (92.8%) had bilateral retinoblastoma whereas the remaining one (7.2%) patient had unilateral retinoblastoma at presentation. Of the 27 eyes that were studied in this group, nine (33.3%) presented with Group B ICIOR followed by groups C, E and D ICIOR. Six (22.2%) eyes were enucleated. The other modalities of management in this group included chemoreduction, focal therapy and plaque brachytherapy. Two (7.4%) of the eyes had high risk histopathological features and these patients underwent adjuvant chemotherapy (Table 3).  Table 3 Table 3 Clinical profile of retinoblastoma patients with RB1 nonsense mutations. 3.4. Splice mutation  Seven of the 49 patients had splice mutations accounting for 14.3% of all the mutations. There were five male and two female patients. Six out of seven (85.7%) patients had bilateral retinoblastoma whereas one (14.3%) had unilateral retinoblastoma at presentation. Of the 13 eyes that were studied in this group, five (38.5%) presented with Group B ICIOR followed by groups C, D and E ICIOR. Two (15.4%) eyes were enucleated. The other modalities of management in this group included chemoreduction, focal therapy and plaque brachytherapy. Two (15.4%) of the eyes had high risk histopathological features and these patients underwent adjuvant chemotherapy (Table 4).  Table 4 Table 4 Clinical profile of retinoblastoma patients with RB1 splice mutations. 3.5. Missense mutation  Five of the 49 patients had nonsense mutations accounting for 10.2% of all the mutations. There were three male and two female patients. Four of the five patients (80%) had bilateral retinoblastoma whereas the remaining one (20%) had unilateral retinoblastoma at presentation. Of the nine eyes that were studied in this group, four (44.4%) presented with Group B ICIOR followed by groups D, E and C ICIOR. Two (22%) eyes were enucleated. The other modalities of management in this group included chemoreduction and focal therapy. One (11.1%) of the eyes had high risk histopathological features and this patient underwent adjuvant chemotherapy (Table 5).  Table 5 Table 5 Clinical profile of retinoblastoma patients with RB1 missense mutations. Go to: 4.\\u2003Discussion In the present study a practical correlation between clinico-pathological features of retinoblastoma with genetic mutational status was attempted. The caveat in such studies is that several different categories of mutation such as deletions or insertions of different sizes, and nonsense mutations have the same impact at the protein level. Essentially, the protein would be truncated in case of all frame-shifting and nonsense mutations, and so would be grossly defective or non-functional or would have undergone degradation either at the mRNA or protein level.  Alonso et al. (2001) studied the spectrum of germ line RB1 mutations in a Spanish population and tried to bring out the phenotypic and molecular implications of these mutations. They found 29 mutations in 49 hereditary retinoblastoma patients, most of them (62%) not reported earlier. Sixty nine percentages of these were nonsense mutations and the remainder splice mutations. Splice mutations were associated with delayed onset at presentation (32 months on average) whereas nonsense mutations had early onset (8.7 months on average). In comparison our study showed that 23 (46.94%) of the mutations identified were not reported earlier to the best of our knowledge, however early onset retinoblastomas in our study were seen more with large deletions (mean age 13.64 months, range 3\\u201322 months) and nonsense mutations (mean age 14.43 months, range 5\\u201326 months).  Taylor et al. (2006) were the first to study genotype\\u2013phenotype correlations in hereditary familial retinoblastoma. This was the first large scale study of familial retinoblastoma with high level of homogeneity in the clinical and genetic analysis of patients and their relatives thereby allowing reliable intra-familial genotype\\u2013phenotype correlations. They studied the mutations and correlated them with demographic features, carrier and disease status and disease eye ratio. In comparison since most of the patients (90% or more) presenting in our clinic are sporadic cases, our series comprised sporadic and a few familial cases. We took into consideration clinical manifestations, stages at presentation and modalities of management.  Another factor in our study that limits interpretation is the relatively small numbers of patients within each of the sub-groups of patients having deletions, nonsense, splice and missense mutations. Group E ICIOR stage at presentation was noted in 40% of patients with large deletions and 33% with small deletions. However, 38.5% of patients with splice mutations and 44.4% of patients with missense mutations presented with Group B ICIOR. Twenty five percentages of eyes with large deletions had high risk features on histopathology and one patient among them developed metastasis. The tumor stages could have also been influenced by delayed presentation, however. More accurate data on age at onset, and comparison of cases with early presentation would enhance correlations. The current study despite its limitations, may be helpful in promoting further larger studies and provides a starting point for looking at correlations. We also feel the need for a larger more uniform sample size with longer follow up to identify the genetic factors that modify the relationship between genotype and phenotype in retinoblastoma.\",\n          \"Bruton\\u2019s tyrosine kinase (BTK), a member of the TEC family of kinases, plays a crucial role in B-cell maturation and mast cell activation. Although the structures of the unphosphorylated mouse BTK kinase domain and the unphosphorylated and phosphorylated kinase domains of human ITK are known, understanding the kinase selectivity profiles of BTK inhibitors has been hampered by the lack of availability of a high resolution, ligand-bound BTK structure. Here, we report the crystal structures of the human BTK kinase domain bound to either Dasatinib (BMS-354825) at 1.9 A\\u02da resolution or to 4-amino-5-(4-phenoxyphenyl)-7H-pyrrolospyrimidin- 7-yl-cyclopentane at 1.6 A\\u02da resolution. This data provides information relevant to the development of small molecule inhibitors targeting BTK and the TEC family of nonreceptor tyrosine kinases. Analysis of the structural differences between the TEC and Src families of kinases near the Trp-Glu-Ile motif in the N-terminal region of the kinase domain suggests a mechanism of regulation of the TEC family members. Keywords: Bruton\\u2019s tyrosine kinase; BTK; Dasatinib; Celera compound; TEC-family; crystal structure Introduction Bruton\\u2019s tyrosine kinase (BTK) is a member of the TEC kinase family, nonreceptor tyrosine kinases that play important roles in T-cell receptor-, B-cell receptor-, and Fcc-receptor-mediated signaling. BTK participates in signal transduction from B-cell antigen receptors resulting in phospholipase C-c2-mediated calcium mobilization1 which, in turn, affects pre-B-cell functional maturation and expansion. Since BTK is required for B-cell function, it is an important target for the potential treatment of inflammatory diseases that involve B-cell activation. Mutations in the human BTK gene are responsible for Abbreviations: BTK, Bruton\\u2019s tyrosine kinase; BTK-KD, BTK kinase domain; ITK, interleukin-2 inducible T-cell kinase; PH, pleckstrin homology; SH, Src homology; TEC, tyrosine kinase expressed in hepatocellular carcinoma; XLA, X-linked agammaglobulinemia. Disclosure: All authors (with the exception of M. J. Romanowski) are employees of Biogen Idec and own company stock. Michael J. Romanowski\\u2019s current address is Novartis Institutes for Biomedical Research, Cambridge, MA 02139. *Correspondence to: Laura F. Silvian, Biogen Idec, Inc., 12 Cambridge Center, Cambridge, MA 02142. E-mail: laura.silvian@biogenidec.com Published by Wiley-Blackwell. VC 2010 The Protein Society PROTEIN SCIENCE 2010 VOL 19:429\\u2014439 429 X-linked agammaglobulinemia (XLA), a male immunodeficiency that results in a deficit of mature B cells and serum immunoglobulin.2,3 Several compounds that inhibit BTK kinase activity in biochemical assays have been described in the literature and differ in their kinase selectivity profiles. One weak compound, LFM-A13 (a-cyano-bhydroxy-b-methyl-N-(2,5-dibromophenyl)-propenamide) is a BTK inhibitor with an IC50 of 2.5 lM in a biochemical assay, but also inhibits PLK3 and JAK2.4\\u20136 However, it was found to be somewhat specific for BTK, exhibiting 100-fold higher IC50 values for related tyrosine kinases such as JAK1, HCK, EGFR, and insulin-receptor kinase (IRK).7 Another compound, Dasatinib ([N-(2-chloro-6-methylphenyl)- 2-(6-(4-(2-hydroxyethyl)piperazin-1-yl)-2-methylpyrimidin-4-ylamino)thiazole-5-carboxamide] or BMS- 354825) [Fig. 1(A)], originally used to target BCRAbl, has been shown to bind to BTK with an IC50 of 5 nM8 but also binds to other kinases such as SRC family members (HCK, SRC, and CSK), and ephrin receptors, FGR, PDGFRa, and YES.9 BTK was identified as a target of Dasatinib through pull-down experiments in the CML cell line K562.8 The reversible Celera compound, 3-cyclopentyl-1-(4-phenoxyphenyl)-1H-pyrazolo pyrimidin-7-amine,10 was recently described by Pan et al. 11 as a potent inhibitor of unphosphorylated BTK (8.2 nM IC50 in a biochemical assay). However, it also inhibits Lck and Src with IC50 values of 2 and 70 nM, respectively.10 It is chemically similar to the commercially available 4-amino-5-(4- phenoxyphenyl)-7H-pyrrolo[2,3-d]pyrimidin-7-yl-cyclopentane [B43; Fig. 2(A)] described as a potent inhibitor of Lck.12 Finally, an irreversible inhibitor from Pharmacyclics11 is currently in Phase I for B-cell lymphomas. It is expected to bind irreversibly to Cys481 in the BTK kinase domain active site (0.72 nM IC50 in a biochemical assay) and its selectivity profile is better than the reversible binder because it exhibits greater selectivity against Lck, which lacks this cysteine (>1000-fold selectivity in a biochemical assay). Future design of potent, specific BTK inhibitors would be facilitated by the structures of these compounds bound to BTK, to discern whether there are regions surrounding the ligand that are unique to this kinase. BTK is composed of several domains: an N-terminal pleckstrin homology (PH) domain, a prolinerich TEC homology domain, two SRC homology domains (SH3 followed by SH2), and a C-terminal kinase domain (BTK-KD). Mutations in all domains of human BTK have been found to lead to XLA and missense mutations have been found in all domains except for the SH3 domain.13 Structures have been Figure 1. BTK-KD Y551E/Dasatinib crystal structure. A: Chemical structure of Dasatinib. B: Electron density (2Fo-Fc map at 1 sigma) for Dasatinib within a surface representation of the BTK protein in the human BTK-KD-Y551E/Dasatinib complex. C: Overall view of the BTK kinase domain bound to Dasatinib. Inhibitor (magenta), amino-terminal lobe (green), carboxy-terminal lobe (gray), C-helix (red), hinge (blue), and activation loop (yellow). D: Close-up of the active site and residues within 5 A\\u02da of the bound Dasatinib. The surface is colored according to the convention of base (blue) and acid (red). 430 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain solved for the kinase domains of apo-murine BTK7 and human ITK,14 but a high-resolution structure of a full-length protein with regulatory domains is not available. Low-resolution structures of BTK solved by small angle X-ray scattering have revealed an extended, linear arrangement of the SH3, SH2, and kinase domains, which contrasts with structures of autoinhibited full-length Src and Abl kinases in which a more compact arrangement of the SH2 and SH3 domains allows for the SH2 domain to bind near the C-terminal tail of the kinase domain.15 Structural studies of the Src family of tyrosine kinases have revealed that these proteins can adapt two conformations: an autoinhibitory state of the protein, referred to as an \\u2018\\u2018assembled regulatory domain\\u2019\\u2019 conformation, and an active, more open, structure, where the SH2 domain does not interact with the unphosphorylated C-terminal tail.16 Here, we describe the 1.94 A\\u02da resolution crystal structure of the human BTK-KD Y551E mutant bound to Dasatinib and a 1.6 A\\u02da resolution crystal structure of the unphosphorylated human BTK-KD bound to B43. We observe that the two structures differ in the orientation of the C-helix, similar to conformational changes observed in Src kinase family members that are locked into active or inactive states. Both BTK-KD structures reveal ordered density for the WEX motif at the N-terminus of the kinase domain, where X is a hydrophobic residue. The location of the tryptophan side chain at the base of the C-helix provides an explanation for how the WEX motif acts as an important regulatory element for the TEC family of kinases, similar to its role in regulation of the Src family of kinases, and suggests that the two families have a similar mechanism of regulation. Results Protein purification and characterization BTK-KD and BTK-KD Y551E were purified to  95% purity using a simple, three-step process utilizing two successive glutathione-Sepharose chromatography steps followed by size exclusion chromatography. Mass spectrometry indicated that the majority of the wild-type BTK-KD and BTK-KD Y551E was intact and unphosphorylated (wild-type calculated mass \\u00bc 32,641.6 Da, measured mass \\u00bc 32,643 Da; mutant calculated mass \\u00bc 32,607.5 Da, measured mass \\u00bc 32,609 Da), although 2 and 8%, respectively, were missing the first 4 N-terminal residues (wildtype calculated mass \\u00bc 32,317.2 Da, measured mass Figure 2. BTK-KD/B43 crystal structure. A: Chemical structure of B43. B: Electron density (2Fo-Fc map at 1 sigma) for B43 within a surface representation of the BTK protein in the human BTK-KD-B43 complex. C: Overall view of the BTK kinase domain bound to B43. Inhibitor (green), amino-terminal lobe (green), carboxy-terminal lobe (gray), C-helix (red), hinge (blue), and activation loop (yellow). D: Close-up of the active site and residues within 5 A\\u02da of the bound B43. The surface is colored according to the convention of base (blue) and acid (red). The active site cysteine is shown with a yellow surface. Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 431 \\u00bc 32,319 Da and mutant calculated mass \\u00bc 32,283.2 Da, measured mass \\u00bc 32,285 Da). Crystal structures The crystal structures of BTK-KD Y551E/Dasatinib and BTK-KD/B43 complexes were determined using 1.94 A\\u02da diffraction data (Rfree \\u00bc 25.8%) and 1.6 A\\u02da diffraction data (Rfree \\u00bc 23.1%), respectively (Table I). The electron density maps clearly revealed the positions of the ligands. Figure 1(B) depicts the electron density for Dasatinib and Figure 2(B) reveals the electron density for B43. The human BTK-KD-Y551E/Dasatinib complex structure exhibits a \\u2018\\u2018C-helix in,\\u2019\\u2019 active conformation in which the catalytic lysine forms H-bonds to Glu445 of the C-helix [Fig. 1(C,D)]. The bilobal fold of the BTK KD is similar to that reported for other tyrosine kinase structures: the N-terminal lobe (residues 392\\u2013473) consists of five antiparallel b-sheets and two a-helices and contains the glycine-rich loop that covers the active site. The C-terminal lobe (residues 482\\u2013659) consists of nine a-helices and two b-strands and the activation loop. Finally, the hinge region (residues 474\\u2013481) connects the two lobes of the kinase and contains the active site cysteine (Cys481). The ordered region of the Dasatinib cocrystal structure encompasses residues 392\\u2013658 but electron density for parts of the activation loop (residues 542\\u2013558) and residues 435\\u2013441 is not visible. However, the DFG-portion of the activation loop is visible and is in the so-called \\u2018\\u2018DFG-in\\u2019\\u2019 conformation. The unphosphorylated BTK-KD/B43 cocrystal structure, in contrast, displays a \\u2018\\u2018C-helix-out\\u2019\\u2019 inactive conformation [Figs. 2(C) and 3(B)]. Most of the kinase is ordered except the tip of the glycine rich loop (residues 410\\u2013414). The activation loop displays a \\u2018\\u2018DFG-in\\u2019\\u2019 conformation and is completely ordered. It is composed of two alpha helices, in which Arg544 within the first helix of the activation loop forms a salt bridge to Glu445 of the C-helix. Dasatinib interactions Dasatinib makes several hydrogen bonds to the hinge and occupies a hydrophobic pocket behind the Thr474 gatekeeper residue [Fig. 1(B,D)]; similar to the previously reported structures of Dasatinibbound Abl (PDB ID: 2GQG),17 Lyn (PDB ID: 2ZVA),18 and cSrc (PDB ID: 3G5D).19 Its H-bond interactions to the hinge are described in Figure 1(D), including interactions with the backbone carbonyl and amide of 477, water-mediated interactions with the Tyr476 side chain, and an interaction between the Thr474 side chain and the compound amide nitrogen. The glycine-rich loop of BTK-KD curls toward Dasatinib to form a hydrophobic interaction with Phe413 [Fig. 1(D)]. The ortho-chloro, methyl phenyl substituent of Dasatinib is twisted to enter a hydrophobic pocket, composed of both hydrophobic and hydrophilic residues Met449, Val458, Leu460, Ile472, Lys430, Glu445, and Ser538. Finally, Glu445 of the C-helix forms a salt bridge with the catalytic Lys430; the epsilon amino group of this lysine is poised to make a pi-cation interaction with the Dasatinib ortho-chloro, methyl phenyl ring [Fig. 1(D)]. B43 interactions The B43 compound makes several hydrogen bonds to the hinge and occupies a hydrophobic pocket behind the Thr474 gatekeeper residue [Fig. 2(D)]. The 4-amino pyrrolopyrimidine of B43 occupies the position of the adenine ring of ATP and the cyclopentyl ring occupies the space generally occupied by the ATP ribose in typical protein kinase structures. The 4-amino pyrrolopyrimidine makes many interactions with the hinge; the exocyclic amine directly interacts Table I. Data Collection and Refinement Statistics BTK-KD/B43 complex BTK-KD Y551E/Dasatinib complex Data collection Space group P21212 P21212 Cell dimensions a (A\\u02da ) 72.5 73.4 b (A\\u02da ) 104.3 104.6 c (A\\u02da ) 38.0 38.1 Wavelength (A\\u02da ) 1.54 0.98 Resolution (A\\u02da ) 50\\u20131.6 50\\u20131.94 Rsym (%)a 8.0 (36.3)a 8.3 (61.9) I/r(I) a 18.5 (4.6)a 32.1 (3.3) Multiplicity 3.8 (4.1)a 6.8 (6.6) Total no. reflections/no. unique reflections 405,732/39,020 490,862/22,496 Completeness (%)a 98.1 (98.5)a 99.2 (99.2) Refinement Resolution (A\\u02da ) 24.5\\u20131.6 28.4\\u20131.94 No. reflections 36,293 21,141 Rwork (Rfree) b (%) 19.2 (23.1) 19.9 (25.8) No. molecules per asymmetric unit 1 1 No. atoms Protein 2161 2016 Ligand 28 33 Water 495 164 Average B-factors (A\\u02da c ) Protein atoms 25.8 23.4 Ligand 30.5 41.7 Water 53.5 35.7 R.m.s.d Bond lengths (A\\u02da ) 0.007 0.024 Bond angles () 1.14 2.05 Ramachandron plot % Allowed 98.5 98.2 % Generously allowed 1.5 1.4 % Not allowed 0 0.4 Rsym \\u00bc R|Ihkl  <Ihkl>|/RIhkl. a The value in parentheses is for the highest resolution bin (approximate interval, 0.1 A\\u02da ). b Rwork \\u00bc RIhkl| |Fo  |Fc| |/RIhkl|Fo| for all data except 5%, which is used for the Rfree calculation. 432 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain with the gatekeeper Thr474 hydroxyl and the backbone carbonyl of 475, the N-3 of the pyrimidine accepts a hydrogen bond from the backbone amine of Met477, and the N-1 of the pyrimidine forms a water-mediated hydrogen bond network to the hydroxyl of Tyr476 and the backbone carbonyl of Ala 478 (Fig. 2). The distal phenyl group of the phenoxyphenyl is twisted 38 out of plane of the phenylether, such that it enters a hydrophobic pocket composed of only hydrophobic residues, Phe442, Met449, Leu460, Ile472, Phe540, and Leu542, and is in proximity to Asp539 of the \\u2018\\u2018DFG\\u2019\\u2019 motif. Phe540 of this motif forms a face-to-edge pi-stacking interaction with the phenoxyphenyl group of B43. Asp539 of the DFG-motif forms a salt bridge with the catalytic Lys430 but does not form direct hydrogen bond interactions with the compound. Discussion Relevance of structures for drug discovery The structures of the human BTK-KD Y551E/Dasatinib and BTK-KD/B43 complexes we report here differ from the publicly available structure of apo-murine BTK-KD (PDB ID: 1K2P) and are arguably more relevant for drug discovery for diseases in which inhibition of BTK may be desired. When the apo-mouse BTK structure is superimposed on the human BTK-KD/B43 structure (RMSD of 1.19 A\\u02da over 230 aligned a-carbons), the biggest differences are observed in the activation loop and in the glycine-rich loop. The activation loop of the mouse apoBTK-KD structure adapts an extended configuration with Tyr551 pointed toward solvent. In the mouse apo-BTK structure, the glycine loop also caves into the active site and occludes the ATP binding pocket. Because the mouse and human BTK-KD\\u2019s are 98.3% identical, and only four amino acids are replaced in the mouse sequence (Q379K, K433R, K625R, and T653S), it is likely that the kinase domain flexibility observed in the apo-murine BTK-KD structure is due to a lack of occupancy of a compound in the active site, rather than due to an intrinsic structural difference between the mouse and human species. Selectivity of compound inhibition For both Dasatinib and the reversible Celera compound, the size and hydrogen bonding nature of the gatekeeper residue of a given kinase generally correlates with its degree of biochemical inhibition9 (Table II). Most of the kinases that are inhibited by 10 lM Dasatinib with a Kd < 1 nM, or that are inhibited by 10 lM Celera compound with less than 5% residual activity, have a threonine gatekeeper. A valine residue in this gatekeeper position is tolerated for the Celera compound binding, but is not as well tolerated for Dasatinib binding to the Ret and KDR kinases (Table II). Because the threonine gatekeeper forms H-bond interactions with both compounds, it is possible that the H-bonding binding energy plays a greater role in binding Dasatinib compared to the Celera compound. An alternative explanation for the poor binding of Dasatinib to valine gatekeeper containing kinases KDR and Ret is that there are differences in side chains within 5 A\\u02da of the compound (Table III). In particular, one residue in the back pocket that forms close hydrophobic interactions with Dasatinib in BTK is Met449, which is replaced by a leucine in KDR and Ret. Because the back pocket in the Dasatinib cocrystal structure Figure 3. Differences between the \\u2018\\u2018inactive\\u2019\\u2019 B43 and \\u2018\\u2018active\\u2019\\u2019 Dasatinib-bound conformation of BTK. A: Overlay of BTK-KD bound to B43 (green), ITK bound to staurosporin (magenta, PDB ID: 1SNU) and P38a (yellow, PDB ID: 1P38). This indicates that the larger phenylalanine gatekeeper of ITK compared to the Thr474 gatekeeper of BTK prevents B43 from binding. It also indicates that the shorter hinge of P38a blocks the B43 binding site. B: Overlay of B43 structure (green) and Dasatinib structure (magenta) indicates that they superimpose well near the hinge but differ near the C-helix, resulting in a shift of >10 A\\u02da in the conformation of the Glu445 side chain and the residues that surround the hydrophobic back pocket. Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 433 is composed of mixed hydrophobic and hydrophilic residues, Dasatinib may have a greater reliance on Met449 compared to B43, whose back pocket is completely surrounded by hydrophobic residues. Either explanation (i.e., different energy considerations for H-bonding to a threonine gatekeeper for different compounds or differences in the makeup of side chains that line Dasatinib\\u2019s unique hydrophobic pocket), could explain why Dasatinib does not bind as well to Ret and KDR. The exception to the rule of requiring a small gatekeeper for compound binding is p38a, EGFR, and NIMA (never in mitosis gene A)-related kinase 11 (NEK11) kinases, which have threonine gatekeepers, but are only moderately inhibited by both small molecules. P38a kinase has a shorter hinge, and thus its reduced affinity can be ascribed to a smaller binding site [Fig. 3(A)]. Similarly, there are differences in the other residues within 5 A\\u02da of the two small molecules (Table III), which could account for the differences in affinity for NEK11 and EGFR. Both differ in the residues that make up the hydrophobic pocket behind the threonine gatekeeper, with EGFR substituting a cysteine for Val458 and NEK11 substituting a larger phenylalanine for Leu460 and a leucine for Met449. Furthermore, NEK11 contains a glycine substitution for Ser538, a residue which is likely to enhance the flexibility of the DFG motif which follows (Table III). In contrast, Abl, Lyn, and Src, kinases which are inhibited well by both small molecules, show no significant variation in the residues that make up the hydrophobic pocket. In these kinases Ser538 is replaced with alanine, a residue with similar conformational flexibility, and is not likely to effect the flexibility of the activation loop (Tables II and III). As the size of the gatekeeper residue increases from threonine to phenylalanine [as in ITK, Fig. 3(A)], methionine (as in Igf1R, JAK1, JNK3, and SYK) (Table II), or leucine (as in Aurora A) (Table II) neither inhibitor reduces kinase activity to the full extent because both binding modes require insertion of the compound past the bulky gatekeeper into the hydrophobic back pocket. Table II. Gatekeeper Residue in Various Kinases, % Activity for Celera Compound and Kd Values for Dasatinib Kinase Gatekeeper % Activity in the presence of 10 lM Celeraa Dasatinib Kd (nM) b Ack1 Ser 3 6 Abl Thr 3 0.5 Bmx Thr 1 0.0 BTK Thr 1 1.0 CSK Thr 0 0.2 EGFR Thr 20 100 FGR Thr 3 0.5 FYN Thr 3 0.7 Hck Thr 5 0.3 Kit Thr 4 0.6 Lck Thr 13 0.2 Lyn Thr 4 0.6 P38a Thr 57 30 Src Thr 1 0.2 Yes Thr 1 0.3 SIK Thr 17 ND RIPK2 Thr 4 30 cRAF Thr 11 600 EphB1 Thr 65 0.4 NEK11 Thr 64 7000 RET Val 0 200 KDR/VEGFR2 Val 6 3000 RSK1 Leu 80 ND RSK2 Leu 83 ND AurA Leu 101 2000 IGF1R Met 87 ND JAK2 Met 88 ND JNK3 Met 76 ND Syk Met 50 3000 Itk Phe 18 ND ND, not determined. a Upstate kinase profiler. b Taken from Ref. 9. Table III. Residues Within 5 A\\u02da of Compounds in the BTK-KD/B43 and BTK-KD Y551E/Dasatinib Structures NTD GRL UPH LWH HYP AVL 428b 430c 408b 409d 413d 416b 472d 474a,c 475b 476b 477b 479d 480b 481c 458c 460b 449b 538b 539b 540b 542b BTK A K L G F V I T E Y M N G C V L M S D F L NEK11 Ve K L G F V I T E YCe Ge Re De V Fe Le Ge DFVe EGRF A K L G F V Le T Qe Le M Fe G CCe LMTe DFL ABL A K L G Ye V I T E FMYe G Ne V L MAe DFL LYN A K L G F V I T E Y M Ke G Se V L MAe DFL SRC A K L G F V I T E Y M Ke G Se V L MAe DFL RET A K L G F V Le Ve E YAe Ye G Se I e L Le SDFL KDR A K L G F V Ve Ve E Fe Ce Fe G Ne VLLe SDFL NTD, N-terminal domain; GRL, glycine rich loop; UPH, upper hinge; LWH, lower hinge; HYP, hydrophobic pocket; AVL, activation loop. a Gatekeeper residue. b Residue interacts with both B43 and Dasatinib in BTK. c Residue only interacts with Dasatinib in BTK. d Residue only interacts with B43 compound in BTK. e Differences in sequence from BTK sequence. 434 PROTEINSCIENCE.ORG Crystal Structures of Human BTK Kinase Domain Overall conformational changes In the BTK-KD Y551E/Dasatinib complex structure, the Gly-rich loop of BTK partially shields Dasatinib from solvent, as found in the Abl kinase structure (data not shown). Because of the curled-in glycinerich loop in the BTK structure, a hydrophobic interaction is formed between Phe413 and the exocyclic methyl of the Dasatinib pyridine ring [Fig. 1(D)], and between Gln412 and Asp539 of the \\u2018\\u2018DFG\\u2019\\u2019 motif. In every other respect, however, the overall conformation of the BTK-KD/Dasatinib structure is similar to the known Dasatinib-bound conformations of Lyn,18 cSrc,19 and Abl,17 including the compound\\u2019s H-bond interactions with the hinge and the position of its C-helix. The nonvisible residues differ between the two BTK structures, which can be correlated with the different interactions made with different compounds or by the mutation in the activation loop. The tip of the glycine-rich loop is disordered in the B43 structure but is ordered in the Dasatinib structure, while the activation loop is disordered in the Dasatinib structure but is well ordered in the B43 structure. The curled glycine-rich loop forms van der Waals contacts with Dasatinib, between a methyl substituent and the Phe413 side chain, whereas no similar interactions take place between B43 and the BTK-KD protein residues and the glycine-rich loop is disordered. One might expect that the Y551E mutation in the activation loop of the BTK-KD Y551E/Dasatinib structure is responsible for the activation loop disorder; the mutated residue is electrostatically incompatible with the conformation of the activation loop seen in the BTK-KD/B43 structure. In the BTK-KD/B43 structure, the Y551 residue is in close proximity to an Asp521 side chain; this is likely to be electrostatically repelled by mutation of the tyrosine to a glutamate. While it is often hard to pinpoint why flexible regions of crystal structures are disordered, it appears that formation of critical molecular interactions produces ordered electron density for the more flexible regions of BTK. Comparison of the structures of the human BTK-KD-Y551E/Dasatinib complex and the BTKKD/B43 complex reveals a change of conformation from catalytically \\u2018\\u2018active\\u2019\\u2019 to \\u2018\\u2018inactive.\\u2019\\u2019 The Dasatinib complex is more similar to the ATP-bound conformation of most kinases, in which a conserved glutamate from the C-helix forms a salt bridge to the catalytic lysine. In fact, no crystals could be formed with the unphosphorylated, wild-type BTK kinase construct, prompting us to make the Y551E mutant as a mimic of the phosphorylated wild-type protein. In contrast, the BTK-KD/B43 complex shows an outward shift of the C-helix [Fig. 3(B)] relative to its position in the Dasatinib structure, the conserved salt bridge from the glutamate to the catalytic lysine breaks, and a large hydrophobic pocket opens behind the gatekeeper residue. The ability of different kinases to adapt a C-helix out conformation might enable the design of specific inhibitors that targets this larger hydrophobic pocket. Furthermore, Cys481 in the active site of BTK-KD could also be exploited to gain kinase selectivity in which a small molecule may be irreversibly bound to this cysteine via a covalent bond.11 To determine the overall similarity of the BTKKD/B43 structure to other kinases, the B43 complex structure was submitted to the Dali-lite server for structure alignment and scoring20 (http://ekhidna. biocenter.helsinki.fi/dali_server/). The top hits, inactive Hck (PDB ID: 1QCF), inactive SRC (PDB ID: 2SRC), inactive ABL (PDB ID: 2G1T), ITK (PDB ID: 1SM2), and mouse BTK (PDB ID: 1K2P), could be aligned with the human BTK over more than 260 a-carbons and with an rmsd of 2.0 A\\u02da or better. The highest scoring hits, excluding the TEC family of kinases, were all inactive conformations of tyrosine kinases from the Src and Abl families, consistent with their overall sequence similarities to human BTK. The conformation of the activation loop and C-helix in the human BTK-KD/B43 structure is very similar to the inactive Src structure (PDB ID: 2SRC) with an rmsd 1.64 A\\u02da over 257 a-carbons; in Src the activation loop forms two alpha helices and occludes access of the substrate peptide. The overall conformation of the BTK-KD Y551E/Dasatinib structure is similar to the active c-Src (CSK) structure (PDB ID: 1Y57) where the activation loop is swung out and the C-helix moves toward the active site (Fig. 4). Comparison of Src and TEC family kinases The phosphorylation-triggered regulation of BTK and Src differ. Unlike the Src family, the TEC family of nonreceptor tyrosine kinases lacks a conserved tyrosine in the C-terminus that could be phosphorylated to then bind to the SH2 domain.16 BTK is regulated by the phosphorylation of two tyrosine residues, Tyr223 in the SH3 domain and Tyr551 in the activation loop of the kinase domain, both of which participate in kinase activation.21 In a recent study of BTK autophosphorylation, the Y551F mutant was shown to have a 5 to 10-fold lower enzymatic activity than the wild-type protein, indicating that this tyrosine plays an important role in BTK activation.22 Furthermore, mutation of a conserved tryptophan in the N-terminal W-E-X motif, in which X is a hydrophobic residue, also appears to effect the activities of the two kinase families differently. In Src, mutation of the Trp to Ala increases kinase activity while in BTK, mutation of the Trp to Ala reduces kinase activity.23 The human BTK structures described here include ordered density for the WEI motif (residues 392\\u2013395), an area which was disordered in the Marcotte et al. PROTEIN SCIENCE VOL 19:429\\u2014439 435 available murine BTK structure (PDB ID: 1K2P) and the human ITK structure (PDB ID: 1SM2). This enables a structural comparison of the TEC family and the Src family kinases in this conserved region. The Trp side chain shifts from being solvent exposed in the \\u2018\\u2018inactive\\u2019\\u2019 BTK-KD/B43 complex structure to being wedged into a pocket behind the inward C-helix in the \\u2018\\u2018active\\u2019\\u2019 BTK-KD Y551E/Dasatinib complex structure [Fig. 4(A,B)]. A structural superposition of the two BTK structures with the inactive SRC (PDB ID: 2SRC) 24 and an active CSK structure (PDB ID: 1Y57) 16 show that the side chain of Trp395 superimposes in the active structures of both kinase families. In the \\u2018\\u2018inactive\\u2019\\u2019 conformations, a lysine or methionine side chain from the rotated C-helix sterically occludes the tryptophan side chain, and the Trp side chains are not superimposable [Fig. 4(B)]. The similar structural shift observed suggests that the Src and Tec families utilize a similar means of activating the conformation of the kinase domain. However, the inactive conformations of the Trp side chain differ and the differential energy required in switching from inactive to active conformations may differ; this could explain why mutation of this residue to alanine produces different regulatory effects on the two proteins. This study reports the structures of human BTKKD with Dasatinib or 4-amino-5-(4-phenoxyphenyl)- 7H-pyrrolo[2,3-d]pyrimidin-7-yl-cyclopentane bound at the active site. Analysis of the binding modes reveals that the size and H-bonding potential of the gatekeeper residue in different kinases and the nature of the residues that make up the hydrophobic pocket behind the gatekeeper residue are critical in dictating if either compound binds. We propose that the presence of a distinct hydrophobic pocket in the B43 structure suggests that one could design compounds with limited selectivity for kinases that can adapt the \\u2018\\u2018C-helix-out\\u2019\\u2019 conformation. Movements in the C-helix, presumably induced by phosphorylation of the activation loop residue Tyr551, can shift BTK into an active conformation and could generate a second metal binding site containing Glu445 that Lin et al. have shown is important for optimal catalytic activity of the Tyr551-phosphorylated BTK.25 Finally, we show that, like the Src-family of kinases, BTK can adapt a similar conformational rotamer of Trp395 in its \\u2018\\u2018active\\u2019\\u2019 conformation, which is correlated with a similar movement of the C-helix. With a growing recognition that BTK plays a key role in many B cell lymphomas and autoimmune diseases, these structures will assist with selective drug design.  The Tec family kinases are tyrosine kinases that function primarily in hematopoietic cells. The catalytic activity of the Tec kinases is positively influenced by the regulatory domains outside of the kinase domain. The current lack of a full-length Tec kinase structure leaves a void in our understanding of how these positive regulatory signals are transmitted to the kinase domain. Recently, a conserved structure within kinases, the \\u2018regulatory spine\\u2019, has been identified that assembles and disassembles as a kinase switches between its active and inactive states. Here we define the residues that comprise the regulatory spine within Tec kinases. Compared to previously characterized systems, the Tec kinases contain an extended regulatory spine that includes a conserved methionine within the C-helix and a conserved tryptophan within the SH2-kinase linker of Tec kinases. This extended regulatory spine forms a conduit for transmitting the presence of the regulatory domains of Tec kinases to the catalytic domain. We further show that mutation of the gatekeeper residue at the edge of the regulatory spine stabilizes the regulatory spine resulting in a constitutively active kinase domain. Importantly, the regulatory spine is preassembled in this gatekeeper mutant rendering phosphorylation on the activation loop unnecessary for its activity. Moreover, we show that the disruption of the conserved electrostatic interaction between Btk R544 on the activation loop and Btk E445 on the C-helix also aids in the assembly of the regulatory spine. Thus, the extended regulatory spine is a key structure that is critical for maintaining the activity of Tec kinases.  Keywords: Itk and Btk, regulatory spine, kinase activation, gatekeeper residue, phosphorylation Go to: INTRODUCTION Protein kinases catalyze the transfer of a phosphate group from ATP to a hydroxyl containing amino acid side chain; either Ser/Thr for serine-threonine kinases or Tyr for tyrosine kinases[1]. The activity of protein kinases is exquisitely regulated within the cell[1; 2; 3; 4]. The switch from the catalytically inactive state of a kinase to the active state is often accompanied by the phosphorylation of a key residue within the activation loop, a large flexible loop that lies between the two lobes of the kinase domain structure[1]. Phosphorylation of the activation loop residue can trigger concerted movements in other mobile elements within the kinase domain such as the C-helix and the DFG motif that brings about the assembly of the catalytically critical residues in the kinase active state[1]. The active state conformations associated with different kinases are nearly identical across the many kinases for which high resolution structures have been solved[2; 3; 5; 6; 7]. It is not surprising that the active states of distinct kinases are very similar since the phospho-transfer chemistry carried out by different kinases is the same. In contrast, it is becoming clear that the structural features associated with the inactive state of various kinases differ widely.  The search for features that are conserved within the structures of active kinases has recently led to the identification of a structure termed the \\u2018regulatory spine\\u2019[8; 9]. The regulatory spine was first identified by Local Spatial Patterns alignment analysis using a set of serine-threonine and tyrosine kinase structures that included PKA as a model kinase[8; 9]. The regulatory spine defines a stretch of amino acid residues that is assembled only in the active state of kinases. Unlike consensus sequences that consist of a continuous stretch of amino acids in the primary structure, the regulatory spine consists of disparate residues that span the N- and C-terminal lobes of the kinase domain. Assembly of the regulatory spine has been proposed to be a crucial step in the activation of protein kinases and this structure is disrupted in structures of kinases in the inactive state. Moreover, the regulatory spine has been proposed as the mechanism by which allosteric effects can be propagated to the kinase active site[8; 9].  The Tec kinases are immunologically related tyrosine kinases which consist of five mammalian members: Itk, Btk, Tec, Txk and Bmx[10; 11]. Tec kinases share similar domain architecture as Src, Abl and Csk family of kinases, in that they have the SH3-SH2-kinase domain cassette[10]. Despite domain similarities, we and others have shown that the regulation of Tec kinases is distinct from Src and Abl kinases[10; 12; 13; 14]. Unlike the Src and Abl kinases, the N-terminal regulatory domains of Tec kinases positively regulate the activity of the kinase domain[14]. While the isolated kinase domains of Src and Abl are active, the isolated kinase domains of the Tec family exhibit poor catalytic activity[13; 14; 15]. Regulation of the Tec kinases is in fact more similar to the Csk family of enzymes, whose N-terminal regulatory domains are essential for the catalytic activity of the kinase domain[16]. Moreover, while the structures of full-length Src and Csk kinases are available, the structure of a full-length Tec kinase remains elusive[2; 17; 18]. This leaves a significant gap in our mechanistic understanding of the regulation of the Tec family kinases.  In this manuscript, we define the regulatory spine within the Tec family kinases and show that mutation of the spine residues leads to different effects in the context of the isolated kinase domain verses full-length Itk. Moreover, we show that the assembly of a stable regulatory spine within members of the Tec kinase family is critically dependent on the presence of the SH2-kinase linker region. The residues that comprise the regulatory spine within PKA are insufficient to promote the assembly of a stable regulatory spine within the Tec kinases. We extend the Tec regulatory spine to include a conserved methionine within the C-helix and a conserved tryptophan residue within the SH2-kinase linker of Tec kinases. Together with the spine residues originally identified within the PKA kinase domain, the conserved methionine and tryptophan residues form a continuous structure that links the kinase active site to the N-terminal regulatory domains of Tec kinases. Furthermore, we show that the stabilization of the regulatory spine by the gatekeeper threonine to methionine mutation eliminates the need for phosphorylation on the activation loop for Btk kinase activity. Thus, the extended regulatory spine is a structure that is critical for the regulation of Tec kinases and if stabilized appropriately, is sufficient for activation in the absence of activation loop tyrosine phosphorylation and in the absence of the non-catalytic Tec regulatory domains.  Go to: RESULTS The regulatory spine controls Tec kinase activity  The regulatory spine within PKA consists of five residues: L95, L106, Y164, F185, and D220[9]. These residues are spread throughout the primary sequence of the PKA kinase domain: L95 is located on the C-helix, L106 is on the N-terminus of the \\u03b24 strand, Y164 is part of the \\u2018HRD\\u2019 motif, F185 is from the \\u2018DFG\\u2019 motif and D220 is on the F-helix within the C-terminal lobe[9]. Alignment of the structure of the Btk and Itk kinase domains with that of PKA shows that the corresponding regulatory spine residues within the Btk and Itk kinase domains should consist of: Btk M449, L460, H519, F540 and D579 and Itk M409, L420, H479, F500 and D539 respectively (Fig. 1a).  Figure 1 Figure 1 Identification of the regulatory spine in Tec kinases To test the role of the predicted spine residues in the Tec kinases, we wished to take advantage of the rapid bacterial expression and purification system that has been developed for the isolated kinase domains of Itk and Btk[19]. One issue that arises, however, is the fact that the isolated kinase domains of the Tec kinases exhibit poor catalytic activity and so the expected loss of function mutations in the regulatory spine would be difficult to characterize given the already low activity of the wild type kinase domains. In order to take advantage of the ease of the bacterial expression system, we needed an isolated Btk and Itk kinase domain with higher catalytic activity. It has been demonstrated previously that introduction of a bulky hydrophobic residue such as isoleucine or methionine at the gatekeeper position activates multiple kinases[20]. Indeed, we find that the Itk F434M and Btk T474M isolated kinase domains are more active when compared to wild-type isolated kinase domain (manuscript in preparation). We therefore probed the importance of the predicted regulatory spine residues in Itk and Btk by mutating them individually to alanine in the context of the Itk F434M and Btk T474M isolated kinase domain mutants. Activity measurements are carried out by monitoring the phosphorylation levels of a peptide substrate in a radioactive assay or by detecting the level of autophosphorylation on the activation loop tyrosine (Y511 in Itk and Y551 in Btk) by western immunoblotting. While phosphorylation of the activation loop tyrosine is achieved by the activity of the Src family kinases Lck and Lyn respectively in vivo, under in vitro conditions Tec kinases autophosphorylate on the activation loop tyrosine[21; 22; 23].  Disruption of the regulatory spine residues by mutation to alanine is predicted to disrupt kinase activity. Indeed, mutation of Btk M449, H519, F540 and D579 and Itk M409, H479, F500 and D539 to alanine within the isolated kinase domain of Itk F434M and Btk T474M leads to decreased phosphorylation on the activation loop tyrosine and drastically reduces the catalytic activity (Fig. 1b, c, d and e). Itk kinase residues such as M409, F500 and H479 are highly conserved with other kinases with well-defined roles in catalysis[8]. Itk M409 is involved in binding the substrate ATP[12]. Itk F500 is part of the \\u2018DFG\\u2019 motif at the start of the activation loop segment and is responsible for stabilizing the conformation of the preceding aspartate, and the C-helix[8]. Itk H479 is part of the \\u2018HRD\\u2019 motif where it serves as a scaffold for D499 and F500[8]. It is therefore not surprising that mutation at these sites leads to a loss in activity. However, Itk D539 is not part of any previously characterized regulatory motif and therefore the drastic reduction in Itk activity upon mutation of this residue to alanine highlights the importance of the regulatory spine in regulating Itk activity. Moreover, since the regulatory spine residues are conserved within the Tec kinase family, this structure would be predicted to be critical for regulating the activity of all Tec kinase family members.  Unexpectedly, mutation of Itk L420 and Btk L460 to alanine, which based on the previous PKA work is predicted to be part of the regulatory spine, failed to inactivate the isolated kinase domains of Itk F434M or Btk T474M (Fig. 1b, c, d and e). In fact, the isolated kinase domains of both Itk L420A/F434M and Btk L460A/T474M double mutants showed an increase in activity when compared to the isolated kinase domain of Itk F434M and Btk T474M, respectively. To ensure that the activating effect of the Leu to Ala mutation was not an artifact of working with the activated (gatekeeper mutant) isolated kinase domains of Itk and Btk, the regulatory spine residues were mutated to alanine in the context of wild-type isolated kinase domains of Itk and Btk. Since the catalytic activity of the isolated kinase domain of Tec kinases is generally poor, the detection of activation loop phosphorylation levels by western immunoblotting is difficult and for Itk, in particular, phosphorylation Y511 is below the detection limit by western immunoblotting. Hence, the activity of the wild-type and regulatory spine mutants of the isolated kinase domains of Itk and Btk are monitored by their ability to phosphorylate a peptide substrate in the radioactive assay.  The results of the regulatory spine mutations in the context of the wild type kinase domains mirror the results for the activating gatekeeper mutants described above. Mutation of Btk M449, H519, F540 and D579 to alanine within the isolated wild type kinase domain of Btk leads to decreased catalytic activity (Fig. 1f). Circular dichroism spectra of the Btk mutants overlay well with that of wild-type Btk showing that the decreased activity of the Btk mutants are not due to unfolding of the kinase (Supp. Fig. 1). Since the activity of the isolated Itk wild-type kinase domain is not significantly above background levels, the resulting activity of the Itk spine mutants: Itk M409A, H479A, F500A and D539A can not be interpreted except to say activity is no greater than wild type (Fig. 1g). However, consistent with our earlier results with the gatekeeper mutant, the Itk L420A and Btk L460A mutants are both more active than wild-type isolated kinase domain of Itk and Btk, respectively (Fig. 1f and g). Interestingly, previous studies on the regulatory spine within the Abl kinase have shown that mutation of Abl L320 to glycine (corresponding to Itk L420 and Btk L460), leads to only a slight decrease in kinase activity of full length Abl[20]. This is also consistent with modeling studies of Abl that have shown that Abl L320 has a modest impact on regulatory spine formation[20]. Thus, Itk L420 and Btk L460 do not play a major role in the assembly of the regulatory spine within the context of the isolated kinase domain of Itk and Btk.  The regulatory spine is not assembled in the structure of phosphorylated Itk kinase domain  Activation of Tec kinases requires the phosphorylation of a conserved tyrosine within the activation loop[12; 21; 23]. Separate high-resolution structures of the Btk kinase domain with Y551 either phosphorylated (active) or unphosphorylated (inactive) have recently been reported[24]. A comparison of the inactive and active states of Btk kinase domain shows clear differences in the region of the regulatory spine (Fig. 2a). The regulatory spine residues assemble into a linear arrangement in the structure of the active Btk kinase, while interactions between the regulatory spine residues (in particular L460, M449 and F540) are disrupted in the structure of inactive Btk kinase. The structural differences in this region between active and inactive Btk are consistent with the role of the regulatory spine as defined previously for PKA (Fig. 2a).  Figure 2 Figure 2 The regulatory spine is assembled only in the active state of kinases Crystal structures of the isolated kinase domain of Itk that is either unphosphorylated or phosphorylated on Y511 in the activation loop are also available[12]. Surprisingly, the available Itk kinase domain structures overlay quite well, with little or no conformational differences between them regardless of the phosphorylation state of the activation loop. Inspection of the Itk kinase domain structures shows that the regulatory spine residues in the Itk kinase domain structures adopt a disrupted configuration similar to the inactive Btk kinase domain (Fig. 2b). An additional hallmark of an active kinase domain structure is the formation of a crucial ion pair between a conserved lysine positioned within the \\u03b23 strand of the kinase domain and a conserved glutamate on the C-helix[1]. The distance between the corresponding residues (Itk K390 and E405) within the structures of both phosphorylated and unphosphorylated Itk kinase domain is on the order of 8.0 \\u00c5 which is significantly greater than the 3 to 4 \\u00c5 distance that is observed in the structures of other active kinases (Fig. 2b)[12]. Indeed, the distance between the same conserved ion pair within Btk, (Btk K430 and E445) changes from 14.6 to 3.8 \\u00c5 upon activation of the kinase (Fig. 2a). These observations suggest that both the phosphorylated and unphosphorylated Itk kinase domains adopt a conformation consistent with the \\u2018inactive\\u2019 enzyme. The inability of the Itk kinase domain to fully assemble into an active conformation, despite being phosphorylated on Y511 in the Itk activation loop, points to additional requirements for activation.  A notable difference between the Itk and Btk crystal structures is that the construct used for the crystallization of the Btk kinase domain included the SH2-kinase linker, while that of the Itk kinase did not (Fig. 2a & b)[12; 24]. We hypothesize that the absence of the SH2-kinase linker within the Itk construct used for crystallography might prevent the assembly of a stable regulatory spine in the kinase domain despite phosphorylation on the activation loop Y511.  Extension of the Tec kinase regulatory spine  We have previously shown that the SH2-kinase linker, the 17 amino acids between the SH2 domain and the kinase domain, is critical for the activity of both Btk and Itk, exerting a positive effect on the catalytic function of the kinase domain[14]. More specifically, a conserved tryprophan (Itk W355, Btk W395) in the SH2-kinase linker, as well as a methionine residue in the C helix (Itk M410, Btk M450) are crucial residues in the Tec family kinase regulatory apparatus (Fig. 3a)[14]. Like the regulatory spine mutants described in Figure 1, point mutations of Itk W355, Btk W395 or Itk M410 to alanine all result in a significant drop in catalytic activity (Fig. 3b). Examination of the crystal structures of Btk shows that Btk W395 and Btk M450 are located at the \\u2018top\\u2019 of the regulatory spine and, in the active Btk structure, serve to extend the hydrophobic packing of the spine residues well into the N-terminus of the kinase domain (Fig. 3c & d). In the inactive Btk structure, repositioning of the W395 and M450 side chains disrupts the extended regulatory spine (Fig. 3c & d) in much the same manner that repositioning of M449 in the inactive Btk structure disrupts the core of the spine (Fig. 2a). The active and inactive Btk structures consist of the same amino acids (residues 382\\u2013659) and differ with respect to the activation loop tyrosine (Y551). In the active structure the activation loop tyrosine is mutated to glutamate (Y551E) to mimic phosphorylated Btk, and for the inactive structure wild type Btk in the unphosphorylated form was used for crystallization[24].  Figure 3 Figure 3 Extension of the regulatory spine within Tec kinases to include the SH2-kinase linker Compared to the regulatory spine that has been defined for PKA[9], the Tec kinase family requires at least two additional residues (Itk W355 & M410) to fully assemble the regulatory spine. The tryptophan residue is located in the SH2-kinase linker region providing an explanation for the positive regulatory role of this non-catalytic region[14]. The available Itk kinase domain structures are entirely consistent with this finding; without the contribution of the SH2-linker region, the regulatory spine of Itk does not assemble into the active conformation. As a result, the structures of the phosphorylated and unphosphorlated Itk kinase domain fragments are very similar to each other and neither resembles an active kinase. Future structures of Tec kinases with and without the SH2-kinase linker in various activation (phosphorylation) states will be required to fully probe this hypothesis.  Itk L420 and Btk L460 are essential for the formation of the extended regulatory spine within full-length Tec kinases  The conserved tryptophan in the SH2-kinase linker is critical for the activity of full-length Tec kinases[14]. Upon activation, the tryptophan side-chain forms part of a hydrophobic pocket that is lined by the conserved methionine on the C-helix and leucine on the \\u03b24 strand (Itk M410, L420 and Btk M450, L460) (Fig. 3d). Although Itk L420 and Btk L460 were shown to not be a part of the regulatory spine in the context of the isolated kinase domain (Fig. 1), the structure of the Btk kinase domain (which contains the SH2-kinase linker region) suggests that this leucine would be critical for the formation of the extended regulatory spine in the context of the full-length kinase. We therefore tested the role of this leucine in the context of full-length Itk.  As shown in Figure 3e, mutation of Itk L420 to alanine within the context of full-length Itk disrupted the activity of Itk. Consistent with our earlier results, mutation of Itk M409, F500 and D539 to alanine within full length Itk also inactivated the kinase. Problems with cloning prevented us from testing the activity of the full-length Itk H479A mutant. Together, Itk M409, L420, H479, F500, D539, M410 and W355 form a conserved allosteric signaling network that is absolutely required for the activity of Itk. Based on sequence conservation, the extended regulatory spine defined here likely controls the activity of the entire Tec kinase family.  Phosphorylation on the activation loop tyrosine is not required for the activity of the Btk T474M mutant  It is well established that activation loop phosphorylation is required for activation of tyrosine kinases[1; 8] and specifically for the Tec kinases, phosphorylation on the activation loop tyrosine results in at least a ten-fold increase in activity[21; 25]. Structures of active and inactive kinase domains from a number of different kinases demonstrate a conserved electrostatic network that switches between the active and inactive conformations (Fig. 4a)[1; 8]. In the inactive state of Btk, the conserved glutamate (Btk E445) on the C-helix (of the Lys-Glu ion pair) is associated with a conserved arginine (Btk R544) on the activation loop[24; 26]. Activation of Btk by phosphorylation on Btk Y551 (the activation loop tyrosine) leads to a specific interaction between pY551 and R544, and a concomitant loss of the association between the R544 and E445[24; 26]. Btk E445 on the C-helix then swings toward the kinase active site and associates with the conserved lysine side chain (Btk K430) to bring about the assembly of the active state of the kinase[24; 26]. The conformational adjustment of the C-helix brings M449 and M450 in line with the other residues of the regulatory spine (Fig. 4a). Thus, phosphorylation on the activation loop tyrosine is an initiating step in the process of regulatory spine assembly.  Figure 4 Figure 4 The regulatory spine is preassembled in the Btk T474M gatekeeper mutant Since mutation of the gatekeeper residue to methioine appears to stabilize the regulatory spine within Btk, we next probed the requirement of activation loop phosphorylation in the context of the T474M gatekeeper mutation. The activity of the BtkT474M/Y551F double mutant was compared to that of the Btk T474M mutant by monitoring phosphorylation of an exogenous substrate. We find that mutation of the activation loop Btk Y551 to phenylalanine has no effect on the activity of the Btk T474M mutant (Fig. 4b & c). The Btk T474M/Y551F double mutant and the Btk T474M single mutant are both active, whereas the wild type Btk kinase requires activation loop phosphorylation; the Btk Y551F single mutant exhibits poor activity (Fig. 4b). These data are consistent with the idea that phosphorylation on the activation loop tyrosine initiates the assembly of the regulatory spine by disrupting the association between Btk E445 and Btk R544. Once pY551 competes with E445 for association with R544, the E445/K430 salt bridge is formed, bringing the C-helix (and importantly M449) into the regulatory spine structure. Under conditions that pre-organize the regulatory spine structure (as in the Btk T474M mutant), phosphorylation on the activation loop Y551 is no longer required to trigger the conformational changes that accompany Btk activation.  We further tested the importance of this switched electrostatic network by mutating Btk R544. We reasoned that mutation of Btk R544 will activate the kinase as inward movement of the C-helix is restrained by the E445:R544 interaction and loss of this electrostatic interaction should facilitate movement of the C-helix. Indeed, mutation of Btk R544 to serine in the context of wild-type isolated kinase domain of Btk leads to a two-fold increase in activity (Fig. 4d & e). In contrast, mutation of Btk R544 to serine in the context of Btk T474M isolated kinase domain does not further activate the kinase (Fig. 4d & e). These results again suggest that pre-assembly of the kinase regulatory spine (by mutation) can overcome the regulatory interactions that normally control kinase activity. Thus, for the wild-type kinase under physiological conditions, phosphorylation on the activation loop sets into motion a cascade of events: (1) formation of the pY551:R544 interaction with concomitant disruption of the Btk E445:R544 interaction, (2) inward movement of the C-helix and formation of the E445:K430 interaction and (3) assembly of the regulatory spine, that culminates in the formation of an active kinase (Fig. 4a).  Go to: DISCUSSION The structures of the isolated kinase domain of Btk in both the active and inactive states have provided significant insight into Tec kinase activation[24; 26]. However, the lack of a full-length structure of any Tec kinase leaves numerous unanswered questions regarding the regulation of Tec kinases. While the SH2 domain and the SH2-kinase linker region of Tec kinases have been shown to be required for the activity of Tec kinases[12; 13; 14], the mechanism by which these regulatory domains positively influence catalytic activity has not been clear. Here we identify an intramolecular allosteric signaling network that extends from the SH2-kinase linker into the kinase domain of Tec kinases. This intramolecular connectivity defines how the non-catalytic Tec regulatory domains, through the SH2-kinase linker region, impinge on the kinase domain and stabilize the regulatory spine. In the absence of the Tec regulatory domains, whether achieved by deletion or conformational changes within the full-length molecule, the regulatory spine is disrupted and kinase activity is inhibited.  Comparing our results to those previously published for PKA shows that the Tec kinases require an extended spine structure and it is this longer regulatory spine that couples kinase activity to regions outside of the kinase domain. Specifically, a conserved methionine in the C-helix and a conserved tryptophan in the SH2-linker swing in to \\u2018cap\\u2019 the spine in the active Btk structure and are critical for the activity of the Tec kinases. Mutation of these extended spine residues disables kinase activity even if the \\u2018core\\u2019 residues defined for PKA are intact.  Further support for the importance of the entire regulatory spine comes from examination of the genetic mutations in Btk that are associated with X-linked agammaglobulinemia (XLA) in humans, a disease that is characterized by impaired B cell development[27]. Several of the extended regulatory spine residues identified in this study are mutated in XLA patients[27]. In addition to nonsense mutations at Btk M449, M450 and W395, which cause premature truncation of the protein, there are also missense mutations at Btk M450, F450 and D579 (M450I, F450S and D579V). These mutations likely destabilize the regulatory spine and prevent Btk kinase activation.  Activation loop tyrosine phosphorylation has long been known to influence kinase activity. In addition to priming the kinase domain for the assembly of the regulatory spine, phosphorylation on the conserved Tyr of the activation loop is also thought to stabilize the substrate-binding site of the activation loop[1]. However, our work has shown that in the context of the activating gatekeeper Btk T474M mutation, loss of Y551 phosphorylation (by mutation to phenylalanine) has no affect on substrate phosphorylation. Pre-assembly of the regulatory spine (mutation of T474M) removes any requirement for activation loop phosphorylation. These results are consistent with earlier studies that focused on phosphorylation of Y551 in Btk[25]. That earlier work did show that phosphorylation on Btk Y551 significantly alters catalytic activity of Btk; kcat of the Btk Y551F mutant is greatly diminished compared to wild-type Btk presumably due to loss of the trigger for regulatory spine assembly. However, the affinity for a peptide substrate is unaffected by Y551 phosphorylation; Km of a peptide substrate for Btk Y551F mutant is the same as wild-type Btk. As well, an alternative direct substrate docking mechanism has been identified for the Tec kinases, in which an SH2 domain within the substrate docks onto the kinase domain outside of the active site and facilitates substrate phosphorylation[28; 29; 30]. It is therefore possible that the peptide substrate-binding region on the activation loop may only play a minor role in substrate recognition in Tec kinases and that activation loop phosphorylation plays one significant role: initiation of spine assembly by altering electrostatic interactions in and around the active site.  We have previously shown that the regulation of Tec family kinases is similar to that of Csk (C-terminal Src kinase)[14]. Unlike the Src and Abl family of kinases, the SH3 and SH2 regulatory domains of Csk and Tec kinases positively influence the catalytic activity of the kinase domain[14; 16]. Like Itk and Btk, the isolated kinase domain of Csk has poor catalytic activity[16; 31; 32]. In order to define the elements within the Csk kinase domain that are responsible for the poor catalytic activity, a mutagenesis study was carried out in which Src:Csk kinase domain chimeras were created[33]. A Src:Csk chimeric kinase domain where the N-terminal lobe of Csk was replaced with the N-terminal lobe of Src was shown to exhibit activity comparable to that of full-length Csk[33]. Further mutagenesis of the N-terminal lobe, identified the C-helix and the \\u03b2 turn between the \\u03b24 and \\u03b25 strands (the \\u2018top\\u2019 of the regulatory spine) as key structural elements that are required for the activity of the isolated kinase domain of Csk[33].  There is remarkable correlation between the elements that are identified in the Csk mutagenesis study and the extended regulatory spine of Tec kinases identified in this study. Since the Csk and Tec family kinases are both positively regulated by domains outside of the kinase domain, it is not surprising that they would share the key determinants of catalytic activity. Indeed, as suggested in the original identification of the kinase regulatory spine[9], this feature likely plays a critical role in all active kinases. Our work confirms this notion for the Tec family kinases, and illustrates that extension of the regulatory spine by two residues couples this regulatory feature to allosteric events occurring outside of the kinase domain. Moreover, our data suggest that despite a high degree of sequence conservation, the five core regulatory spine residues do not assemble into the active conformation in the context of the isolated Itk or Btk kinase domains. The energetics of spine assembly seem to be tailored for each specific regulatory environment; the Tec kinase domains are maintained in their inactive state (spine assembly does not occur even when the activation loop is phosphorylated) until additional regions of the protein (in this case the SH2-linker) are present to drive complete spine assembly forward and fully activate the kinase.\",\n          \"Along-standing problem in oncology is the variability of treatment response observed in early stage clinical trials. Drugs that fail to induce disease regression in most patients or prolong median progression-free survival (PFS) are deemed inactive and often abandoned, even when the drug exhibits profound activity in a small number of patients. We hypothesized that sequencing the tumor genomes of such \\u201coutlier\\u201d patients might identify unique somatic alterations that are the basis of their drug response, information that could in turn inform future clinical development.  We studied the tumor genome of a patient with metastatic bladder cancer who achieved a durable (>2 years) and ongoing complete response to everolimus, a drug targeting the mTORC1 (mammalian target of rapamycin) complex (Fig. 1A). The patient was enrolled in a phase II trial (ClinicalTrials. gov NCT00805 129) that failed to achieve its PFS end point. Whole-genome sequencing of DNA derived from the primary tumor and blood (1) identified 17,136 somatic missense mutations and small insertions and deletions (mutation rate of 6.21 per million bases). Of these, 140 were non-synonymous mutations within protein-coding or noncoding RNA regions of the genome. Structurally, this tumor genome was intact, lacking significant copy number alterations or functional translocations (Fig. 1B). Among confirmed coding mutations were (i) a two-base-pair deletion in the TSC1 (tuberous sclerosis complex 1) gene, resulting in a frameshift truncation (c.1907_1908del, p.Glu636fs), and (ii) a nonsense mutation in the NF2 (neurofibromatosis type 2) gene, creating a premature stop codon (c.863C>G, p.Ser288*). These loss-of-function mutations were noteworthy (table S1) because alterations in these genes have been associated with mTORC1 dependence in preclinical models (2). Sequencing of both genes in a second cohort of 96 high-grade bladder cancers identified five additional somatic TSC1 mutations, whereas no additional NF2 mutations were detected (fig. S1). Although the NF2 mutation was uncommon in bladder cancers, knockdown of NF2 expression in TSC1-null bladder cancer cells was associated with enhanced sensitivity to mTORC1 inhibition (fig. S1).  Fig. 1 Fig. 1 (A) Computed tomography images of the index patient demonstrating complete resolution of metastatic disease (arrows). (B) Somatic abnormalities in the outlier responder\\u2019s genome included (from outside to inside) copy number alterations; mutations ... Because TSC1 is mutated in a subset of bladder cancers (3), we explored whether TSC1 mutation is a biomarker of clinical benefit from everolimus therapy in this disease. We thus analyzed 13 additional bladder cancer patients treated with everolimus in the same trial with a targeted deep sequencing assay designed to interrogate the coding exons of ~200 genes commonly mutated in human cancers [(1) and fig. S1]. This analysis revealed three additional tumors harboring nonsense mutations in TSC1, including two patients who had minor responses to everolimus (Fig. 1C; 17 and 24% tumor regression). A fourth patient with 7% tumor regression had a somatic missense TSC1 variant of unknown functional consequence. In contrast, tumors from eight of the nine patients showing disease progression were TSC1 wild type. Patients with TSC1-mutant tumors remained on everolimus longer than those with wild-type tumors (7.7 versus 2.0 months, P= 0.004) with a significant improvement in time to recurrence (4.1 versus 1.8 months; hazard ratio = 18.5, 95% confidence interval 2.1 to 162, P = 0.001).  These results suggest that mTORC1-directed therapies may be most effective in cancer patients whose tumors harbor TSC1 somatic mutations (4) and demonstrate the feasibility of using whole-genome and capture-based sequencing methodologies in the clinical setting to identify previously unrecognized biomarkers of drug response in genetically heterogeneous solid tumors. Although single-patient anecdotes are often dismissed as failing to provide meaningful clinical evidence, this example illustrates the potential for such cases to inform future clinical development of drugs in molecularly defined populations.\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Test, Train and Cross Validation Split\n"
      ],
      "metadata": {
        "id": "tE-M1VemPn9I"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Splitting the data into train, test, and cv\n",
        "\n",
        "y_true = result['Class'].values\n",
        "\n",
        "result.Gene = result.Gene.str.replace('\\s+', '_')\n",
        "result.Variation = result.Variation.str.replace('\\s+', '_')\n",
        "\n",
        "# x = result.drop(result['Class'], axis  = 1)\n",
        "# print(result.head())\n",
        "x_train, test_df, y_train, y_test = train_test_split(result, y_true, stratify = y_true, test_size = 0.2)\n",
        "train_df, cv_df, y_train, y_cv = train_test_split(x_train, y_train, stratify = y_train, test_size = 0.2)\n"
      ],
      "metadata": {
        "id": "DxjH9tXRPRVf"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "x_train.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "I43C5-0TlZa6",
        "outputId": "35ac6abd-4e18-4409-90e4-543b4e503425"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        ID    Gene Variation  Class  \\\n",
              "3100  3100    RAC1      Q61R      7   \n",
              "1215  1215  PIK3CA     C378R      2   \n",
              "2858  2858   BRCA2    D2665G      5   \n",
              "2598  2598   BRCA1    V1809F      1   \n",
              "802    802   ERBB4     E836K      7   \n",
              "\n",
              "                                                   Text  \n",
              "3100  Mutational hotspots indicate selective pressur...  \n",
              "1215  PIK3CA, encoding the p110A catalytic subunit o...  \n",
              "2858  Mutation screening of the breast and ovarian c...  \n",
              "2598   Abstract  The BRCA1 gene from individuals at ...  \n",
              "802   Tyrosine phosphorylation is important in signa...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7570b069-0a01-4398-9141-ce1d788043c0\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Variation</th>\n",
              "      <th>Class</th>\n",
              "      <th>Text</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3100</th>\n",
              "      <td>3100</td>\n",
              "      <td>RAC1</td>\n",
              "      <td>Q61R</td>\n",
              "      <td>7</td>\n",
              "      <td>Mutational hotspots indicate selective pressur...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1215</th>\n",
              "      <td>1215</td>\n",
              "      <td>PIK3CA</td>\n",
              "      <td>C378R</td>\n",
              "      <td>2</td>\n",
              "      <td>PIK3CA, encoding the p110A catalytic subunit o...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2858</th>\n",
              "      <td>2858</td>\n",
              "      <td>BRCA2</td>\n",
              "      <td>D2665G</td>\n",
              "      <td>5</td>\n",
              "      <td>Mutation screening of the breast and ovarian c...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2598</th>\n",
              "      <td>2598</td>\n",
              "      <td>BRCA1</td>\n",
              "      <td>V1809F</td>\n",
              "      <td>1</td>\n",
              "      <td>Abstract  The BRCA1 gene from individuals at ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>802</th>\n",
              "      <td>802</td>\n",
              "      <td>ERBB4</td>\n",
              "      <td>E836K</td>\n",
              "      <td>7</td>\n",
              "      <td>Tyrosine phosphorylation is important in signa...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7570b069-0a01-4398-9141-ce1d788043c0')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-7570b069-0a01-4398-9141-ce1d788043c0 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-7570b069-0a01-4398-9141-ce1d788043c0');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-026c2c61-aa96-4de0-8162-8bb5d931cbdf\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-026c2c61-aa96-4de0-8162-8bb5d931cbdf')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-026c2c61-aa96-4de0-8162-8bb5d931cbdf button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "x_train",
              "summary": "{\n  \"name\": \"x_train\",\n  \"rows\": 2656,\n  \"fields\": [\n    {\n      \"column\": \"ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 948,\n        \"min\": 1,\n        \"max\": 3320,\n        \"num_unique_values\": 2656,\n        \"samples\": [\n          445,\n          629,\n          837\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 246,\n        \"samples\": [\n          \"FLT3\",\n          \"ROS1\",\n          \"ARID2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Variation\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 2394,\n        \"samples\": [\n          \"G1567D\",\n          \"L46F\",\n          \"R177Pfs*126\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Class\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2,\n        \"min\": 1,\n        \"max\": 9,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          3,\n          2,\n          6\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Text\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1624,\n        \"samples\": [\n          \"Since Notch phenotypes in Drosophila melanogaster were identified 100 years, Notch signaling has been extensively characterized as a regulator of cell fate decisions in a variety of organisms and tissues. However, in the past 20 years, accumulating evidence has linked alterations in the Notch pathway to tumorigenesis. In this Perspective, we discuss the pro-tumorigenic and tumor suppressive functions of Notch signaling and dissect the molecular mechanisms that underlie these functions in hematopoietic cancers and solid tumors. Finally, we link these mechanisms and observations to possible therapeutic strategies targeting the Notch pathway in human cancers.  This year will be the centennial of the discovery of a signaling pathway that has fascinated developmental, molecular, and cancer biologists around the world. Mutant Notch phenotypes in the fly wing were characterized by John S. Dexter 100 years ago (Dexter, 1914) and, rapidly after, Thomas Hunt Morgan identified the mutant alleles (Morgan, 1917). Almost seven decades later, after the molecular biology revolution, Spyros Artavanis-Tsakonas and Michael Young cloned the Notch receptor and attributed the wing-notching phenotype to gene haplo-insufficiency (Kidd et al., 1986; Wharton et al., 1985). These studies brought a revolution in a large number of fields including developmental and stem cell biology, neuroscience, and \\u2013 related to this Perspective \\u2013 cancer biology (Fortini et al., 1993). Indeed, in the early nineties, gain-of-function mutations of the pathway were identified in cancer (Ellisen et al., 1991; Gallahan and Callahan, 1997; Gallahan et al., 1987; Jhappan et al., 1992). A deluge of reports followed, cementing the role of Notch signaling as oncogenic but also tumor suppressive, depending on the context. In this Perspective, we attempt to provide a detailed characterization of Notch functions in both solid and hematopoietic cancers and discuss the molecular mechanisms explaining such functions as well as approaches to target Notch signaling in human cancers.  Go to: A brief description of the Notch signaling pathway There are four Notch receptors (named Notch1\\u20134) in mammals. Notch1 and Notch2 each have 36 EGF-like repeats, while Notch3 and Notch4 have 34 and 29 repeats, respectively, which affect their affinity for corresponding ligands (Haines and Irvine, 2003; Okajima and Irvine, 2002; Rebay et al., 1991) (Figure 1). Notch receptors are single pass type I transmembrane molecules coded by a single precursor that becomes a non-covalently linked heterodimer consisting of an N-terminal extracellular (NEC) fragment and a C-terminal transmembrane-intracellular subunit (NTM) as a result of cleavage by a furin-like protease in the trans-Golgi network (Blaumueller et al., 1997) (Figures 1 and and2).2). The Notch pathway is normally activated upon interactions with ligands such as Delta-like and Jagged, which are also transmembrane proteins containing EGF-like repeats. In mammals, there are three Delta-like ligands (Dll1, Dll3, and Dll4) and two Jagged ligands (Jag1 and Jag2). The Notch pathway gets activated in a strictly controlled fashion: ADAM10/17 metalloproteases cause an S2 cleavage in the receptor, followed by a third cleavage (S3 cleavage) mediated by the presenilin\\u2013\\u03b3-secretase complex, composed of presenilin 1 (PSEN1), PSEN2, nicastrin (NCSTN), presenilin enhancer 2 (PEN2), and anterior pharynx-defective 1 (APH1) (Shah et al., 2005). This series of events releases the intracellular portion of the Notch receptor (termed ICN) that then translocates into the nucleus to mediate target gene activation (De Strooper et al., 1999; Schroeter et al., 1998). Notch-ICN is a transcriptional activator (Bray, 2006) consisting of ankyrin repeats, a RAM (RBP-J\\u03ba associated molecule) domain, a transactivation domain (TAD), a nuclear localization signal (NLS), and a PEST domain regulating protein stability (Figures 1 and and2).2). Notch ligands are also cleaved by \\u03b3-secretase and ADAM metalloprotease complexes, thus providing an additional level of regulation of the pathway (LaVoie and Selkoe, 2003; Six et al., 2003). Despite the overall similarities between the receptors, the differences in the ligand-binding extracellular domains and the transactivation intracellular domains lead to distinct ligand affinities and capacity to activate downstream transcription.  Figure 1 Figure 1 Protein structure and mutations of a typical Notch receptor Figure 2 Figure 2 Overview of the Notch signaling pathway In the nucleus, Notch binds to initially inactive CBF1-Su(H)\\u2013LAG1 (CSL) (aka RBP-J\\u03ba) complexes and mediates their conversion to a transcriptional activator followed by the recruitment of the co-activator protein mastermind-like 1 (MAML1) (Figure 2) (Nam et al., 2006; Wilson and Kovall, 2006; Wu et al., 2000). The ankyrin repeats seem to play an important role for MAML1 recruitment. The list of target genes regulated by Notch is very much dependent on cell type and can include genes whose products are involved in fundamental aspects of cell biology, such as cell cycle regulation (Joshi et al., 2009; Lewis et al., 2007), cellular differentiation, and metabolism (Palomero et al., 2006). Common targets of the pathway include the HES and HEY (Iso et al., 2001a; Iso et al., 2001b; Jarriault et al., 1995) families of transcription repressors as well as MYC transcription factor (Palomero et al., 2006; Sharma et al., 2006; Weng et al., 2006). The binding and function of Notch on DNA appears to be a rapid and dynamic process controlled by the kinase CDK8 and the ubiquitin ligase Fbxw7 leading to Notch phosphorylation, ubiquitination, and its subsequent proteasomal degradation (Fryer et al., 2004; Mukherjee et al., 2005; O'Neil et al., 2007; Thompson et al., 2007), which shuts off the pathway (Figure 2).  Various tools have been developed to study the transcriptional activity of the pathway, such as ChIP-seq and ChIP-chip to map Notch1 binding on the genome (Castel et al., 2013; Ntziachristos et al., 2012; Palomero et al., 2006; Wang et al., 2011a) and mouse models that allow efficient tracing of receptor cleavage/activity in many different tissues (Hansson et al., 2006; Liu et al., 2011; Mizutani et al., 2007; Souilhol et al., 2006). Recently, the group of Artavanis-Tsakonas (Fre et al., 2011; Sale et al., 2013) and our laboratory (Oh et al., 2013) have traced Notch pathway activity in vivo by using reporter systems for Notch receptors expression and Hes1 activity by coupling them to fluorescent proteins (Figure 3). As there are several unanswered questions regarding Notch ligand expression, even under physiological conditions, an exciting next step could involve development of fluorescent tools to probe ligand expression together with pathway activation in real-time within a living organism.  Figure 3 Figure 3 In vivo mapping of Notch pathway activity using a Hes1GFP reporter Go to: NOTCH signaling pathway in cancer The Notch pathway is genetically altered in a large number of hematopoietic and solid tumors (Figure 1). Intriguingly, these alterations can lead to either activation or repression of the pathway depending on the context and the activation status of other potentially oncogenic pathways (Table 1 and Figure 4). Interestingly, it appears that there are multiple and distinct modes of aberrant regulation of the pathway and its targets in cancer. They include activating and inactivating mutations, receptor/ligand over-expression, epigenetic regulation, and effects of post-translational modifications, most notably receptor and ligand fucosylation (especially O-fucosylation) (Haines and Irvine, 2003; Lei et al., 2003; Okajima et al., 2003) and ubiquitination (Fryer et al., 2004; Thompson et al., 2007). We initially discuss T cell acute leukemia, a disease in which Notch has a well-characterized oncogenic role. Subsequently we present several other cases of hematopoietic and solid tumors where Notch has tumor suppressive or oncogenic roles, along with its potential mechanisms of action and partners.  Figure 4 Figure 4 Simplified scheme of Notch interactions with other signaling pathways in cancer Table 1 Table 1 Oncogenic and tumor suppressive roles of Notch signaling in human cancers * Go to: T-Cell Acute Lymphoblastic Leukemia (T-ALL) NOTCH1 is a master transcription factor that controls innate and adaptive immunity and plays an important role in directing hematopoietic development towards T cells (Aifantis et al., 2008; Li and von Boehmer, 2011; Radtke et al., 2013). The very first finding of Notch pathway alterations in cancer comes from the work of Ellisen and colleagues, which was confirmed subsequently by other groups, that revealed a rearrangement between the intracellular part of NOTCH1 (ICN1) and the T cell receptor beta (TRB) locus, leading to high level expression of truncated, constitutively-active NOTCH1 in leukemia (Ellisen et al., 1991). In vitro studies (Capobianco et al., 1997), as well as animal modeling (Girard et al., 1996; Pear et al., 1996) then revealed that ICN1 is a strong oncogenic allele. Most importantly, ten years ago, the Aster and Look laboratories reported the first activating NOTCH1 mutations in human T-ALL, occurring in approximately 50% of all cases (Weng et al., 2004).  The majority of these mutations encompass single amino acid substitutions, insertions and deletions located in exons 26 and 27 of the genetic locus, which encode the N-terminal and C-terminal components of the hetero-dimerization domain respectively. These mutations lead to lower protection of S2 cleavage of Notch, resulting in either ligand independent activation or hypersensitivity of the pathway to ligands. Another rare group of mutations, the Juxtamembrane Expansion Mutants (JME) also augments NOTCH1 activation at the cell membrane (Sulis et al., 2008) (Figure 1). Finally, PEST domain mutants encompass another category of NOTCH1 mutations in 20\\u201325% of T-ALLs. PEST domain alterations lead to truncation or loss of the domain due to frame-shift or nonsense nucleotide substitutions, which impair proteasomal degradation mediated by the ubiquitin ligase FBXW7 and lead to higher ICN1 cellular concentrations (Weng et al., 2004). The importance of ICN1 degradation in physiology becomes more evident by the fact that 15 % of T-ALL cases harbor mutations or deletions in FBXW7 (Asnafi et al., 2009; O'Neil et al., 2007; Thompson et al., 2007). These changes are localized in three arginine residues critical for its interaction with ICN1. Mutations in the PEST domain and FBXW7 do not occur concurrently which implies that they play the same role to increase stability of ICN1 (Asnafi et al., 2009; O'Neil et al., 2007).  The fact that FBXW7 mutations directly affect cells with leukemia-initiating (LIC) properties though the stabilization and overexpression of MYC, another well-characterized substrate of this ubiquitin ligase, further demonstrates that NOTCH and MYC actions are intertwined in cancer cells (King et al., 2013). Interestingly, NOTCH1 mutations in T-ALL were shown to have a favorable prognosis and better outcome post-treatment in a number of studies including the ALL-Berlin-Frankfurt-Munster 2000 study (Breit et al., 2006), a study by the Japan Association of Childhood Leukemia Study that examined NOTCH1 and FBXW7 mutational status in T-ALL and T-cell lymphoblastic lymphoma patients (Park et al., 2009), and the Lymphoblastic Acute Leukemia in Adults (LALA)-94 and the GRAALL-2003 trials (Asnafi et al., 2009). Finally, another report on 134 pediatric patients from the EORTC-CLG 58881 and 58951 protocols concluded that NOTCH1 and FBXW7 mutations associate with improved early chemotherapeutic response and lower minimal residual disease (MRD) levels (Clappier et al., 2010). It remains to be seen whether Notch pathway inhibition will be used successfully to target T-ALL, especially in relapsed disease that is refractory to conventional chemotherapy-based treatments.  Go to: Chronic lymphocytic leukemia (CLL) CLL is the most common leukemia in adults. Recently, it was demonstrated using next generation sequencing-based approaches that 10\\u201312% of CLL cases exhibit activating mutations of NOTCH1, underlining the significance of such mutations as a prognostic marker. The vast majority of these mutations are in the PEST domain, leading to truncated protein variants with a longer half-life (Figure 1). Interestingly, there seems to be a mutational hotspot in this disease, with P2515Rfs being the most prevalent mutation (Fabbri et al., 2011; Puente et al., 2011; Rossi et al., 2012a). Mutations of NOTCH1 are mutually exclusive with TP53 abnormalities and survival outcomes are poor in both cases (Rossi et al., 2012a; Wickremasinghe et al., 2011). NOTCH1 and SF3B1 (a splicing factor) mutations were associated with decreased overall survival, and both retained independent prognostic significance for survival outcomes (Oscier et al., 2013). Mutational activation of NOTCH1 was observed at significantly higher frequency during disease progression towards the high-risk Richter transformation (30%) and chemo-refractory CLL (20%) (Fabbri et al., 2011). This later study and a very recent large-scale clinical analysis of CLL patients (Weissmann et al., 2013) confirmed that NOTCH1 mutations are an adverse prognostic parameter in this disease.  Go to: Lymphoma Non-Hodgkin lymphoma (NHL) is a heterotypic mix of diseases, the most prevalent amongst them consist of Burkitt lymphoma, Follicular lymphoma (FL, the most indolent amongst NHL cases) (Pasqualucci et al., 2014), (Roulland et al., 2011), and diffuse large B cell lymphoma (DLBCL). Burkitt lymphoma, mainly characterized by the upregulation of MYC due to its translocation to the immunoglobulin locus, display recurrent gain-of-function NOTCH1 mutations in 8\\u20139% of patients (Love et al., 2012). FL and DLBCL are malignancies of B cell origin and together comprise 60% of new NHL diagnoses in North America. FL and the germinal center B-cell (GCB) DLBCL subtype are derived from germinal center B cells, whereas the more aggressive activated B-cell (ABC) DLBCL subtype is most likely derived from cells that have exited the germinal center. NOTCH2 is mutated in ~8% of DLBCL cases (Lee et al., 2009). These are mainly gain-of-function mutations affecting the PEST domain (and thus the stability of the protein) as well as copy number alterations (Morin et al., 2011). Interestingly, NOTCH2 is required for B-cell development in the spleen marginal zone environment and has been implicated in splenic marginal zone lymphoma (SMZL) (Kiel et al., 2012; Rossi et al., 2012b), as 20% of SMZL cases exhibit gain-of-function NOTCH2 mutations accompanied by mutations of NOTCH1, SPEN and DTX1 (Rossi et al., 2012b). It was suggested that these genetic changes are associated with adverse prognosis (Kiel et al., 2012). Finally, Jundt and colleagues have characterized an activating role for NOTCH1 in classic Hodgkin lymphoma (Schwarzer et al., 2012; Schwarzer and Jundt, 2011). These authors suggested that NOTCH1 is activated through the upregulation of its ligands within the tumor niche and suppresses genes important for B cell identity, such as E12/E47 and the early B cell factor (EBF) (Jundt et al., 2008). Additional studies are required to better define Notch receptor and ligand expression, targeted signaling pathways in the distinct subtypes of lymphoma.  Go to: Acute myeloid leukemia and myelo-monocytic neoplasms Several years ago, emerging evidence indicated that the Notch pathway could have tumor suppressive roles in various types of tumors, in stark contrast to its oncogenic role in the aforementioned hematopoietic malignancies (Nicolas et al., 2003; Rangarajan et al., 2001). In contrast to the tumorigenic role of NOTCH1 in T-ALL, our laboratory and others have recently characterized a tumor suppressive role of the Notch pathway in myeloid malignancies. We have shown that deletion of nicastrin (Ncstn), an essential component of the \\u03b3-secretase complex, leads to the induction of chronic myelomonocytic leukemia (CMML) (Klinakis et al., 2011), a disease characterized by increased extramedullary hematopoiesis, monocytosis, myeloproliferation, and frequent progression to acute myeloid leukemia (AML). This is a Notch-mediated effect, as compound deletion of Notch1/2 in vivo led to similar effects. This was further attested when analysis of the conditional model for the deletion of FX (the homologue of human GDP-L-fucose synthase) or O-fucosyltransferase 1 (Pofut1) showed myeloid hyperplasia (Yao et al., 2011), underlining the importance of Notch receptor fucosylation for ligand binding and pathway activation. Ablation of MAML1 can lead to similar phenotypes (Chen et al., 2008). Mechanistically, the tumor suppressor role of NOTCH in this disease is mediated by direct repression of the PU.1 and CEBP\\u03b1 promoters by HES1. Subsequent screening of primary CMML samples for Notch pathway mutations showed that NCSTN, Mastermind-like 1 (MAML1), APH1A, and NOTCH2 are mutated and genetically inactivated in about 12% of CMML patients. These mutations are unique to CMML and not found in other myeloproliferative disorders such as polycythemia vera (PV) and myelofibrosis (MF). Notch inactivating mutations co-occurred with other described myeloid mutations in genes such as TET2, FLT3, and ASXL1 (Klinakis et al., 2011). Based on these findings we were able to show that combination of Notch pathway and TET2 inactivation leads to acute myeloid leukemia (AML). AML cells specifically express NOTCH2 on their surface but show no signs of pathway activity. Interestingly, re-activation of the Notch pathway in established AML leads to complete disease remission (Kannan et al., 2013; Lobry et al., 2013). This observation provides a rationale for the use of specific NOTCH2 activating antibodies or specific agonists as a viable therapeutic strategy in this type of leukemia. Mechanistically, there might be several ways to suppress Notch pathway activity in AML. Initially, AML cells might reside in microenvironments that lack Notch ligands. Another putative mechanism is epigenetic silencing, achieved by DNA and histone methylation of target gene promoters/transcriptional start sites. In agreement with this possibility, we found that Notch target genes are characterized by H3K27me3 marks (Lobry et al., 2013) and mice carrying the R132H mutation of isocitrate dehydrogenase 1 (IDH1) (Figueroa et al., 2010; Gross et al., 2010; Xu et al., 2011) develop a myeloproliferative disease characterized by marked DNA hyper-methylation of Notch pathway genes such as Lfng, Maml3, and Hes5 (Sasaki et al., 2012).  Go to: Acute B cell leukemia (B-ALL) Interestingly, Notch signaling also appears to act as a tumor suppressor in B cell ALL (B-ALL) (Zweidler-McKay et al., 2005). In agreement with the AML findings, Notch pathway re-activation leads to growth inhibition and induces apoptosis in human B-ALL cells. In a recent follow-up publication, it was shown that several Notch pathway targets in B-ALL are suppressed by DNA cytosine hyper-methylation on their promoters followed by histone H3K27 and H3K9 trimethylation (Kuang et al., 2013). The parallel between AML and B-ALL is intriguing and can potentially be explained by a recent Notch activity mapping effort (Oh et al., 2013) that demonstrated activity of the pathway in T cell progenitors and pre-erythrocytes and a lack of pathway activation in the B cell and myelo-monocytic lineages. These findings provide support for a key role for NOTCH as a developmental regulator that can determine the fate of progenitors in the hematopoietic system. In this model, NOTCH action needs the addition of other oncogenic stimuli to transform cells. In agreement with this idea, we found that Notch pathway inactivation can lead to increased frequency of granulocyte-monocyte progenitors (GMP), cells that can initiate diseases like CMML and AML upon further alterations (Klinakis et al., 2011).  Go to: Notch signaling in solid tumors A number of recent reviews have thoroughly summarized our knowledge on Notch signaling in solid tumors (Nowell and Radtke, 2013; Ranganathan et al., 2011; South et al., 2012) (see also Table 1). Recent genomic data, including resources from The Cancer Genome Atlas (TCGA), underscore the prevalence and the complexity of Notch pathway alterations in human cancers, although they do not currently provide detailed functional interpretations for these alterations. Our goal in the section below is not to provide an exhaustive list of the solid tumor types in which Notch signaling is altered and the possible consequences of these alterations. Rather we aim to highlight some key observations in a few prominent tumor types and draw some points of discussion from these studies, including the plethora of partners used by Notch (Figure 4) and the distinct roles of the four Notch receptors.  Go to: Breast cancer Breast cancer is a very prevalent form of cancer in which the Notch pathway may act as a tumor suppressor or an oncogene depending on the subtype. One of the first indications that Notch signaling may play a role in solid tumors actually came from experiments with mouse mammary tumor viruses (MMTV). Integration of the MMTV genome next to the \\u201cInt-3\\u201d locus resulted in an activating mutation of Notch4, leading to the constitutive activation of the receptor and breast cancer development (Gallahan and Callahan, 1997; Jhappan et al., 1992; Robbins et al., 1992). Since this seminal discovery, a number of studies have confirmed that activation of Notch signaling plays an oncogenic role in breast cancer (Colaluca et al., 2008; Pece et al., 2004; Robinson et al., 2011; Xu et al., 2012). In breast cancer cells, Notch signaling can be activated by functional interactions with other signaling pathways, including the Ras and the Wnt pathways (Ayyanan et al., 2006; Fitzgerald et al., 2000; Izrailit et al., 2013; Meurette et al., 2009; Weijzen et al., 2002). Recent observations indicate that Notch4 may play a more specific role compared to other Notch receptors in breast cancer stem cells (Harrison et al., 2010). In contrast, a recent study indicates that hyperactivation of NOTCH3 may actually be detrimental to breast cancer cells by inducing senescence (Cui et al., 2013). Interestingly, mammary epithelial cells respond differently to different levels of activation of the Notch pathway (Mazzone et al., 2010). Thus, while accumulating evidence indicates that Notch is pro-tumorigenic in breast cancer, in certain contexts, specific (high) levels of activation may be tumor suppressive; alternatively, different Notch receptors may have unique signaling outputs in mammary epithelial cells or in different subtypes of breast cancer. Once more, the notion of a \\u201cdifferentiation switch\\u201d could explain the many faces of Notch signaling in this type of tumor.  Go to: Lung cancer Lung adenocarcinoma (LAC) is a major subtype of lung cancer. Initial observations suggested that Notch signaling promotes the expansion of LAC cells in culture (Dang et al., 2003; Eliasz et al., 2010; Haruki et al., 2005). More recent in vivo studies demonstrate that Notch signaling is a key promoter of LAC development and maintenance (Allen et al., 2011; Licciulli et al., 2013; Maraver et al., 2012) and that NOTCH3 plays a unique role in the self-renewal of LAC tumor-propagating cells (Zheng et al., 2013). Expression of JAG2 at the surface of lung adenocarcinoma cells leads to homotypic interactions with Notch receptors and promotes the metastatic potential of these LAC stem cells (Yang et al., 2011). Thus, while mutations and other alterations may not be frequent in LAC (Westhoff et al., 2009), Notch pathway activity correlates significantly with worse survival in lung cancer patients (Hassan et al., 2013; Zheng et al., 2013) and activation of Notch may be important for the sustained growth of LAC. Targeting NOTCH3 and/or JAG2 may benefit a very large number of lung cancer patients worldwide.  Squamous cell lung carcinoma (SqCC) is the second major type of non-small cell lung cancer. In stark contrast to LAC, Notch signaling is thought to be a tumor suppressor of SqCC development, as evidenced by the identification of loss-of-function mutations in human tumors (Wang et al., 2011b). These mutations mainly cluster in the EGF-like repeat region of NOTCH1 and thus have the potential to disrupt ligand binding or to produce truncated receptors (Figure 1). While functional validation for these observations is still missing owing to the current lack of appropriate mouse models, numerous observations indicate that inactivation of Notch signaling promotes the development of squamous cell carcinoma in other tissues, including in cutaneous and head-and-neck tumors (Agrawal et al., 2012; Pickering et al., 2013; Proweller et al., 2006; Rothenberg and Ellisen, 2012; Wang et al., 2011b). These observations suggest that loss of Notch pathway activity may be critical for the growth of tumor cells with squamous differentiation characteristics.  Small cell lung carcinoma (SCLC) is a neuroendocrine subtype of lung cancer, representing a smaller fraction of lung cancer cases (~12\\u201315%) but with the highest mortality rate. Genomic studies have failed to identify recurrent mutations in the Notch pathway in SCLC (Peifer et al., 2012; Rudin et al., 2012). However, early observations indicated that hyperactivation of Notch signaling blocks the cell cycle of SCLC cells (Sriuranpong et al., 2001; Sriuranpong et al., 2002). A tumor suppressive role for Notch in SCLC is supported by evidence that Notch may play a similar role in other neuroendocrine tumors such as medullary thyroid carcinoma (Cook et al., 2010). Thus far, however, no functional evidence has been obtained in vivo that activation of Notch may block SCLC development or maintenance, and it is still possible that subpopulations of cells in SCLC tumors may display some Notch activity and contribute to SCLC growth (Kluk et al., 2013; Salcido et al., 2010).  Thus, three different subtypes of lung cancer display strikingly different roles for Notch signaling in cancer development, from an active oncogenic role with rare genetic alterations in LAC to tumor suppressor with inactivating mutations in SqCC and then possibly tumor suppressive with no sign of mutations in SCLC. It is possible that these differences are related to the role of Notch in cell fate decisions during lung embryonic development.  Go to: Liver cancer Genome sequencing analyses did not reveal recurrent mutations in Notch pathway genes in hepatocellular carcinoma (HCC), a leading cause of cancer-related deaths worldwide (Fujimoto et al., 2012; Guichard et al., 2012). Nevertheless, Notch signaling has been of interest to liver cancer biologists because of the prominent role of Notch signaling in liver development, including mutations in NOTCH2 or JAG1 in patients with Alagille syndrome (syndromic bile duct paucity) (McDaniell et al., 2006; Oda et al., 1997). Haploinsufficient mutations in a specific ligand and a specific receptor in the Notch pathway would suggest that Notch signaling could play very context-dependent and level-dependent roles in liver tumors. Initial observations suggested that low levels of Notch correlates with high activity of the Wnt pathway, a major oncogenic pathway in HCC (Wang et al., 2009). Also, high levels of active Notch1 may inhibit the expansion of HCC cells (Qi et al., 2003), and deletion of Notch1 in the liver of mice results in hyperproliferative hepatocytes, suggesting a tumor suppressive role for Notch in HCC (Croquelois et al., 2005). Similarly, Notch signaling has a tumor suppressive effect in HCC initiated by inactivation of the RB pathway (Viatour et al., 2011). However, other reports have more recently provided evidence that Notch signaling is active and oncogenic in HCC (Dill et al., 2013; Tschaharganeh et al., 2013; Villanueva et al., 2012), and possibly important for the development of tumors following hepatitis B virus infection (Jeliazkova et al., 2013). These observations suggest that the role of Notch signaling in HCC may be different in the distinct molecular subgroups of this cancer type and underscore the need to further explore the molecular contexts associated with tumor suppressive or oncogenic roles of Notch in the liver.  In contrast to the complex roles of Notch signaling in HCC, accumulating evidence supports a pro-tumorigenic role for Notch signaling in cholangiocarcinoma (CCC). Mutations of the Notch repressor FBXW7 are found in a subset of human tumors (Akhoondi et al., 2007). Similar to the disruption of bile ducts in Alagille patients, activation of Notch2 in liver progenitors and adult hepatocytes promotes biliary tubulogenesis (Jeliazkova et al., 2013). Finally, constitutive activation of NOTCH1 is sufficient to initiate CCC development in mice (Zender et al., 2013).  It is likely that the sometimes contradictory consequences of Notch activation in liver cells are due to a combination of the strength of the downstream signal, the timing of the activation, the cell type in which this activation occurs, and the receptor involved (Ortica et al., 2013). There seems to be a consensus that higher Notch levels in liver progenitors favors bile duct differentiation versus hepatocytic differentiation. Possibly activation of Notch (e.g. NOTCH2) in these progenitors promotes CCC while suppressing HCC (Guest et al., 2013). It is also possible that Notch switches from a suppressive role in the early stages of HCC development to a more oncogenic role. Although it was proposed that Notch signaling plays a role in liver cancer invasion and metastasis (Lim et al., 2011; Zhou et al., 2013), more work is required to further support this notion.  Go to: Colorectal cancer The intestinal epithelium possesses an unprecedented self-renewal rate that appears to be linked to a high susceptibility to malignant transformation. Notch signaling has been known for many years now to be involved in both the control of homeostatic self-renewal in stem cell populations and the development of colorectal cancer (CRC) (Fre et al., 2005; Radtke and Clevers, 2005; van Es et al., 2005). While mutations in NOTCH genes are rare, Notch signaling is overexpressed or constitutively activated in CRC in part because of mutations in regulators of Notch signaling, including in FBXW7 (although FBXW7 clearly controls other cellular pathways beyond Notch) (Akhoondi et al., 2007; Babaei-Jadidi et al., 2011; Camps et al., 2013; Miyaki et al., 2009; Sancho et al., 2010; Zhu et al., 2013). In addition, Notch activation has been linked to activation of Wnt signaling and Hippo/YAP signaling in CRC cells, although the various levels of crosstalk between these pathways are still not fully understood (Camargo et al., 2007; Fre et al., 2009; Kim et al., 2012; Kwon et al., 2011; Peignon et al., 2011; Rodilla et al., 2009; Tschaharganeh et al., 2013). In particular, Jagged1, expressed on tumor cells themselves or produced from endothelial cells, is thought to be a key ligand for Notch activation in CRC cells (Lu et al., 2013; Rodilla et al., 2009; Tschaharganeh et al., 2013). Another Notch ligand, DLL4, plays a non-cell autonomous role in CRC development in large part by controlling the development of blood vessels necessary for tumor growth (Fischer et al., 2011; Ridgway et al., 2006). Expression of miR-34a in CRC stem cells may help control Notch output and generate a bimodal Notch response (Bu et al., 2013). Finally, Notch signaling may play a crucial role not only in the early stages of CRC development by controlling the fate of stem cells and cancer stem cells but also at the later stages of tumor invasion and metastasis (Sonoshita et al., 2011).  Go to: Pancreatic cancer The major and most lethal type of pancreatic cancer is pancreatic ductal adenocarcinoma (PDAC). An early study detected evidence of Notch pathway activation in PDAC and showed that Notch lies downstream of TGF\\u03b2 during ductal metaplasia, an early stage of PDAC development (Miyamoto et al., 2003). Mouse genetics studies have demonstrated that activation of Notch signaling cooperates with oncogenic K-Ras to promote both initiation and dysplastic progression from acinar cells by inducing their rapid reprogramming to a duct-like phenotype (De La et al., 2008). Indeed, pharmacological inhibition of Notch signaling slows the progression of the disease in mutant mice and prevents the expansion of some human PDAC cell lines (Cook et al., 2012; Mizuma et al., 2012; Plentz et al., 2009), possibly in part because of an inhibition of PDAC stem cells (Bailey et al., 2013). Genetic inactivation of Notch2, but not Notch1 (Avila et al., 2012; Mazur et al., 2010), inhibits PDAC development initiated by oncogenic K-Ras. In fact, loss of Notch1 function may even promote PDAC development, although the basis of this observation remains unknown (Hanlon et al., 2010).  Go to: Melanoma The Notch pathway has been found to be active in melanoma (Asnaghi et al., 2012). NOTCH1 appears to promote disease progression (Rangarajan et al., 2001; Zhang et al., 2012) and growth of melanocytes under hypoxic conditions (Bedogni et al., 2008). There are no documented gain-of-function mutations affecting the pathway in this disease (Hodis et al., 2012), suggesting that the pathway might be affected through transcriptional and epigenetic control possibly through contrasting actions of BRN2, a possible activator, and MITF, which acts as a repressor of the Notch pathway (Thurber et al., 2011). Whatever the mechanism of activation, recent pre-clinical studies have reported that \\u03b3-secretase inhibitors (GSI) can reduce the tumor initiating potential and suggested that GSI combination with chemotherapy could be a useful new therapeutic approach in melanoma (Huynh et al., 2011).  In conclusion, Notch signaling plays distinct roles in different types of tumors, both solid and liquid (hematopoietic). The presented list is by no means exhaustive but provides us with an extensive overview of Notch signaling in cancer. Most of these studies are relatively recent emphasizing the increased interest in the study of Notch signaling in cancer during the last decade. Mechanistically, more work is required to pinpoint specific molecular pathways and gene targets in each tumor type, but emerging technologies and most notably DNA and RNA next generation sequencing-based approaches will continue to help us further dissect the role of this pathway in tumor initiation and progression.  Go to: A perspective on two decades of Notch-centered cancer research: Remaining intriguing questions This brief overview of some of the most common and/or lethal human cancers, both hematopoietic and solid, highlight several key aspects of Notch signaling in cancer development that hold true in other tumors in which Notch signaling is also altered, including myeloma, prostate, ovarian, skin, and brain cancers. Obviously, there are several outstanding questions that have to be addressed to not only help us better understand pathway function in cancer but also enable more efficient therapeutic targeting (see below). An initial question is whether Notch pathway mutations are tumor-initiating or tumor-propagating. Most likely, both types of mutations can be described, depending on the tumor type. We discussed an intriguing example in myeloid neoplasms where Notch signaling loss of activity seems to expand the frequency of leukemia-initiating cells (LIC) but requires secondary mutational events to lead to full-blown disease (Klinakis et al., 2011; Lobry et al., 2013). A similar scenario might play out in BALL, as it was shown that Notch activity directs lymphocyte progenitors exclusively to the T cell lineage, at the expense of B cell differentiation (Pui et al., 1999; Radtke et al., 1999). On the other hand, it is intriguing to ask whether NOTCH1 activating mutations in T-ALL occur to simply define lineage, by locking cells in a specific differentiation status (T cell in this case), or to truly transform the cells. Further studies that can genetically sequence both leukemia and normal stem cell/progenitor populations, preferably at the single cell level, might address such questions.  One particularly interesting aspect of Notch signaling in cancer progression that has been emerging in the last few years is its potential impact on metastasis, which may be linked to the role of Notch in cancer stem cells (see (Giancotti, 2013) for a recent discussion). Early studies had identified JAG1 expression as a marker of metastatic prostate cancer (Santagata et al., 2004) and found a role for Notch in the control of epithelial-mesenchymal transitions (EMT) (Timmerman et al., 2004). Indeed, JAG1 expression on tumor cells may help promote the spread of breast cancer cells to the bone microenvironment by activating Notch signaling in bone cells (Sethi et al., 2011). Activation of Notch during EMT and metastasis may be under the control of the miR-200 microRNA (Brabletz et al., 2011; Yang et al., 2011). This pro-metastatic function of Notch signaling may be promoted by its crosstalk with the machinery responding to hypoxic environments (Sahlgren et al., 2008; Yeung et al., 2011). Furthermore, an increasing number of studies connect Notch signaling to molecules and pathways involved in tumor invasion and metastatic growth, including Tenascin C (Oskarsson et al., 2011) and regulators of polarity (McCaffrey et al., 2012) in breast cancer.  As discussed above, Notch signaling in tumor cells has been involved in various aspects of angiogenesis in multiple studies, especially via the DLL4 and JAG1 ligands (Benedito et al., 2009; Li and Harris, 2005; Phng and Gerhardt, 2009; Zeng et al., 2005). In particular, the Notch ligand DLL4 is upregulated in the angiogenic vasculature in response to VEGF and blockade of DLL4 was shown to lead to markedly increased non-productive tumor vascularity, which inhibits tumor growth (Noguera-Troise et al., 2006; Ridgway et al., 2006). While these observations seemed promising clinically (Hoey et al., 2009), long-term blockade of DLL4 leads to the development of vascular neoplasms (Yan et al., 2010), potentially limiting the therapeutic potential of DLL4-blocking strategies. Thus, activation of Notch signaling may contribute to tumor spread via multiple mechanisms, including by maintaining the self-renewal of cancer stem cells, by directly contributing to the cellular processes involved in tumor invasion (e.g. EMT and response to hypoxia) (Wang et al., 2011c), by controlling neo-vascularization, as well as by playing a key role in the metastatic niche. Such issues could be more important in solid tumors than leukemia, however, it is intriguing to define Notch-ligand expressing niches in different types of hematopoietic tumors and test whether ligand expression is important for leukemia cell homing to different tissues and response to drug treatments. For example, targeting the expression or function of a specific ligand could affect NOTCH1-expressing T-ALL homing and metastasis. As most cancer patients die from metastatic disease, it will be important in the near future to continue to investigate the molecular and cellular basis of tumor spread in connection with Notch signaling.  Go to: Therapeutic targeting of the Notch pathway in tumors As proteolytic cleavage of NOTCH receptors by the presenilin/\\u03b3-secretase complex is a prerequisite for the activation of signaling (in the absence of downstream activating mutations), small molecule GSI efficiently blocks NOTCH1 activity in T-ALL cells and has been proposed as a molecular targeted therapy for the treatment of this disease (Aster and Blacklow, 2012; Palomero and Ferrando, 2009). However, animal studies have shown that systemic inhibition of NOTCH signaling results in \\u201con-target\\u201d gastrointestinal toxicity because of the accumulation of secretory goblet cells in the intestine due to alterations in the differentiation of intestinal stem cells following Notch inactivation. Phase 1 clinical trials further confirmed these treatment side effects. As a result, inhibition of the pathway using GSI alone may not be the most viable therapeutic choice in the future. An alternative to the use of single GSI treatment is the combinatorial use of glucocorticoids and GSI, where glucocorticoids ameliorate the GSI-induced gut toxicity by inducing the expression of Cyclin D2, protecting the animals from developing intestinal goblet cell metaplasia (Real et al., 2009).  Notch signaling targeting is, however, not restricted to the usage of GSI. Alpha-secretase inhibitors (ASI) against the ADAM10/17 metalloproteases that mediate receptor S2 cleavage are available (Zhou et al., 2006) and are currently being tested (Purow, 2012). Furthermore, using phage display technology, pharmaceutical companies have generated highly specialized antibodies against NOTCH1 and NOTCH2 that act mainly through stabilization of the negative regulatory region (NRR) of the receptors, and the protection from proteolytic cleavage, thus inhibiting the production of ICN1/2 (Wu et al., 2010). These antibodies lead to lower levels of gastrointestinal toxicity and other side effects emanating from pan-Notch pathway inhibition achieved by GSI. Selective blocking of NOTCH1 inhibits tumor growth in pre-clinical models through at least two mechanisms: inhibition of cancer cell growth and deregulation of angiogenesis. Soluble extracellular fractions of Notch receptors and ligands can also act as decoys and inhibit the pathway in a dominant-negative manner. A Notch1 decoy decreased tumor cell viability in xenograft models (Funahashi et al., 2008). However, under different conditions, a DLL1 decoy can play either an activating or inhibitory role (Hicks et al., 2002). Thus, a better understanding of the dynamics by which decoys work is needed before they can be considered as a viable therapeutic strategy.  Other types of experimental inhibitors entail synthetic peptides that mimic MAML1 but lack its active domains. Despite the use of these peptides to serve basic research purposes, their use for therapeutic purposes is still limited. Moellering et al. generated a synthetic, cell-permeable, alpha-helical peptide (SAHM1) blocking MAML1 recruitment and NOTCH-mediated transcription as it binds with high affinity to the interface on the NOTCH-CSL transactivation complex (Moellering et al., 2009). Treatment of human T-ALL cell lines and a mouse model of NOTCH1-driven T-ALL with SAHM1 resulted in strong, NOTCH-specific inhibition of cell proliferation, and leukemia progression (Moellering et al., 2009).  Another intriguing idea for the treatment of tumors that are induced by NOTCH and depend on pathway activity is to not target the Notch pathway itself but focus on its signaling targets. Several such efforts are currently underway. Briefly, we have recently demonstrated in vivo T-ALL remission when we target: a) the NOTCH1-induced IKK kinase complex with a pivotal role in controlling the NF-kB pathway which-in turn-is strongly related to NOTCH in leukemia (Figure 4) (Dan et al., 2008; Espinosa et al., 2010; Vilimas et al., 2007), b) the CyclinD:CDK4/6 kinase complex, hyperactivated in this type of acute leukemia (Sawai et al., 2012), and c) the bromodomain-containing protein BRD4 (King et al., 2013). BRD proteins can be transcriptional co-activators and share common binding patterns with T-ALL oncogenes NOTCH1 and MYC in promoters and enhancers of key genes for the induction and progression of the disease. Bradner and colleagues recently modified a thienodiazepine molecule so that it inhibits binding of BRD to the acetylated residues of histone H4 (Filippakopoulos et al., 2010). We were able to show that such drugs can target both NOTCH1- and MYC-regulated transcription in T-ALL, leading to complete disease remission in vivo. Such, \\u201cepigenetically\\u201d-targeted therapies might be particularly attractive considering the ability of Notch to alter locus accessibility and initiate transcription. We have recently connected NOTCH1 binding to loss of H3K27me3 on target promoters and demonstrated an antagonism between NOTCH1 binding and polycomb complex 2 (PRC2) recruitment and activity (Ntziachristos et al., 2012). Based on these findings, H3K27me3 demethylation inhibitors might be an attractive therapy option in NOTCH1-induced T-ALL (or CLL). Finally, recent evidence suggests that it may be possible to inhibit Notch signaling by interfering with its trafficking in cancer cell secretory pathways (Ilagan and Kopan, 2013; Kramer et al., 2013).  While a number of \\u201canti-Notch\\u201d strategies are emerging, it may be as important to specifically activate Notch in tumors where activation of the Notch pathway is tumor suppressive. As discussed above for AML, in the case where tumor cells express a Notch receptor (NOTCH2) but do not show signs of pathway activation, providing a ligand for these receptors, or treating with activating antibodies may be sufficient to activate the pathway in certain contexts and inhibit tumor growth. Furthermore, in cases where Notch receptors are not expressed (e.g. they are transcriptionally silenced), or if some of their key target genes are silenced, approaches to de-repress the expression of these genes may be useful to slow cancer growth (Stockhausen et al., 2005).  Go to: Future directions in the understanding and treatment of Notch-induced tumors The NOTCH pathway has been the intense focus of cancer researchers for the last two decades. Unfortunately, there are still no FDA-approved, Notch-targeted therapies. Retrospectively, this is not surprising, as we now know that the pathway plays key roles in several tissues, including adult differentiating and regenerating tissues, explaining the potential side-effects of general inhibitors of the Notch pathway such as GSI. The critical question is whether one can successfully target the Notch pathway to significantly inhibit cancer growth. Another major conundrum comes from possible distinct roles for Notch at several stages of the tumorigenic process, an idea that was not been thoroughly examined. Furthermore, it is likely that inhibition of Notch signaling in tumors initiated by Notch activating mutations will have a therapeutic effect, as tumors are often addicted to early initiating events. However, in tumors where alterations in Notch pathway members occur late during tumor evolution, tumors may rapidly invent ways around the targeting of Notch.  We would suggest that specificity should be the key for future attempts to target Notch activity in cancer cells: one should have a complete map of both Notch receptor and ligand expression in different cancers and their microenvironments to be able to use antibodies or other small molecules that specifically inhibit only the relevant molecules. Targeted (Notch-focused) sequencing of tumors is also important to provide a clear idea of the type of mutation and its potential impact on pathway activity. Importantly, a large number of tumors containing Notch activating mutations, like the ICN1 translocation, cannot be treated with GSI. In contrast, receptor-specific antibody agonists could be of significance clinical value for tumors in which Notch signaling has a tumor suppressive function. Myeloid neoplasms are a cancer subtype that could benefit from targeted pathway activation as we have shown that in such tumors, the pathway is inactive but the NOTCH2 receptor is expressed on the surface of the cells and can be activated by ligand binding, leading to cell death. Strategies to activate Notch in some cancer are worth testing, first in pre-clinical models and hopefully in the near future in patients. Moreover, both Notch agonists and antagonists could also be used in combination with current treatments, including chemotherapy and more recent targeted therapies. A recent intriguing example of such treatments is the combination of anti-DLL4 antibodies with either chemotherapy or Avastin or VEGF traps to target tumor angiogenesis (Lobov et al., 2011; Noguera-Troise et al., 2006). Another one is the combination of Notch receptor inhibition (using GSI or antibodies) and glucocorticoids for the treatment of T-ALL (Real et al., 2009).  As with other signaling pathways involved in cancer, identification and targeting of Notch-interacting partners and targets could be pivotal for the development of anti-tumor therapy protocols. Some attempts have been made to identify such genes/proteins using whole proteome (mass spectrometry) (Yatim et al., 2012) and genome/transcriptome (RNA-seq, gene array, ChIP-seq for NOTCH1 and HES1) (Ntziachristos et al., 2012; Wang et al., 2011a) approaches. These studies suggest that, apart from a small fraction of \\u201cuniversal\\u201d targets, including members of the HES and HEY families, Notch pathway activity controls the expression of a large number of tissue and cell-type specific gene targets. Indeed, we have previously shown that NOTCH2-HES1 signaling can regulate the expression of CEBPA and PU1, two key regulators of myeloid differentiation, but these genes are not affected by Notch pathway regulation in T-ALL. Thus, potential targeting of the function of tissue-specific Notch pathway targets could offer more targeted therapies with fewer side effects. In a similar fashion, it will be intriguing to define the biochemical composition of the nuclear Notch complex in different tissues to see whether there is specificity that can guide small molecule inhibition efforts.  Overall, it is fair to say that it took the scientific community almost a century to reach to the point that the basic molecular tenets of Notch signaling are well understood. Similarly, although we know for the last two decades that Notch signaling is involved in cancer, only recently we developed the means (small molecules, antibodies) to effectively target pathway activation in this disease. There is still a significant need for further research efforts that will better define the pathway and propose drugs or drug combinations that can affect Notch signaling specifically in cancer, avoiding harmful side-effects and improving both survival and quality of life for patients with Notch-induced tumors.  Purpose: To identify and characterize novel, activating mutations in Notch receptors in breast cancer and to determine response to the gamma secretase inhibitor (GSI) PF-03084014.  Experimental Design: We used several computational approaches, including novel algorithms, to analyze next-generation sequencing data and related omic datasets from The Cancer Genome Atlas (TCGA) breast cancer cohort. Patient-derived xenograft (PDX) models were sequenced, and Notch-mutant models were treated with PF-03084014. Gene-expression and functional analyses were performed to study the mechanism of activation through mutation and inhibition by PF-03084014.  Results: We identified mutations within and upstream of the PEST domains of NOTCH1, NOTCH2, and NOTCH3 in the TCGA dataset. Mutations occurred via several genetic mechanisms and compromised the function of the PEST domain, a negative regulatory domain commonly mutated in other cancers. Focal amplifications of NOTCH2 and NOTCH3 were also observed, as were heterodimerization or extracellular domain mutations at lower incidence. Mutations and amplifications often activated the Notch pathway as evidenced by increased expression of canonical Notch target genes, and functional mutations were significantly enriched in the triple-negative breast cancer subtype (TNBC). PDX models were also identified that harbored PEST domain mutations, and these models were highly sensitive to PF-03084014.  Conclusions: This work suggests that Notch-altered breast cancer constitutes a bona fide oncogenic driver segment with the most common alteration being PEST domain mutations present in multiple Notch receptors. Importantly, functional studies suggest that this newly identified class can be targeted with Notch inhibitors, including GSIs. Clin Cancer Res; 21(6); 1487\\u201396. \\u00a92015 AACR.  Translational Relevance  Although the Notch pathway is reportedly activated in breast cancer, the molecular mechanisms leading to its aberrant activation are poorly understood, hampering the clinical development of Notch inhibitors. In this study, we demonstrate that the Notch pathway is activated via multiple mutational mechanisms primarily involving the PEST domain of NOTCH1, NOTCH2, and NOTCH3. Collectively, approximately 13% of triple-negative breast cancer exhibits a genetic alteration coupled with pathway upregulation, and these alterations may serve as biomarkers to identify patients most likely to respond to Notch inhibitors.  Introduction  The Notch pathway is a highly conserved developmental pathway responsible for a variety of cell fate decisions (1\\u20133). The pathway is activated during normal breast development, and has been implicated as a key driver in breast cancer (4\\u20137). This has motivated the development of Notch inhibitors, including gamma secretase inhibitors (GSI), as the gamma secretase (GS) complex is required to cleave and activate all four Notch receptors (8). However, pinpointing where and how this complex pathway that has redundancy at several nodes is activated and identifying robust biomarkers of response represent a critical gap. Integrated omic datasets have recently been generated on hundreds of human tumors and relevant preclinical models, which now provide an opportunity to explore this question in detail. Within breast cancer, there is an urgent need to identify new therapeutic strategies for triple-negative breast cancer (TNBC) in particular, which is associated with a poor prognosis, lacks effective therapies, and does not have a well-established catalog of oncogenic drivers (9).  Notch is normally cleaved at the S1 site while trafficking in the Golgi and forms a bipartite receptor held together by noncovalent interactions within the heterodimerization (HD) domain. This domain is close to the membrane on the extracellular side of the cell and is flanked by site2 (S2) toward the C-terminus and a negative regulatory region (NRR) toward the N-terminus. In the unstimulated state, the NRR prevents access to and cleavage of the S2 site by an ADAM metalloprotease. Upon binding by a ligand expressed on an adjacent cell, a conformational change of the NRR exposes the S2 site, thus allowing its cleavage. The GS complex then mediates S3 cleavage within the transmembrane domain liberating the Notch intracellular domain (NICD), which translocates to the nucleus and regulates the transcription of target genes. An important mechanism of NICD regulation is protein turnover. It is normally degraded quickly due to the PEST domain located at the C-terminus (1, 10).  There is a growing list of mechanisms whereby Notch is activated by mutation or complex genetic rearrangements in cancer and human diseases. In T-cell acute lymphoblastic leukemia (T-ALL), translocations that remove a large portion of the extracellular domain (ECD) and heterodimerization domain mutations involving NOTCH1 disrupt the normal function of the ECD/NRR and effectively lead to ligand-independent Notch activation (11, 12). Inactivating PEST domain mutations in NOTCH1 that increase NICD1 half-life and Notch signaling are also frequently observed in T-ALL as well as chronic lymphocytic leukemia, splenic marginal zone lymphoma (SMZL), and mantle cell lymphoma (13\\u201315). Similar NOTCH2 PEST domain mutations are present in diffuse large B-cell lymphoma (DLBCL), SMZL, and Hadju\\u2013Cheney syndrome (16\\u201320). More recently, chromosomal rearrangements involving NOTCH1 and NOTCH2 were identified in breast cancers (21). The common feature across the genomic rearrangements that were shown to activate Notch signaling was the removal of a large portion of the ECD leading to ligand-independent activation, which is reminiscent of translocations and heterodimerization domain mutations in T-ALL. However, mutations or rearrangements involving the PEST domain have not been previously characterized in breast cancer.  Materials and Methods  Cell lines and antibodies  All cell lines were obtained from the ATCC before 2009 (over 5 years ago, exact date not known). The HCC1599 cell line was short tandem repeat (STR) authenticated by the authors (May 2013) using the StemElite assay (Promega) at the University of Arizona Research Laboratory. The other cell lines used in this study were not STR authenticated as most of the samples were from older studies. The HCC1599 and HPB-ALL cells were grown in RPMI-1640 medium and supplemented with 10% FBS, 50 IU/mL penicillin/0.05 mg/mL streptomycin. The remaining cell lines were grown in media recommended by the suppliers with supplements, including HEPES buffer, sodium pyruvate, nonessential amino acids, penicillin\\u2013streptomycin, ITS, and glutamine.  The primary antibodies used were anti-NICD1 (CST #4147), anti-Notch1 (CST #3608), and anti-GAPDH (CST #2118) antibodies (Cell Signaling Technology).  In vivo studies  All in house in vivo studies were conducted in compliance with the Guide for Care and Use of Laboratory Animals and were approved by the Pfizer Global Research and Development Institutional Animal Care and Use committee. The authorization to use animals in the CERFE facilities at XenTech for the HBCx PDX model was obtained by The Direction des Services V\\u00e9t\\u00e9rinaires, Minist\\u00e8re de l'Agriculture et de la P\\u00eache, France (agreement no. B-91-228-107). The animal care and housing were in accordance with European Convention n\\u00b0 STE123. All experiments were performed in accordance with French legislation concerning the protection of laboratory animals, and in accordance with a currently valid license for experiments on vertebrate animals, issued by the French Ministry for Agriculture and Fisheries to Dr. Truong-An Tran, Study Director, Xentech (no. A 91-541 dated December 21, 2010; validity, 5 years). Tumor-bearing athymic nude mice were dosed twice daily at 140 mg/kg on a 12-day on, 4-day off schedule for two cycles. The MAXF1162 model was run at Oncotest, and all experiments were approved by the local authorities and were conducted according to all applicable international, national, and local laws and guidelines. Tumor-bearing nude mice were dosed twice daily at 140 mg/kg on a 10-day on, 4-day off schedule for three cycles. The AA1077 patient-derived and HCC1599 cell line xenograft models were run at Pfizer in SCID-Bg mice. Mice were dosed at 110 mg/kg twice daily for 9 days (AA1077) or at 120 mg/kg twice daily for 12 days (HCC1599). To evaluate efficacy, mice with palpable tumor sizes were randomly assigned to different groups, and the mean value of the tumor size was matched between the groups. Differences between the vehicle- and PF-03084014\\u2013treated groups were statistically significant by the Student t test. The percentage of tumor regression was calculated using the following formula 100 \\u00d7 [1 \\u2212 (treated final volume/treated initial volume)] and the percentage of tumor growth inhibition was calculated using the following formula 100 \\u00d7 [1 \\u2212 (treated final volume \\u2212 treated initial volume)/(vehicle final volume \\u2212 vehicle initial volume)]. For pharmacodynamic studies, tumor-bearing mice received 100 to 140 mg/kg PF-03084014 twice daily for 2 days before terminal collection. Some pharmacodynamic groups received one dose on the second day. The tumors were harvested 4 to 6 hours after the last dose, snap frozen and pulverized in a liquid nitrogen\\u2013cooled mortar before analysis.  DNA/RNA analysis  Genomic DNA and total RNA were prepared from cell pellet or frozen tumor tissue with the Qiagen DNeasy Blood and Tissue Kit (Cat. #69504) and the Qiagen RNeasy Mini Kit (Cat. 74104), following the manufacturer's protocol. Junction PCR was then performed to verify the break point of genomic DNA. Primer sequence for each particular sample is listed in the Supplementary Material. Total RNA were subjected to direct Quantitative RT-PCR (see below), or treated with calf intestinal phosphatase (CIP) and tobacco acidic pyrophosphatase (TAP), and then reverse transcribed to amplify the 5\\u2032 end messenger RNA sequence of NOTCH1 in HCC1599. CIP and TAP were included in the FirstChoice RLM-RACE Kit from Life Technologies (cat. no. AM1700).  Transcriptomic sequencing (RNA-seq) of PDX models  RNA-seq was performed on PDX models with 100 bp (base pair) paired-end reads. Raw RNA-seq reads were filtered using Xenome (22) to remove potential reads from contaminating mouse cells. Non-mouse reads were then aligned to human reference genome using TopHat2 (23). About 226.2, 253.9 and 298.3 million mapped reads were generated for HBCx14, AA1077, and MAXF1162, respectively.  TCGA data acquisition  Tier-2 mutation, somatic copy number, and mRNA expression data (RNA Seq V2 RSEM) from The Cancer Genome Atlas (TCGA) invasive breast carcinoma cohort were obtained from the TCGA data portal and Memorial Sloan Kettering Cancer Center's cBio portal (24). Raw Affymetrix SNP 6.0 array data were also downloaded from the TCGA data portal. Prealigned RNA-seq data (in BAM format) were downloaded via The Cancer Genomics Hub (25), dbGaP accession number PHS000178, version phs000178.v8.p7. A total of 956 tumors with complete mutation, copy-number, and gene-expression data were analyzed (Supplementary Table S1).  Gene-expression Analysis  Gene-expression profiles of the external PDX panel were generated using Affymetrix U133Plus2 arrays. CEL files were provided by the vendor (Xentech). Raw intensity data were processed by GC Robust Multiarray average background adjustment, quantile normalization, and median-polish summarization to generate the probe-level data in R. Normalized probe level data were further summarized into gene level using the GSEA CollapseDataset function.  Differential expression analysis of the Notch pathway genes in the TCGA breast cancer cohort were performed on the RNA-seq based Transcript per million metric from the RSEM results provided in the TCGA tier-3 data. Of the 130 TNBC tumors included in this cohort, 21 were selected as Notch altered that include patients with simple mutations or complex alterations in the heterodimerization or PEST domain of NOTCH1, NOTCH2, or NOTCH3, and those with focal amplification of NOTCH2 or NOTCH3 (inferred copy number >4). The Notch-nonaltered group included 50 TNBC tumors with no mutation, nor alteration, nor amplification (inferred copy number <2.5) in all Notch receptors. Note, that we did not include in this analysis TNBC tumors that harbor mutations or alterations in non-hotspot regions, and those with mild or broad copy-number gains, as they represent a \\u201cgray zone\\u201d that may blur the signal differentiating the two groups. Differential expression was calculated using a two-sample t test. Multiple hypothesis testing was controlled by FDR using the Benjamini\\u2013Hochberg method.  Nanostring Analysis  Nanostring technology (26) was used to measure the RNA transcript levels using the nCounter assay according to the manufacturer's recommended protocols. Briefly, transcript-specific capture and detection probes were designed and manufactured by the Nanostring Technologies and 100 ng of total RNA was hybridized to nCounter probe sets for 16 hours at 65\\u00b0C. Samples were processed using an automated nCounter Sample Prep Station (NanoString Technologies, Inc.). Cartridges containing immobilized and aligned reporter complex were subsequently imaged and counted on an nCounter Digital Analyzer (NanoString Technologies, Inc.) set at 1,155 fields of view. Reporter counts were analyzed and normalized using NanoString's nSolver analysis software version 1.  Quantitative RT-PCR  Total RNA was isolated using the miRNeasy Mini Kit (Qiagen). Two micrograms of RNA per reaction was used to generate cDNA using the High Capacity cDNA Reverse Transcription Kit (Life Technology). Q-PCR was performed in triplicate by using ABI PRISM 7900HT Sequence Detection System with Taqman Universal PCR Master Mix (Life Technology). Primer/probes for Hes4, Hey1, Hey2, HeyL, Myc, NRARP, CCND1, and Notch3 were purchased from Life Technology. The expression level was normalized to GAPDH. Data normalized using hypoxanthine-guanine phosphoribosyltransferase and peptidylprolyl isomerase E yielded similar results.  Results  Identification of Notch receptor mutations and focal amplifications in the TCGA breast cancer dataset  To identify potential patient selection biomarkers for the GSI PF-03084014, we mined the TCGA breast cancer data for mutations and alterations involving Notch receptors (27). A number of mutations were reported by the standard TCGA pipeline or discovered internally in each receptor (Supplementary Figs. S1 and S2 and Supplementary Table S1). Of the 956 tumor samples analyzed with complete gene expression, copy-number, and mutation data, there were 42 mutations in NOTCH1, NOTCH2, or NOTCH3, 25 of which either clustered in the heterodimerization domain or led to a disruption of the PEST domain (Fig. 1A and Table 1). The PEST domain is a negative regulatory domain that is responsible for degrading the active NICD and, when disrupted, leads to an increase in the NICD half-life. Although PEST domain mutations in Notch are established oncogenic events in leukemias and lymphomas, they have not been characterized in breast cancer. The majority of PEST domain mutations were nonsense mutations or frameshifting indels, and were therefore predicted to truncate the normal protein sequence. The heterodimerization mutations on the other hand, occurred at highly conserved amino acid residues (28). Similar mutations are present in NOTCH1 in T-ALL and activate Notch signaling (12). Although no mutations in this cohort were recurrent, several other mutations disrupted the protein at the same amino acid location as in other established Notch-driven cancers as well as Hadju\\u2013Cheney syndrome, a genetic disorder characterized by pathway-activating NOTCH2 PEST domain mutations (Table 1).  Figure 1. Download figureOpen in new tabDownload powerpoint Figure 1. Notch receptor mutations and focal amplifications in TCGA invasive breast cancer dataset exhibit pathway activation and are enriched in the triple-negative subtype. A, lollipop graph depicting simple mutations and complex alterations in NOTCH1, 2, and 3 in the TCGA dataset clustered in or near the heterodimerization or PEST domains. The green dotted line indicates in-frame deletion (NOTCH1) or in-frame fusion (BRD4\\u2013NOTCH3), both of which are predicted to produce the NICD. Protein domains were obtained via a Pfam search (http://pfam.sanger.ac.uk/search) of the wt protein sequences. B, examples of TCGA tumors harboring focal amplifications of NOTCH2 (top) or NOTCH3 (bottom) copy-number (CN) ratio in log2 were calculated using data from the matched normal sample as reference. C, expression heatmap of Notch pathway genes in 21 Notch\\u2013altered TN tumors compared with 50 nonaltered TN tumors. See Materials and Methods for the selection of the two groups. Unscaled expression was shown to the right of the heatmap in which the circle indicates median expression and the line indicates the range of expression. The dotted red line indicates background expression level. D, Notch-altered tumors are enriched in TNBC and Notch-altered TNBC tumors are more likely to exhibit increased Notch pathway activity. BCs, breast cancers; NA, Notch altered. View inlineView popup Table 1. NOTCH1, NOTCH2, and NOTCH3 hotspot mutations in TCGA breast tumors More complex structural variants were not provided by the standard TCGA pipelines. We therefore obtained the prealigned RNA-seq data from the TCGA breast cancer study and applied a suite of in-house algorithms, collectively called TopNotch, that are based on a local, de novo transcript assembly approach. This analysis identified six additional candidate alterations in five tumors predicted to activate the Notch pathway. Four alterations disrupted the PEST domains in NOTCH1, NOTCH2, or NOTCH3 (Table 1, Fig. 1A; Supplementary Fig. S3). In terms of molecular mechanisms, three tumors harbored large deletions whereas one tumor harbored a translocation, all of which were predicted to remove the entire PEST domain (Table 1). One tumor harbored both a large PEST domain\\u2013disrupting deletion as well as a fusion with the BRD4 gene in which exons 26-33 of NOTCH3 (which encode NICD3) were fused downstream of exon 1 of a noncanonical BRD4 transcript (Ensembl transcript ID: ENST00000360016). The fusion was predicted to lack the entire NOTCH3 ECD and produce an in-frame NICD3 with an intact GS site. Furthermore, we identified a tumor harboring a deletion of exons 21-27 in NOTCH1, which removes the NRR and heterodimerization domain. The PEST domain breakpoints identified from this analysis were further validated in the whole-exome sequencing data from the same patients because they were all located within the last exon of the Notch genes, including the matched normal samples to confirm their somatic status (see Materials and Methods; Supplementary Figs. S4\\u2013S6 and Supplementary Table S2).  We also analyzed the somatic copy-number alteration data for amplifications of Notch receptors and found a number of tumors with focal amplification of NOTCH2 or NOTCH3 (Fig 1B; Supplementary Fig. S1 and Supplementary Table S1). Focal amplification often led to overexpression of the receptor. Notably, similar to mutational patterns observed in other well-established oncogenic drivers, 37% of Notch mutations co-occurred with copy-number gain >3 at the locus (Table 1). Furthermore, 67% of the mutations exhibited very high expression of the Notch receptor (93rd percentile or greater), including all but one of the mutations that were coincident with copy-number gain. The mutation and expression pattern is most striking for NOTCH3 in which of seven mutated tumor samples, four exhibit copy-number gain, two of which are focal, and all have either 97th or 99th percentile expression of NOTCH3. In contrast with NOTCH1, NOTCH2, and NOTCH3, no compelling mutations or amplifications were found in NOTCH4 (Supplementary Fig. S2). In summary, we identified 43 breast cancers in the TCGA cohort that carry either somatic mutations in a hotspot domain (ECD/HD or PEST) of NOTCH1-3, or somatic focal amplification of NOTCH2 or NOTCH3 with concomitant increase in receptor expression (collectively referred to as \\u201cNotch altered\\u201d hereafter). Our analysis demonstrates a broad spectrum of potentially activating genetic alterations in Notch receptors is present in breast cancer, most often involving the PEST domain.  Notch-mutant and -amplified cancers are associated with the triple-negative subtype and pathway activation  To determine whether these potentially activating Notch alterations were associated with a breast cancer subtype, TCGA data were analyzed for Notch, ESR1, PGR, and ERBB2 status. Of the 956 tumor samples analyzed, 130 (or 13.6%) were determined to be TNBC (see Materials and Methods). Strikingly, 21 of the 43 Notch-altered breast cancers are TNBCs (a 3.6-fold enrichment, P = 6.89E\\u22126 by Fisher exact test, Table 1 and Fig. 1D). This strong enrichment and the high unmet medical need in TNBC prompted us to focus on our subsequent analyses in this subtype. To determine whether Notch alteration resulted in increased Notch pathway activity, we compared Notch pathway gene expression in the 21 Notch\\u2013altered triple-negative tumors with Notch wild-type (wt) triple-negative tumors (Fig. 1C and Supplementary Table S3). Overall, a subset of Notch pathway and target genes, including NOTCH3, HES1, HEY2, MYC, CCND1, HES4, NRARP, and NOTCH1, exhibited significant overexpression (FDR <0.05) in the Notch-altered tumors, and were among the top 4% most upregulated genes in Notch-altered TNBCs when all genes were queried (Supplementary Table S3). One of the more provocative results in this dataset is the strong upregulation of the Notch pathway target genes MYC and CCND1 in the Notch-altered breast tumors, providing insight into both the mechanism, whereby Notch activation may lead to an oncogenic phenotype, as well as how these classical breast cancer genes are, in fact, upregulated in this subset of breast tumors (29, 30). In addition, mutations in PTEN or FBXW7, which have been shown to confer resistance to a GSI in NOTCH1-mutant T-ALL (31, 32), do not co-occur with Notch alterations in this dataset.  As canonical Notch target genes HES4 and HEY2 were differentially expressed between the Notch-altered and wt triple-negative tumors, we used overexpression of one or both of these target genes as an indicator of Notch pathway activation (Supplementary Fig. S7). Of the 21 TNBCs with Notch alterations of interest, 17 showed evidence of Notch pathway activation. Outside TNBC, however, only eight of the 22 Notch-altered tumors showed evidence of pathway activation (Fig. 1D). This suggests that Notch alterations are more likely to be functionally relevant in the TNBC subtype, and we estimate that 13% of TNBC falls into this newly identified Notch-altered oncogenic driver class.  Patient-derived xenograft models harboring Notch receptor PEST mutations are sensitive to PF-03084014  To determine whether in vivo models harboring Notch receptor PEST domain mutations are sensitive to PF-03084014, we sequenced patient-derived xenograft (PDX) models using RNA-seq and whole-genome sequencing. We identified three PDX models all of which had distinct genomic rearrangements that truncated the PEST domain of NOTCH1 or NOTCH2 (Fig. 2A; Supplementary Figs. S8 and S9, and Supplementary Table S4). The MAXF1162 model harbored a translocation that fused the last exon of NOTCH2 with intronic sequence from NBPF8, which disrupted the NOTCH2 PEST domain at amino acid 2320. Similar to several Notch receptor mutations in the TCGA data, this model also harbored a focal amplification of the NOTCH2 locus (Supplementary Fig. S8). The AA1077 model harbored a transcript that fused part of the last exon of NOTCH1 with intronic sequence from NOTCH1 between exons 30 and 31, and disrupted the PEST domain at amino acid 2249. Using WGS data from this model, we confirmed that this was due to a tandem duplication between intron 30 and exon 34 of NOTCH1 (Supplementary Fig. S8). Finally, the HBCx-14 PDX model harbored a heterozygous 10 bp (frameshift) deletion in the PEST domain (disrupting the coding sequence beginning at amino acid 2462) and a homozygous ECD deletion (Supplementary Fig. S9 and Supplementary Table S4). Although the ECD was deleted, the GS site was still present in the mutated protein. All three models were highly sensitive to PF-03084014, including more than 50% tumor regression in the MAXF1162 model, which was similar to the NOTCH1 ECD deleted HCC1599 cell line xenograft model previously known to be sensitive to GSI (Fig. 2B; refs. 33, 34). The MAXF1162 model was characterized as an HER2-amplified model, whereas the other models were triple negative.  Figure 2. Download figureOpen in new tabDownload powerpoint Figure 2. Breast cancer models harboring Notch alterations are sensitive to PF-03084014. A, genomic alterations in Notch receptors in PDX models and the HCC1599 cell line model. The MAXF1162 model harbors a fusion that disrupts the NOTCH2 PEST domain (red lollipop) as well as an amplification of the locus. The AA1077 model harbors a partial tandem duplication of NOTCH1 that disrupts the NOTCH1 PEST domain. The HBCx-14 model harbors an ECD deletion and a PEST domain frameshifting deletion in NOTCH1. The HCC1599 harbors an ECD deletion in NOTCH1. See Supplementary Figs. S8 and S9 for details of the alterations at genomic and transcript levels. B, Notch altered models are sensitive to PF-03084014. Data, average tumor volume + SEM. The MAXF1162 model exhibited 65% tumor regression, the AA1077 model exhibited 88% tumor growth inhibition, the HBCx-14 model exhibited 60% tumor growth inhibition, and the HCC1599 model exhibited 50% tumor regression. All models were statistically significant by the Student t test (P < 0.05). To compare these responses in Notch-mutant models to Notch wt models, we analyzed a panel of eight xenograft models that were previously tested for sensitivity to PF-03084014 and for which we had treated and untreated samples that could be used for sequencing and pharmacodynamic gene-expression analyzes (Supplementary Fig. S10; ref. 33). No activating Notch hotspot domain alterations were found in RNAseq or whole-genome sequencing data from four PDX models [HBCx-5 (HER2+), HBCx-17 (triple negative), HBCx-12B (triple negative), and AA0869 (triple negative)] and four cell line xenograft models [HCC1937 (triple negative), HCC1806 (triple negative), BT-474 (HER2+), and MDA-MB-231 (triple negative)]. Responses in these Notch wt models were very heterogeneous ranging from little to no effect to slight tumor regression. On the other hand, all Notch-mutant models responded to PF-03084014, including greater than 50% tumor regression in two of four models. In summary, we identified mutations in PDX models representative of the TCGA mutations in terms of location, predicted functional consequence, spectrum of molecular mechanisms, and coincidence of multiple events. These data, therefore, confirm and extend our findings from the TCGA analysis and importantly, demonstrate that relevant preclinical models harboring Notch alterations are particularly sensitive to PF-03084014.  We next explored the functional consequences of the NOTCH1 mutations on NICD1, the activated form of the receptor. Mutations that remove the ECD domain are predicted to result in ligand-independent activation of the receptor. Indeed, similar to the HCC1599 model, the HBCx-14 model exhibited robust NICD1 expression that was inhibited by PF-03084014 and lacked full-length NOTCH1 (Fig. 3A\\u2013C). On the other hand, PEST domain\\u2013truncating mutations in the AA1077 and HBCx-14 models produced a lower molecular weight NICD1 species as predicted. In contrast, NOTCH1 wt models did not exhibit strong expression of NICD1 nor were lower molecular weight species present (Supplementary Fig. S11).  Figure 3. Download figureOpen in new tabDownload powerpoint Figure 3. NOTCH1 mutations alter full-length and NICD1 protein and NICD1 half-life. A, NICD1 Western blot analysis in Notch-altered models treated with or without PF-03084014. \\u201c*\\u201d indicates NICD1 species of lower molecular weight than wt NICD1. HPB-ALL is a T-ALL model that is known to generate an NICD1 species of lower molecular weight due to a PEST domain mutation. To better visualize the wt and mutant NICD1 bands in the newly discovered NOTCH1 mutant models, a 15-\\u03bcg lysate was loaded for the HPB-ALL and HCC1599 models, 50 \\u03bcg was loaded for the AA1077 model, and 30 \\u03bcg was loaded for the HBCx-14 model. B, NICD1 Western blot analysis on the NOTCH1 PEST\\u2013truncated HBCx-14 model (lane 6) alongside a panel of lysates from TNBC PDX models from the same collection. The HCC1599 cell line xenograft model \\u00b1 PF-03084014 for 2 days at 100 mg/kg twice daily was included as controls. A 50-\\u03bcg lysate was loaded for all lysates except 25 \\u03bcg for the HCC1599 lysates. The red arrow indicates the lower molecular weight NICD1 species in the HBCx-14 model. C, NOTCH1 Western blot analysis using an antibody that recognizes the NOTCH1 transmembrane and full-length species. Bottom, lighter exposure. The red arrows indicate the lower molecular weight species in the HBCx-14 model and the absence of detectable full-length protein in either the HBCx-14 model or the HCC1599 model. The same amount of lysate was loaded in a separate gel as in B on the same day; therefore, no additional loading control was included for this gel. D, 2 and 12 days treatment with PF-03084014 at 140 mg/kg twice daily. The PEST-mutated NICD1 was less diminished after drug treatment relative to the wt band. E, quantification of bands in D. Data, average band intensity \\u00b1 SEM relative to 12 days vehicle and normalized to GAPDH. S3, cleavage site recognized by the GS complex; PEST, protein domain rich in proline (P), glutamic acid (E), serine (S), and threonine (T). We further explored the consequence of the PEST domain mutation in the HBCx-14 PDX model. In this model, a 2-day treatment with PF-03084014 dramatically reduced the wt NICD1 band. In contrast, the mutated NICD1 band appeared to be only slightly reduced (Fig. 3D). This is consistent with the prediction that PEST-mutant NICD1 has a longer half-life. To quantify this effect, we measured the average intensity of wt and mutated NICD1 bands in the HBCx-14 model from three animals treated for either 2 or 12 days with PF-03084014. Short-term treatment reduced the wt band to approximately 50% of untreated levels whereas the mutated band was not significantly reduced (Fig. 3D and E). In 12-day treated mice, wt NICD1 was reduced to less than 3% of untreated wt levels whereas mutant NICD1 was present at approximately 15% of untreated mutant levels. These data demonstrate that mutations truncating the PEST domain in breast cancer models can increase the protein half-life of NICD1.  Activating alterations in Notch receptors would be predicted to upregulate direct transcriptional targets of the pathway, including the Hes and Hey family transcription factors, similar to the TCGA gene-expression analysis. Indeed, in an initial survey of gene-expression levels of Notch pathway genes across the panels of in vivo models, we observed strong overexpression of HES4, HEY2, HEY1, and/or HEYL in three of the four Notch-mutant models (Fig. 4A). The MAXF1162 and HBCx-14 models both harbored multiple genetic events in the altered Notch receptor, and in both cases the mutated receptor itself exhibited the highest mRNA expression among the panel of the xenograft screened (Fig 4A and B). We next used quantitative RT-PCR to analyze Notch target gene expression between Notch-mutated and Notch wt xenograft models treated with or without PF-03084014. Treatment with drug reduced the expression of nearly all Hes and Hey target genes in the Notch-altered models that exhibited strong overexpression at baseline demonstrating Notch alterations indeed drive the Notch pathway in these models and that PF-03084014 can effectively repress this hyperactivated transcriptional program (Fig. 4C and Supplementary Fig. S12). In addition, the Notch target genes MYC, NRARP, CCND1, and NOTCH3 were also nearly always downregulated by PF-03084014 (Fig. 4C and Supplementary Fig. S13). In contrast, in Notch wt models, Notch target genes very rarely exhibited strong overexpression at baseline. Moreover, although treatment with PF-03084014 downregulated Notch target genes in some Notch wt models, the number of targets regulated by PF-03084014 and the magnitude of downregulation was less, on average, than in the Notch-altered models (Fig. 4C and Supplementary Figs. S12 and S13). Taken together, these functional studies in preclinical in vivo models demonstrate PEST domain mutations in Notch receptors activate the Notch pathway, confer sensitivity to PF-03084014, and provide strong rationale for a personalized medicine strategy for Notch inhibitors in Notch-altered TNBC.  Figure 4. Download figureOpen in new tabDownload powerpoint Figure 4. Notch-mutant breast cancer models often exhibit increased Notch pathway expression sensitive to PF-03084014. A, models rank ordered from left to right using an HES4, HEY2 two gene signature score across an internal in vivo (left) and an external PDX panel (right). B, relative NOTCH2 expression normalized against \\u03b2-actin across the panel of internal in vivo models. C, i, quantitative RT-PCR of key Notch target genes at baseline in Notch wt and Notch-altered models demonstrate Notch-altered models exhibit high level expression of one or more genes, the HES and HEY family genes in particular. C, ii, fold change from baseline following 2 days of treatment with PF-03084014. Notch target genes were consistently downregulated by PF-03084014 treatment in Notch-mutant models and were downregulated to a greater degree compared with Notch wt models. Nanostring digital gene-expression data were used for the internal in vivo model; Affymetrix microarray gene-expression data were used for the external PDX model panel; qRT-PCR data were used for C. Discussion  In this study, a broad spectrum of activating mutations were discovered in NOTCH1, NOTCH2, and NOTCH3, including missense mutations, nonsense mutations, small indels, large deletions, and chromosomal translocations that disrupt either the ECD/HD or PEST domains. This repertoire of mutational mechanisms involving the PEST domains is in contrast with the molecular mechanisms of previously known PEST domain mutations in leukemias and lymphomas that primarily consist of point mutations and small indels (13\\u201315, 17\\u201319). We also found evidence for heterodimerization or ECD mutations similar to what have been previously reported; however, these mutations appear to have a lower prevalence compared with mutations involving the PEST domain.  An interesting observation from this study is the nature of the Notch mutations and alterations. Typically, activating oncogenic mutations is restricted to a few hotspots within a gene, and inappropriately activate a signal transduction cascade: for example, mutations in KRAS at codons 12, 13 or 61, or the L858R mutation or exon 19 deletion of EGFR. Conversely, tumor-suppressor genes such as TP53 are often altered at multiple locations along the gene. Notch receptors are somewhat unique in that the PEST domains and ECD/HD, in essence, negatively regulate the active form of Notch, the NICD. Therefore, inactivation by mutation or deletion of either domain activates the Notch pathway.  The coincidence of mutations with amplification and overexpression of the Notch receptor is similar to other established oncogenes, and suggests that there is pressure to select for high levels of Notch pathway activity. Increased receptor expression may be important, given the unique mechanism of Notch activation, which lacks a true signal amplification step. GS cleavage generates the active NICD molecule, which directly translocates to the nucleus and activates a downstream transcriptional program. Therefore, the maximal activity of the pathway is directly linked to Notch receptor concentration. In the case of the MAXF1162 model that harbors a PEST domain mutation and focal amplification, the receptor amplification may be required to boost the baseline signaling, whereas the PEST domain mutation would be predicted to further increase the duration of the active Notch signal.  The number and complexity of Notch mutations may preclude a definitive immediate interpretation for every alteration observed in the TCGA breast cancer dataset. Reliable functional studies capable of measuring Notch pathway activation, transformation potential, and sensitivity to pathway inhibition will be required to better understand all of the mutations and alterations examined in this study. However, the observation that many of these mutations are recurrent in other cancers or diseases in which pathway activation is an established pathogenic event provides strong evidence that these alterations are operative in breast cancer. Moreover, the increased expression of Notch pathway target genes in Notch-altered human tumors and sensitivity to a GSI in Notch-altered preclinical models provide compelling evidence that many of these mutations are oncogenic.  This dataset is timely, as several Notch pathway inhibitors that target various points in the pathway are currently in early clinical development. Given the spectrum of Notch alterations observed, it should be anticipated that each drug will produce different responses depending on the nature of the mutations. For instance, inhibitory antibodies against Notch receptors will likely only work against tumors in which the domain recognized by the antibody remains intact after the genomic alteration. On the other hand, GSIs should inhibit activating Notch alterations in which the GS cleavage site remains intact, which includes nearly all of the activating Notch alterations described in breast cancer thus far. Collectively, these data suggest an appreciable fraction of patients with TNBC harbor oncogenic Notch alterations that may be effectively treated with targeted inhibitors.  Notch signaling is often considered a model hematopoietic proto-oncogene because of its role as the main trigger of T cell acute lymphoblastic leukemia (T-ALL). Although its role in T-ALL is well characterized and further supported by a high frequency of activating NOTCH1 mutations in T-ALL patients, it still remains an open question whether the effects of Notch signaling are causative in other types of cancer, including solid tumors. Growing evidence supported by recent studies unexpectedly shows that Notch signaling can also have a potent tumor suppressor function in both solid tumors and hematological malignancies. We discuss the intriguing possibility that the pleiotropic functions of Notch can be tumor suppressive or oncogenic depending on the cellular context.  Notch signaling is a highly evolutionarily conserved pathway implicated in diverse functions during embryogenesis and in self-renewing tissues of the adult organism. These functions include the maintenance of stem cells, cell fate specification, proliferation, and apoptosis (Artavanis-Tsakonas, 1988; Leong and Karsan, 2006). In mammals, there are four Notch receptors (Notch1-4), three Delta-like ligands (Dll1, Dll3, and Dll4), and two ligands of the Jagged family (Jag1 and Jag2). When membrane-bound receptors interact with cognate ligands on an adjacent cell, two consecutive proteolytic cleavages of the receptor are initiated, freeing the intracellular portion of Notch to enter the nucleus and activate the transcription of target genes. The first cleavage (S2) in the heterodimerization domain (HD) by ADAM10 (A disintegrin and metalloprotease 10) generates the substrate for the second cleavage (S3) by the \\u03b3-secretase complex. Canonical Notch signaling requires the formation of a complex with a transcription factor of the CSL (CBF-1/Su(H)/Lag-1) family, CBF-1/RBP-Jk/KBF2 in mammals. CBF-1 binds DNA in a sequence-specific manner and acts as a repressor of transcription in the absence of Notch signaling. Displacement of co-repressors bound to CBF-1 by intracellular Notch (ICN) allows the recruitment of co-activators, such as MamL1 (Mastermind Like-1), and histone acetyltransferases, such as p300, to create a short-lived transcriptional activation complex. Recent genome-wide chromatin immunoprecipitation arrays and sequencing have identified a large number of genes that can be regulated directly by Notch (Palomero et al., 2006; Hamidi et al., 2011). Many of these target genes may be cell type specific, but there are a few well characterized transcriptional targets of ICN-CBF1, including the HES (hairy enhancer of split) family of transcription factors, Notch-related ankryin repeat protein (NRARP), c-MYC, and DTX1 (Deltex1; Weng et al., 2006).  Notch as an oncogene  The first evidence for the involvement of Notch signaling in cancer came from T-ALL. T-ALL is a neoplastic disorder accounting for \\u223c10\\u201320% of all acute lymphoblastic leukemias. In 1991, Ellisen et al. (1991) identified a t(7;9)(q34;q34.4) translocation in T-ALL patients, which resulted in fusion of the 3\\u2032 region of NOTCH1 into the TCR\\u03b2 locus and consequent overexpression of the active form of Notch1 (ICN1). This translocation appeared to be rare, found in <1% of T-ALL cases. However, 13 yr later, Weng et al. (2004) identified activating NOTCH1 mutations in \\u223c56% of T-ALL cases examined, introducing NOTCH1 mutation as the main oncogenic lesion in T-ALL. Two major hotspots of mutations were characterized: mutations in the HD domain that induce ligand-independent activation, and mutations in the PEST (proline-glutamate-serine-threonine-rich) carboxy-terminal domain that increase stability of ICN1 (Thompson et al., 2007). Additionally, inactivating mutations were identified in FBW7, an E3 ubiquitin ligase responsible for ICN1 degradation and subsequent termination of Notch signaling (Malyukova et al., 2007; Maser et al., 2007; O\\u2019Neil et al., 2007; Thompson et al., 2007). Of note, animal modeling suggested that NOTCH1 mutations (HD or PEST) are either insufficient to induce disease or are very weak oncogenes in T-ALL, even when they are overexpressed (Chiang et al., 2008). We recently generated knockin mice carrying human NOTCH1 mutant alleles, and our studies also indicated that none of these mutations are sufficient to induce disease (unpublished data).  After the discovery of its involvement in T-ALL, Notch signaling was also implicated in various solid tumors, including breast cancer, medulloblastoma, colorectal cancer, non\\u2013small cell lung carcinoma (NSCLC), and melanoma (Ranganathan et al., 2011). The oncogenic potential of Notch activation in solid tumors was first observed in mouse mammary tumor virus (MMTV)\\u2013driven breast cancer. The integration of MMTV in specific loci of the host genome resulted in dysregulated expression of adjacent genes and subsequent outgrowth of tumorigenic clones. Characterization of one of these loci revealed expression of a truncated constitutively active form of Notch4 (Gallahan et al., 1987). In mouse models, Notch activation can clearly drive mammary tumors, and in human breast cancer, increased expression of Notch or Jag1 correlates with poor prognosis (Reedijk et al., 2005). However, few activating mutations of the Notch pathway have been found in solid tumor patients, with most being observed in NSCLC (Westhoff et al., 2009).  Two recent studies have identified activating NOTCH1 mutations in chronic lymphocytic leukemia (CLL), a frequent adult leukemia (Fabbri et al., 2011; Puente et al., 2011). CLL is characterized by variable clinical presentation and progression but can be divided into two major subtypes: one with mutated immunoglobulin genes (IGV(H)), and another more aggressive form with nonmutated IGV(H). Both studies identified NOTCH1-activating mutations (mainly a frame shift mutation at codon 2515) predicted to impair Fbw7-induced Notch1 degradation. Although the overall frequency was not dramatic (from 8.3 to 12.2%), these NOTCH1 mutations were primarily found in patients with the more clinically aggressive nonmutated IGV(H) subtype of CLL (20.4%) in Richter syndrome (31.0%), and in chemorefractory CLL (20.8%). These results suggest that although NOTCH1 mutations are not pathognomonic or causative of CLL, they are associated with poor prognosis and could define a distinct clinical subtype for therapeutic intervention.  Notch as a tumor suppressor  Although Notch activation (especially at higher levels as conferred by ICN1 expression) can be oncogenic, there is growing evidence that components of the same pathway may have growth-suppressive functions in other hematopoietic cells, skin, and pancreatic epithelium, as well as in hepatocytes.  In the skin, Notch receptor and ligand expression was found largely in the suprabasal cells, and in vitro data suggested that Notch activation induces differentiation and cell cycle arrest (Lowell et al., 2000; Rangarajan et al., 2001; Nguyen et al., 2006). Conditional deletion of NOTCH1 in the skin resulted in a significant increase of the basal epidermal layer (Rangarajan et al., 2001). Consistent with a tumor-suppressive function for Notch in the skin, NOTCH1 loss of function resulted in spontaneous basal cell carcinomas that appeared in older mice and sensitization to chemically induced skin carcinogenesis (Nicolas et al., 2003). This work also suggested that Notch acts as a tumor suppressor in the skin through suppression of the Wnt and Sonic\\u2013hedgehog pathways. A subsequent study indicated that the tumorigenic effect of Notch1 deletion is the result of a non\\u2013cell autonomous defect in the integrity of the skin barrier (Demehri and Kopan, 2009). Thus, mechanistically, tumor inhibition in the skin may involve feedback with the microenvironment in addition to cross talk between Notch and other signaling pathways.  In this issue of the Journal of Experimental Medicine, Viatour et al. (2011) propose a novel tumor suppressor role for Notch signaling in hepatocellular carcinoma (HCC). HCC is one of the most devastating cancers, with >600,000 deaths/yr worldwide, and is strongly associated with prior hepatitis virus B or C infection. To gain further insights into the mechanism driving initiation and progression of HCC, the authors generated a mouse model of the disease by deleting the retinoblastoma protein (RB) and its two related family members p107 and p130 in mouse liver. These triple KO (TKO) mice developed liver cancer with histological and molecular features typical of human HCC. In their model, inactivation of the RB pathway led to the expansion of the stem/progenitor compartment in the liver. The authors propose that these adult progenitor cells are the tumor-initiating cells of HCC after RB inactivation. In corroboration with previous findings showing that hyperactivation of E2F and Myc signals are sufficient to induce HCC, both pathways were up-regulated in the TKO mice.  Using whole transcriptome profiling and gene set enrichment analysis, Viatour et al. (2011) showed that the Notch pathway was also up-regulated in TKO mice, suggesting an oncogenic role for Notch signaling in HCC development. Unexpectedly, inhibition of Notch signaling in TKO mice using DAPT, a potent \\u03b3-secretase inhibitor, led to accelerated HCC development. And enforced activation of Notch signaling using ICN1 led to cell cycle arrest and apoptosis in primary HCC cells isolated from TKO mice, as well as in human HCC cell lines. To further address the clinical relevance of these observations, the authors looked at Notch activation status in a cohort of patients. They found that patients with better survival showed significantly higher expression of Notch-related genes, including HES1. Taken together, these data strongly support a potential tumor suppressor role for Notch signaling in HCC.  Our laboratory has recently found that conditional Notch loss-of-function through the deletion of Nicastrin (NCSTN), an essential component of the \\u03b3-secretase complex, or compound deletion of NOTCH1 and NOTCH2, resulted in a myeloproliferative syndrome with common features of the human disease chronic myelomonocytic leukemia (CMML; Klinakis et al., 2011). Whole transcriptome analysis revealed that Notch signaling inhibited a monocytic/granulocytic differentiation program in an early multipotential progenitor. This was at least partially mediated by direct repression of the PU.1 and C/EBP\\u03b1 promoters by HES1. Sequencing of Notch pathway genes revealed that \\u223c12% of CMML patients harbored inactivating mutations in NCSTN, MAML1), APH1A, or NOTCH2. These mutations were unique to CMML and were not found in other myeloproliferative disorders such as Polycythemia vera and myelofibrosis. Analogous to the tumor-suppressive function of Notch in epithelial cells and HCC, these studies suggested that Notch signaling may also act to prevent uncontrolled proliferation and transformation of myeloid cells during hematopoietic development.  Furthermore, two recent studies of head and neck squamous cell carcinoma (HNSCC), the sixth most common cancer worldwide, identified mutations affecting Notch receptors. Agrawal et al. (2011) identified 28 different NOTCH1 mutations in 21/120 patients (17.5%). 11 of these mutations were nonsense or insertion/deletions predicted to result in loss of function, supporting a tumor-suppressive function for Notch in HNSCC. The remaining 17 were missense mutations, mostly within the extracellular EGF-like repeats that are required for receptor\\u2013ligand interaction. A study by Stransky et al. (2011) identified NOTCH1 mutations in 11% of patients analyzed and NOTCH2 or NOTCH3 mutations in an additional 11% of patients. Mutations identified in this study were nonsense, missense, or insertion/deletions targeting the extracellular domain of the Notch receptors and therefore predicted to be loss-of-function mutations. The significance of these mutations in HNSCC requires further validation; nevertheless, they implicate NOTCH1 as a tumor suppressor in HNSCC.  Finally, in B cell malignancies, Notch was also reported to suppress growth and induce apoptosis, providing additional evidence that Notch could act as a tumor suppressor in hematopoietic cells (Zweidler-McKay et al., 2005). Another recent study suggested a similar tumor suppressor role for Notch signaling in neuroblastoma (Zage et al., 2011). However these data were mostly generated from in vitro studies using enforced overexpression of ICN1 or stimulation with recombinant Notch ligand and will require further in vivo validation. It will thus be important to test the role of Notch in these disease models using in vivo genetic approaches.  Conclusion/future directions  Given that Notch is involved in an array of fundamental processes both during embryonic development and in adult tissues, it is perhaps not surprising that aberrant Notch signaling can result in a wide range of pathological consequences. The oncogenic function of Notch in lymphocytes and mammary tissue, versus the growth-suppressive role in HCC, CMML, HNSCC, and skin, highlights the intriguing dual role of a single signaling pathway (see Table I for a listing of Notch function in selected cancer types). Indeed, depending on the cellular context, Notch may promote stem cell maintenance or induce terminal differentiation.  Table I. Table I. Dual role of Notch signaling in cancer The detailed mechanistic explanation of this duality of action remains under investigation. We propose that in both cases (oncogenic and tumor suppressive function), Notch signaling mainly targets programs of stem and progenitor cell differentiation, acting as a cell fate determinant. By affecting normal differentiation, Notch could set the stage for additional mutations and eventual cell transformation. For example, in myeloid leukemia, defective Notch signaling (caused either by mutations or by gene silencing) commits stem cells and multipotential progenitors to the granulocytic/monocytic progenitor (GMP) fate, expanding the pool of putative leukemia-initiating cells (LICs). Additional oncogenic lesions, such as TET2 mutations (Moran-Crusio et al., 2011), could transform these cells and lead to the initiation of monocytic/granulocytic leukemia. On the flip side, Notch1-activating mutations direct progenitors toward the T cell lineage but are not sufficient for the induction of T-ALL in the absence of additional oncogenic lesions. Another possibility could be that Notch signaling is involved in terminal differentiation of multipotential progenitors, suppressing the accumulation of a potential cancer-initiating population. This latter mechanism could be involved in the induction of HCC, as Notch signaling seems to be activated in committed progenitors, defined by Viatour et al. (2011) as an HCC-initiating population, and not before their commitment.  Further detailed studies are required to integrate such hypotheses, including a detailed analysis of Notch receptor interactions with ligands in specialized tissue microenvironments. Moreover, with growing interest in clinical applications of \\u03b3-secretase inhibitors (GSIs) and blocking antibodies to Notch ligands, it is vital to understand all possible systemic consequences of Notch/RBP-J inhibition (Real et al., 2009). Whereas inhibition of Notch may have clinical efficacy where Notch has an oncogenic role, activation of Notch using peptides or antibodies should be evaluated as a therapeutic target in malignancies where NOTCH plays a tumor-suppressive role. Very rare cases of human T cell acute lymphoblastic leukemia (T-ALL) harbor chromosomal translocations that involve NOTCH1, a gene encoding a transmembrane receptor that regulates normal T cell development. Here, we report that more than 50% of human T-ALLs, including tumors from all major molecular oncogenic subtypes, have activating mutations that involve the extracellular heterodimerization domain and/or the C-terminal PEST domain of NOTCH1. These findings greatly expand the role of activated NOTCH1 in the molecular pathogenesis of human T-ALL and provide a strong rationale for targeted therapies that interfere with NOTCH signaling. T-ALL is an aggressive cancer that preferentially affects children and adolescents. It is commonly associated with acquired chromososomal translocations and other genetic or epigenetic abnormalities, which lead to aberrant expression of a select group of transcription factors (1). NOTCH1 was discovered as a partner gene in a (7;9) chromosomal translocation found in <1% of T-ALLs (2). It encodes a transmembrane receptor that is required for the commitment of pluripotent progenitors to T cell fate (3) and the subsequent assembly of pre\\u2013T cell receptor complexes in immature thymocytes (4).  Cleavage of pro-NOTCH1 by a furinlike protease during transit to the cell surface (5) produces a NOTCH1 heterodimer comprised of noncovalently associated extracellular (NEC) and transmembrane (NTM) subunits (6). The heterodimerization domain (HD) responsible for stable subunit association consists of a 103 amino acid region of NEC (HD-N) and a 65 amino acid region in NTM (HD-C) (7). Physiologic activation of NOTCH receptors occurs when ligands of the Delta-Serrate-Lag2 (DSL) family bind to the NEC subunit and initiate a cascade of proteolytic cleavages in the NTM subunit. The final cleavage, catalyzed by \\u03b3-secretase (8, 9), generates intracellular NOTCH (ICN), which translocates to the nucleus and forms a large transcriptional activation complex that includes proteins of the Mastermind family (10\\u201312).  Prior work has shown that enforced NOTCH1 signaling is a potent inducer of T-ALL in the mouse (13\\u201315) and is required to sustain the growth of a human t(7;9)-positive T-ALL cell line (16). To investigate the possibility of a more general role for NOTCH signaling in human T-ALL, we tested T-ALL cell lines lacking the t(7;9) for NOTCH dependency by treating these cells with a \\u03b3-secretase inhibitor (17). Of 30 human T-ALL cell lines tested, 5 showed a G0/G1 cell-cycle arrest that equaled or exceeded that of T6E, a reference NOTCH1-dependent murine T-ALL cell line (Fig. 1A). This drug-induced growth suppression was abrogated by retroviral expression of ICN1 (Fig. 1B) and reproduced (fig. S1) by retroviral expression of dominant negative Mastermindlike-1 (16). These results indicated that the growth of these five cell lines depends on NOTCH-transduced signals.    Download high-res image  Open in new tab  Download Powerpoint Fig. 1. Identification of T-ALL cell lines that require NOTCH signals for growth. (A) Effects of compound E, a \\u03b3-secretase inhibitor (GSI), on cycling cell fractions (S and G2/M). After treatment for 4 to 8 days with compound E (1 \\u03bcM) or dimethyl sulfoxide (DMSO) carrier (mock), the DNA content of propridium iodide-stained cell populations was determined by flow cytometry. (B) Abrogation of \\u03b3-secretase inhibitor-induced cell-cycle arrest by MSCV-GFP-ICN1 retrovirus. Starting 2 days after transduction with empty MSCV-GFP or MSCV-GFP-ICN1, cells were treated with either compound E (GSI, 1 \\u03bcM) or DMSO carrier (mock) for 7 to 10 days. After staining with DRAQ5, the proportion of growth-arrested cells (G0 + G1 fractions) in the GFP\\u2013 and GFP+ subpopulations was determined by flow cytometry.  Because physical dissociation of the NOTCH extracellular domain has been linked to receptor activation (6, 18), we reasoned that the HD domain of NOTCH1 (7) could be the site of gain-of-function mutations. A second logical candidate region for oncogenic mutations is the negative regulatory PEST sequence lying at the C terminus of the NOTCH1 NTM (19), as retroviral insertions that cause deletion of this region have been reported in murine T-ALL (14, 15). Remarkably, sequencing revealed mutations that involve both the HD-N domain and the PEST domain in four of the five NOTCH-dependent cell lines (summarized in Fig. 2). Missense mutations affecting HD-N caused nonconservative changes at amino acid positions that are invariant in vertebrate NOTCH1 receptors (fig. S2). One cell line, DND-41, had two different HD-N mutations within the same NOTCH1 allele. The PEST mutations were short insertions or deletions causing shifts in reading frame that are predicted to result in partial or complete deletion of the PEST domain (fig. S4). Sequencing of cDNAs revealed that the HD-N and PEST domain mutations lie in cis in the same NOTCH1 allele in each of the four cell lines tested (fig. S5). Normal NOTCH1 cDNA clones were also identified in each cell line, indicating that both alleles are expressed. This is consistent with Western blot analysis (Fig. 3), which revealed that cell lines with HD-N and PEST domain mutations contained a polypeptide of the expected size of NTM and additional smaller polypeptides. We also sequenced a subset of T-ALL cell lines that were insensitive to the \\u03b3-secretase inhibitor. This revealed NOTCH1 mutations in 9 out of 19 non-responsive cell lines, including three lines with dual HD and PEST domain mutations (table S1). The failure of all cell lines with mutations to respond to \\u03b3-secretase inhibitors may result from these cell lines having been maintained in tissue culture for many years.    Download high-res image  Open in new tab  Download Powerpoint Fig. 2. NOTCH1 HD and PEST domain mutations in human T-ALL. A schematic representation of human NOTCH1 shows the distribution and frequency of HD-N, HD-C, and PEST domain mutations in NOTCH-dependent T-ALL cell lines (black arrowheads) and primary T-ALL samples (white arrowheads). NEC, NOTCH1 extracellular domain; LNR, Lin/NOTCH repeats; HD-N and HD-C, N-terminal and C-terminal halves, respectively, of the heterodimerization domain; N\\u2122, NOTCH transmembrane subunit; TM, transmembrane domain; RAM, RAM domain; ANK, ankyrin repeat domain; TAD, transcriptional activation domain. The arrow denotes the site of furin cleavage.    Download high-res image  Open in new tab  Download Powerpoint Fig. 3. Western blot analysis of lysates from NOTCH-dependent T-ALL cell lines. Whole-cell extracts were analyzed with a rabbit polyclonal antibody raised against the transcriptional activation domain of NOTCH1 (13). The lower panel is a longer exposure of the same blot that accentuates the presence of additional NOTCH1 polypeptides of smaller size than the normal NTM. Each lane contains 25 \\u03bcg total protein.  We also identified frequent NOTCH1 HD and PEST domain mutations in primary T-ALL samples obtained from the bone marrow of 96 children and adolescents at the time of diagnosis (summarized in Fig. 2). At least one mutation was identified in 54 tumors (56.2%). Mutations were seen in tumors associated with misexpression of HOX11 (2 of 3 cases), HOX11L2 (10 of 13, or 77%), TAL1 (12 of 31, or 39%), LYL1 (9 of 14, or 64%), MLL-ENL (1 of 3), or CALM-AF10 (1 of 2) (table S2), which together define the major molecular subtypes of T-ALL (1). The HD domain mutations in primary tumors were clustered in a \\u201chot spot\\u201d spanning residues 1574 to 1622 of HD-N and included each of the three L to P missense mutations originally identified in the NOTCH-dependent T-ALL cell lines, as well as deletions of 1 to 2 residues and short \\u201cin-frame\\u201d insertions (fig. S2). In addition, a smaller number of missense mutations were observed in HD-C, which again involved highly conserved residues (fig. S3). PEST domain mutations included insertions or deletions that induced a shift in reading frame and point mutations that created premature stop codons (fig. S4). In contrast with T-ALLs, B cell ALLs (B-ALLs) (n = 89) showed no mutations in these regions of NOTCH1 (20). Mutations were also absent from four remission bone marrow samples obtained from patients whose T-ALLs harbored NOTCH1 mutations (20), indicating that these mutations are acquired within the malignant clones.  To prove that HD domain mutations found in T-ALL patients have effects on function, NOTCH-sensitive reporter-gene assays were conducted in human U2OS cells (Fig. 4). Single L to P mutations within the HD-N domain at residues 1575, 1594, or 1601 caused a 3- to 9-fold increase in luciferase activity, whereas a T-ALL\\u2013associated PEST deletion at position 2471 resulted in \\u223c1.5- to 2-fold increase. More strikingly, each HD mutation and the same PEST domain truncation in cis resulted in 20- to 40-fold increases; in contrast, the same mutations in trans produced lesser effects close to the average of each mutation acting alone. The synergistic interaction of HD and PEST domain mutations in cis is consistent with a model (fig. S6) in which (i) HD domain mutations enhance \\u03b3-secretase cleavage and increase the rate of production of ICN1 and (ii) truncations that remove the PEST domain increase ICN1 half-life (19). The intermediate levels of activation produced by these mutations in trans presumably reflect competition between relatively weak and strong gain-of-function NOTCH1 polypeptides for factors required for processing and signaling. The stimulatory effects of mutated transmembrane NOTCH1 polypeptides were completely abrogated by a \\u03b3-secretase inhibitor (Fig. 4), which indicates a requirement for proteolysis at the juxtamembrane site of NTM for signal transduction. In contrast, the stimulation produced by ICN1, which is constitutively nuclear, was unaffected by \\u03b3-secretase inhibition (Fig. 4).    Download high-res image  Open in new tab  Download Powerpoint Fig. 4. HD and PEST domain mutations activate NOTCH1 signaling synergistically. Human U2OS cells were transiently co-transfected in 24-well format with the indicated pcDNA3 plasmids, a NOTCH-responsive luciferase reporter gene, and an internal Renilla luciferase internal control plasmid, as described previously (13). Twenty-five ng of pcDNA3 plasmid was used per well, except for experiments with pcDNA3-ICN1, in which 5 ng of plasmid per well was used. \\u201cDeltaP\\u201d denotes the presence of a deletion removing NOTCH1 residues 2473 to 2556. Normalized luciferase activities in whole-cell lysates were determined in triplicate and expressed relative to the activity in lysates prepared from vector-transfected cells. Error bars represent standard deviations.  Several factors may explain the high frequency of NOTCH1 mutations in T-ALL. The requirement for NOTCH1 signals during several stages of normal early T cell development provides a functional basis for its frequent involvement. Unlike the t(7;9), which is created during attempted V-D-J\\u03b2 rearrangement in committed T cell progenitors, the common point mutations and insertions described here could occur in multipotent hematopoietic progenitors, which normally express NOTCH1 (21). This would be predicted to induce daughter cells to adopt a T cell fate (22) and thereby increase the pool of cells at risk for additional leukemogenic events, such as synergistic mutations affecting NOTCH1 and misexpression of other critical transcription factors. The NOTCH1 mutations we describe here are currently specific to human T-ALL among vertebrate cancers, but a mutation involving the putative HD domain of the NOTCH homolog GLP-1 causes massive germ-cell proliferation in Caenorhabditis elegans (23), which suggests that such mutations have a highly conserved capacity to cause abnormal growth in specific cellular contexts.  Of potential clinical relevance, our findings identify the NOTCH pathway as a rational molecular therapeutic target in T-ALL. Although up to 75% of T-ALL patients are currently cured with very intensive cytotoxic chemotherapy regimens (24), new therapies are needed for patients with refractory disease, and less toxic, more efficacious drug combinations would be generally beneficial. Potent, specific inhibitors have already been developed (25) as a result of the involvement of \\u03b3-secretase in the production of amyloidogenic peptides in patients with Alzheimer's disease. On the basis of the results reported here, clinical trials are warranted to test the efficacy and potential side effects (26) of this class of NOTCH-pathway inhibitor in patients with T-ALL.\",\n          \"ABSTRACT Germline mutations that inactivate the tumor suppressor gene BRCA1 are associated with an increased risk of cancers of the breast and other tissues, but the functional consequence of many missense variants found in the human population is uncertain. Several predictive methods have been proposed to distinguish cancer-predisposing missense mutations from harmless polymorphisms, including a small colony phenotype (SCP) assay performed in the model organism, yeast (Saccharomyces cerevisiae). The goal of this study was to further evaluate this colony size assay. We constructed 28 missense mutations throughout the C-terminal 305 amino acid residues of BRCA1. Mutated proteins were expressed in yeast and evaluated using the SCP assay. We conclude there is as yet no evidence the assay can identify inactivating mutations upstream of the BRCT repeats. However, within and between the BRCT repeats, results of the assay are in general agreement with predictions based on structural modeling, other in vitro and in vivo assays, and cross-species sequence conservation. Thus, the yeast assay appears to provide confirmatory in vivo evidence to aid in characterizing some BRCA1 missense variants. INTRODUCTION Approximately 2.5\\u20135% of breast cancer cases can be attributed to germline mutations in the BRCA1 gene,1,2 which also confer increased risk of ovarian and prostate cancer. 3,4 BRCA1 has the hallmarks of a tumor suppressor gene. The wild type allele is characteristically lost in tumors of germline mutation carriers.5,6 Also, transfection of wild type BRCA1 into human tumor cell lines results in growth inhibition through cell cycle arrest.7,8 Thus, germline BRCA1 mutations are expected to result from gene inactivation. Indeed, over 80% of BRCA1 sequence variants catalogued in the Breast Cancer Information Core (BIC) database9 are frameshift or nonsense mutations resulting in premature translation termination, regulatory mutations, or splice site alterations.10 Since the deletion of only eleven C-terminal amino acid residues is known to result in predisposition to breast cancer,11 most or all of these mutations very likely inactivate BRCA1 function as well, as has been confirmed in numerous cases.9 The risk associated with missense variants, of which over 350 have been described, is much more difficult to predict. Many at-risk individuals currently receive a BRCA1 genetic test result that is ambiguous.12 In a few cases, the testing of samples from many family members provides evidence of linkage between a particular missense allele and increased cancer risk. However, genetic evidence is often either unavailable or inconclusive. About half of the mutations reported to the BIC are from a single family or individual, and because noninherited breast cancer is common, clusters of affected family members may occur by chance. This uncertainty has substantial health implications. Confirmed BRCA1 mutation carriers may elect presymptomatic interventions such as prophylactic mastectomy or oophorectomy and/or more frequent and aggressive screening.13 Should cancer be discovered, increasing evidence suggests that BRCA1-associated tumors are genetically and phenotypically distinct from sporadic tumors and respond differently to treatment.14-18 Thus, a test or series of tests capable of distinguishing cancer-predisposing BRCA1 alleles from neutral polymorphisms would be of great value. Despite extensive study, the precise biochemical function of the BRCA1 protein is unclear. Through its association with proteins involved in DNA repair, transcription and chromatin modification, it has been proposed to play a role in the maintenance of genome integrity.19-22 This role is likely to be multi-faceted, making a biochemical assay for BRCA1 functionality challenging. In the absence of conclusive genetic or biochemical means of classification, most missense variants of BRCA1 in the BIC database are listed as \\u201cunclassified variants.\\u201d Tentative predictions of functionality can be made on the basis of sequence conservation between www.landesbioscience.com Cancer Biology & Therapy 453 species and structural modeling. Complete sequences of BRCA1 orthologs are available for the chimpanzee,23 dog,24 mouse,24 rat,25 cow,26 chicken,27 and African clawed frog, Xenopus laevis. 28 Alteration of residues highly conserved between these species and humans are expected to be more likely to affect BRCA1 function. A comparison of several additional species has been made for an internal region of the gene.29 Structural predictions can be made for two regions of the BRCA1 protein: the N-terminal RING finger30 and the two C-terminal BRCT domains. The three-dimensional structures of the two BRCT domains of the human and rat have been solved by X-ray crystallography.31,32 Each BRCT domain contains a four-stranded parallel \\u03b2 sheet surrounded by three \\u03b1 helices. Sequence alterations predicted to disrupt the folding of one of the BRCT domains can be considered more likely to affect the protein\\u2019s function and therefore to be diseaseassociated. Direct in vitro biophysical measurement of the structural consequences of some missense mutations has also been reported.33,34 While each of these predictive methods is valuable, an in vivo assay would provide additional information. It has been shown that transfected wild type BRCA1 decreases sensitivity to ionizing radiation of a BRCA1- mouse cell line, whereas disease-associated missense alleles do not.35 Missense alleles have also been shown to affect chromatin unfolding in an engineered mammalian cell line.36 These assays shows great promise, but are complex, and results from them have as yet been reported only for a small number of already classified alleles. Other, simpler assays may also contribute to characterizing the growing number of sequence variants. Two in vivo activities associated with the C-terminal region of BRCA1 have been identified when this region is expressed in the yeast Saccharomyces cerevisiae. First, fused to a DNA-binding domain, it allows transcriptional activation of heterologous promoters.37-41 Second, it confers a small colony phenotype (SCP) on yeast transformants.42 This growth phenotype, discovered serendipitously in the course of two-hybrid experiments, is enhanced by fusion of the BRCT domains to the transcription activation domain of yeast GAL4 and the SV40 large T antigen nuclear localization signal. Expression in yeast of human BARD1, which contains two BRCT domains, also results in growth inhibition.43 A further upstream region of BRCA1 (amino acid residues 1293-1559) also exerts a negative effect on yeast growth as well as a transcription activation function in yeast and mammalian cells.44 Thus, these two yeast phenotypes may be mechanistically related. Both of the BRCA1 yeast phenotypes are abolished by truncations or missense mutations known to be disease-associated, but are insensitive to mutations known to represent benign polymorphisms. The phenotypic differences do not result from altered protein stability, as all the variants were shown by western blotting to be present in equivalent amounts.37,39,42 Although the biochemical basis of the small colony phenotype is not yet understood, its manifestation appears to rely on the expression of a functionally intact BRCA1 C-terminal region. Apparently, both BRCT domains are required, since truncation of the C-terminal domain or missense mutations in either abrogates the phenotype. It has therefore been proposed that the small colony and transcriptional activation phenotypes may offer simple assays predicting the functional status of BRCA1 sequence variants in the BRCT region.45 The purposes of the experiments reported here were to further characterize the SCP assay and to evaluate its potential in BRCA1 sequence variant classification. MATERIALS AND METHODS Site-Directed Mutagenesis. To construct each desired mutation, two complementary oligonucleotide primers were synthesized, containing the required base pair change(s) in the center. One of two alternate methods was then used. The Splice Overlap Extension method was performed as described by Vallejo et al46 using the mutagenic primers and the flanking primers ACTSEQ2 (5\\u2019-CTAGCTTGGGTGGTCATATG-3\\u2019) and ACTSEQ3 (5\\u2019-CGGGGTTTTTCAGTATCTACG-3\\u2019) and the plasmid pADBRCA1\\u2206242 as template. The mutated BRCA1 PCR product was digested with EcoRI and BamHI and inserted into the same sites of the plasmid pACT2\\u2019.42 Alternatively, the \\u201cQuick Change\\u201d method was applied directly on the plasmid pAD-BRCA1\\u22062, as recommended by the manufactuer (Stratagene, Inc.). DNA Sequencing. For each plasmid, the entire BRCA1 insert was sequenced on both strands to confirm the presence of the targeted nucleotide substitution and the absence of any untargeted mutations, using either an ABI 310 or ABI 377 automated DNA sequencer (Applied Biosystems, Inc.). Sequences were assembled using Sequencher software (Gene Codes Corp.). Yeast Transformation. The S. cerevisiae strain L40 (MATa his3D200 trp1-901 leu2-3,112 ade2 LYS2::(lexAop)4-HIS3 URA3: :(lexAop)8-lacZ)47 was transformed using standard lithium acetate procedures48 and spread on synthetic yeast medium lacking leucine. Colonies were grown approximately 48 hours at 30\\u02daC before restreaking. Colony Size Quantification. Two randomly selected colonies from each initial transformation plate were restreaked and incubated at 30\\u02daC exactly 54 hours. Two or more colonies from each streak were separately resuspended in sterile distilled water using a pipettor. The cells were counted using a hemocytometer. RESULTS Using PCR-mediated site-directed mutagenesis, we introduced 28 missense variants within the C-terminal 305 amino acid residues of BRCA1 (Table 1). Each variant was inserted into a translational fusion expression vector, as previously described.42 Each plasmid, as well as several controls, was transformed into the yeast strain L40.47 Expression of the SCP was found to be quite strain-dependent, and L40 provided a clearly visible distinction between small and large colonies (Fig. 1). Colony size was determined as described in Materials and Methods. The results are presented in Figure 2. Mutations Upstream of the BRCT Repeats. Humphrey et al.42 reported that most of the growth inhibition is associated with expression of the C-terminal 213 amino acid residues (1651-1863), which contain the BRCT domains. However, expression of a longer BRCA1 fragment (305 C-terminal residues; 1559-1863) caused an additional two-fold reduction in colony size. Mammalian cross-species sequence conservation in this 92 residue region upstream of BRCT-N (the N-terminal BRCT) is considerably lower than within the BRCT domains, and no meaningful sequence alignment is apparent between human BRCA1 and the chicken and Xenopus orthologs.27,28 No confirmed disease-associated missense allele has been mapped to this region. We tested whether missense mutations within this region could detectably alter the SCP. One previously tested mutation (P1637L)42 and ten new mutations in this region were tested. As shown in (Table 1), seven of these eleven alleles have been detected in humans. The other four, including the double mutation LK1600/01EE, were selected to drastically alter (from hydrophobic or basic to acidic) residues conserved between the six sequenced mammalian orthologs. As shown in Figure 2, none of these eleven mutations caused a significant increase in the size of the yeast colonies. One of the mutations, S1613C, has been reported to abolish transcriptional activation in yeast.40 We conclude from our results that there is as yet no evidence that the SCP assay can identify inactivating missense mutations in the region upstream of the BRCT domains. In contrast, at least one predicted inactivating mutation has been mapped to this region using the yeast transcription activation assay. YEAST ASSAY FOR BRCA1 MISSENSE VARIANT CHARACTERIZATION 454 Cancer Biology & Therapy 2004; Vol. 3 Issue 5 www.landesbioscience.com Cancer Biology & Therapy 455 YEAST ASSAY FOR BRCA1 MISSENSE VARIANT CHARACTERIZATION Mutations within the BRCT Region. Eleven of the eighteen novel point mutations within the BRCT region showed increased yeast colony size, compared with yeast expressing wild type BRCA1, indicating a probable disruption of BRCT structure and/or function. Among these mutations, there was an approximately 2.5-fold range of colony size. The previouslytested control 11-residue truncation allele Y1853X showed the greatest difference from wild type BRCA1. It is possible that colony size correlates with severity of the missense mutations, but further study will be required to determine this. The correlation between level of cross-species sequence conservation and effect of mutation on the SCP is striking. Mutations altering eight of the nine residues conserved in all eight species (Table 1) result in abrogation of the SCP. V1653M is the one exception. Conversely, mutations altering only two (1652 and 1783) of the nine positions at which multiple residues are \\u201callowed\\u201d result in such abrogation. Substitutions at position 1652 can result in opposite effects in the yeast transcription activation assay, depending on the substituted residue; M1652K abolishes transcription activation,38 whereas M1652I has no effect.40 M1652I also has no effect on the SCP, 42 whereas our results suggest M1652K and M1652T are deleterious. Thus, it appears mutations at this position need to be especially carefully evaluated. Furthermore, results of the SCP assay for these eighteen alleles are consistent with nearly every other prediction that has been made, based on structural modeling, in vitro biophysical characterization and/or transcription activation phenotype (see allele-specific references in Table 1). The only exceptions are four disagreements with the predictions of Mirkovic et al.49 which are based on a set of structure-based rules. D1739E, M1783T and V1833M were predicted by these rules to be likely benign replacements, whereas the SCP assay indicates possible functional and/or structural defects. Conversely, the structure-based algorithm predicts a deleterious effect of V1804D, a mutation that shows a similar SCP to wild type BRCA1. Of particular interest is the modest aspartate to glutamate substitution at residue 1739, between the two BRCT repeats. Although the human crystal structure in the region of this residue is unresolved,31 the corresponding region of the rat structure shows a short \\u03b2 sheet, predicted to stabilize the interface of the two BRCT repeats.32 Determination of the structural consequences of this modest substitution may reveal the basis for its effect on the SCP and resolve the status of its likely disease association. DISCUSSION Our results suggest that the structural integrity of the BRCT region (approximately the most C-terminal 213 amino acid residues) of BRCA1 is the most critical determinant in the yeast SCP assay. None of the tested missense mutations in the 92 residue region upstream of BRCT-C, including ones affecting conserved positions and one that abolishes transcription activation in yeast and mammalian cells, has a detectable effect on growth. Thus, there is as yet no evidence that the SCP assay may be useful in characterizing variants upstream of residue 1652. Within the BRCT region, mutations that abolish the SCP are interspersed with those having no effect. Seven mutations at five positions display an SCP similar to wild type BRCA1. All but one (V1653M) alter residues incompletely conserved between the orthologs of eight vertebrate species and, in fact, three result in substitution of a residue allowed in another species at the corresponding position (T1720I, V1804A, P1806A). With one exception (V1804D49), the SCP assay predictions for these alleles are consistent with all other available predictions. Thus, the evidence favors assignment of these variants as benign polymorphisms. However, these amino acid substitutions may affect functional attributes not detectable by the SCP assay; therefore, Figure 1. Relative colony sizes. The yeast strain L40 was transformed with various plasmids. Individual transformants were restreaked into sectors of the same plate and incubated at 30\\u02daC for 54 hours. Clockwise starting at top: pAD-BRCA1\\u22062 (wild type), G1788V, K1601E, pACT2\\u2019 (vector). Figure 2. Quantification of colony sizes. Number of cells per yeast colony transformed with the vector pACT2\\u2019, pAD-BRCA1\\u22062 (wild type BRCA1), or the BRCA1 plasmid carrying the indicated mutation. Error bars indicate standard deviation. its value in characterizing polymorphisms is uncertain. Elucidation of the biochemical basis for the SCP and its relation to BRCA1\\u2019s function in human cells may help clarify this issue. In contrast, we identified eleven variants at nine positions within BRCT-N, BRCT-C or in the interval between them that alleviate the SCP. All eleven affect residues completely conserved between the BRCA1 genes of all eight vertebrates, with two exceptions: M1652 and M1783 are both substituted with I in the chicken.27 All have been predicted by various means to disrupt folding and/or function of the BRCT domain. Thus, the SCP assay provides confirmatory in vivo evidence for the deleterious nature of these mutations. Although caution must be exercised in applying these results directly to clinical decision-making, we conclude that the yeast SCP assay may be useful in conjunction with other information to evaluate selected alleles. This confirmation of the potential value of the SCP assay highlights the need to understand its biochemical basis. This understanding Table 1 MUTATIONS CHARACTERIZED IN THIS STUDY Secondary Structure, Predicted Structural References for Functional Mutation Allowed Residues* Detected in Humans Consequences Predictions Upstream of BRCT L1600E L N Unknown K1601E K N \\u201c LK1600/01EE LK N \\u201c K1606E Q, K, R Y \\u201c A1608V A, E Y \\u201c S1613C S, N Y \\u201c 40 A1615D A N \\u201c M1628V M, S, V, R Y \\u201c M1628T M, S, V, R Y \\u201c P1637L P Y \\u201c 42 A164IT A, S Y \\u201c BRCT Region M1652T M, I, F Y Adjacent to \\u03b2 sheet 1, BRCT-N 32, 40#, 50 M1652K M, I, F N \\u201c 32, 38, 40#, 49, 50 V1653M V Y \\u201c 49, 50 S1655F S Y \\u201c 50 T1720A T, V, I Y \\u03b1 helix 3, BRCT-N 38#, 49, 50 T1720I T, V, I N \\u201c 38#, 49, 50 V1736A V Y Between BRCT domains 38#, 49 D1739E D Y \\u201c 49 D1739G D Y \\u201c 49, 50 M1775R M Y \\u03b1 helix 1, BRCT-C/ 31-34, 36, 37, 40, 49, 51 surface, disrupts interface M1783T M, I Y \\u03b1 helix 1, BRCT-C/ 34, 49-51 intrahelical, destabilizes fold G1788V G Y Loop between \\u03b1 helix 1 32, 34, 44#, 49-51 and \\u03b2 sheet 2, BRCT-C, incorrect folding V1804A V, T, A, S N Loop between \\u03b2 51 sheets 2 and 3, BRCT-C V1804D V, T, A, S Y \\u201c 49-51 P1806A P, L, A, S Y \\u201c 38, 49, 50 V1808A V Y \\u03b2 sheet 3, BRCT-C, 34, 49-51 destabilizes fold V1810G V Y \\u201c 49, 50 N1819Y N, S, A, T, R N \\u03b1 helix 2, BRCT-C 49, 51 V1833M V Y \\u03b2 sheet 4, BRCT-C, 34, 38#, 49-51 destabilizes fold Y1853X Y Y 11 residue truncation Those italicized and underlined abrogate the small colony phenotype, and thus are predicted to negatively affect BRCA1 structure and/or function. *Upstream of BRCT: allowed residues in mammals; within BRCT region: allowed residues in mammals, chicken and Xenopus. #This reference provides information on a different substitution at this residue, or an adjacent insertion. YEAST ASSAY FOR BRCA1 MISSENSE VARIANT CHARACTERIZATION 456 Cancer Biology & Therapy 2004; Vol. 3 Issue 5 www.landesbioscience.com Cancer Biology & Therapy 457 will not only further define the limits of the assay\\u2019s usefulness, but may also shed some light on the normal function of BRCA1 in human cells  Mutations in BRCA1 and BRCA2 account for the majority of hereditary breast and ovarian cancers, and therefore sequence analysis of both genes is routinely conducted in patients with early-onset breast cancer. Besides mutations that clearly abolish protein function or are known to increase cancer risk, a large number of sequence variants of uncertain signifi cance (VUS) have been identifi ed. Although several functional assays for BRCA1 VUSs have been described, thus far it has not been possible to conduct a high-throughput analysis in the context of the full-length protein. We have developed a relatively fast and easy cDNA-based functional assay to classify BRCA1 VUSs based on their ability to functionally complement BRCA1-defi cient mouse embryonic stem cells. Using this assay, we have analyzed 74 unclassifi ed BRCA1 missense mutants for which all predicted pathogenic variants are confi ned to the BRCA1 RING and BRCT domains.  SIGNIFICANCE: BRCA1 VUSs are frequently found in patients with hereditary breast or ovarian cancer and present a serious problem for clinical geneticists. This article describes the generation, validation, and application of a reliable high-throughput assay for the functional classifi cation of BRCA1 sequence variants of uncertain signifi cance. Cancer Discov; 3(10); 1142\\u201355. \\u00a92013 AACR.  Authors\\u2019 Affi liations: 1 Division of Molecular Pathology and Cancer Genomics Centre and 2 Department of Pathology, The Netherlands Cancer Institute Amsterdam, The Netherlands; and 3 TaconicArtemis GmbH, Cologne, Germany  Note: Supplementary data for this article are available at Cancer Discovery Online (http://cancerdiscovery.aacrjournals.org/).  Current address for C. Klijn: Genentech Inc., South San Francisco, California.  Corresponding Author: Jos Jonkers, The Netherlands Cancer Institute, Plesmanlaan 121, Amsterdam, 1066 CX, The Netherlands. Phone: 31-20- 5122000; Fax: 31-20-5122050; E-mail: j.jonkers@nki.nl doi: 10.1158/2159-8290.CD-13-0094 \\u00a92013 American Association for Cancer Research.  INTRODUCTION  Germline loss-of-function mutations in BRCA1 and BRCA2 are known to result in an approximately tenfold increased lifetime risk of developing breast or ovarian cancer. Thus far, no other genes have been identifi ed with such a strong link to hereditary breast and ovarian cancer (HBOC), and in the past decades, many women have been screened for germline mutations in BRCA1 or BRCA2 . This has resulted in the identifi cation of numerous pathogenic mutations as well as a large number of sequence variants for which the clinical relevance is not clear. In the most recent publication of the ENIGMA (Evidence-based Network for the Interpretation of Germline Mutant Alleles) group, an international consortium for the evaluation of BRCA1 or BRCA2 sequence variants, a total of 1,273 unique BRCA1 variants is mentioned ( 1 ), but this number may increase even further because of the implementation of high-throughput sequencing methods. Of the BRCA1 variants of uncertain signifi cance (VUS), 920 are nontruncating exonic mutations that may affect protein function or mRNA splicing, but for which there is not enough linkage information to indicate whether they are pathogenic. To aid genetic counseling of individuals with BRCA1/2 VUS, both genetic and functional classifi cation methods have been developed. Genetic analysis of BRCA1 or BRCA2 VUSs relies on cosegregation with disease, cooccurrence with known pathogenic mutations, and family history of cancer. These data have been integrated into computational models to calculate the likelihood that a VUS is disease-causing ( 2 ). Additional in silico analysis of the evolutionary conservation of the amino acids affected by the mutation and the predicted impact of the mutant amino acids on protein folding are also implemented in such models. Functional assays do not rely on preexisting data, but directly test the effect of BRCA1/2 VUSs on known functions of the encoded proteins ( 3 ). Although this may seem relatively straightforward, it can be diffi cult to extrapolate data from functional assays into cancer risks for patients. For instance, in most assays only part of the BRCA1 protein is analyzed. In addition, some of the more elegant assays are technically demanding and not suitable to analyze large numbers of mutations. We reasoned that a good functional assay should fulfi ll three basic requirements: (i) it should investigate the biologic effects of a BRCA1 VUS in the context of the full-length protein; (ii) it should be conducted under normal physiologic conditions in a noncancerous cell type; and (iii) it should be based on a highly standardized and reproducible protocol.  These considerations led us to develop a functional assay based on physiologic expression of full-length human BRCA1 cDNA in mouse embryonic stem cells that are genetically engineered to allow conditional deletion of endogenous Brca1 . Mutant BRCA1 cDNAs are generated using site-directed mutagenesis (SDM) and introduced in a defi ned genomic locus of mouse embryonic stem cells by recombinase-mediated cassette exchange (RMCE). In this way, we have analyzed 86 BRCA1 variants for their effects on cell proliferation and drug sensitivity, including 74 clinically relevant VUSs.  RESULTS  Generation of Mouse Embryonic Stem Cells Expressing BRCA1 Sequence Variants  Although BRCA1-defi cient tumor cells proliferate rapidly in situ , loss of BRCA1 in normal cells leads to a severe proliferation defect ( 4 ). We decided to make use of this phenotype for the functional analysis of BRCA1 variants in Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  1144\\u2002|\\u2002CANCER DISCOVERY OCTOBER 2013 www.aacrjournals.org RESEARCH ARTICLE Bouwman et al. Brca1 -selectable conditional knockout (SCo) mouse embryonic stem cells ( 5 ). These cells carry one Brca1 -null allele and a selectable conditional Brca1SCo allele, which contains, in addition to loxP sites around exons 5\\u20136, a split puromycin resistance marker that is activated upon Cre-mediated deletion of exons 5\\u20136. They also contain a CreERT2 allele in the Rosa26 locus, which allows for controlled activation of Cre via administration of 4-hydroxy tamoxifen (4-OHT). Thus, BRCA1-defi cient embryonic stem cells can be easily obtained via 4-OHT\\u2013induced inactivation of the Brca1SCo allele and subsequent selection for puromycin resistance ( Fig.  1 ; ref.  5 ). To allow effi cient integration of human BRCA1 variants in one specifi c genomic locus, we supplied the other Rosa26 allele with F3 and Frt recombination sites for RMCE by the site-specifi c recombinase Flp ( Fig.  1 and Supplementary Fig. S1; ref. 6 ). Cells that have undergone successful RMCE can be selected because they express a truncated neomycin selection marker under control of the endogenous Rosa26 promoter, which further increases targeting effi ciency. Flpmediated recombination ensures single-copy integration of BRCA1 expression constructs at the same Rosa26 locus, thus avoiding position-effect variegation and copy number\\u2013 dependent differences in expression.  RMCE vectors were equipped with a human BRCA1 cDNA expression construct, which was modifi ed using SDM to introduce defi ned mutations in BRCA1 . The focus of our analysis was on Dutch and Belgian VUSs that were found in families with HBOC. We also included a number of variants  Figure 1. \\u2002 Schematic overview of the RMCE procedure in R26 CreERT2/RMCE ; Brca1 SCo/\\u0394 embryonic stem cells. Before the introduction of a human BRCA1 cDNA, R26 CreERT2/RMCE ; Brca1 SCo/\\u0394 embryonic stem cells are mouse BRCA1-profi cient and sensitive to both neomycin and puromycin. Targeting of singlecopy human BRCA1 cDNA variants to the Rosa26 locus by Flp RMCE results in expression of human BRCA1 and neomycin resistance. Addition of 4-OHT leads to CreERT2-mediated deletion of mouse Brca1 exons 5 and 6, resulting in loss of mouse BRCA1 protein and concomitant expression of puromycin from the phosphoglycerate kinase (PGK) promoter. This enables selection of mouse BRCA1-defi cient, human BRCA1-expressing R26 CreERT2/hBRCA1 ; Brca1\\u0394/\\u0394 embryonic stem cells that can be used in functional complementation assays. R26 CreERT2/RMCE;Brca1SCo/\\u0394 R26CreERT2/hBRCA1;Brca1SCo/\\u0394 R26CreERT2/hBRCA1;Brca1\\u0394/\\u0394 3 5\\u20136 3 5\\u20136 7\\u20138 3 3 14 7\\u20138 7\\u20138 3 7\\u20138 5\\u20136 Introduction of human BRCA1 by RMCE 3 14 SA SA Brca1SCo5\\u20136 Brca1\\u03945\\u201313 R26CreERT2 R26RMCE Brca1\\u03945\\u20136 Brca1\\u03945\\u201313 R26CreERT2 R26hBRCA1 Brca1SCo5\\u20136 Brca1SCo5\\u20136 Brca1\\u03945\\u201313 R26CreERT2 R26RMCE Brca1\\u03945\\u201313 R26CreERT2 R26hBRCA1 3 14 CreERT2 CreERT2 CreERT2 CreERT2 Puro Puro Puro PGK PGK PGK Puro PGK zsgreen Hyg FLP zsgreen Hyg FLP SA SA NeoR Selection hBRCA1\\u2013, mBrca1+, neoS, puroS hBRCA1+, mBrca1+, neoR, puroS hBRCA1+, mBrca1\\u2013, neoR, puroR pA hBRCA1 hBRCA1 EF1\\u03b1 EF1\\u03b1 \\u03945\\u2032 neo \\u03945\\u2032 neo \\u03945\\u2032 neo EF1\\u03b1 hBRCA1 Cre-mediated deletion of mouse Brca1 by 4-OHT addition 3 14 PuroR Selection SA SA SA SA Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  OCTOBER 2013CANCER DISCOVERY\\u2002|\\u20021145 Functional Analysis of BRCA1 Sequence Variants RESEARCH ARTICLE that were previously classifi ed using functional assays ( 7, 8 ) or a multifactorial likelihood model ( 9 ), as well as M1400V, L1407P, and M1411T, which have been reported to attenuate the interaction between BRCA1 and PALB2 ( 10 ). To allow validation of our functional complementation assay, we included a series of eight BRCA1 variants that are known to be deleterious or neutral according to the Breast Cancer Information Core database (BIC; http://research.nhgri .nih.gov/bic/ ; Supplementary Table S1). These controls include the well-known pathogenic BRCA1 founder mutations 185delAG and 5382insC and the neutral polymorphisms Y105C, R866C, and E1250K. The BIC designation is supported by the classifi cation according to the International Agency for Research on Cancer ( http://iarc.fr/ ), which includes the Align-GVGD score that indicates biophysical and evolutionary alterations ( http://agvgd.iarc.fr/ ; Supplementary Table S1; refs. 11, 12 ). Align-GVGD scores vary between C0 (likely neutral) and C65 (likely deleterious). Sequence-verifi ed constructs that contained the intended VUS, but no other BRCA1 mutations, were transfected into R26 CreERT2/RMCE;Brca1 SCo/\\u0394 embryonic stem cells to undergo RMCE ( Fig.  2 ). Neomycin-resistant clones were pooled and RMCE was confi rmed by PCR analysis. Subsequently, protein expression of the human BRCA1 variants was analyzed by Western blot analysis with a human BRCA1-specifi c antibody ( Fig. 2 and Supplementary Fig. S2). Comparison with BRCA1 protein levels in embryonic stem cells expressing human BRCA1 from a bacterial artifi cial chromosome (BAC), which is known to rescue embryonic lethality of Brca1 -null mice ( 7 ,  13 ), showed that transcription of the human BRCA1 cDNA from the EF1A gene promoter results in physiologic levels of BRCA1 protein (Supplementary Fig.  S1). Most BRCA1 variants were expressed at equal levels, allowing comparison of their functional activities. A number of C-terminal BRCA1 mutants showed low levels of expression (Supplementary Table S2; Supplementary Fig. S2). Real-time RT-PCR analysis showed that decreased protein expression was not caused by decreased mRNA expression (Supplementary Fig.  S3), suggesting that the low abundance of BRCA1 protein results from posttranslational events. In fact, for most of these variants, mutation-associated protein instability has already been documented in previous publications (refs. 8 , 14 , 15 ; Supplementary Table S2).  Functional Complementation Assay of BRCA1 Sequence Variants in Mouse Brca1 -Null Embryonic Stem Cells  As a fi rst functional test, we assayed the ability of BRCA1 variants to restore the proliferation defect of switched R26 CreERT2/RMCE ;Brca1 SCo/\\u0394 embryonic stem cells ( Fig.  2 ). Expression of endogenous mouse Brca1 was shut off through overnight induction of Cre activity by 4-OHT, and 7 days  Figure 2. \\u2002 Workfl ow for the functional classifi cation of BRCA1 sequence variants in Brca1 -null embryonic stem (ES) cells. Outline of the generation of mouse Brca1 -defi cient embryonic stem cells expressing human BRCA1 variants and functional complementation assays. Indicated are the experimental steps and the time it takes one person to analyze 20 mutants. Introduction of BRCA1 variants in R26CreERT2/RMCE;Brca1SCo/D ES cells BRCA1 Construct + G418 Selection of clones with correct RMCE BRCA1 Expression analysis SDM and sequencing RMCE in ES cells x + 4-OHT + Puromycin Functional assays Functional assays Cre-Mediated deletion of Brca1SCo allele Proliferation and cisplatin sensitivity assays Selection of Brca1\\u0394SCo/\\u0394 ES cells Timecourse for the analysis of 20 BRCA1 variants Weeks 0 1 2 7 3 8 4 5 6 SDM and RMCE Functional assays Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  1146\\u2002|\\u2002CANCER DISCOVERY OCTOBER 2013 www.aacrjournals.org RESEARCH ARTICLE Bouwman et al. after switching, cells were plated in 96-well plates for proliferation assays, and were analyzed using Sulphorhodamine B staining. For each group of mutants tested, we included positive and negative controls consisting of embryonic stem cells containing, respectively, wild-type human BRCA1 cDNA and an empty RMCE vector. BRCA1 variants were evaluated on their ability to support growth compared with these controls (Supplementary Tables S3 and S4). Although we tested 28 mutations in the central domain encoded by exon 11 (aa 224\\u20131366), BRCA1 variants that were unable to rescue the proliferation defect of Brca1 -null mouse embryonic stem cells to BRCA1 wild-type levels were confi ned to the conserved N- and C-terminal domains of BRCA1.  Cisplatin Sensitivity Assay for Classifi cation of BRCA1 Variants  Although the ability of BRCA1 variants to support proliferation appears to be indicative of VUS function, results were not always clear-cut (Supplementary Tables S3 and S4). BRCA1 is known to be important for DNA interstrand crosslink (ICL) repair through mechanisms that are both dependent ( 5 ) and independent ( 16 ) of its function in homologous recombination (HR). The role of BRCA1 in ICL repair is stressed by the occurrence of genetic reversion mutations restoring BRCA1 protein expression in platinum-resistant ovarian tumors in BRCA1 -mutation carriers ( 17, 18 ). We therefore decided to conduct a 96-well\\u2013based cisplatin sensitivity assay to allow a more stringent evaluation of BRCA1 VUS functionality.  Half-maximal inhibitory concentrations (IC 50 ) of cisplatin were determined using a resazurin cell viability assay, and BRCA1 variants were again classifi ed in comparison to wildtype BRCA1 and an empty RMCE vector. To obtain corrected cisplatin IC 50 values, we fi tted a log-logistic curve constrained at 1 and 0. We excluded a fi t if the residual squared error (RSE) exceeded 0.1. We then applied a Bayesian predictor to classify BRCA1 VUS as pathogenic or benign. Most variants that showed less than wild-type activity in the proliferation assay also scored as functionally impaired in the cisplatin sensitivity assay and were classifi ed as deleterious ( Table 1 ,  Table 1. \\u2002 Functional classifi cation of BRCA1 VUS based on cisplatin response Variant DNA change Type of mutation a Classifi cation S4F c.11C>T VUS Not Clear M18T c.53T>C VUS Deleterious 185delAG c.68_69delAG Deleterious control Deleterious b K45Q c.133A>C VUS Neutral C61G c.181T>G Deleterious control Deleterious C64G c.190T>G VUS Deleterious b D67Y c.199G>T VUS Neutral Y105C c.314A>G Neutral control Neutral N132K c.396C>A VUS Neutral P142H c.425C>A VUS Neutral L147F c.441G>C VUS Neutral L165P c.494T>C VUS Neutral R170W c.508C>T VUS Neutral S186Y c.557C>A VUS Neutral V191I c.571G>A VUS Neutral T231M c.692C>T VUS Neutral D245V c.734A>T VUS Neutral L246V c.736T>G VUS Neutral V271L c.811G>C VUS Neutral S308A c.922A>G; c.923G>C Artifi cial h Not Clear L358R; C360R; E362H c.1073T>G; c.1078T>C; c.1084G>C; c.1086G>T Artifi cial Neutral L668F c.2002C>T VUS Neutral D695N c.2083G>A VUS Neutral P798L c.2393C>T VUS Neutral N810Y c.2428A>T VUS Neutral T826K c.2477C>A VUS Neutral (continued) Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  OCTOBER 2013CANCER DISCOVERY\\u2002|\\u20021147 Functional Analysis of BRCA1 Sequence Variants RESEARCH ARTICLE (continued) Variant DNA change Type of mutation a Classifi cation R841Q c.2522G>A VUS Neutral Y856H c.2566T>C VUS Neutral R866C c.2596C>T Neutral control Neutral S988A c.2962T>G Artifi cial Neutral M1008I c.3024G>A VUS Neutral E1060A c.3179A>C VUS Neutral S1101N c.3302G>A VUS Neutral K1110del c.3328_3330delAAG VUS Neutral S1140G c.3418A>G VUS Neutral E1214K c.3640G>A VUS Neutral N1236K c.3708T>G VUS Neutral E1250K c.3748G>A Neutral control Neutral L1267S c.3800T>C VUS Neutral E1282V c.3845A>T VUS Neutral S1297del c.3891_3893delTTC VUS Neutral S1301R c.3903T>A VUS Neutral E1346K c.4036G>A VUS Neutral V1378I c.4132G>A VUS Neutral M1400V c.4198A>G VUS Neutral L1407P c.4220T>C VUS Neutral b M1411T c.4232T>C VUS Neutral R1443G c.4327C>G VUS Neutral S1448G c.4342A>G VUS Neutral S1486C c.4456A>T VUS Neutral S1497A c.4489T>G Artifi cial Neutral V1534M c.4600G>A VUS Neutral R1589P c.4766G>C VUS Neutral M1628T c.4883T>C VUS Neutral S1651P c.4951T>C VUS Not Clear S1651F c.4952C>T VUS Not Clear M1652I c.4956G>A VUS Neutral S1655F c.4964C>T VUS Deleterious b H1686R c.5057A>G VUS Deleterious H1686Q c.5058T>A VUS Deleterious V1688del c.5062_5064delGTT VUS Deleterious T1691I c.5072C>T VUS Not clear R1699W c.5095C>T Deleterious control Deleterious b R1699Q c.5096G>A VUS Deleterious b G1706E c.5117G>A VUS Deleterious b G1706A c.5117G>C VUS Neutral A1708E c.5123C>A Deleterious control Deleterious W1718C c.5154G>T VUS Deleterious T1720A c.5158A>G VUS Neutral E1735K c.5203G>A VUS Not clear  Table 1. \\u2002 Functional classifi cation of BRCA1 VUS based on cisplatin response (Continued) Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  1148\\u2002|\\u2002CANCER DISCOVERY OCTOBER 2013 www.aacrjournals.org RESEARCH ARTICLE Bouwman et al.  Fig. 3 ). Variants showing increased cisplatin sensitivity were tested at least twice before they were classifi ed. The positive and negative controls classifi ed as expected, although the known pathogenic truncation mutation 5382insC scored as neutral in one of three assays, stressing the need for repeat experiments. Also BRCA1 variants that were previously tested in other assays conducted as expected. It should be noted that the artifi cial variants S308A and S1497A rescued proliferation and cisplatin responses of mouse Brca1 - defi cient embryonic stem cells in BAC complementation assays, but are predicted to be deleterious based on their effects on embryonic stem cell differentiation and their response to \\u03b3-irradiation, respectively ( 7 ). The V1804D mutation scored as a neutral variant in our assay, which is in line with most published data ( 8, 9 ), except for the results of an embryonic stem cell\\u2013based BAC complementation assay ( 7 ). Again, all BRCA1 mutations that were classifi ed as deleterious were confi ned to regions encoding the conserved N- and C-terminal domains ( Fig.  4 ), despite the observation that deletion of the central region encoded by exon 11 leads to genetic instability in mice ( 12 ). Remarkably, the three mutations that diminish the interaction between BRCA1 and PALB2 ( 10 ), just C-terminal of the region encoded by exon 11, had no effect on cisplatin sensitivity in this assay. Of note, there was in general good correlation between our classifi cation and the Align-GVGD score (Supplementary Table S4). Notable exceptions were the neutral control variant R866C, which validated our assay but scored as likely deleterious (C65) using Align-GVGD, and G1770V, which scored as likely neutral (C0) by Align-GVGD but was classi- fi ed as deleterious in our assay. Our assay also classifi ed some variants for which the Align-GVGD scores were less clear, emphasizing the usefulness of functional assays to complement in silico analysis.  We also analyzed the possible effects of all BRCA1 variants on mRNA splicing, which may have deleterious consequences but cannot be assessed in our cDNA-based assay. The predictive value for exonic variants outside the consensus splice sites is questionable ( 19, 20 ), but four missense mutations were present in existing splice sites (Supplementary Table S4). Although c.5154G>T (encoding W1718C) was deleterious, three of these variants were classifi ed as neutral in our assay. One of them, c.441G>C (encoding L147F), had an increased probability to damage the splice donor site of exon 7. Another variant, c.5072C>T (encoding T1691I), might also affect splicing, but no defect was measured in blood samples of mutation carriers ( 19 ). The third exonic splice site mutation that was classifi ed as neutral in our cisplatin sensitivity assay, c.133A>C (K45Q), is not predicted to lead to aberrant splicing.  Results from Proliferation and Cisplatin Sensitivity Assays Correlate with HR Activity of BRCA1 Variants  BRCA1 and BRCA2 are involved in DNA repair via HR ( 12 , 21 ). Together with non-homologous end joining (NHEJ), HR forms the cellular defense against DNA double-strand breaks (DSB), a severe type of DNA damage that is lethal if unrepaired. Although HR is essentially error-free, NHEJ is error-prone, and therefore defects in HR are known to lead to genomic instability. Although it is not clear whether other Variant DNA change Type of mutation a Classifi cation V1736A c.5207T>C VUS Not clear b D1739G c.5216A>G VUS Deleterious D1739V c.5216A>T VUS Deleterious H1746Q c.5238C>G VUS Not clear R1753T c.5258G>C VUS Not clear b 5382insC c.5266dupC Deleterious control Not clear b L1764P c.5291T>C VUS Deleterious b C1767S c.5300G>C VUS Neutral G1770V c.5309G>T VUS Deleterious b W1782C c.5346G>T VUS Neutral A1789T c.5365G>A VUS Deleterious E1794D c.5382G>C VUS Neutral V1804D c.5411T>A VUS Neutral P1812R c.5435C>G VUS Neutral W1837R c.5509T>C VUS Deleterious H1862L c.5585A>T VUS Neutral a Type of mutation indicates if a variant is a VUS according to the BIC database (http://research.nhgri.nih.gov/bic/). b Functionally impaired in the DR-GFP and/or combined PARP inhibitor/cisplatin sensitivity assay.  Table 1. \\u2002 Functional classifi cation of BRCA1 VUS based on cisplatin response (Continued) Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  OCTOBER 2013CANCER DISCOVERY\\u2002|\\u20021149 Functional Analysis of BRCA1 Sequence Variants RESEARCH ARTICLE  Figure 3. \\u2002 Waterfall chart of cisplatin IC 50 values normalized for wild-type human BRCA1 and empty RMCE vector controls. Corrected cisplatin IC 50 values for all tested human BRCA1 variants and controls, including repeat experiments. BRCA1 variants are classifi ed as functionally impaired or neutral when the corrected IC 50 values are similar to either the empty RMCE vector or the wild-type human BRCA1 controls ( P < 0.05). 4.0 \\u20133.0 \\u20132.0 \\u20131.0 0.0 1.0 2.0 3.0 Corrected cisplatin IC50 (a.u.) Mutant hBRCA1 Wild-type hBRCA1 Empty RMCE vector Positive control Negative control  Figure 4. \\u2002 Predicted pathogenic BRCA1 amino acid substitutions are confi ned to the evolutionarily conserved N- and C-terminal domains. Schematic representation of the BRCA1 protein with the positions of variants classifi ed as neutral (green) or deleterious (red) indicated. Positive and negative controls are depicted by open pinheads and the known deleterious truncation mutations 185delAG (N-terminal) and 5382insC (C-terminal) are marked in blue. The amino acid sequences of the evolutionarily conserved RING (N-terminal) and BRCT (C-terminal) domains are specifi ed to show the exact positions of deleterious (red) and neutral (green) variants. The amino acid substitutions or nonsense (X) mutations of positive and negative controls are encircled. RING domain BRCT domains Deleterious VUS Neutral VUS Deleterious truncation mutant Deleterious control variant Neutral control variant 1646 1706 1766 91 109 1826 1863 61 31 1 MDLSALRVEEVQNVINAMQKILECPICLEL VNKRMSMVVSGLTPEEFMLVYKFARKHHITLTNLITEETTHVVMKTDAEFVCERTLKYFL IKEPVSTKCDHIFCKFCMLKLLNQKKGPSQ GIAGGKWVVSYFWVTQSIKERKMLNEHDFEVRGDVVNGRNHQGPKRARESQDRKIFRGLE CPLCKNDITKRSLQESTRFSQLVEELLKII CAFQLDTGLEYANSYNFAK ICCYGPFTNMPTDQLEWMVQLCGASVVKELSSFTLGTGVHPIVVVQPDAWTEDNGFHAIG QMCEAPVVTREWVLDSVALYQCQELDTYLIPQIPHSHY T Q E C A X Y C GG F R - Q W Q I E A V R L S C D DR T V P G X Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  1150\\u2002|\\u2002CANCER DISCOVERY OCTOBER 2013 www.aacrjournals.org RESEARCH ARTICLE Bouwman et al. functions of BRCA1 are also important for tumor suppression ( 16 , 22 , 23\\u201326 ), its role in HR is likely to be relevant. Therefore, we used the direct repeat (DR)-GFP assay ( 27 ) to measure the effects on HR for a subset of our BRCA1 VUSs. A number of predicted pathogenic variants and controls were shuttled into R26 CreERT2/RMCE;Brca1 SCo/\\u0394 ;Pim1 DR-GFP/wt embryonic stem cells carrying the DR-GFP reporter in the Pim1 locus. Expression of endogenous mouse Brca1 was switched off, and cells were transfected with a plasmid encoding the I-SceI meganuclease as well as an mCherry fl uorescent marker to control for transfection effi ciency. Repair of I-SceI\\u2013induced DNA DSBs in DR-GFP via HR leads to expression of GFP, which can be monitored by fl ow cytometry. All but one of the predicted pathogenic variants tested resulted in defective HR, thereby confi rming our functional classifi cation ( Fig. 5 ). The only predicted deleterious variant that did not signifi cantly differ from wild-type BRCA1 , R1699Q, seemed to support intermediate levels of HR activity.  PARP Inhibitor Sensitivity Assay for Classifi cation of BRCA1 Variants with Intermediate Activity  Deleterious effects of variants with intermediate or partial activity may escape detection in certain functional assays. Although cisplatin sensitivity assays allow robust and  Figure 5. \\u2002BRCA1 sequence variants classifi ed as pathogenic do not restore HR. R26 CreERT2/hBRCA1; Brca1 SCo/\\u0394;Pim1 DR-GFP/wt embryonic stem cells carrying the DR-GFP reporter gene in the Pim1 locus and mutant BRCA1 or controls in the Rosa26 locus were switched using 4-OHT and transfected with a vector expressing I-SceI and mCherry. Transfected cells were analyzed for GFP expression as a measure of HR activity. Expression of BRCA1 wild-type (WT) cDNA resulted in increased HR compared to the empty RMCE vector (Vector) control. Error bars indicate the SD between the results of three independent I-SceI transfections. Signifi cantly decreased HR activity compared with the wild-type control (green line) is indicated. GFP-Positive cells (%) 0.0 0.5 1.0 1.5 2.0 2.5 3.0 3.5 4.0 4.5 ** ns ** * * ns * * * * * * * ns ns WT Vector 185delAG C64G V271L S1651F S1655F R1699Q R1699W G1706A G1706E V1736A R1753T 5382insC L1764P G1770V reproducible classifi cation of several functionally impaired BRCA1 variants, assays using other compounds may have additional value. It is known that BRCA1- and BRCA2- defi cient cells are extremely sensitive to PARP1 inhibition ( 28, 29 ), leading to a larger dynamic range between BRCA2- defi cient cells and isogenic BRCA2-profi cient controls than for cisplatin ( 30 ). We therefore tested complementation of PARP inhibitor sensitivity for a number of BRCA1 mutants and the BRCA1 wild-type control. Given the unexpected neutral effects of the M1400V, L1407P, and M1411T mutations in the PALB2 interaction domain, we decided to include these variants in this series, as well as the R1699Q and V1736A variants that have recently been shown to confer (intermediate) breast and ovarian cancer risk ( 31, 32 ). To allow direct comparison of results from different assays, we repeated the cisplatin sensitivity and proliferation assays in parallel to the olaparib sensitivity assay. Our results indicate that BRCA1 wild-type and empty vector controls indeed show a larger difference in sensitivity for olaparib than for cisplatin ( Fig.  6 ). However, this increase in dynamic range is accompanied by an increased variation between repeat experiments for BRCA1-profi cient samples. Nevertheless, the functional defect of the R1699Q and V1736A mutations becomes more evident, and there also Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  OCTOBER 2013CANCER DISCOVERY\\u2002|\\u20021151 Functional Analysis of BRCA1 Sequence Variants RESEARCH ARTICLE  Figure 6. \\u2002 PARP inhibitor sensitivity assay of BRCA1 sequence variants. R26 CreERT2/hBRCA1 ;Brca1 SCo/\\u0394 or R26CreERT2/RMCE; Brca1 SCo/\\u0394 embryonic stem cells carrying mutant BRCA1 , BRCA1 wild-type (WT), or empty RMCE vector (Vector) controls in the Rosa26 locus were switched using 4-OHT and assayed for sensitivity to cisplatin or the PARP inhibitor olaparib. The cytotoxicity assays were conducted in parallel and data were normalized to the average of the wild-type controls. Error bars indicate the SD between the results of biologic triplicates for which the cells were independently switched. Signifi cant deviation from the average IC 50 values of the wild-type control (green line) is indicated. IC50 Relative to WT controls (%) 0 20 40 60 80 100 120 140 160 Cisplatin Olaparib * **** **** * * ** *** **** ns ns ns ns WT Vector M1400V L1407P M1411T R1699Q V1736A seems to be a less than wild-type response for the PALB2 interaction mutant L1407P. In the concurrent proliferation analysis, R1699Q and V1736A both show an intermediate functional defect (Supplementary Fig.  S4). Also the M1411T mutation seems to affect the response to PARP inhibition, but the difference with BRCA1 wild-type is not signifi cant. Interestingly, the L1407P and M1411T variants have previously been shown to be more defective than M1400V in a gene conversion assay ( 10 ). It should also be noted that, in contrast with the large-scale classifi cation experiments, the cisplatin sensitivity assay conducted in parallel to the PARP inhibitor assay identifi ed signifi cant functional defects for L1407P and V1736A.  DISCUSSION  Over the past few years several functional assays for classifi cation of BRCA1 VUSs have been developed. Several of these assays are restricted to functions of the BRCA1 protein that reside in the evolutionarily conserved RING or BRCT domains. Examples include in vitro transactivation assays for BRCT peptides ( 8 ) and measurement of ubiquitin ligase activity for protein fragments encompassing the N-terminal RING domain ( 33 ). Other assays were designed to evaluate the functions of full-length mutant BRCA1 protein, either by monitoring general effects on proliferation or response to DNA damage ( 7 ), or by directly focusing on the role of BRCA1 in DNA repair via HR ( 34 ).  Because BRCA1 VUSs are not restricted to regions encoding the N- or C-terminal domains, and given the observation that interaction between these domains is required for recruitment of BRCA1 to damaged DNA ( 35 ), functional assays for the full-length protein would be ideal. In principle, such assays can be conducted in cell lines derived from BRCA1 - mutated tumors, but there are indications that the outcome of assays for BRCA1 function depends on the cellular context. As absence of BRCA1 leads to loss of cellular viability, it is thought that additional mutations are required for BRCA1- associated tumorigenesis. For example, loss of p53 alleviates the consequences of BRCA1 defi ciency both in vitro and in vivo ( 4 ) and is common in BRCA1-defi cient tumors ( 36, 37 ). Also, depletion of 53BP1 is known to suppress the defects caused by BRCA1 defi ciency ( 5 , 38 , 39 ). Therefore, aberrations in BRCA1-defi cient tumor cells may mask functional defects of BRCA1 VUSs. We reasoned that assays in normal cells that can be depleted from endogenous BRCA1 expression are most likely to reveal functional defects of BRCA1 VUSs. The usefulness of this approach has been previously shown by Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  1152\\u2002|\\u2002CANCER DISCOVERY OCTOBER 2013 www.aacrjournals.org RESEARCH ARTICLE Bouwman et al. BAC complementation assays in mouse embryonic stem cells ( 7 , 40 ). However, mutagenesis of large BAC clones by recombineering and functional complementation of cells with these mutant constructs is time-consuming and technically demanding, and therefore cannot be conducted in a high-throughput setting. We therefore set out to develop a cDNA-based functional complementation assay in mouse embryonic stem cells that is easier to control and scale up and more suitable for routine functional classifi cation of BRCA1 sequence variants.  One advantage of BAC transgenics in complementation assays is that genes are expressed at physiologically relevant levels. This is the result of low copy-number integrations and the presence of natural regulatory elements required for proper gene expression. We decided to use RMCE to allow single-copy integration of BRCA1 cDNAs at one specifi c genomic locus. The use of RMCE effectively prevents multiple or partial integrations, concatemers, and position-effect variegation. As a result, all variants are expressed at equal levels. Moreover, transcription of BRCA1 cDNAs from the EF1A gene promoter results in physiologic levels of BRCA1 protein, comparable with those observed in embryonic stem cells stably transfected with a BAC containing the human BRCA1 locus. Indeed, the wild-type human BRCA1 cDNA was able to complement Brca1 -null embryonic stem cells in cellular proliferation, drug sensitivity, and HR assays. Our RMCE strategy uses BRCA1 cDNA constructs in which mutations can be swiftly introduced using SDM, enabling a higher throughput than introduction of mutations via BAC recombineering. In addition, the introduction of BRCA1 cDNAs via RMCE obviates the need to analyze multiple embryonic stem cell clones for correct integration and expression.  As a proof of principle, we used our functional complementation assay to analyze exonic BRCA1 VUSs that were identifi ed in families with HBOC in the Netherlands and Belgium, as well as a set of previously analyzed BRCA1 variants. A number of mutations resulted in reduced BRCA1 protein levels, most likely because of structural destabilization. In all cases, this led to diminished capacity for functional complementation. Also, several variants that gave rise to normal BRCA1 protein levels were unable to rescue the proliferation defect and cisplatin sensitivity of Brca1 -null embryonic stem cells. As BRCA1 loss of function mutations are associated with increased cancer risk, variants that score as functionally impaired in our embryonic stem cell assay system may be causally involved in tumor formation. This notion is supported by the fact that seven of eight known pathogenic or neutral control variants in our validation series were correctly classifi ed by the cisplatin sensitivity assay. The pathogenic 5382insC truncation mutation could not be classifi ed because it scored as neutral in one of three transfection series. This was probably due to technical reasons, as the 5382insC mutation did not restore HR activity in Brca1 -defi cient embryonic stem cells, in contrast with R1699Q, which was recently shown to confer intermediate risk of HBOC ( 32 ).  Our assay system yielded ambivalent results for nine other variants: S4F, S308A, S1651P, S1651F, T1691I, V1736A, E1735K, H1746Q, and R1753T. T1691I and E1735K were classifi ed as functionally impaired in only one cisplatin sensitivity test, whereas values from repeat experiments could not be taken into account because of RSE values above 0.1. S4F, S1651P, S1651F, V1736A, H1746Q, and R1753T were differently classifi ed in repeat experiments, which may refl ect technical fl aws or intermediate activity of these variants. Of note, the V1736A mutation was recently identifi ed as a pathogenic variant with hypomorphic activity in DNA repair ( 31 ). Although we did not measure signifi cant HR activity of V1736A in a DR-GFP gene conversion assay, intermediate activity of this variant is supported by the results of the proliferation assays and additional cisplatin sensitivity assays. S1651F showed HR activity similar to wild-type BRCA1 , whereas R1753T was HR-defi cient. S308A is an artifi cial mutation of a BRCA1 phosphorylation site that was able to support proliferation and resistance to DNA damage in an embryonic stem cell-based BAC complementation assay. However, S308A-complemented embryonic stem cells did show increased apoptosis when cultured in embryoid bodies ( 7 ), indicating a partial defect that might explain the ambivalent results for this mutation in our assays.  Our results show that BRCA1 variants should ideally be assayed in triplicate to avoid misclassifi cation. This also applies to BRCA1 VUS that we classifi ed as neutral, most of which were tested only once because in our current study we focused on variants that showed functional impairment. The striking restriction of unambiguously predicted pathogenic mutants to the terminal RING and BRCT domains suggests that some plasticity is allowed for the central domain of BRCA1. However, our dataset is still limited and more experiments are required to gain insight into the function of this domain.  Our cDNA-based system allows for several additional functional assays that have been described previously for BAC transgenic embryonic stem cells ( 7 ). These include assays for defects during in vitro and in vivo embryonic stem cell differentiation but also treatments with other cytotoxic agents. As a proof of principle, we investigated the activity of a number of mutants in the response to the PARP inhibitor olaparib. These included M1400V, L1407, and M1411T, which were previously shown to impair PALB2 binding and have a negative effect on BRCA1 function ( 10 ). Although the differences in sensitivity of BRCA1-defi cient versus BRCA1-profi cient embryonic stem cells are larger for olaparib than for cisplatin, increased variation between repeat experiments allowed us to identify functional defects only for L1407 and not for the other two variants in the PALB2 interaction domain, M1400V and M1411T. However, the R1699Q and V1736A variants, which are known to have hypomorphic activity, clearly showed a defect in the response to olaparib. Interestingly, both mutations are in the BRCT domain, and it has recently been shown that mutation of this domain especially confers sensitivity to PARP inhibition ( 41 ). Together, our results show that PARP inhibitor sensitivity assays may have added value, especially for the classifi cation of BRCA1 VUSs with intermediate phenotypes.  Platinum drugs and PARP inhibitors are selectively toxic to BRCA1-defi cient cells because they target HR defi ciency. Although the role of BRCA1 in HR is thought to be essential for maintaining genomic integrity and preventing accumulation of (oncogenic) mutations, other activities may also contribute to its tumor suppression function. These activities Downloaded from cancerdiscovery.aacrjournals.org on May 26, 2017. \\u00a9 2013 American Association for Cancer Research. Published OnlineFirst July 18, 2013; DOI: 10.1158/2159-8290.CD-13-0094  OCTOBER 2013CANCER DISCOVERY\\u2002|\\u20021153 Functional Analysis of BRCA1 Sequence Variants RESEARCH ARTICLE may include the HR-independent role for BRCA1 in ICL repair, which has been attributed to the facilitation of FANCD2 accumulation at cross-linked DNA ( 16 ). Nevertheless, we observed a good correlation between the results of the cisplatin sensitivity assay and the results of the DR-GFP HR assay. All BRCA1 variants that failed to restore the cisplatin response in Brca1 -null embryonic stem cells were also defective in catalyzing gene conversion, thereby confi rming our functional classifi cation. It will be interesting to see if this holds true for all BRCA1 variants or whether there are also pathogenic mutations that have no effect on HR.  As with any other in vitro approach, our functional complementation assay system might still fail to identify all pathogenic variants because it does not necessarily recapitulate all aspects of BRCA1 function in vivo . A limitation of our cDNA-based assay is also that it cannot be used to investigate effects on mRNA splicing. Although algorithms have been designed to predict possible splice defects, the consequences of mutations outside of the consensus splice sites especially require functional validation experiments. For this purpose, BAC complementation assays ( 7 ), minigene-based splicing assays, or BRCA1 transcript analysis of patient blood samples ( 19 ) may be instrumental. However, transacting factors also affect splicing, and these may be tissue specifi c ( 42 ). A possible solution would be to determine the presence of BRCA1 splice variants in tumor tissue from BRCA1 VUS carriers and use this information to generate a cDNA construct for analysis of the functional consequences. To evaluate BRCA1 VUSs, there remains a need for multifactorial models that combine results from functional assays and in silico analyses with genetic evidence and other information from mutation carriers. This also includes DNA copy number data from tumors from BRCA1 VUS carriers, as it is known that BRCA1- associated breast tumors show distinct genomic aberrations ( 43, 44 ). Our functional assay system does however provide a robust and easily implementable tool for the functional characterization of large numbers of BRCA1 VUSs within the context of the full-length protein. It is our hope that our assay system will fi nd its way to clinical genetics laboratories where it can be used to aid genetic counseling. Ideally, these tests should be coordinated on an international level and in close collaboration with the ENIGMA consortium.    Genetic screening of the breast and ovarian cancer susceptibility gene BRCA1 has uncovered a large number of variants of uncertain clinical significance. Here, we use biochemical and cellbased transcriptional assays to assess the structural and functional defects associated with a large set of 117 distinct BRCA1 missense variants within the essential BRCT domain of the BRCA1 protein that have been documented in individuals with a family history of breast or ovarian cancer. In the first method, we used limited proteolysis to assess the protein folding stability of each of the mutants compared with the wild-type. In the second method, we used a phosphopeptide pull-down assay to assess the ability of each of the variants to specifically interact with a peptide containing a pSer-X-X-Phe motif, a known functional target of the BRCA1 BRCT domain. Finally, we used transcriptional assays to assess the ability of each BRCT variant to act as a transcriptional activation domain in human cells. Through a correlation of the assay results with available family history and clinical data, we define limits to predict the disease risk associated with each variant. Forty-two of the variants show little effect on function and are likely to represent variants with little or no clinical significance; 50 display a clear functional effect and are likely to represent pathogenic variants; and the remaining 25 variants display intermediate activities. The excellent agreement between the structure/function effects of these mutations and available clinical data supports the notion that functional and structure information can be useful in the development of models to assess cancer risk. Introduction Since the cloning of BRCA1 in 1994, a large effort has been made to sequence the BRCA1 genes of women who are at increased risk for early-onset breast and ovarian cancers (1). Corresponding Author: J.N. Mark Glover or Alvaro N.A. Monteiro, Department of Biochemistry, University of Alberta, Edmonton, Alberta, T6G 2H7, Canada. Phone: 780-492-2136; Fax: 780-492-0886; mark.glover@ualberta.ca or alvaro.monteiro@moffitt.org. Reprints and Subscriptions: To order reprints of this article or to subscribe to the journal, contact the AACR Publications Department at pubs@aacr.org. Note: Supplementary data for this article are available at Cancer Research Online (http://cancerres.aacrjournals.org/). Disclosure of Potential Conflicts of Interest: No potential conflicts of interest were disclosed. NIH Public Access Author Manuscript Cancer Res. Author manuscript; available in PMC 2011 June 15. Published in final edited form as: Cancer Res. 2010 June 15; 70(12): 4880\\u20134890. doi:10.1158/0008-5472.CAN-09-4563. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript This effort has revealed several mutations that are strongly linked to cancer. Unfortunately, a large number of rare BRCA1 variants have been uncovered in the human population for which risk assessment has been problematic, due to a lack of informative pedigree data linking each mutation with disease risk. BRCA1 encodes a large, 1,863\\u2013amino acid nuclear protein that plays a critical role in the response of cells to genotoxic stress (2\\u20134). BRCA1 is thought to act as an essential mediator protein in DNA damage\\u2013induced nuclear signaling events, in which it interacts with phosphorylated partner proteins such as the DNA helicase, BACH1, the nuclease CtIP, and another signaling protein, Abraxas, to relay signals generated from chromatin surrounding the damage to downstream targets such as DNA repair proteins and factors involved in cell cycle regulation (5\\u201311). Interactions with phosphorylated partner proteins are mediated by a tandem pair of repeats at the COOH terminus of BRCA1, termed BRCT repeats (12\\u201314). The BRCT repeat region can also act as a transcriptional activation domain when linked to a sequence-specific DNA binding module, and this activity may contribute to the ability of BRCA1 to regulate the expression of genes such as p21 and GADD45 (15,16) and modulate the activity of other transcription factors such as p53 and ER (17,18). The NH2-terminal region of the protein contains a RING domain, which forms a heterodimeric complex with the RING domain of BARD1 to form a ubiquitin ligase (19\\u201321). Cancer-associated mutations tend to cluster in the RING and BRCT repeat regions, showing the critical role of these domains in BRCA1 tumor suppression. The BRCA1 BRCT domain specifically interacts with phosphorylated protein targets containing the motif pSer-x-x-Phe. Structural studies have revealed that recognition involves a conserved phosphoserine recognition pocket in the NH2-terminal BRCT repeat, composed of Ser1655, Gly1656, and Lys1702, which each supply ligands to recognize the phosphate (12\\u201314). The phenylalanine residue at the +3 position of the peptide target is recognized by a largely hydrophobic groove formed at the interface between the repeats. Phosphopeptide recognition is likely a common function of BRCT domains in many proteins associated with DNA damage\\u2013mediated signal transduction. For example, the phosphorylated form of the histone variant, H2AX, which serves as a critical chromatin mark of DNA double-strand breaks, is specifically bound by the tandem BRCT repeats of MDC1 to initiate DNA doublestrand break signaling (22,23). Several studies have attempted to predict the cancer risks associated with unclassified missense variants in BRCA1 through bioinformatics analysis based on multiple sequence alignment data and protein structure prediction, as well as analyses of the clinical and family history data available for some of these mutations (1,24\\u201328). These studies indicate that the RING and BRCT domains, which are the most highly conserved regions of BRCA1, likely contain the vast majority of cancer-associated mutations. Here, we use in vitro protein folding, phosphopeptide-binding, and cell-based transcriptional assays to assess the structural and functional consequences of a set of 117 distinct missense variants that have been uncovered in the BRCA1 BRCT domain in the human population. Our results indicate that nearly one half of the BRCT missense variants result in significant structural and/or functional defects that are likely associated with an increased cancer risk. Materials and Methods Proteolysis assays pLM1 plasmid (0.25 \\u03bcg) encoding the BRCT variants was used directly as template for protein synthesis reactions with the TNT-Quick in vitro transcription/translation system (Promega). Both wild-type and mutant proteins were produced by in vitro transcription/ translation because many mutants could not be expressed in soluble form in Escherichia coli. The stability of the protein fold for the BRCA1 BRCT mutants was tested using Lee et al. Page 2 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript proteolysis assays, as previously reported (29). The measurements were normalized against the internal background, and the control measurements taken in the same day. The missense variants were tested in triplicate. Phosphopeptide-binding assays The BRCT variants were translated and labeled as in proteolysis assays. The ability of the BRCA1 BRCT mutants to bind pSer-containing peptides was tested using peptide-binding assays, as previously reported (14). Bead-immobilized peptide affinity resin was prepared by incubating a 10-fold molar excess of a biotinylated phosphopeptide (SRSTpSPTFNK) and the corresponding unphosphorylated peptide (SRSTSPTFNK) with streptavidin agarose beads (SigmaAldrich). Seven mutants, with the wild-type BRCA1-BRCT (1,646\\u20131,858) as a positive control, and M1775R and A1708E as negative controls were examined at a time. The intensities of the bands were measured and the backgrounds were corrected using ImageQuaNT (Amersham Biosciences), and the measurements were normalized against the wild-type measurement taken in the same day. The missense variants were tested in triplicate. Transcriptional assays The construct containing wild-type BRCA1 amino acids 1,646 to 1,859 (wt) was used as a positive control and variants M1775R and Y1853X as negative controls. HEK293T cells were seeded in 24-well tissue culture plates, 1 \\u00d7 105  cells per well, and 24 hours later were transfected in triplicates with Fugene 6 according to the manufacturer's instructions (Roche). The vectors containing the variants were cotransfected with pG5Luc, which contains a firefly luciferase reporter gene driven by GAL4-responsive binding sites, and phGR-TK, which contains a Renilla luciferase gene driven by a constitutive TK basal promoter used as an internal control. Cells were harvested 24 hours after transfection, and luciferase activity was measured using the Dual-Luciferase Reporter Assay System (Promega) following the manufacturer's instructions. Activity was plotted as a percentage of the wt BRCA1 activity. Thirty-four of 117 variants had already been assessed in previous articles, albeit in different amino acid contexts (amino acids 1,396\\u20131863 or aa 1,560\\u20131863; refs. 30\\u201332). In only one case (G1706A), the qualitative assessment differed because it was considered of moderate effect previously and of no effect in the present study. This shows the high reproducibility of the transcriptional assays. Align-GVGD classification of BRCA1 BRCT variants The 117 BRCT variants were analyzed with the Align-GVGD algorithm at three depths of alignment. Nine mammalian BRCA1 sequences, chicken and one additional sequence (either frog, puffer fish, or sea urchin), were used for the alignment (for more detail, see ref. 33). A-class scores were assigned for variants in which the scores from at least two of three alignments agreed. Definition of a validation set To compile our validation set (Supplementary Table S1; Table 1), we identified variants that had been previously analyzed and classified by genetic (1) or integrative methods (25, 34\\u2013 40). Integrative methods use a combination of data from different sources, including cooccurrence with known deleterious mutations, personal and family history of patients carrying the variant, and cosegregation of the variant with disease within pedigrees, to assign an overall risk score. Sufficient data were available to assign 14 of the BRCA1 BRCT variants as associated with disease, whereas 10 of the variants were predicted to be neutral Lee et al. Page 3 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript (Table 1). Where possible, the probability of pathogenicity of each of these variants was classified according to the IARC scheme (Supplementary Table S1; ref. 41). We calculated specificity, sensitivity, and 95% confidence interval measures for each of the structure and functional assays based on a comparison with this variant set (Supplementary Table S2). A description of how the BRCA1 BRCT variant set was obtained is given in Supplementary Materials and Methods. Results Generation of a large set of BRCA1 BRCT missense variants Many of the BRCA1 variants that have been uncovered through worldwide sequencing efforts are now deposited within the Breast Cancer Information Core database (http://research.nhgri.nih.gov/bic/) as well as the Kathleen Cuningham Foundation Consortium for research into Familial Breast Cancer database (http://www.kconfab.org). To facilitate the assessment of the structural and functional defects associated with these mutations, we cloned all the currently available missense variants (Fig. 1) into an E. coli expression vector that we used previously for our structural and functional work of the BRCA1 BRCT domain. This vector places the BRCA1 BRCT coding sequence (corresponding to residues 1,646\\u20131,859) under control of a T7 promotor. Using this plasmid system, recombinant protein can be expressed in E. coli and purified using established procedures for detailed biochemical and structural studies. The T7 promoter can also be used for in vitro transcription/translation of the recombinant protein. The in vitro transcription/translation system is rapid and allows the generation of multiple proteins in parallel and, because the recombinant protein is specifically labeled with 35S-methionine, no purification is required. All variants can be successfully made in this system, regardless of their folding stability. This is a further advantage over E. coli expression, in which we have found that only well-folded variants can be expressed in soluble form and purified. Thus, we have used in vitro transcribed/translated material for our structural stability and peptidebinding assays described below. We also evaluated the possibility that some of these variants could result in splicing abnormalities. We used the NNSplice 0.9 algorithm (42) to predict the effects of these changes on the efficiency with the BRCT coding region (see Supplementary Materials). Determination of the protein folding defects associated with BRCT missense variants We used limited proteolysis to determine the stability of the protein fold in the set of BRCA1 BRCT missense variants using in vitro transcribed/translated material (Supplementary Fig. S1). The wild-type protein is highly resistant to digestion by trypsin, elastase, or chymotrypsin, indicating that the in vitro produced BRCT protein is stably folded (43). On the other hand, a pilot study involving 25 of the BRCA1 BRCT missense variants showed that certain single amino acid substitutions in the BRCT domain of BRCA1 can destabilize its structure, leading to increased susceptibility to trypsin-mediated proteolysis (29). The assay is sufficiently sensitive to discriminate wild-type protein from the strongly destabilized variant A1708E as well as the moderately destabilized variant M1775R. Proteolysis of each of the missense variants was carried out at a series of trypsin concentrations, and the amount of full-length protein remaining was visualized by SDSPAGE. The results were compared with those of the wild-type protein as well as the M1775R and A1708E variants carried out in parallel. We quantified the percentages of fulllength protein remaining following digestion at different concentrations of trypsin and determined the severity of the destabilizing effects by comparing the percentage of protein remaining following digestion at 10 \\u03bcg/mL trypsin to the controls (Supplementary Fig. S2A). We defined highly destabilizing mutations as those with similar rates of digestion to Lee et al. Page 4 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript the A1708E control (<5% of the protein remaining after digestion with 10 \\u03bcg/mL of trypsin). Moderately destabilizing mutations were defined as those with a similar stability to M1775R (between 5% and 65% of the protein remaining after digestion with 10 \\u03bcg/mL of trypsin). Finally, structurally stable variants were defined as those with a stability similar to the wild-type protein (>65% of the protein remaining after 10 \\u03bcg/mL of trypsin digestion). Based on these criteria, 26 of 117 of the missense mutations showed high sensitivity to tryptic digestion; 30 of 117 showed intermediate sensitivity; and 61 of 117 showed a stability that was indistinguishable from the wild-type (Supplementary Figs. S2A and S3). We also noted that the C1697R mutant showed a particularly low protein expression level. Analysis of the phosphopeptide binding activity of BRCT missense variants The BRCA1 BRCT domain interacts with BACH1 and several other proteins through the selective recognition of the peptide motif pSer-x-x-Phe (10\\u201314,44). All mutations that have been strongly linked to disease using family-based data, such as A1708E and M1775R (1,35), fail to bind specifically to pSer-x-x-Phe\\u2013containing peptides as shown in a pilot study that analyzed the peptide binding activities of 25 BRCA1 BRCT domain variants (14). This indicates the functional importance of the BRCA1 BRCT domain as a phosphopeptidebinding module in tumor suppression. We examined the ability of the BRCT domains containing missense variants to bind peptides derived from BACH1 using a pull-down assay in which biotinylated peptide was bound to streptavidin beads. We tested binding to a phosphorylated version of the peptide containing the BRCA1 target sequence pSer-Pro-Thr-Phe or an unphosphorylated version as a control. The results were compared with experiments with the wild-type BRCA1 BRCT as well as with the peptide binding\\u2013deficient variants A1708E and M1775R carried out in parallel. We present two analyses of the data. In the first analysis, we define \\u201cbinding activity\\u201d as the fraction of the input protein that bound to the phosphorylated peptide, normalized against the binding of the wild-type protein (Supplementary Fig. S2B). In the second analysis, we define \\u201cbinding specificity\\u201d as the ratio of the variant bound to the phosphopeptide compared with the nonphosphorylated counterpart, normalized to the wildtype control (Supplementary Fig. S2C). Classified in this way, we observed a range of binding activities and specificities. As observed for the protein stability assays, the variants tended to cluster near either 100% or 0% activity, with relatively few variants displaying intermediate binding activities/specificities (Supplementary Fig. S3). Thirty-eight of 117 of the variants displayed binding activities >80% of wild-type levels, whereas 35 of 117 variants displayed binding activities <20% of wild-type. The binding specificity analysis revealed 29 of 117 of the variants exhibited >80% of wild-type levels, whereas 50 of 117 of the variants exhibited <20% of wild-type levels. Analysis of the transcriptional activation activity of BRCA1 BRCT variants The transcriptional activation activity of the BRCA1 BRCT domain was the first identified activity of this region of the protein (15,16). Assays for BRCA1 BRCT transcriptional activation have been established in human and yeast cell systems (45). In either method, the BRCT domain is fused to a heterologous DNA binding domain that recognizes a defined regulatory sequence within the promoter of the reporter. Pilot studies involving subsets of the BRCA1 BRCT variants showed that mutations that have been associated with cancer using family-based data abrogate, or markedly diminish, the expression of the reporter gene (31,32,46). Moreover, the fact that mutations that destabilize the BRCT fold are strongly associated with a loss of transactivation function indicates that this assay is a monitor of the integrity of the BRCT domain (29). Lee et al. Page 5 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript As observed for the protein stability and phosphopeptide-binding assays, a wide range of transcriptional activities were observed for the variant set, with the observed values tending to cluster either at very low or very high activity levels (Supplementary Figs. S2D and S3). Thirty-eight of 117 variants displayed transcriptional activity levels at >80% of wild-type, whereas 54 of 117 were found to be <20% of wild-type levels. Correlation of structural and functional data with family history and clinical data The use of functional data in cancer risk assessment is predicated on rigorous validation of the results against sets of variants for which cancer risks are well defined. Easton and colleagues (35) have assigned risks based on genetic data for a large number of BRCA1 variants. We used these data, together with available genetic information for several other variants (25,34\\u201340), to create a validation set of 24 variants, consisting of 14 deleterious mutations and 10 neutral variants (Table 1). We plotted the protein stability, peptide binding, and transcriptional activity data against the combined log10 likelihood ratio scores for the subset of 19 variants analyzed by Easton and colleagues (Fig. 2; likelihood ratio). In general, this analysis revealed a clear correlation between structural/functional defects and cancer risk (35). A comparison of all the genetically characterized variants with the experimental results is shown in Table 1. This comparison was used to define the sensitivities and specificities of the functional assays. Sensitivity ranged from 81.8% for phosphopeptide binding activity to 100% for phosphopeptide binding specificity and transcription assays. Specificity ranged from 75% for proteolysis assays to 87.5% for transcription assays, to 100% for binding activity assays (Supplementary Table S2). Although overall there is a good agreement between the assessed cancer risks and the functional activities of the variants, there are several variants in which cancer risk assessment and the structure/function results do not agree. For example, a significant number of cancer-associated (IARC classes 5 and 4) variants are not structurally destabilized as determined by the proteolysis assay. Each of these variants (R1699W, M1775R, and M1775K) is highly defective for specific phosphopeptide recognition and transcriptional activity. Structural studies reveal that both Arg1699 and Met1775 make critical direct contacts to the phosphopeptide (12\\u201314), and structural investigations of the M1775R and M1775K variants reveal subtle structural rearrangements that block the +3 specificity pocket (37,47). Two of the cancer-associated (IARC class 5) variants, I1766S and G1788V, were highly defective in the proteolysis and transcriptional assays; however, they showed significant levels of nonspecific peptide binding, independent of phosphorylation. It is likely that the unfolded state of these variants exposes hydrophobic residues that are responsible for these nonspecific interactions, as suggested previously (14). G1706E, one of the cancer-associated (IARC class 5) variants, is defective in the proteolysis and peptidebinding assays; however, it shows an intermediate level of transcriptional activity. The neutral (IARC class 1) variant V1736A exhibited the least agreement between the different assays, showing significant sensitivity to proteolysis and a strong transcriptional defect while showing intermediate/strong levels of phosphopeptide interaction. Considering the overall high specificity and sensitivity obtained in the assays (Supplementary Table S2), we have defined limits to predict disease association for the uncharacterized variants based on the biochemical and transcriptional assay data (Table 1; Fig. 2). It is important to stress that the definition of these limits should not be used for classifying variants in terms of their pathogenicity in the absence of additional data. The cancer-associated (IARC classes 5 and 4) variants in general exhibited protein stability levels <58% of the wild-type, phosphopeptide binding activities <32% of wild-type, phosphopeptide specificity binding activities <21% of wild-type, and transcriptional activation function <35% of wild-type levels. Conversely, the neutral variants (IARC classes 1 and 2) in general had protein stability levels >59% of the wild-type, phosphopeptide Lee et al. Page 6 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript binding activities >67% of wild-type, phosphopeptide specificity binding activities >49% of wild-type, and transcriptional activation function >76% of wild-type levels. Variants falling between these limits are defined as uncertain. Using these limits, we have defined the structure and functional activities of each of the variants (Fig. 3). Comparison of results from different assays and overall classification Next, we compared the results from the different assays for the entire set of variants (Fig. 3). In 64 of 117 (54.7%) of the variants, there was complete concordance between the different assays, with 35 of 117 (29.9%) giving protein stability, peptide binding, and transcription activities within the limits that correspond to neutral variants in IARC classes 1 and 2, whereas 29 of 117 (24.8%) exhibited values in the range corresponding to cancer-associated (IARC classes 4 and 5) variants for each of the assays. Based on these results, we have classified the variants into groups based on their functional effect (Table 2). Although classification is straightforward for those variants in which there is a complete agreement between the different assays, for the variants in which there is not a complete agreement, classification is more complex. The largest class of discordant results includes variants that exhibit defects in binding and transcription but little or no folding defect (6 of 117 variants or 5.1%). These variants likely directly affect the phosphopeptide binding surface and may only affect residues that are exposed to the solvent preserving the overall fold structure. Because phosphopeptide recognition is critical for the tumorsuppressive function of BRCA1 (14,48), we have therefore included these variants in the strong functional effect class. Likewise, severely destabilized variants that show significant levels of phosphorylation-independent peptide binding but a strong transcriptional defect have also been categorized as strong functional effect (Table 2). We have also described eight of the variants as of \\u201cmoderate functional effect\\u201d in which a preponderance of the data points to a loss of function. These include variants with intermediate levels of peptide binding specificity (W1837G and A1843P) or transcriptional activity (V1696L, G1706E, and H1746N), and strong defects in all of the other assays. In addition, we have included three additional mutants in this group that show poor agreement between the different assays but are likely to directly affect the peptide-binding surface. S1655F is predicted to impede recognition of the phosphoserine in the bound peptide, whereas G1656D could perturb the structure of the phosphoserine-binding pocket, and the introduction of a negatively charged side chain in this region would be expected to further inhibit phosphopeptide recognition. R1699L would remove a critical determinant for recognition of the phenylalanine residue at the +3 position of the phosphopeptide, and indeed, the similar R1699W variant is pathogenic (IARC class 5; Table 1). We predict a given variant to be neutral when each of the assays gives values similar to wild-type within the limits defined by comparison with the set of variants that have been characterized through family history and genetic data. In addition, we define a set of seven variants as \\u201clow functional effect\\u201d that display an intermediate activity for one of the assays and wild-type levels for the other three. Finally, there is a group of 25 variants (21.4%) that we define as \\u201cuncertain.\\u201d In general, these variants show a poor level of agreement between the different assays, and a significant number of assay results in the intermediate range. Correlation of mutant classification with predictions from Align-GVGD The Align-GVGD method (http://agvgd.iarc.fr/agvgd_input.php) uses protein multiple sequence alignment (PMSA) data to provide cancer risk estimates for BRCA1 mutations (33). In this method, each missense variant is classified based on the observed amino acid Lee et al. Page 7 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript fluctuation at that position in the PMSA (the Grantham variance, GV) as well as a comparison of the amino acid substitution with the range of amino acids observed at that position (Grantham difference, GD). In this way, seven classes of variants (C0, C15, C25, C35, C45, C55, and C65) that represent a graded series of GV-GD scores were defined. Variants in the C0 class correspond to variants that are well conserved in the PMSA and are therefore predicted to be neutral, whereas the C65 class contains missense variants that deviate from the PMSA and likely associated with an increased cancer risk. To assess the accuracy of the Align GV-GD method, we compared Align GV-GD scores to the results for the structure/function assays (Fig. 4). For the variants with no functional effect, a large majority (28 of 35, 80%) falls into the C0 class, three are classified as C15, and two are classified as C25. Only one of them, G1706A, received a high score in Align GV-GD, C55. Another variant at this position (G1706E) scored as C65, but this variant was classified with moderate functional effect by structure/function assays. These results suggest substitution of a small side chain at position 1,706 is tolerated, but not substitution of the larger, charged Glu. For the variants with strong functional effect, 25 of 42 (59.5%) are in the C65 class and 16 are spread throughout classes C15 to C55 (Fig. 4B). One missense variant (A1752P) was classified C0, but this variant shows strong defect in all assays. The substitution of a proline at position 1,752 likely destabilizes the linker \\u03b1-helix, which may be responsible for the strong folding defect associated with this variant. Overall, there is strong agreement between Align GV-GD predictions and the results of functional assays, indicating that sequence alignment\\u2013based predictive methods can offer a valuable measure of risk assessment for variants in which there is no direct functional data. Discussion Here, we present the first comprehensive assessment of the structural and functional consequences of 117 BRCA1 variants corresponding to the complete set of BRCT missense variants detected to date in the human population (100% coverage). Our protein structure assay, peptide binding, and transcriptional assays show a remarkable degree of agreement and give us confidence in the validity of these results. Correlation of this experimental data with the available clinical and family history data for a subset of these variants defines limits that can be used to tentatively predict functional effect that is likely to be reflected in the disease risk associated with these variants. In this way, we show that 42.7% of the variants fall into the \\u201cclear functional effect\\u201d or \\u201cmoderate functional effect\\u201d categories (Table 2) and are likely associated with an increased cancer risk, whereas 35.9% fall into the \\u201cno functional effect\\u201d or \\u201clow functional effect\\u201d categories. The remaining 21.4% falls into the uncertain category and could perhaps represent low-penetrance variants. The fact that a large proportion of variants tested displayed some functional effect lends support to the idea of applying higher prior probabilities for causality to variants found in the BRCT domains (35). The BRCA1 BRCT variant set provides a useful set of reagents for further detailed study of variants of particular interest. For example, variants that are moderately well folded based on proteolytic stability can often be purified and studied in more detail. Circular dicroism spectroscopy and differential scanning calorimetry have been used to quantitatively assess the fold stability of V1696L, R1699W, M1775K, M1775R, M1783T, V1808A, V1809F, and V1833M (49\\u201351). Although our proteolytic assay indicated that the stability of two of these variants, R1699W and V1808A, were indistinguishable from wild-type, the more sensitive thermodynamic studies of the purified forms of these variants reveal a small but significant reduction in fold stability. In addition, quantitative assays of phosphopeptide binding, such as isothermal titration calorimetry, can reveal the detailed thermodynamics of peptide-BRCT interactions (50,51). The structural rearrangements induced by three of these variants, M1775R (47), M1775K (37), and V1809F (14), have been visualized through Lee et al. Page 8 Cancer Res. Author manuscript; available in PMC 2011 June 15. NIH-PA Author Manuscript NIH-PA Author Manuscript NIH-PA Author Manuscript crystallographic structural analysis. Each of the variants affect the +3 specificity pocket formed at the interface between the two BRCT repeats. For M1775R and M1775K, the substituted positively charged side chain adopts an orientation that fills the specificity pocket and blocks the binding of the Phe side chain at the +3 position of the target peptide. Intriguingly, the V1809F substitution, which occurs far from the +3 pocket, induces a series of conformational adjustments that results in the movement of Met1775 to fill the specificity pocket, explaining the loss of peptide binding specificity and transcriptional activity associated with this variant. Nonsynonymous single nucleotide polymorphisms (SNP) that lead to amino acid substitutions are likely to have diverse phenotypic consequences and implications for human disease. Although computational methods have been proposed to assess the effect of individual SNPs (52), relatively little has been done at the experimental level to provide insight into the effects of SNPs. The database of BRCA1/BRCA2 mutations provides an important first look at the kind of genetic diversity that exists in the human population. Another example of a large mutation database is that assembled for mutations found in p53 from human tumors. Detailed studies of the stability of the p53 DNA binding domain indicate that, like the BRCA1 BRCT domain, this relatively unstable domain is sensitive to mutations that destabilize the protein fold in addition to other classes of mutations that do not significantly affect protein fold stability but instead modify DNA-binding or proteinprotein interaction surfaces (53). Thus, rapid, general assays of protein structure and function such as the proteolysis assay used here could potentially prove useful for the rapid, high-throughput assessment of the effect of SNPs on protein function. Ultimately, unambiguous classification of variants, as proposed by the IARC Unclassified Genetic variants Working Group (41), will need a comprehensive integrative model to incorporate data from several sources. Robust integrative models have been proposed but they still lack a coherent framework to incorporate functional data. A major roadblock in the development of such a model was the lack of an established series of functional assays with standardized controls, validated against available genetic data. The present study fills this gap and allows for the development of better predictive models. Although a systematic large-scale classification of all BRCA1 variants into the IARC classes has not yet been done, it is expected that a large number will fall into class 3 (uncertain), due to the lack of genetic-based data. Conceivably, even large-scale efforts may not be able to gather sufficient genetic data on individual variants because most are extremely rare. Thus, we anticipate that the use of structural and functional information will be instrumental to move variants from the class 3 into classes that provide a basis for better risk assessment and clinical decisions. In addition, the fact that variants can now also be functionally and structurally rationalized will provide an impulse for further functional dissection of the tumor-suppressive activities of BRCA1.  Germ line inactivating mutations in BRCA1 confer susceptibility for breast and ovarian cancer. However, the relevance of the many missense changes in the gene for which the effect on protein function is unknown remains unclear. Determination of which variants are causally associated with cancer is important for assessment of individual risk. We used a functional assay that measures the transactivation activity of BRCA1 in combination with analysis of protein modeling based on the structure of BRCA1 BRCT domains. In addition, the information generated was interpreted in light of genetic data. We determined the predicted cancer association of 22 BRCA1 variants and verified that the common polymorphism S1613G has no effect on BRCA1 function, even when combined with other rare variants. We estimated the specificity and sensitivity of the assay, and by meta-analysis of 47 variants, we show that variants with <45% of wild-type activity can be classified as deleterious whereas variants with >50% can be classified as neutral. In conclusion, we did functional and structure-based analyses on a large series of BRCA1 missense variants and defined a tentative threshold activity for the classification missense variants. By interpreting the validated functional data in light of additional clinical and structural evidence, we conclude that it is possible to classify all missense variants in the BRCA1 COOH-terminal region. These results bring functional assays for BRCA1 closer to clinical applicability.  Go to: Introduction The breast and ovarian cancer predisposition gene BRCA1 displays large allelic diversity with several thousand different alleles documented thus far (Breast Cancer Information Core Database).16 This is consistent with the expected allelic structure of genes that determine rare monogenic diseases (1). A large portion of the documented alleles have had their disease association determined by inference from genetic and biochemical evidence that indicates that even very small truncations of 11 COOH-terminal amino acids result in a protein with compromised function (2\\u20134). Furthermore, deletion analysis suggests that truncation of only eight amino acids may abrogate function (5). However, a significant number of BRCA1 alleles, mostly containing missense changes, have proved more difficult to assess for disease association. These are termed unclassified variants (UCV) or variants of uncertain significance. The lack of conclusive genetic information is primarily due to low frequency of each individual UCV.  The problem generated by these UCVs is widely recognized and the need to provide risk assessment to individuals in high-risk families has brought several advances aimed at classifying these variants as follows: (a) Bayesian methods to analyze pedigrees (6); (b) use of information from interspecies sequence variation (7\\u20139); (c) integrated methods combining information from different sources in a comprehensive framework (10, 11); (d) functional assays to assess the effect of amino acid changes on protein function (12\\u201316); (e) methods based on co-occurrence with a deleterious mutations (17); and (f) structure-based analysis to generate computation prediction models (18, 19). Although still far from clinical application, these methods have provided important information.  Here we apply a transcription-based assay to assess the effect of 22 variants (R1443G, V1534M, D1546N, L1564P, P1614S, E1644G, S1655F, L1664P, T1700A, G1706E, V1713A, V1736A, G1738R, G1738E, R1753T, I1766S, L1764P, Q1785H, G1788D, E1794D, V1804D, and P1806A) on the activity of the BRCA1 COOH terminus (amino acids 1,396\\u20131,863) and interpret the findings in light of all available clinical, genetic, and structural information. These UCVs were found in individuals from families with breast or ovarian cancer in which no other deleterious mutation in BRCA1 or BRCA2 was found, and thus it is not clear whether they are causally related to disease. We also assessed the function of constructs representing the combination of the common polymorphism S1613G with other rare variants. The transcription-based method, used as a monitor for the integrity of the BRCT domain and its flanking regions, has shown an excellent agreement with existing genetic data (12, 13).  Several findings with significant implication for genetic counseling and risk assessment emerged from this study, which included a meta-analysis of 47 BRCA1 UCVs. The study determined tentative activity thresholds for functional classification of UCVs, identified regions less tolerant of mutations, and raised the testable hypothesis that there are no moderate-risk BRCA1 missense variants.  Go to: Materials and Methods Constructs  Control constructs containing the wild-type (wt) BRCA1, S1613G, M1775R, and Y1853X were previously described (12). Mutations R1443G, V1534M, D1546N, L1564P, P1614S, E1644G, S1655F, L1664P, T1700A, G1706E, V1713A, V1736A, G1738R, G1738E, R1753T, I1766S, L1764P, Q1785H, G1788D, E1794D, V1804D, and P1806A were introduced by splicing by overlapping extension PCR (20) using plasmid p385-BRCA1 as template. Primers sequences are available on request. For each mutation, both products (5\\u2032 and 3\\u2032 regions) were combined and used as a template for a final round of PCR using 24ENDTand UX13 primers (12). To obtain the double mutants S1613G/V1534M, the same procedure and primers described above were used with the pCDNA3\\u03b2HA-BRCA1 (gift from Ralph Scully, Beth Israel Deaconess Medical Center, Department of Medicine, Harvard Medical School, Boston, MA), which contains the S1613G polymorphism, as template. For the S1613G/H1402Y, S1613G/L1407P, S1613G/M1628V, and S1613G/T1685I double mutants, previously described primers (12) were used with the pCDNA3\\u03b2HA-BRCA1 plasmid as template. The final PCR products were then digested with BamH1 and EcoR1 and ligated to pLex9 or pGBT9 vectors. All mutations were confirmed by sequencing. To obtain GAL4-DBD fusions in a mammalian expression vector, pGTB9 constructs were digested with HindIII and BamH1, then a 1.8-kb band was isolated and ligated into equally digested pCDNA3.  Transcription assay in yeast and in mammalian cells  The transcriptional assays were done essentially as described (12, 21). Briefly, Saccharomyces cerevisiae strain EGY48 was cotransformed with the effector plasmid pLex9, which contains a fusion of LexA DNA binding domain and BRCA1 amino acids 1,396 to 1,863, with different variants and the plasmid reporter pRB1840, which contains a lacZ gene under the control of one LexA operator (22). At least three individual clones for each variant were tested for liquid \\u03b2-galactosidase assays using o-nitrophenyl-\\u03b2-d-galactopyranoside and the assays were done in triplicates. Activity was determined as a comparison to wt BRCA1 and S1613G (positive controls) or to M1775R and Y1853X (negative controls). For assays in mammalian cells, we used pG5Luc as a reporter and transfections were normalized with an internal control, phGR-TK (Promega, Madison, WI), which contains a Renilla luciferase gene under a constitutive TK basal promoter. Transfections were done with human 293T cells in triplicate using Fugene 6 (Roche, Indianapolis, IN), harvested 28 h posttransfection, and luciferase activity was measured using a dual luciferase assay system (Promega). Western blots were incubated with \\u03b1-GAL4 DBD mouse monoclonal antibody (Clontech, Mountain View, CA) or \\u03b1-LexA DBD rabbit polyclonal antibody (Upstate, Charlottesville, VA). We used the SAS application package to calculate confidence intervals for validation.  Structural analysis  For residues located at the BRCT domain, we did an analysis based on the 1t29 structure (23) and a sequence alignment of 13 BRCA1 orthologues created with SAM-T2K homology search software (24) and subsequently hand-edited. The species used in the alignment are Homo sapiens (AAA73985), Pan troglodytes (AAG43492), Gorilla gorilla (AAT44835), Pongo pygmaeus (AAT44834), Macaca mullata (AAT44833), Canis familiaris (AAC48663), Bos taurus (AAL76094), Monodelphis domestica (AAX92675), Mus musculus (AAD00168), Rattus norvegicus (AAC36493), Gallus gallus (AAK83825), Xenopus laevis (AAL13037), Tetraodon nigroviridis (AAR89523). We used the RenderByAttribute routine in the molecular visualization program Chimera to color each residue position according to its percent conservation in the alignment and to visualize patches of conserved residues on the BRCT domain surfaces (25). For each of the eight BRCT UCVs, we computationally replaced the wt side chain with that of the variant and optimized the conformation of the variant backbone and side chain atoms with the mutate model routine in MODELLER (26). Hydrogen bonds in wt and variant protein models were calculated with Chimera FindHBond routine using default parameters. A more detailed description of the method and a computational analysis of 36 BRCT UCVs that integrates features of protein tertiary structure, evolutionary conservation, and amino acid residue properties can be found in a companion paper to our current study (27).  Go to: Results Functional analysis of missense variants  The location of the variants studied, as well as the negative and positive controls, is indicated by arrowheads in Fig. 1. Two known deleterious/high-risk variants, M1775R and Y1853X, were used as negative (i.e., loss-of-function) controls and S1613G (a neutral polymorphism) and wt BRCA1 (amino acids 1,396\\u20131,863) were used as positive controls (12). Because African Americans and Hispanics may disproportionately receive uninformative results (28, 29), we paid particular attention to UCVs found in minority populations. Four variants (P1614S, T1700A, Q1785H, and E1794D) have been documented in African Americans, two (V1713A and G1788D) in Hispanics (ref. 30; Breast Cancer Information Core Database), and two (L1564P and V1804D) in both ethnic groups (30, 31).  Figure 1 Figure 1 Location of variants in BRCA1. Location of variants (closed black triangles) and of negative (open triangles, underlined) and positive (closed gray triangles, gray font) controls. Gray boxes, BRCT domains; DBD, GAL4 DNA binding domain. Six variants (R1443G, V1534M, D1546N, L1564P, P1614S, and E1644G) that lie upstream of the BRCTdomains were investigated for their effect on transcription (Fig. 2). This region displays relatively low conservation across other BRCA1 orthologues with no recognizable structural motif and was therefore expected to be more tolerant to changes (32). Variants R1443G, P1614S, and E1644G showed transcription activation levels equal or higher than wt BRCA1 whereas V1534M, D1546N, and L1564P had lower activity (between 60% and 80% of the wt activity) in yeast (Fig. 2A). In mammalian cells, variants V1534M, D1546N, P1614S, and E1644G showed transcription activation activity comparable to wt (within 1.5 SD) whereas variants R1443G and L1564P had reduced activity (\\u223c55% of wt; Fig. 2B).  Figure 2 Figure 2 Functional analysis of missense variants in BRCA1. A, quantitative transcriptional assay in yeast cells. Cells were cotransformed with a LexA-responsive \\u03b2-galactosidase reporter gene (diagram shown above the graph) and a LexA DBD fusion to residues ... Sixteen UCVs in the BRCTdomains (S1655F, L1664P, T1700A, G1706E, V1713A, V1736A, G1738R, G1738E, R1753T, L1764P, I1766S, G1788D, Q1785H, E1794D, V1804D, and P1806A) were also tested. Variants S1655F, T1700A, G1706E, V1713A, V1736A, G1738R, G1738E, R1753T, L1764P, and 1766S displayed markedly decreased activity with <40% of the wt activity in yeast and <20% in mammalian cells (Fig. 2A and B). The Q1785H and E1794D variants displayed activity comparable to wt in both yeast and mammalian cells. Variants L1664P and V1804D showed activity comparable to the wt in yeast but \\u223c60% to 80% in mammalian cells, whereas variant P1806A showed activity comparable to the wt in mammalian cells but reduced (64%) in yeast cells. Finally, variant G1788D showed between 30% and 40% activity in yeast or mammalian cells.  Expression levels were comparable for all variants in yeast cells (Fig. 2C). In mammalian cells, several variants showed decreased expression levels, suggesting instability of the protein product (Fig. 2D). Nevertheless, protein levels in mammalian cells seem to be a poor predictor of overall activity as V1534M or Q1785H, for example, showed very low levels but with activity comparable to wt (Fig. 2B and D).  As a first approach, we arbitrarily considered >80% and <40% of wt activity as the thresholds to classify the variants as neutral or deleterious, respectively (12). Using this threshold, our data indicate that (a) V1534M, P1614S, E1644G, Q1785H, and E1794D do not represent high-risk variants and are likely to be neutral; (b) S1655F, T1700A, G1706E, V1713A, V1736A, G1738R, G1738E, R1753T, L1764P, and I1766S represent deleterious/high-risk variants; (c) variants R1443G, L1546N, L1564P, L1664P, V1804D, and P1806A do not represent high-risk variants, although we cannot rule out the possibility that they may represent moderate-risk variants; and (d) variant G1788D does not represent a neutral variant, although we cannot rule out the possibility that it may represent a moderate-risk variant instead of a deleterious one. Importantly, despite variation in levels in yeast and mammalian assays, no variant presented clearly conflicting results (>80% in one test and <40% in another).  Analysis of double mutants  All previously published assays (3, 12, 13, 33) were done in the context of one molecular haplotype that corresponds to the wt sequence (designated as haplotype 1 in ref. 17). However, there are several frequent BRCA1 polymorphisms, such as S1613G, and of 10 common BRCA1 haplotypes, S1613G is present in five (17). Common haplotypes containing S1613G variants do not contribute to disease predisposition, and the presence of S1613G has been determined to have no effect on protein function (12, 33, 34). However, the role of S1613G has not been analyzed in the context of rare haplotypes. To examine whether the co-occurrence in cis of the S1613G polymorphism with other variants could affect activity, we combined it with neutral variant H1402Y, located at the coiled-coil motif, because it has been found in combination with S1613G (haplotype 2 in ref. 17). We also combined it with a predicted deleterious variant, L1407P, in the same motif (12). In addition, we arbitrarily chose M1628V, located in the vicinity of the polymorphism, and two other variants, V1534M and T1685I, located at approximately the same distance from the polymorphism. The presence of S1613G did not affect the activity of deleterious (S1613G/L1407P and S1613G/T1685I) or neutral (S1613G/H1402Y, S1613G/V1534M and S1613G/M1628V) variants (Fig. 3).  Figure 3 Figure 3 Functional analysis of double mutants in BRCA1. A, quantitative assay in human 293T HEK cells done as described in Fig. 2. No difference was found in the activity between pairs of rare variant and rare variant combined with S1613G polymorphism. B, protein ... Clinical data  Pedigrees corresponding to six variants analyzed here illustrate the difficulties in inferring causality even in large kindreds (Supplementary Fig. S1).  V1534M The proband in this Italian family (M933) was diagnosed with ovarian and bilateral breast cancer and tested positive for the variant. Loss of heterozygosity (LOH) and sequencing analysis in one of the proband's breast tumors revealed loss of the wt allele. We also analyzed DNA from a histologically cancer-free specimen of breast tissue from the sister diagnosed with breast cancer at age 40 years and it revealed wt alleles only. Therefore, the LOH analyses from these two individuals are not conclusive.  D1546N In family 230, four individuals were tested and two women with early-onset breast cancer do not carry the variant. However, it is possible that the variant derives from the mother's side of the family and the information from the father's side of the family is not relevant to conclude the absence of segregation.  P1614S The African American family 2593 presents with six cases of breast cancer and one ductal carcinoma in situ and also carries the H2116R BRCA2 UCV. The proband and her sister with ductal carcinoma in situ tested positive for the variant, suggesting that this variant could account for the disease.  E1664G In addition to the proband, we tested the mother who was found not to carry the variant. Excluding the possibility of nonpaternity, this suggests that the variant is neutral because no cancer was reported in the deceased father and his family.  V1736A Three specimens from two independent tumors from the proband of AUS, who had bilateral breast cancer, tested positive for the variant. Two specimens from the left breast tumor showed LOH of the mutant allele and one specimen from the right breast tumor showed no LOH.  V1804D For family 1008, only two individuals affected with breast cancer were tested and both were shown to carry the variant. The proband's paternal side on family 922 is Native American. All three individuals affected with breast cancer tested carry the variant.  Structural analysis  To structurally rationalize the loss or retention of BRCA1 transcriptional activation in our panel of missense variants, we have divided the variants located at the BRCT domains into four categories: (a) putative disruption of the BRCT fold hydrophobic core; (b) putative disruption of the BRCA1 BRCT interaction with phosphorylated protein partners; (c) putative disruption of binding sites on the protein surface; and (d) no evidence for functional effect (Table 1).  Table 1 Table 1 Structure-based analysis of BRCA1 variants in the BRCT domains Validation  To evaluate the specificity and sensitivity of the transcription assay, we identified missense variants in the COOH terminus of BRCA1 that have been classified as either deleterious or neutral based on (a) the co-occurrence with other deleterious variants because compound heterozygous mutation leads to embryonic lethality in humans with high statistical probability (9, 17) and (b) integrated methods using multifactorial likelihood models (9\\u201311, 17).17 The thresholds for the overall odds for or against causality in the likelihood models are arbitrary. Thus, we choose to use both highly stringent (neutral: overall combined odds >100:1 against causality; deleterious: overall combined odds >1,000:1 for causality; ref. 10) and less stringent (neutral: >10:1 against causality; deleterious: >10:1 for causality) thresholds. Fourteen variants were classified as neutral and 10 variants classified as deleterious using less stringent threshold (Table 2). A more stringent threshold reduces the number of classified neutral and deleterious variants to 13 and 6, respectively.  Table 2 Table 2 Cross validation of the transcription assay with genetic data The transcription assay correctly classified all 24 variants (Table 2; Fig. 4). Because our values for sensitivity and specificity are both 100%, there is no defined upper bound for our confidence intervals. We can exclude with 95% confidence that the sensitivity is not lower than 69% (based on 10 of 10 samples) or 54% (based on 6 of 6 samples; using a more stringent threshold) and the specificity is not lower than 77% (based on 14 of 14 samples) or 75% (based on 13 of 13 samples; using a more stringent threshold).  Figure 4 Figure 4 Missense variants and risk. A, hypothetical correlation between protein activity and cancer risk assuming a continuous relationship. B, hypothetical correlation between protein activity and cancer risk assuming a discrete relationship. C, functional analysis ... Missense variants as a group  One important question about risk determination using functional assays is how to interpret quantitative results. It is possible that risk is inversely correlated with protein activity in a continuous fashion (Fig. 4A). In that case, variants will present as a continuous series of high-risk, intermediate-risk, and low-risk variants. However, it is also possible that variants will be either high or low risk with virtually no moderate-risk variant found in the population (Fig. 4B). To begin to address this problem, we have plotted the activity (and its range of variation) of all the 47 variants tested thus far in the transcription activation assay in a quantitative manner (Fig. 4C). Next, we identified a validated deleterious variant that displayed the highest activity (R1699W; Table 2; Fig. 4C) and a validated neutral variant that displayed the lowest activity (L1564P; Table 2; Fig. 4C). Interestingly, these variants provide a very narrow intermediate range of activity (<50% and >45%). This result suggests that the assay classifies variants either as high or low risk with no variant considered as moderate risk.  Go to: Discussion Reliable classification of BRCA1 alleles containing missense changes remains a top priority for risk assessment for breast and ovarian cancer. Unfortunately, it seems clear that in the case of most rare variants, no single data source is informative enough to unambiguously classify them into neutral or deleterious (10, 19). To contribute to the classification of these problematic alleles, we have developed a functional test based on transcription activation that has provided a reproducible and standardized way to assess the functional effect of these variants on protein function (12, 13, 21). The transcription assay is a monitor for the integrity of the BRCT domain, and there is emerging evidence that it can also reliably predict functional effect in a region preceding the BRCT domains including the coiled-coil motif (3, 5, 12, 13, 33). Recent experimental evidence has indicated that BRCT domains are specialized motifs that bind phosphorylated peptides (23, 35\\u201341). Importantly, the BRCT region involved in phosphopeptide recognition colocalizes with a region that is critical for transcription activation in the heterologous system we use (18) and may underline a structural basis for the correlation between the transcription assay and the integrity of the BRCT domain.  One problem with a quantitative approach for classification is at which activity level one should draw a cutoff value. To address this problem, we analyzed all variants tested to date and identified the ones that had been classified using integrated methods. The identification of the neutral variant showing the lowest activity (L1564P) and of the deleterious variant with the highest activity (R1699W) led to the surprising finding that, given the small interval between these variants, there were only two classes (high risk and low risk) with no intermediate (moderate risk) class (Fig. 4). This supports the assumption used in the proposition of the integrated method that variants were either high or low risk (10). Whereas this may reveal a limitation of the assay, it may also reflect a biological property of BRCA1 alleles. Further studies are needed to investigate this notion with important implications for risk assessment.  Whereas it has been estimated that \\u223c13% of individuals undergoing testing receive uninformative reports due to the finding of an UCV, the picture is likely to be more somber for members of minority populations. In a recent study, it was shown that African Americans had a larger number of UCVs (46%) than Caucasians (12%; ref. 29). Therefore, we analyzed eight variants (L1564P, P1614S, T1700A, V1713A, Q1785H, G1788D, E1794D, and V1804D) that have been found in African Americans and/or Hispanics (30, 31, 42).  Using the threshold of activity defined in Fig. 4 (>50% for neutrals and <45% for deleterious), variants T1700A, V1713A, and G1788D displayed activity compatible with a deleterious classification, whereas L1564P, P1614S, Q1785H, E1794D, and V1804D displayed activity compatible with a neutral classification (Fig. 4). For the L1564P, P1614S, and V1804 variants, the functional results are consistent with co-occurrence analyses indicating that they are neutral variants (Table 1; ref. 17). Structural analysis supports a neutral classification for Q1785H and V1804D and a deleterious classification for T1700A, V1713A, and G1788D (Table 2). Interestingly, the T1700 residue is part of the phosphoserine binding pocket in which it makes hydrogen bonding interactions with the serine hydroxyl group, and it may play an important role in binding specificity (35, 36, 41). These results also highlight the difficulty in relying on limited pedigree data that may suggest cosegregation with the cancer phenotype (Supplementary Fig. S1).  We also did functional tests on 14 additional variants and functionally classified six (R1443G, V1534M, D1546N, E1644G, L1664P, and P1806A) as neutral and eight (S1655F, G1706E, V1736A, G1738R, G1738E, R1753T, L1764P, and I1766S) as deleterious (Fig. 4). The V1534M classification is supported by co-occurrence data (Table 2) whereas the D1546N classification is supported by clinical data, specifically from family 230 (Supplementary Fig. S1) in which at least two affected women have been shown not to carry the variant. Structural analysis is consistent with the classification of all BRCT variants (S1655F, L1664P, G1706E, V1736A, G1738R, G1738E, R1753T, L1764P, I1766S, and P1806A; Table 1). However, caution should be exercised in the classification until further evidence for these thresholds is obtained.  S1613G is a common polymorphism that displays a wide geographic occurrence but does not make a significant contribution to breast or ovarian cancer risk (34). Its high allele frequency, ranging from 0.2 to 0.55 in various populations, makes it probable that haplotypes may exist with S1613G in combination with other neutral or deleterious rare variants. The S1613G variant is present in 5 of the 10 most common BRCA1 haplotypes (17). This raised the question of whether there could be an interaction between these missense changes that confer different properties to these haplotypes. We investigated this by comparing naturally occurring and hypothetical haplotypes containing a combination of neutral and deleterious variants. We found that haplotypes containing Gly1613 were no different than haplotypes containing Ser1613 in terms of activity in the functional assays. Although we cannot rule out that the transcriptional assay may not be sensitive to detect small changes and that these haplotypes may behave differently in vivo, our results suggest that S1613G does not modify the risk conferred by other missense variants in cis. This, however, may not hold true for other common polymorphisms.  In conclusion, we have tested a series of common and rare UCVs of BRCA1 with an emphasis on variants found in minority ethnic groups. Combining functional, structural, and clinical information, we classified these alleles as either deleterious or neutral. We have also shown that the S1613G polymorphism does not alter the effect of other rare variants. Importantly, our analysis of 47 variants allowed us to define a tentative threshold of activity for classification and to propose the hypothesis that these rare missense variants confer either high or low risk but not moderate risk. If proved, this hypothesis will have profound implications for genetic counseling.  As a discipline, genetic counseling has seen a tremendous transformation. In the recent past, genetic counseling focused almost exclusively on pediatric syndromes that, by and large, are completely penetrant and do not present confounding phenocopies. Presently, issues of genetic counseling for cancer have come to the forefront. Data from Ontario illustrate this trend, with breast cancer consultations surpassing the number of consultations for all other reasons, including pediatric conditions and fertility issues (43). Many important problems surface about the determination of individual risk of cancer when incomplete penetrance and frequent phenocopies confuse the picture, as is the case with BRCA1. Thus, extensive family genotyping, complementary methods for detection, and classification of alleles become paramount.\",\n          \"Metastasis is the leading cause of death for gastric carcinoma (GC). An epigenetic biomarker panel for predicting GC metastasis could have significant clinical impact on the care of GC patients. The main purpose of this study is to characterize the methylation differences between GCs with and without metastasis.  Experimental Design  Genome-wide DNA methylation profiles between 4 metastatic and 4 non-metastatic GCs and their surgical margins (SM) were analyzed using methylated-CpG island amplification with microarray. The methylation states of 73 candidate genes were further analyzed in GC patients in a discovery cohort (n=108) using DHPLC, bisulfite-sequencing, and MethyLight. The predictive values of potential metastasis-methylation biomarkers were validated in GC patient cohorts in China (n=330), Japan (n=129), and Korea (n=153).  Results  The GC genome showed significantly higher proportions of hypomethylation in the promoter and exon-1 regions, as well as increased hypermethylation of intragenic fragments when compared to SMs. Significant differential methylation was validated in the CGIs of 15 genes (Ps<0.05) and confirmed using bisulfite-sequencing. These genes included BMP3, BNIP3, CDKN2A, ECEL1, ELK1, GFRA1, HOXD10, KCNH1, PSMD10, PTPRT, SIGIRR, SRF, TBX5, TFPI2, and ZNF382. Methylation changes of GFRA1, SRF and ZNF382 resulted in up- or down-regulation of their transcription. Most importantly, the prevalence of GFRA1, SRF, and ZNF382 methylation alterations was consistently and coordinately associated with GC metastasis and the patients\\u2019 overall survival throughout discovery and validation cohorts in China, Japan and Korea.  Conclusion  Methylation changes of GFRA1, SRF, and ZNF382 may be a potential biomarker set for prediction of GC metastasis.  Keywords: gastric cancer, metastasis, DNA methylation Go to: Introduction Gastric carcinoma (GC) is the second leading cause of cancer death throughout the world (1). Global statistics showed that in 2008 alone, nearly 989,000 people were diagnosed with GC, and approximately 464,000 people died from this disease (2). Currently, GC prognosis is primarily determined based on the clinical data and pathological stages of patients at the time of diagnosis and treatment (3). However, successful management of GC patients is still hampered by the lack of highly sensitive and specific biomarkers capable of predicting prognosis and likelihood of metastasis. Epigenetic alterations, including aberrant DNA methylation changes, may play an important role in gastric carcinogenesis as indicated by the increased hypermethylation of tumor suppressor genes in GC patients (4\\u20136). Given their important functions in cancer initiation and progression, methylation changes are being investigated as potential biomarkers for the early detection of cancers, the prediction of cancer progression, and the prediction of chemotherapeutic sensitivity (7).  Recent advances in high-throughput technologies have significantly expanded our capability of interrogating genome-wide DNA methylation changes in cancer (6, 8, 9). Methylated-CpG island amplification with microarray (MCAM) is one of the most powerful tools available for displaying differential methylation related to pathogenesis (10). A number of DNA methylome studies have been reported in a variety of primary cancers, including GC. However, few studies have been conducted to vigorously validate the methylation changes of the candidate genes at the single molecule level in numerous tumor samples (6, 11). Therefore, despite the long list of differentially methylated genes in GC patients, a promising DNA methylation biomarker has not yet reached to the clinical utility.  In the present study, genome-wide DNA methylation analysis using the MCAM assay was performed in GCs (10). A large number of differentially methylated regions were identified between GCs and their corresponding surgical margin (SM). In addition, differential methylation profiles between metastatic and non-metastatic GCs were identified. Most importantly, the methylation status of promoter CpG islands (CGIs) from 73 candidate genes was characterized using denatured high performance liquid chromatography (DHPLC) in 48 pairs of gastric samples from GC and non-cancer patients (12). The predictive values of three potential metastasis-related candidates were further validated in multiple cohorts from China, Japan, and Korea following the Reporting Prognostic Tumor Marker Study (REMARK) guidelines. We demonstrated that the methylation status of GFRA1, SRF and ZNF382 could be used as potential synergistic biomarkers for the prediction of GC metastasis.  Go to: Materials and Methods Patient characteristics and sample collection  A total of 504 patients with GC from 3 academic medical centers in China, Japan, and Korea were included in this study. The study was approved by the local Institution Review Boards (IRB) at each institution, and all patients were given written informed consent unless the IRB permitted a waiver. The 2003 UICC-TNM (tumor-node-metastasis) system was used for the classification of GCs (13). 330 Chinese GC inpatients that underwent surgical treatment at Peking University Cancer Hospital & Institute between 1999 and 2006 were enrolled in the discovery and validation cohorts based on the following criteria: a) availability of frozen, fresh GC and SM samples; b) follow-up available for at least 5 years; c) falls into the proper pathological TNM (pTNM) stages as described in the results section. In the validation cohort from Korea, 153 GC inpatients that received surgical treatment were selected from Seoul National University Hospital during 2004 with a follow-up of at least 3 years. Paraffin-embedded SM samples were used in the Korea study. The validation cohort from Japan included 78 GC inpatients that acquired surgical treatment between 1995 and 2002 with a follow-up of at least 5 years, as well as an additional 79 GC patients between 2010 and 2011 who did not have survival data. The SM samples were not available for these Japanese patients. GCs were classified as cardiac or non-cardiac in terms of location (14). Patients with pre-operative chemotherapy were not included in the discovery or independent validation cohorts. Normal/gastritis biopsies (NorG) from 56 outpatients at Peking University Cancer Hospital were used as the cancer-free controls.  Study design  The discovery patient cohort from Peking University Cancer Hospital consisted of 54 randomly selected patients with non-metastatic GCs and 54 matched patients with distant metastatic GCs. Among them, 8 paired GC and the corresponding SM samples from patients with or without distant and lymph metastasis were analyzed using MCAM on a customized Agilent promoter array. The clinical and histological features of these 8 patients can be found in Supplementary Table S1. The remaining GC and SM samples from 100 patients were used for the characterization of 73 CGIs using DHPLC and bisulfite clone sequencing. The methylation states of the three most promising candidate CGIs were analyzed in three analogous independent validation cohorts from China (n=222), Japan (n=129), and Korea (n=153). The overall study design is outlined in Supplementary Fig. S1. Genomic DNA was isolated using phenol/chloroform extraction.  Cell lines and culture  MKN74 cell line was kindly provided by Dr. Yasuhito Yuasa at Tokyo Medical and Dental University in 2010; RKO cell line, from Dr. Guoren Deng, at University California in San Francisco in 2001; AGS, by Dr. Chengchao Shou in 2009, HeLa and MGC803, by Dr. Yang Ke in 2004, at Peking University Cancer Hospital. All cells were grown in monolayer in appropriate medium supplemented with 10% FBS and maintained at 37\\u00b0C in humidified air with 5% CO2. These cell lines were tested and authenticated by Beijing JianLian Genes Technology Co., LTD before they were used in this study. STR patterns were analyzed using Goldeneye\\u212220A STR Identifiler PCR Amplification Kit. Gene Mapper v3.2 software (ABI) was used to match the STR pattern with the online databases of National Platform of Experimental Cell Resources for Sci-Tech for MGC803 cell and the American Type Culture Collection (ATCC) for other cells.  Genome-wide analysis of DNA methylation in GC tissues using MCAM  Genomic DNA (2 \\u03bcg) from 8 pairs of fresh GC and SM samples was analyzed using the MCAM approach (10). Briefly, genomic DNA was digested consecutively with SmaI and XmaI which cut unmethylated and methylated CCCGGG sites, respectively. The XmaI digestion produces sticky ends that can be ligated to linkers, while SmaI digestion results in blunt ends that are unable to be ligated to linkers. The ligation-mediated PCR products from GC and SM samples were purified and labeled with Alexafluor647 or 555, respectively, using the Bioprime Plus Array CGH Indirect Genomic Labeling kit (Invitrogen, Carlsbad, CA) according to the manufacturers\\u2019 instructions. The labeled DNA was co-hybridized to a custom designed Agilent Oligonucleotide Array, and the slides were washed and scanned as described previously (15). Data was extracted using the Feature Extraction Tool (Agilent Technologies, Santa Clara, CA) and exported for further analysis. The custom designed Agilent oligonucleotide array was designed using Agilent eArray service (https://earray.chem.agilent.com/earray). The array consisted of approximately 99,028 probes (44\\u201360mers) that covered 29,879 in silico SmaI-digested DNA fragments (>60bp and <2000bp) in the human genome. The probes were tiled within each fragment with 100bp spacing. The methylation states of 6177 genes were determined using this custom methylation array.  Microarray data normalization and probe/gene selection  The raw array data was processed and normalized by the Beijing CO-FLY Bioinformatic Company. Background model adjustment was carried out using the minimum normalization algorithm. Systematic differences between arrays were normalized using the quantile method as described (16, 17). The methylation array data, as well as the probe information, have been deposited into the Gene Expression Omnibus under accession number GSE47724.  The mean intensity of the normalized array hybridization (methylation) signal of each probe for sex-related chromosomes and autosomes in the SM samples from 4 males and 4 females (Supplementary Fig. S2A\\u2013C) were analyzed. As expected, the intensities of 784 of 2390 X chromosome-probes (32.8%) were significantly higher in the female samples than the male samples (Student\\u2019s t-test, P<0.05); in contrast, 35 of 87 Y chromosome-probes (40.2%) were significantly higher in the male samples when compared to the female samples. These sex-specific differences were only observed in 1250 of 96,550 (1.3%) probes in the 22 autosomes. These results confirmed that the quality of the normalized data is sufficient to differentiate sex-specific DNA methylation and suitable for studying GC- or metastasis-related methylation changes.  The methylation signal ratio ([GC]/[SM]) was calculated for each array probe. The Student\\u2019s paired t-test (P<0.01) was used to identify the differentially methylated probes between GC and SM samples from the 8 patients analyzed. The Mann-Whitney U-test (P<0.029) was used to identify the metastasis-specific differentially methylated probes between the 4 metastatic and 4 non-metastatic GC patients. The methylation ratio data including the adjusted P-values for each probe are included in the data file S1.  The difference between GC-related hypermethylated and hypomethylated probes was calculated for each sliding window (sequence or region) using 51 probe-matched fragments, which included the target probe along with 25 probes both upstream and downstream of the target. Probes near the centromeres and telomeres of each chromosome were not included due to the absence of the 25 upstream or downstream probes. The numerical differences for 99K probes were charted to display the detailed regional methylation trend (or net methylation signal) for the corresponding chromosome arm.  Identification of differentially methylated candidate genes  In order to identify GC and metastasis-specific differentially methylated candidate genes for further evaluation, the promoter and exon-1 regions were focused on due to their known inverse correlation to epigenetic repression of gene transcription. The differentially methylated probes in these regions were defined as the top-100 probes and used in hierarchical clustering analysis and preparation of a heatmap, when their P-values were less than 0.05, and their absolute mean difference-values were within the top 100. Candidate genes were selected from these GC- or metastasis-related probes according to their function information in the public databases.  Hot-start PCR and DHPLC analysis  CpG-free universal primer sets and bisulfite-modified DNA (18) were used to amplify the genes of interest. The PCR reaction mixture (30 \\u03bcl) included 20 ng DNA template, 0.15 mmol/L dNTP, 0.15 \\u03bcmol/L of each primer, and 0.9 U of HotStart Taq DNA polymerase (Qiagen GmbH, Hilden, Germany). The PCR products were then analyzed quantitatively by DHPLC using the WAVE\\u00ae DNA Fragment Analysis System (12, 19). PCR products of hypermethylated and hypomethylated genes were separated using a DNASep\\u00ae analytical column (Transgenomic) at the corresponding partial denaturing temperature as listed in the supplementary materials and methods). M.SssI-methylated genomic DNA, obtained from blood samples, was used as a positive control. A sample containing a methylated PCR product peak was defined as methylation-positive and used to calculate methylation positive rate [ratio of methylation-positive sample number to total sample number]. The peak areas corresponding to the methylated and unmethylated PCR products were used to calculate the percentage of methylated copies [proportion of hypermethylated copies = methylation-peak area/ total peak area] for each gene analyzed.  MethyLight  The methylation states of GFRA1, SRF, and ZNF382 were determined using the MethyLight assays. Gene-specific probes labeled with 6FAM and TAMRA were employed to quantify the relative copy number of methylated alleles compared to the COL2A1 control (20). The sequences of the primer set and gene-specific probes can be found in the supplementary materials and methods.  Statistical analysis  The SPSS 16.0 Trend-test and Pearson\\u2019s Chi-square test were used to analyze the difference in methylation frequency between GC and SM samples and between metastatic and non-metastatic GC samples. Student\\u2019s paired t-test, Kruskal-Wallis H-test, and One-Way ANOVA were used to identify differentially methylated regions between the different groups of samples. The Mann-Whitney U-test and Student\\u2019s t-test were used to analyze the association between the percentage of methylated copies and the clinicopathological features. All statistical tests were two-sided, and P<0.05 was considered statistically significant. The cutoff value was calculated according to the receiver operating characteristic (ROC) curve using the percentage of methylated copies to predict GC metastasis. The log-rank test was used to compare survival time between groups. Cox-proportional hazards models were used to identify independent predictors of survival (month) with adjustment for relevant clinical covariates. Functional annotation of the differentially methylated regions was performed using EpiExplorer (21).  Go to: Results Genome-wide analysis of GC-related differential DNA methylation  In order to identify differentially methylated genes related to GC development and metastasis, genome-wide DNA methylation analysis was conducted in 8 pairs of GC and SM samples using the MCAM assay utilizing a 99K custom-designed Agilent oligonucleotide microarray as described above (10). Through this method, 9860 probes in 4047 genes were identified with significant methylation differences between the 8 GC and 8 SM samples (Paired t-test, P<0.01). Of the differentially methylated probes, 4177 showed hypermethylation (42%; [GC]>[SM]), while the remaining probes were hypomethylated (58%; [GC]<[SM]) (Data file S1). Nearly half of the hypomethylated probes (49%) were found to be within a 102~3 bp region of the transcription start site (TSS), while 42% of the hypermethylated probes were within a 103~4 bp region of the TSS (Supplementary Fig. S2D; P<0.0000001). When compared to the hypermethylated probes, the hypomethylated probes showed a considerably higher GC content than the hypermethylated ones (Median, 0.68 vs. 0.50), indicating the hypomethylation lies mainly in typical CGIs (Supplementary Fig. S2E). The promoter and exon-1 regions showed significantly higher proportions of hypomethylation to hypermethylation (26.8% vs. 23.4% for the promoter, P=1.2\\u00d710\\u22124; 13.8% vs. 3.0% for exon-1, P=5.5\\u00d710\\u221275) in GCs compared to SMs. The opposite trend was seen in the intragenic regions, which showed a significantly lower proportion of hypomethylation to hypermethylation (32.1% vs. 41.5%, P=5.4\\u00d710\\u221222) (Fig. 1A). A heatmap displaying the top-100 differentially methylated probes between GCs and SMs in the promoter and exon-1 regions is provided in Fig. 1B.  Figure 1 Figure 1 Distribution of probes with significant GC-related differential methylation changes in the human GC genome Most GC-related differentially methylated probes were clustered in specific chromosomal regions, especially sub-telomeric regions (Data File S2). Although the presence of SmaI/XmaI restriction sites primarily determined the distribution patterns of probes with GC-related methylation changes, certain chromosomal locations showed increased hypomethylation with little to no overlapping hypermethylation. After being normalized with respect to the probe density, chromosomes 7, 8, and 20 were clearly shown to harbor multiple long-range hypermethylated domains. In contrast, most regions in chromosomes 3, 4, 14, 15, and 18 were found to be more favorable to long-range hypomethylation (Fig. 1C).  Genome-wide analysis of GC metastasis-related differential DNA methylation  Among the 8 pairs of GC and SM samples analysed, half were metastatic GCs and the other half were sex-, age-, location-, and differentiation-matched non-metastatic control GCs (Supplementary Table S1). The MCAM analysis identified 8553 probes that were differentially methylated between the metastatic and non-metastatic GC groups (Mann-Whitney U-test, P<0.029). Among these metastasis-related candidate probes, 623 probes corresponded to 480 genes that overlapped with the GC-related genes identified above. A heatmap displaying the top 100 metastasis-related, differentially methylated probes is provided in Fig. 1D.  Identification in 15 GC-related aberrantly methylated genes  From the list of differentially methylated CGIs, 63 candidate genes were selected for further analysis based on their known functions and statistical significance of differential methylation signals between metastatic and non-metastatic GCs or between GCs and SMs (Supplementary Table S2). Ten known tumor-related genes that were not included in the oligonucleotide array were also selected as complementary and control genes for the validation study. The CGIs of these 73 genes were amplified using CpG-free primer sets. The bisulfite-PCR products were then analyzed using DHPLC to quantify the methylation levels of these CGIs in the 8 paired GC and SM samples (Fig. 2A; Data file S3). Differential methylation was observed in 37 CGIs between the 8 pairs of samples (Supplementary Table S2, underlined). The methylation levels of these 37 CGIs were further examined in additional 40 pairs of GC and SM samples, as well as 56 NorG samples. Significant differential methylation between paired GC/SM and NorG samples was observed in 15 CGIs (Ps<0.05; Table 1). The number of samples with hypermethylated CGIs in the promoter and exon-1 of BMP3, BNIP3, ECEL1, HOXD10, KCNH1, PSMD10, PTPRT, SRF, TBX5, TFPI2, and ZNF382 gradually increased from the NorG \\u2192 SM \\u2192 GC samples (Trend or Chi-square test, Ps<0.040). These results suggest that hypermethylation of these 11 genes may play significant roles in GC development. Furthermore, the GC samples showed a significantly higher percentage of hypermethylated CDKN2A and GFRA1 (Ps<0.050) and significantly lower levels of methylation in ELK1 and SIGIRR when compared to the SM samples.  Figure 2 Figure 2 DNA methylation of GFRA1, SRF, and ZNF382 in GC samples Table 1 Table 1 Prevalence of CGI methylation in gastric mucosa samples containing various pathological changes from GC patients and non-cancerous control patients The positive rate of methylation in CDKN2A and PSMD10 was significantly higher in the GC and SM samples than was seen in the NorG samples. In contrast, the positive-rate and proportion of methylated ELK1 and GFRA1 in the NorG samples were strikingly higher than in the GC and SM samples, indicating that hypomethylation of these genes occurs in gastric carcinogenesis as field effects. Furthermore, the positive rates of BNIP3, KCNH1, and ZNF382 methylation in the GC samples were more than 3-times higher than the SM and NorG samples (29% vs. 7~4%, 42% vs. 4~14%, and 69% vs. 18~23%, respectively). Based on this information, these genes are most likely involved in GC-specific methylation changes.  The methylation states of these CGIs were further confirmed using traditional bisulfite sequencing. The bisulfite sequencing results were consistently in agreement with the DHPLC analysis (Fig. 2B and Supplementary Fig. S3). In addition, quantitative MethyLight assays using fresh or formalin-fixed paraffin-embedded tissue samples further validated the DHPLC results (Supplementary Fig. S4A; Spearman test, P<0.020). To understand if methylation changes in these CGIs affect gene expression, the mRNA levels of SRF, ZNF382, and GFRA1 were analyzed in matched tissue samples using quantitative RT-PCR. The qRT-PCR results showed that mRNA expression of all three genes was inversely correlated with the prevalence of methylation in their CGIs (Supplementary Fig. S4B; Spearman test, P<0.050).  Confirmation of GC metastasis-related DNA methylation markers  Among the above 48 pairs of GC and SM samples, 24 pairs were from patients with lymphatic and distant metastasis, and 24 pairs were from sex-, age-, location-, and GC differentiation grade-matched patients without metastasis. Thus, the methylation states of these 15 CGIs were further analyzed to determine if they are associated with GC metastasis. DHPLC results showed that the methylation states of the BMP3, GFRA1, SRF, and ZNF382 CGIs were significantly different between metastatic and non-metastatic GC samples. The proportion of methylated BMP3 and GFRA1 was lower in the metastatic GC samples than the non-metastatic GC samples (median, 1.8% vs. 5.9%; 8.6% vs. 38.6%; Mann-Whitney U-test, Ps<0.040). The positive rate of SRF and ZNF382 methylation was also lower in the metastatic GC samples than the non-metastatic GC samples (4% vs. 33%; 54% vs. 79%, P=0.020/0.066). Therefore, the relationship between GC metastasis and methylation of these four CGIs was tested in additional GC and SM samples obtained from Chinese patients (n=50~60). When these samples were taken together as a discovery cohort, the relationship between GC lymph/distant metastasis and the methylation changes in GFRA1, SRF, and ZNF382 was statistically significant (Table 2); however, such an association was not observed for BMP3 (data not shown).  Table 2 Table 2 SRF, ZNF382, and GFRA1 methylation prevalence comparison in SM and GC samples from Chinese patients in the discovery cohort with various clinicopathological characteristics To investigate whether the methylation status of the three potential biomarkers mentioned above had an impact on overall survival, Kaplan-Meier analysis was performed on each gene individually. Results showed that the overall survival of GC patients with GFRA1 or ZNF382 methylation-high (cut-off value: percentage of methylated copies >26.4% for GFRA1 or 1.3% for ZNF382) or SRF methylation-positive was elongated when compared to methylation-low or methylation-negative patients in the discovery cohort (log-rank test, P=0.068, Fig. 3A; P=0.010, Fig. 3B; P=0.001, Fig. 3C, respectively). Sub-stratification analysis revealed that SRF methylation was only correlated with overall survival in patients with non-cardiac GCs (P<0.033) but not with cardiac GCs (P=0.146). Therefore, only non-cardiac GC patients were included in the survival analysis in the following SRF methylation validation cohorts.  Figure 3 Figure 3 Kaplan-Meier survival curves of GC patients with different GFRA1, SRF, and ZNF382 methylation states The predictive value of these methylation markers for GC metastasis was further confirmed using three independent validation cohorts in China (n=222), Japan (n=129), and Korea (n=153). Because the proportion of both methylated and unmethylated alleles of CGIs can be quantitatively and simultaneously determined using DHPLC, this method was consistently used to detect the methylation levels within these CGIs in freshly-frozen gastric samples from Chinese and Japanese patients. However, MethyLight was used to analyze the paraffin-embedded samples from the Korean patients, as fresh samples were not available. Results from these cohorts showed that the methylation positive rates of GFRA1, SRF, and ZNF382 were inversely and significantly correlated with pTNM stage and lymph metastasis in all three cohorts (Table 3). Kaplan-Meier analysis also showed that the overall survival of GC patients with higher methylation levels of GFRA1 and SRF CGIs was consistently longer than those without methylation of these two genes across all three validation cohorts (Fig. 3A and 3B). However, correlation between ZNF382 methylation and overall survival of GC patients was not statistically significant in all three validation cohorts (Fig. 3C). These results indicate that ZNF382 methylation may be a weak GC metastasis biomarker when compared with GFRA1 and SRF methylation.  Table 3 Table 3 Comparison of SRF, ZNF382, and GFRA1 methylation positive rates in GC patients with various clinicopathological characteristics in the Chinese, Japanese and Korean validation cohorts* In addition, after adjustment for age, sex, differentiation, location, pTNM stage, and vascular embolus, GFRA1 or SRF methylation was still an adequate prognostic indicator in multivariate analysis among all patients in these validation cohorts (hazard ratios [HR]=0.543 or 0.395; 95%CI [0.304\\u20130.938] or [0.165\\u20130.945]; n=300 or 452).  Sub-stratification analysis showed that the overall survival of stage-I&II GC patients with methylated SRF was significantly elongated when compared to SRF methylation-negative patients in all four cohorts (HR=0.357, 95%CI [0.164\\u20130.778], n=198). Similar difference was also observed for GFRA1 or ZNF382 methylation-high, but not statistically significant (HR=0.608 or 0.498, 95%CI [0.336\\u20131.099] or [0.243\\u20131.023], n=173 or 167). Among GC patients from Korea whose histological types of GCs were available, GFRA1 methylation-high was significantly associated with low-risk of metastasis of both intestinal- and diffuse-types of GCs (positive rate: 82.4% and 76.5% for non-metastatic GCs; 43.2% and 43.1% for metastatic GCs, Ps<0.05). GFRA1 methylation-high was also significantly correlated with longer overall survival of diffuse-type GC patients (HR=0.482, 95%CI [0.247\\u20130.938], n=67). However, ZNF382 methylation-high was significantly associated with low-risk of metastasis of intestinal-type GCs (93.8% vs. 62.5%, P=0.036), but not diffuse-type GCs.  Synergic analysis of three methylation markers  In order to investigate if a combination of the methylation markers (GFRA1, SRF, and ZNF382) has a synergistic effect on predicting GC metastasis, the merged data was reanalyzed in the above 4 patient cohorts. As expected, the number of patients with one or more methylated genes among the three-gene panel was significantly decreased in GC samples with lymph/distant metastasis (Fig. 3D left; linear-trend test, P<0.00001; one gene vs. two genes, P=0.046). The sensitivity and specificity of 2~3 positive-methylation changes of 3 genes for detection of non-metastatic GCs were 60% and 67%, respectively. The positive and negative predictive values were 57% and 69%. In addition, multivariate analysis also showed that the number of combined methylation changes of GFRA1, SRF, and ZNF382 was an independent predictor of overall survival for GC patients (n=246) after adjusting for the pTNM stage, GC location, differentiation, vascular embolus, age, and sex (HR=0.734; 95%CI [0.562\\u20130.958]) (Fig. 3D right). The pTNM stage and GC location were also independent survival factors (HR=3.608; 95%CI [2.648\\u20134.917] and HR=2.723; 95%CI [1.608\\u20134.613], respectively). These results suggest that using a combination of this three-gene panel may function as a synergic biomarker set for predicting GC prognosis.  GFRA1, SRF, and ZNF382 expression changes in gastric carcinogenesis  The protein expression of the three genes in the paired GC and SM samples in both regular tissue sections and tissue microarray (TMA) were analyzed using the IHC assay as described in the supplementary methods (22). IHC analysis revealed that GFRA1 expression was predominantly observed in the cytoplasm of stromal cells, especially in the vessel cells in GCs (Supplementary Fig. S5A). Among 38 pairs of IHC-informative cases, the proportion of GCs with strong GFRA1-staining was significantly higher than SMs (24/38 vs. 12/38, P<0.01). Among 28 pairs of informative cases, the proportion of GCs with strong ZNF382-staining in epithelial cells was lower than SMs (4/28 vs. 11/28, P<0.07) (Supplementary Fig. S5B). Statistically significant association was not observed between GFRA1 (or ZNF382) staining and clinical parameters, such as invasion, lymph metastasis, embolus, differentiation, and overall survival. SRF-staining was only observed in the nucleus of some stromal fibroblasts and smooth muscle cells in both regular GC and SM sections (Supplementary Fig. S5C). Therefore, SRF expression was not further examined using TMA.  Go to: Discussion Over a four-year period, a comprehensive epigenetic biomarker discovery and validation study involving over 500 patient samples from three large academic medical centers in China, Japan, and Korea had been conducted. The biomarker discovery effort started off with a genome-wide analysis of differentially methylated genes between metastatic and non-metastatic GCs in a small number of patient samples. The microarray-based methylation profiling identified a large number of GC-specific and metastasis-specific candidate genes that were differentially methylated. From the list of differentially methylated genes, a step-by-step elimination process identified a 15-gene panel associated with GC/metastasis-specific DNA methylation changes. The 15 genes were validated using multiple independent methods from a discovery cohort of GC patient samples. Finally, a methylation biomarker-set consisting of GFRA1, SRF, and ZNF382 was validated for the prediction of GC metastasis and patients\\u2019 overall survival in four cohorts from China, Japan, and Korea. This novel epigenetic biomarker set may be used in the decision making process for personalized post-operational therapy. To our knowledge, this is the first such study which specifically focuses on the metastasis of gastric cancer.  A large number of genome-wide DNA methylation studies have been reported for many different tumor types in recent years (6, 10, 11). However, most of the studies failed to perform large-scale and in-depth follow-up studies to validate the candidate genes discovered through the genome-wide analyses. As a result, few methylation markers have been developed from the large number of DNA methylation studies published so far. The present study represents the most comprehensive and quantitative characterization of DNA methylation biomarkers in GC to date. Moreover, the three methylation biomarkers associated with GC metastasis and patients\\u2019 survival were validated not only in multiple cohorts, but also in freshly frozen and paraffin-embedded samples using several independent methods such as DHPLC and MethyLight. The vigorous testing performed in this study ensures the high reliability and feasibility of these novel biomarkers in different clinical settings.  It has been previously reported that 2,540 of 17,800 tested genes are differentially expressed between 80 pairs of GC and SM samples. Furthermore, it was found that there are four times as many upregulated genes in GCs than there are downregulated genes (1983 vs. 557; GSE27342) (23). Therefore, the frequent DNA hypomethylation in the promoter and exon-1 regions of the GC methylome observed in this study may account for the prevalent increase in gene expression. In fact, an increasing number of studies have reported reactivation of proto-oncogenes by DNA hypomethylation in several cancers (24\\u201326).  Long-range epigenetic silencing and large epigenetic structures have been reported in different cancers (27\\u201329). Differential long-range hypermethylation and hypomethylation trends may be related to cancer/tissue-specific DNA methylation (11, 30). In the present study, it was found that chromosomes 7, 8, and 20 appeared more favorable for long-range hypermethylation (or amplification of methylated-regions). In contrast, chromosomes 3, 4, 14, 15, and 18 had an affinity for long-range hypomethylation (or deletion of methylated-regions). Further studies are warranted to determine which of these long-range hypermethylated and hypomethylated- regions are GC-specific changes and which are changes across cancer-types.  Most of the 15 aberrantly methylated genes identified in GCs are involved in cell proliferation, differentiation, apoptosis, adhesion, and embryonic development (Supplementary Table S2). Previous reports have demonstrated that silencing of BMP3, BNIP3, CDKN2A, HOXD10, TFPI2, and ZNF382 via methylation correlates with both the development and progression of cancers (12, 19, 31\\u201334). Similar associations were also observed in GC samples used in the present study. Methylation changes of KCNH1, PSMD10, and SRF in cancer tissues have not previously been reported. Furthermore, BNIP3, KCNH1, and ZNF382 methylation levels were more than 3-times higher in the GC samples than in SM and NorG samples. It is needed to study whether methylation of these genes may affect their expression states in gastric carcinogenesis. In addition, though TBX5 and ELK1 methylation is not associated with GC metastasis, the overall survival of GC patients with methylated TBX5 or ELK1 was longer than those without methylation (P=0.017 or 0.003; data not shown). Because some methylation changes may occur in both GC and SM samples from cancer patients, more GC-related methylation changes could potentially be identified if the NorG samples were used as the normal stomach reference.  Among three genes identified with GC development- and metastasis-related methylation changes, GFRA1 is a cell surface GDNF/neurturin receptor and a tyrosine kinase that is normally expressed in the nervous system and kidney. However, this gene is over-expressed in gut neural crest stem cells and in many cancers (35\\u201341). The present study provides the first evidence that hypomethylation of GFRA1 CGIs may account for its overexpression in cancers. SRF is a master regulator of myogenesis and multiple cellular processes including cell proliferation and migration. Furthermore, SRF is known to play important roles in the epithelial-mesenchymal transition and experimental invasion through cancer and stromal cells (42\\u201348). The present study shows, for the first time, that methylation in the exon-1 region of its CGIs may epigenetically inactivate SRF transcription. Most importantly, we found that SRF methylation was correlated with overall survival in non-cardiac GC patients, but not in cardiac GC patients. It is well known that H. pylori infection increases risk of non-cardiac GC, but not cardiac GC (49). The incidence of cardiac GC is also gradually increased in Western countries coincided with a decrease in prevalence of H. pylori infection, (50). Therefore, whether H. pylori infection contributes to SRF methylation and its biological subsequence warrants future study.  GFRA1 and SRF are two crucial genes in the GDNF-GFRA-RET-RAS-MEK-ERK-ELK-SRF pathway involved in cell migration and cancer invasion (37, 41, 46\\u201348). Therefore, epigenetic alterations of GFRA1 and SRF may play important roles in GC metastasis through modulating this important pathway. ZNF382 is a candidate tumor suppressor gene, and its methylation is associated with GC development (34). However, its link with cancer metastasis has not previously been reported. In the present study, it was found that the methylation status of GFRA1, SRF or ZNF382 was consistently and significantly associated with GC metastasis and patients\\u2019 overall survival in multiple cohorts from different populations, suggesting that they may be used as potential biomarkers for predicting GC metastasis and prognosis. Most importantly, the combination of the three markers was not only identified as an independent survival factor, but also as a strong synergistic biomarker set helping to distinguish metastatic GCs from non-metastatic GCs. The TMA analysis of GFRA1 and ZNF382 from 40 GC patients failed to demonstrate statistically significant association of their protein expression with clinicopathological parameters and overall survival of these patients; however, up-regulation of GFRA1 protein and down-regulation of ZNF382 was indeed observed in the GCs compared to SMs, which is in agreement with hypo- and hyper-methylation of GFRA1 and ZNF382 observed in GCs. Our results suggest that DNA methylation analysis mighty be a more suitable diagnostic tool than IHC for these genes. To further prove the clinical utility of this marker panel on early prediction for GC metastasis, a prospective followup study among non-metastatic GC patients is being conducted.  In conclusion, through a comprehensive and collaborative epigenetic biomarker discovery effort, we have demonstrated that the DNA methylation changes of GFRA1, SRF, and ZNF382 were coordinately associated with GC metastasis and overall patient survival, and this three gene panel has potential to be used as a synergistic biomarker set capable of improving the prognosis and treatment for GC patients.   Statement of translational relevance  Gastric carcinoma (GC) is the second leading cause of cancer deaths in the world, with many occurring in East Asia. To identify DNA methylation biomarkers for prediction of GC metastasis, scientists and oncologists from China, USA, Japan, and Korea have carried out a five-year collaborative study to profile differential methylation patterns in metastatic and non-metastatic GCs and perform an in-depth characterization of methylation changes in the CpG islands of 73 candidate genes. From this study, we established a methylation biomarker-set composed of three genes GFRA1, SRF, and ZNF382 that could be used to synergistically predict GC metastasis and patients\\u2019 overall survival from multiple patient cohorts in China, Japan, and Korea. The established marker set will be a useful clinical tool for decision-making on personalized post-operational therapy that is currently not available.  STAT3 overactivation is a common event in many cancers, including head and neck squamous cell carcinoma (HNSCC), where STAT3 represents a promising therapeutic target. HNSCC is not characterized by frequent kinase mutations, in contrast to some malignancies where mutational activation of kinases upstream of STAT3 is common. Instead, STAT3 may be activated by loss-of-function (LOF) of negative regulators of STAT3, including by promoter hypermethylation of PTPRT. Here we first analyzed The Cancer Genome Atlas (TCGA) data and determined that the PTPRT promoter is frequently hypermethylated in several cancers, including HNSCC (60.1% of tumors analyzed) in association with downregulation of PTPRT mRNA expression and upregulation of pSTAT3 expression. These findings were confirmed in an independent cohort of HNSCC tumors by methylation-specific PCR (MSP) and immunohistochemistry. We demonstrate that PTPRT promoter methylation and gene silencing is reversible in HNSCC cells, leading to PTPRT-specific downregulation of pSTAT3 expression. We further show that PTPRT promoter methylation is significantly associated with sensitivity to STAT3 inhibition in HNSCC cells, suggesting that PTPRT promoter methylation may serve as a predictive biomarker for responsiveness to STAT3 inhibitors in clinical development.  Keywords: head and neck cancer, phosphatase, methylation, biomarker Go to: Introduction Head and neck squamous cell carcinoma (HNSCC) is an invasive malignancy with more than 40,000 expected diagnoses and more than 8,000 expected deaths in 2014 in the United States alone. 1 Most cancers, including HNSCC, are characterized by constitutive activation of signal transducer and activator of transcription 3 (STAT3) via phosphorylation of tyrosine 705 (Y705). STAT3 is a transcription factor and potent oncoprotein that activates or maintains many cancer phenotypes including abnormal growth, proliferation, angiogenesis, and invasion as well as evasion of apoptosis and the anti-tumor immune response. 2 While aberrant activation of kinases upstream of STAT3 likely contributes to constitutive activation of STAT3 in cancer, the role of loss-of-function (LOF) of downstream regulators, especially protein tyrosine phosphatases (PTPs), remains incompletely understood.  The PTP superfamily is a large group of enzymes that, in close coordination with tyrosine kinases, tightly regulates diverse signaling pathways by catalyzing the removal of a phosphate group from specific signaling proteins. The largest class of PTPs in the human genome is comprised of the receptor-like PTP (PTPR) family, which contains 21 distinct members. PTPR family members span the membrane once and contain one or two intracellular catalytic domains, as well as a modular extracellular region that typically contains several protein-protein interaction domains. A role for PTPRs in the context of cancer is increasingly apparent, with many members implicated as oncogenes or tumor suppressor genes. 3 We recently reported that members of the PTPR family are frequently somatically mutated across all solid tumors analyzed, including HNSCC, with PTPR type T (PTPRT) representing the single most commonly mutated PTP in HNSCC and across all cancers. 4 Importantly, pSTAT3 is a validated direct substrate of PTPRT, and LOF mutations of PTPRT lead to increased pSTAT3 expression and enhanced HNSCC cell growth/survival. 4, 5 As these LOF PTPRT mutations are found in relatively few HNSCC tumors (5.6% of tumors analyzed), we sought to determine if loss of PTPRT expression by aberrant promoter methylation contributes to pSTAT3 overexpression and sensitivity to STAT3 inhibition in HNSCC.  Herein, we report a high frequency of aberrant PTPRT promoter hypermethylation in HNSCC and other cancers. We demonstrate that PTPRT promoter methylation significantly downregulates PTPRT expression, with an associated increase in expression of the PTPRT substrate pSTAT3 in HNSCC. We show that this methylation is reversible, leading to specific downregulation of pSTAT3 in HNSCC cells. Further, we demonstrate a correlation between PTPRT promoter methylation and sensitivity to STAT3 inhibition in HNSCC cell lines, suggesting that PTPRT methylation may serve as a predictive biomarker of responsiveness to STAT3 inhibitors currently in clinical development.  Go to: Results Frequent PTPRT promoter hypermethylation leads to decreased PTPRT mRNA expression  To assess aberrant PTPRT promoter methylation in HNSCC, we analyzed TCGA data derived from the Illumina HumanMethylation450 platform. We first determined which CG dinucleotide methylation event was most negatively correlated with PTPRT mRNA expression (Figure 1A). We then defined aberrant hypermethylation as a fractional methylation level (beta value) at least three standard deviations above the mean methylation level of the same genetic locus in organ-matched normal tissue samples and found that 60.1% (256/426 tumors analyzed) of HNSCC tumors were hypermethylated (Figure 1B). By this stringent measure, hypermethylated tumors exhibit significantly decreased PTPRT mRNA expression levels as determined by RNA-Seq (Figure 1C), suggesting the validity of the above definition and that PTPRT hypermethylation has the expected biologic effect. In contrast, copy number alterations of the PTPRT gene are relatively infrequent and are not significantly associated with altered PTPRT mRNA expression (Supplemental Figure 1). As human papilloma virus (HPV) infection is an etiologic and prognostic factor in a subset of HNSCC, we sought to determine if PTPRT promoter hypermethylation is associated with HPV status and observed no significant association (P = 1.00, Fisher\\u2019s exact test; PTPRT promoter hypermethylation in 21/36 [58.3%] HPV-positive tumors versus 145/243 [59.7%] HPV-negative tumors), suggesting that HPV infection is not a driver of PTPRT promoter methylation. 6  Figure 1 Figure 1 Frequent PTPRT promoter hypermethylation is associated with downregulation of PTPRT mRNA in HNSCC tumors In order to validate TCGA findings in an independent HNSCC human cohort, we performed methylation-specific PCR (MSP) on 45 formalin-fixed, paraffin-embedded oral squamous cell cancers with primers directed at the promoter region of PTPRT (representative analysis in Figure 2A). Using this semi-quantitative analysis, a similar high frequency of methylation was observed in this cohort (71.1%, 32/45 tumors analyzed; Figure 2B), further suggesting that PTPRT promoter methylation represents a common mechanism of PTPRT downregulation in HNSCC.  Figure 2 Figure 2 The PTPRT promoter is frequently methylated in an independent cohort of HNSCC tumors The PTPRT promoter is frequently hypermethylated across human cancers  Further analysis of TCGA data reveals that the PTPRT promoter is frequently hypermethylated across a broad array of cancer types when hypermethylation is defined as a fractional methylation level (beta value) at least three standard deviations above the mean methylation level of the same genetic locus in organ-matched normal tissue samples. The highest incidence of PTPRT promoter hypermethylation occurs in colon adenocarcinoma (78.7%, 289/367 tumors analyzed), while HNSCC exhibits the second highest incidence (60.1%) (Figure 3A). Of the cancers analyzed, four exhibit significant downregulation of PTPRT mRNA in hypermethylated tumors (HNSCC, colon adenocarcinoma, lung adenocarcinoma, and breast invasive carcinoma; P < 0.05), suggesting a functional role for aberrant PTPRT promoter methylation across several cancer types (Figure 3B).  Figure 3 Figure 3 The PTPRT promoter is frequently hypermethylated across cancer types in association with downregulated PTPRT mRNA PTPRT expression is inversely associated with STAT3 activation in HNSCC  Additional analysis of TCGA data indicates that PTPRT mRNA expression is inversely correlated with pSTAT3 expression as determined by reverse-phase protein array (RPPA) (Figure 4A). Since pSTAT3 is known to be a direct substrate of PTPRT, we hypothesized that PTPRT promoter methylation may contribute to increased STAT3 activation. 5 Indeed, representative immunohistochemical staining of an independent cohort of HNSCC tumors demonstrated that those with PTPRT promoter methylation, as determined by MSP, express elevated levels of pSTAT3 relative to tumors without PTPRT methylation (Figure 4B). Together, these findings suggest that PTPRT promoter methylation may be a frequent mechanism that contributes to STAT3 activation in HNSCC.  Figure 4 Figure 4 PTPRT promoter methylation and mRNA expression are associated with pSTAT3 expression PTPRT promoter methylation is reversible, leading to PTPRT-specific pSTAT3 downregulation  To further investigate the association between PTPRT promoter methylation and STAT3 activation, we selected a HNSCC cell line (Cal27) that exhibits nearly complete methylation as determined by MSP. Treatment of Cal27 cells with 5-azacytidine (5-aza), a non-specific DNA methyltransferase (DNMT) inhibitor, led to an increase in the unmethylated signal as determined by MSP (Figure 5A), concurrent with overexpression of endogenous wild-type PTPRT and downregulation of pSTAT3 expression (Figures 5B and 5C). Transfection of Cal27 cells with shRNA directed against PTPRT (shPTPRT) following 5-aza treatment also resulted in an increase in the unmethylated signal (Figure 5A), but the expression of PTPRT was significantly downregulated relative to 5-aza treatment alone. Further, pSTAT3 expression remained unaffected under these conditions (Figures 5B and 5C), suggesting that 5-aza-mediated downregulation of pSTAT3 is dependent upon demethylation and re-expression of PTPRT in these cells. These results provide mechanistic evidence that PTPRT promoter methylation contributes directly to STAT3 activation in HNSCC by downregulation of PTPRT expression.  Figure 5 Figure 5 PTPRT promoter methylation is reduced by 5-azacytidine treatment, leading to PTPRT-specific pSTAT3 downregulation in HNSCC cells PTPRT promoter methylation is associated with increased sensitivity to STAT3 inhibition in HNSCC cells  To evaluate whether PTPRT promoter methylation may serve as a predictive biomarker for STAT3 targeted therapies, we first determined EC50 values for selective STAT3 inhibitors in a panel of 8 HNSCC cell lines (Figures 6A and 6B) that exhibit varying levels of PTPRT promoter methylation as determined by MSP (summarized in Figures 6C and 6D). Figures 6E and 6F illustrate that PTPRT promoter methylation and sensitivity to Stattic (a STAT3 SH2 domain inhibitor) or JSI-124 (a JAK/STAT3 pathway inhibitor) are significantly correlated (P < 0.05), indicating that HNSCC tumors that exhibit PTPRT promoter hypermethylation may be most sensitive to treatment with STAT3 inhibitors. A similar trend was observed with the JAK2 inhibitor LY2784544, though this observation was not statistically significant (Supplemental Figure 2). Notably, we observe no direct correlation between PTPRT promoter methylation and pSTAT3 expression in this cell line panel, suggesting that high pSTAT3 expression per se may not be predictive of sensitivity to STAT3 inhibitors. Thus, inhibition of STAT3 directly rather than through upstream kinases may be more efficacious in tumors with PTPRT promoter hypermethylation.  Figure 6 Figure 6 Increased PTPRT promoter methylation is associated with increased sensitivity to STAT3 inhibition We next sought to determine whether PTPRT methylation may serve as a predictive biomarker of sensitivity to STAT3 inhibition in a heterotopic tumorgraft model of HNSCC. We inoculated 16 nude mice with 3\\u00d7106 BICR 18 cells (which exhibit total PTPRT unmethylation) on one flank, and 3\\u00d7106 UMSCC 1 cells (which exhibit near total PTPRT methylation) on the other flank. After 10 days, the unmethylated cells had not formed any tumors, while the methylated cells had formed palpable masses in 15/16 mice (mean volume = 100.3 mm3). As BICR 18 cells are the only HNSCC cells identified to date with total (or near total) PTPRT unmethylation, we are therefore unable to determine the relative sensitivities of unmethylated versus methylated HNSCC cells in vivo. Nevertheless, starting on day 10 post-inoculation, 8 mice were randomized to receive 50 mg/kg Stattic by oral gavage five times per week, with serial tumor measurements performed three times per week. After 9 days of treatment, a large and statistically significant (P < 0.01) reduction in tumor volume was observed in the Stattic treatment arm relative to the no treatment group (Supplemental Figure 3), confirming that a PTPRT-methylated HNSCC heterotopic tumorgraft is indeed sensitive to STAT3 inhibition in vivo.  Go to: Discussion The survival rate for HNSCC patients has remained stagnant in recent decades despite advances in the understanding of the biological underpinnings of this disease and improved therapeutic strategies. In 2006, the Food and Drug Administration approved cetuximab, a monoclonal antibody targeting the epidermal growth factor receptor (EGFR), for the treatment of HNSCC patients. 7 Unfortunately, cetuximab has produced limited success in the clinic, at least in part due to the lack of predictive biomarkers, including EGFR expression or copy number gain. 8 There is therefore a need for the identification of both novel targets for pharmacologic inhibition and biomarkers for sensitivity to emerging therapeutics. STAT3 is a promising target for pharmacologic inhibition, and STAT3 inhibitors are being tested in early phase clinical trials. 2 We previously reported that STAT3 activation contributed to cetuximab resistance in HNSCC preclinical models and tumors from patients treated on cetuximab-containing protocols. 9 It is increasingly apparent that most molecular targeted agents will be most effective in subgroups of patients identified by specific genomic/transcriptomic/proteomic characteristics. The identification of biomarkers that identify those individuals most likely to benefit from these agents would allow the design of more scientifically rational clinical trials.  Here we report that the PTPRT promoter is frequently hypermethylated in HNSCC and other cancers in association with decreased PTPRT expression. A high rate of PTPRT promoter hypermethylation has also been reported in sporadic colorectal cancer, where nearly all tumors analyzed were methylated in contrast to no observed methylation in matched normal tissues, as well as in hepatocellular carcinoma, where the PTPRT methylation levels progressively increase from adjacent tissue to tumor. 10, 11 Our present findings indicate that PTPRT promoter methylation may represent a common event across many cancer types, suggesting that the development of a therapeutic strategy informed by this event may be of wide clinical benefit.  To date, the functional signaling consequences downstream of PTPRT promoter methylation have not been described, and as such, no strategies to mitigate the effects of this event have been proposed. Here we demonstrate that PTPRT promoter methylation is associated with a decrease in PTPRT mRNA expression, which in turn is associated with an increase in expression of the PTPRT substrate pSTAT3. We further evaluated PTPRT promoter methylation in an independent cohort of HNSCC and observed a similar frequency to that observed by TCGA, thus confirming the high incidence of this event. We also confirm that PTPRT methylation significantly contributes to pSTAT3 upregulation in HNSCC cells. Together, these studies suggest that HNSCC cells and tumors that exhibit high levels of PTPRT promoter methylation may be more dependent on STAT3 signaling for growth and survival, thus offering the opportunity to exploit PTPRT methylation as a predictive biomarker of sensitivity to STAT3 inhibition. Importantly, we have established an association between PTPRT promoter methylation and sensitivity to inhibitors of the STAT3 signaling pathway. These studies therefore suggest that screening for PTPRT promoter methylation may ultimately allow the identification of patients who are most likely to benefit from treatment with STAT3 targeting agents and provide a rationale for designing future clinical trials that select for patients with a high degree of PTPRT promoter methylation.   The activity of phosphatases could be influenced by genetic, as well as epigenetic alterations. In our study, we have investigated the methylation status of four PTPRs: PTPRM, PTPRT, PTPRR and PTPRZ1, which were pre-selected using microarray techniques as being alternatively methylated in sporadic colorectal cancer (CRC). The analyses were carried out on 131 surgical specimens obtained from sporadic CRC patients. The methylation status of the four genes was examined using methyl specific PCR (MSP). The analysis of promoter methylation using an Illumina 27K microarray revealed four protein tyrosine phosphatases PTPRM, PTPRT, PTPRR and PTPRZ1 as being hypermethylated with \\u03b2-value \\u22650.2 and P\\u22640.05. Subsequent analysis using MSP confirmed these observations-the frequency of promoter methylation was significantly higher in tumor cells compared with matched normal tissue for each of the analyzed genes. There was no association observed between the methylation status of PTPRs and either CIMP, K-ras (codon 12) and BRAF (exon 15, V600E) mutations or tumor localization (proximal/distal). The results of our study show a statistically significant difference between promoter methylation in cancerous and healthy tissue. This result supports the hypothesis that the PTPR family has an important role in the etiology of CRC. MEDLINE The activity of phosphatases could be influenced by genetic, as well as epigenetic alterations. In our study, we have investigated the methylation status of four PTPRs: PTPRM, PTPRT, PTPRR and PTPRZ1, which were pre-selected using microarray techniques as being alternatively methylated in sporadic colorectal cancer (CRC). The analyses were carried out on 131 surgical specimens obtained from sporadic CRC patients. The methylation status of the four genes was examined using methyl specific PCR (MSP). The analysis of promoter methylation using an Illumina 27K microarray revealed four protein tyrosine phosphatases PTPRM, PTPRT, PTPRR and PTPRZ1 as being hypermethylated with beta-value >= 0.2 and P <= 0.05. Subsequent analysis using MSP confirmed these observations-the frequency of promoter methylation was significantly higher in tumor cells compared with matched normal tissue for each of the analyzed genes. There was no association observed between the methylation status of PTPRs and either CIMP, K-ras (codon 12) and BRAF (exon 15, V600E) mutations or tumor localization (proximal/ distal). The results of our study show a statistically significant difference between promoter methylation in cancerous and healthy tissue. This result supports the hypothesis that the PTPR family has an important role in the etiology of CRC\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 27
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print('Number of data points in train data: ', train_df.shape[0])\n",
        "print('Number of data points in test data: ', test_df.shape[0])\n",
        "print('Number of data points in cross validation data: ', cv_df.shape[0])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "pQ_Y_XDxRdRH",
        "outputId": "f286d671-87d3-44db-f453-9b1064211555"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of data points in train data:  2124\n",
            "Number of data points in test data:  665\n",
            "Number of data points in cross validation data:  532\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "-train_df['Class'].value_counts().values"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FHxVzbzCy_xU",
        "outputId": "5b5a932f-fe6b-42ac-c175-220175324e0a"
      },
      "execution_count": 29,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([-609, -439, -363, -289, -176, -155,  -57,  -24,  -12])"
            ]
          },
          "metadata": {},
          "execution_count": 29
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# print(np.round((train_class_distribution.values[1]/train_df.shape[0]*100), 3))"
      ],
      "metadata": {
        "id": "HmUqIJmBz8GM"
      },
      "execution_count": 31,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Distribution of y_i in train, test, cv datasets\n",
        "train_class_distribution = train_df['Class'].value_counts().sort_index()\n",
        "test_class_distribution = test_df['Class'].value_counts().sort_index()\n",
        "cv_class_distribution = cv_df['Class'].value_counts().sort_index()\n",
        "\n",
        "my_colors = 'rgbkymc'\n",
        "train_class_distribution.plot(kind = 'bar')\n",
        "plt.xlabel('Class')\n",
        "plt.ylabel('Data points per class')\n",
        "plt.title('Distribution of yi in train data')\n",
        "plt.grid()\n",
        "plt.show()\n",
        "\n",
        "sorted_yi = np.argsort(-train_class_distribution.values)\n",
        "for i in sorted_yi:\n",
        "    print(f'Number of data points in class {i+1} : {train_class_distribution.values[i]} ({np.round((train_class_distribution.values[i]/train_df.shape[0]*100), 3)}) %')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 623
        },
        "id": "4XP736vcSkV3",
        "outputId": "d941cebd-1658-48ed-bbca-357fde57c34c"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAHCCAYAAAAJowgXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAABKDklEQVR4nO3deVhWdf7/8dcNAgoIuAHigmsq7mkqarmEoqJmMdO45ParnFFcKTPLEtfMmdLMbRZHM7MsWyx3RM3JsAxbNS1LxVHBRgNUFBHO748u7m93oHLghhtOz8d1cdn5nOXz/tzniK/OdtsMwzAEAABgUW6uLgAAAKAkEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAk+Li4mSz2Uqlr+7du6t79+726b1798pms2njxo2l0v+oUaNUr169UumrqC5fvqxHHnlEwcHBstlsmjx5crG3abPZFBcXd8tlTp48KZvNpjVr1hS7v9KQd+zs3bu3xPsqzOcHlCbCDn7X1qxZI5vNZv+pWLGiQkJCFBkZqSVLlujSpUtO6efs2bOKi4vTF1984ZTtOVNZrq0w5s+frzVr1mjs2LF69dVXNXz4cFeXVGTLly8vN+GpJBw5ckRxcXE6efKkq0uBxVRwdQFAWTB79mzVr19f2dnZSklJ0d69ezV58mS9+OKLev/999WqVSv7sjNmzNCTTz5pavtnz57VrFmzVK9ePbVp06bQ6+3cudNUP0Vxq9r++c9/Kjc3t8RrKI7du3erU6dOmjlzptO2efXqVVWocOtfj6Ghobp69ao8PDyc1u/y5ctVvXp1jRo1ymnbzHPPPffo6tWr8vT0dPq2neXIkSOaNWuWunfvXubPKKJ8IewAkvr27av27dvbp6dPn67du3erf//+GjhwoL799ltVqlRJklShQoXb/kNYXJmZmfL29nb5P0zO/Ie8pJw/f15hYWFO3WbFihVvu0zemUBXuXLlinx8fAq9vJubm0vrBVyJy1jATfTs2VPPPPOMTp06pXXr1tnbC7pnJz4+Xl27dlVAQIB8fX3VpEkTPfXUU5J+uVfirrvukiSNHj3afsks73JF9+7d1aJFCyUlJemee+6Rt7e3fd3f3rOTJycnR0899ZSCg4Pl4+OjgQMH6vTp0w7L1KtXr8AzBL/e5u1qK+ienStXruixxx5TnTp15OXlpSZNmuhvf/ubDMNwWM5ms2n8+PF677331KJFC3l5eal58+bavn17wR/4b5w/f14PP/ywgoKCVLFiRbVu3VqvvPKKfX7ePSgnTpzQli1b7LXf7BJIt27d1Lp16wLnNWnSRJGRkQ61F+WenVGjRsnX11dnzpzRoEGD5Ovrqxo1aujxxx9XTk7OLbdXr149HT58WB9++KF9LHn7Ke9y64cffqhx48YpMDBQtWvXliSdOnVK48aNU5MmTVSpUiVVq1ZNf/zjH/N9DgXds5N37B05ckQ9evSQt7e3atWqpYULF96y1jxZWVmaMmWKatSoocqVK2vgwIH673//m2+5wtS4Zs0a/fGPf5Qk9ejRw/4Z5NW7adMmRUVFKSQkRF5eXmrYsKHmzJlz288VkDizA9zS8OHD9dRTT2nnzp169NFHC1zm8OHD6t+/v1q1aqXZs2fLy8tLx48f1/79+yVJzZo10+zZs/Xss89qzJgxuvvuuyVJnTt3tm/jwoUL6tu3rwYPHqyHHnpIQUFBt6xr3rx5stlsmjZtms6fP6/FixcrIiJCX3zxhf0MVGEUprZfMwxDAwcO1J49e/Twww+rTZs22rFjh6ZOnaozZ85o0aJFDst/9NFHeueddzRu3DhVrlxZS5YsUXR0tJKTk1WtWrWb1nX16lV1795dx48f1/jx41W/fn299dZbGjVqlNLS0jRp0iQ1a9ZMr776qqZMmaLatWvrsccekyTVqFGjwG0OHz5cjz76qL755hu1aNHC3n7w4EF99913mjFjRqE/t1vJyclRZGSkOnbsqL/97W/atWuXXnjhBTVs2FBjx4696XqLFy/WhAkT5Ovrq6efflqS8h0H48aNU40aNfTss8/qypUr9vo//vhjDR48WLVr19bJkye1YsUKde/eXUeOHJG3t/ct6/3555/Vp08fPfDAA3rwwQe1ceNGTZs2TS1btlTfvn1vue4jjzyidevWaejQoercubN2796tqKiofMsVpsZ77rlHEydO1JIlS/TUU0+pWbNmkmT/c82aNfL19VVsbKx8fX21e/duPfvss8rIyNBf//rXW9YJyAB+x1avXm1IMg4ePHjTZfz9/Y22bdvap2fOnGn8+q/OokWLDEnGTz/9dNNtHDx40JBkrF69Ot+8bt26GZKMlStXFjivW7du9uk9e/YYkoxatWoZGRkZ9vY333zTkGS89NJL9rbQ0FBj5MiRt93mrWobOXKkERoaap9+7733DEnG3LlzHZb7wx/+YNhsNuP48eP2NkmGp6enQ9uXX35pSDJefvnlfH392uLFiw1Jxrp16+xt169fN8LDww1fX1+HsYeGhhpRUVG33J5hGEZaWppRsWJFY9q0aQ7tEydONHx8fIzLly871D5z5sxbbu/EiRP5PreRI0cakozZs2c7LNu2bVujXbt2t62xefPmDvsmT95x2rVrV+PGjRsO8zIzM/Mtn5iYaEgy1q5da2/LO3b27Nljb8s79n69XFZWlhEcHGxER0ffstYvvvjCkGSMGzfOoX3o0KH5Pr/C1vjWW2/lq/FW2/jzn/9seHt7G9euXbtlrQCXsYDb8PX1veVTWQEBAZJ+Oc1e1Jt5vby8NHr06EIvP2LECFWuXNk+/Yc//EE1a9bU1q1bi9R/YW3dulXu7u6aOHGiQ/tjjz0mwzC0bds2h/aIiAg1bNjQPt2qVSv5+fnpxx9/vG0/wcHBGjJkiL3Nw8NDEydO1OXLl/Xhhx+art3f31/33XefXn/9dfslt5ycHG3YsEGDBg0ydf/L7fzlL39xmL777rtvO+bCePTRR+Xu7u7Q9uszednZ2bpw4YIaNWqkgIAAHTp06Lbb9PX11UMPPWSf9vT0VIcOHQq1jyTlOxYKevS/uDX+dhuXLl3S//73P919993KzMzU0aNHC7UN/H4RdoDbuHz5skOw+K0//elP6tKlix555BEFBQVp8ODBevPNN00Fn1q1apm6Gblx48YO0zabTY0aNSrxR3ZPnTqlkJCQfJ9H3qWGU6dOObTXrVs33zaqVKmin3/++bb9NG7cWG5ujr+ibtZPYY0YMULJycn6z3/+I0natWuXUlNTnfq4esWKFfNdSivMmAujfv36+dquXr2qZ5991n4PVfXq1VWjRg2lpaUpPT39ttusXbt2vnvQCruP3NzcHMKs9Mv9T86uUfrlcvH9998vf39/+fn5qUaNGvaQVtht4PeLe3aAW/jvf/+r9PR0NWrU6KbLVKpUSfv27dOePXu0ZcsWbd++XRs2bFDPnj21c+fOfP8nfrNtONvNXnyYk5NTqJqc4Wb9GL+5mbm0REZGKigoSOvWrdM999yjdevWKTg4WBEREU7royQ/24KOkwkTJmj16tWaPHmywsPD5e/vL5vNpsGDBxcqcJfGPipujWlpaerWrZv8/Pw0e/ZsNWzYUBUrVtShQ4c0bdq0Mv96BLgeYQe4hVdffVWSHJ7UKYibm5vuvfde3XvvvXrxxRc1f/58Pf3009qzZ48iIiKc/sbl77//3mHaMAwdP37c4X1AVapUUVpaWr51T506pQYNGtinzdQWGhqqXbt26dKlSw5nd/IuI4SGhhZ6W7fr56uvvlJubq7D2Z3i9uPu7q6hQ4dqzZo1ev755/Xee+8VeGnIVYpynGzcuFEjR47UCy+8YG+7du1agfvemUJDQ5Wbm6sffvjB4WzOsWPHilzjzca/d+9eXbhwQe+8847uuecee/uJEyeKOQr8XnAZC7iJ3bt3a86cOapfv76GDRt20+UuXryYry3v5XxZWVmSZL8fxFn/AK1du9bhPqKNGzfq3LlzDk/PNGzYUAcOHND169ftbZs3b873iLqZ2vr166ecnBwtXbrUoX3RokWy2Wy3fXqnsPr166eUlBRt2LDB3nbjxg29/PLL8vX1Vbdu3Yq87eHDh+vnn3/Wn//8Z12+fNnhfhVX8/HxMX2MuLu75zsL8/LLL5f4I9l5+3rJkiUO7YsXL863bGFrvNmxmBdGf72N69eva/ny5UWqHb8/nNkBJG3btk1Hjx7VjRs3lJqaqt27dys+Pl6hoaF6//33b/kyttmzZ2vfvn2KiopSaGiozp8/r+XLl6t27drq2rWrpF+CR0BAgFauXKnKlSvLx8dHHTt2LPAejMKoWrWqunbtqtGjRys1NVWLFy9Wo0aNHB6Pf+SRR7Rx40b16dNHDz74oH744QetW7cu3z0WZmobMGCAevTooaefflonT55U69attXPnTm3atEmTJ0/Ot+2iGjNmjP7+979r1KhRSkpKUr169bRx40bt379fixcvvuU9VLfTtm1btWjRQm+99ZaaNWumO++80yk1O0O7du20YsUKzZ07V40aNVJgYKB69ux5y3X69++vV199Vf7+/goLC1NiYqJ27dp1y0f7naFNmzYaMmSIli9frvT0dHXu3FkJCQk6fvx4kWts06aN3N3d9fzzzys9PV1eXl7q2bOnOnfurCpVqmjkyJGaOHGibDabXn31VZddDkU55LLnwIAyIO+R3rwfT09PIzg42OjVq5fx0ksvOTzinOe3j54nJCQY9913nxESEmJ4enoaISEhxpAhQ4zvvvvOYb1NmzYZYWFhRoUKFRweWe7WrZvRvHnzAuu72aPnr7/+ujF9+nQjMDDQqFSpkhEVFWWcOnUq3/ovvPCCUatWLcPLy8vo0qWL8dlnn+Xb5q1q++2j54ZhGJcuXTKmTJlihISEGB4eHkbjxo2Nv/71r0Zubq7DcpKMmJiYfDXd7JH430pNTTVGjx5tVK9e3fD09DRatmxZ4OPxhX30/NcWLlxoSDLmz59f4HwV49FzHx+ffMv+9pi5mZSUFCMqKsqoXLmyIcm+n271ioSff/7Z/jn5+voakZGRxtGjR/N9zjd79LygY6+g/V6Qq1evGhMnTjSqVatm+Pj4GAMGDDBOnz6d7/MrbI2GYRj//Oc/jQYNGhju7u4O9e7fv9/o1KmTUalSJSMkJMR44oknjB07dtz0UXXg12yGQTQG8Pvy0ksvacqUKTp58mSBT4wBsBbCDoDfFcMw1Lp1a1WrVk179uxxdTkASgH37AD4Xbhy5Yref/997dmzR19//bU2bdrk6pIAlBLO7AD4XTh58qTq16+vgIAAjRs3TvPmzXN1SQBKCWEHAABYGu/ZAQAAlkbYAQAAlsYNypJyc3N19uxZVa5c2emv9QcAACXDMAxdunRJISEh+b44+NcIO5LOnj2rOnXquLoMAABQBKdPn1bt2rVvOp+wI9lfPX/69Gn5+fmVaF/Z2dnauXOnevfuLQ8PjxLtq7QxtvLLyuNjbOUTYyufSntsGRkZqlOnzm2/Qoawo//7pl0/P79SCTve3t7y8/Oz5EHO2MonK4+PsZVPjK18ctXYbncLCjcoAwAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAAS6vg6gIAANZW78ktptfxcje0sIPUIm6HsnJsptY9uSDKdH+wNpef2Tlz5oweeughVatWTZUqVVLLli312Wef2ecbhqFnn31WNWvWVKVKlRQREaHvv//eYRsXL17UsGHD5Ofnp4CAAD388MO6fPlyaQ8FAACUQS4NOz///LO6dOkiDw8Pbdu2TUeOHNELL7ygKlWq2JdZuHChlixZopUrV+qTTz6Rj4+PIiMjde3aNfsyw4YN0+HDhxUfH6/Nmzdr3759GjNmjCuGBAAAyhiXXsZ6/vnnVadOHa1evdreVr9+fft/G4ahxYsXa8aMGbrvvvskSWvXrlVQUJDee+89DR48WN9++622b9+ugwcPqn379pKkl19+Wf369dPf/vY3hYSElO6gAABAmeLSsPP+++8rMjJSf/zjH/Xhhx+qVq1aGjdunB599FFJ0okTJ5SSkqKIiAj7Ov7+/urYsaMSExM1ePBgJSYmKiAgwB50JCkiIkJubm765JNPdP/99+frNysrS1lZWfbpjIwMSVJ2drays7NLarj2Pn79p5UwtvLLyuNjbK7n5W6YX8fNcPjTjLL+eZSX/VYUpT22wvbj0rDz448/asWKFYqNjdVTTz2lgwcPauLEifL09NTIkSOVkpIiSQoKCnJYLygoyD4vJSVFgYGBDvMrVKigqlWr2pf5reeee06zZs3K175z5055e3s7Y2i3FR8fXyr9uAJjK7+sPD7G5joLOxR93Tntc02vs3Xr1qJ3WIrK+n4rjtIaW2ZmZqGWc2nYyc3NVfv27TV//nxJUtu2bfXNN99o5cqVGjlyZIn1O336dMXGxtqnMzIyVKdOHfXu3Vt+fn4l1q/0SwqNj49Xr1695OHhUaJ9lTbGVn5ZeXyMzfVaxO0wvY6Xm6E57XP1zGduyso19zTWN3GRpvsrTeVlvxVFaY8t78rM7bg07NSsWVNhYWEObc2aNdPbb78tSQoODpYkpaamqmbNmvZlUlNT1aZNG/sy58+fd9jGjRs3dPHiRfv6v+Xl5SUvL6987R4eHqV24JVmX6WNsZVfVh4fY3Mds4+OO6ybazO9fln+LH6trO+34iitsRW2D5c+jdWlSxcdO3bMoe27775TaGiopF9uVg4ODlZCQoJ9fkZGhj755BOFh4dLksLDw5WWlqakpCT7Mrt371Zubq46duxYCqMAAABlmUvP7EyZMkWdO3fW/Pnz9eCDD+rTTz/VP/7xD/3jH/+QJNlsNk2ePFlz585V48aNVb9+fT3zzDMKCQnRoEGDJP1yJqhPnz569NFHtXLlSmVnZ2v8+PEaPHgwT2IBAADXhp277rpL7777rqZPn67Zs2erfv36Wrx4sYYNG2Zf5oknntCVK1c0ZswYpaWlqWvXrtq+fbsqVqxoX+a1117T+PHjde+998rNzU3R0dFasmSJK4YEAADKGJd/XUT//v3Vv3//m8632WyaPXu2Zs+efdNlqlatqvXr15dEeQAAoJxz+ddFAAAAlCTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDTCDgAAsDSXhp24uDjZbDaHn6ZNm9rnX7t2TTExMapWrZp8fX0VHR2t1NRUh20kJycrKipK3t7eCgwM1NSpU3Xjxo3SHgoAACijKri6gObNm2vXrl326QoV/q+kKVOmaMuWLXrrrbfk7++v8ePH64EHHtD+/fslSTk5OYqKilJwcLA+/vhjnTt3TiNGjJCHh4fmz59f6mMBAABlj8vDToUKFRQcHJyvPT09XatWrdL69evVs2dPSdLq1avVrFkzHThwQJ06ddLOnTt15MgR7dq1S0FBQWrTpo3mzJmjadOmKS4uTp6enqU9HAAAUMa4/J6d77//XiEhIWrQoIGGDRum5ORkSVJSUpKys7MVERFhX7Zp06aqW7euEhMTJUmJiYlq2bKlgoKC7MtERkYqIyNDhw8fLt2BAACAMsmlZ3Y6duyoNWvWqEmTJjp37pxmzZqlu+++W998841SUlLk6empgIAAh3WCgoKUkpIiSUpJSXEIOnnz8+bdTFZWlrKysuzTGRkZkqTs7GxlZ2c7Y2g3lbf9ku7HFRhb+WXl8TE21/NyN8yv42Y4/GlGWf88yst+K4rSHlth+7EZhmH+SCohaWlpCg0N1YsvvqhKlSpp9OjRDqFEkjp06KAePXro+eef15gxY3Tq1Cnt2LHDPj8zM1M+Pj7aunWr+vbtW2A/cXFxmjVrVr729evXy9vb27mDAgAAJSIzM1NDhw5Venq6/Pz8brqcy+/Z+bWAgADdcccdOn78uHr16qXr168rLS3N4exOamqq/R6f4OBgffrppw7byHtaq6D7gPJMnz5dsbGx9umMjAzVqVNHvXv3vuWH5QzZ2dmKj49Xr1695OHhUaJ9lTbGVn5ZeXyMzfVaxO24/UK/4eVmaE77XD3zmZuycm2m1v0mLtJ0f6WpvOy3oijtseVdmbmdMhV2Ll++rB9++EHDhw9Xu3bt5OHhoYSEBEVHR0uSjh07puTkZIWHh0uSwsPDNW/ePJ0/f16BgYGSpPj4ePn5+SksLOym/Xh5ecnLyytfu4eHR6kdeKXZV2ljbOWXlcfH2FwnK8dcWHFYN9dmev2y/Fn8Wlnfb8VRWmMrbB8uDTuPP/64BgwYoNDQUJ09e1YzZ86Uu7u7hgwZIn9/fz388MOKjY1V1apV5efnpwkTJig8PFydOnWSJPXu3VthYWEaPny4Fi5cqJSUFM2YMUMxMTEFhhkAAPD749Kw89///ldDhgzRhQsXVKNGDXXt2lUHDhxQjRo1JEmLFi2Sm5uboqOjlZWVpcjISC1fvty+vru7uzZv3qyxY8cqPDxcPj4+GjlypGbPnu2qIQEAgDLGpWHnjTfeuOX8ihUratmyZVq2bNlNlwkNDdXWrVudXRoAALAIl79nBwAAoCQRdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKU5JeykpaU5YzMAAABOZzrsPP/889qwYYN9+sEHH1S1atVUq1Ytffnll04tDgAAoLhMh52VK1eqTp06kqT4+HjFx8dr27Zt6tu3r6ZOner0AgEAAIqjgtkVUlJS7GFn8+bNevDBB9W7d2/Vq1dPHTt2dHqBAMyp9+QW0+t4uRta2EFqEbdDWTm2Qq93ckGU6b4AoLSZPrNTpUoVnT59WpK0fft2RURESJIMw1BOTo5zqwMAACgm02d2HnjgAQ0dOlSNGzfWhQsX1LdvX0nS559/rkaNGjm9QAAAgOIwHXYWLVqkevXq6fTp01q4cKF8fX0lSefOndO4ceOcXiAAAEBxmA47Hh4eevzxx/O1T5kyxSkFAQAAOJPpe3ZeeeUVbdnyfzdAPvHEEwoICFDnzp116tQppxYHAABQXKbDzvz581WpUiVJUmJiopYtW6aFCxeqevXqnN0BAABljunLWKdPn7bfiPzee+8pOjpaY8aMUZcuXdS9e3dn1wcAAFAsps/s+Pr66sKFC5KknTt3qlevXpKkihUr6urVq86tDgAAoJhMn9np1auXHnnkEbVt21bfffed+vXrJ0k6fPiw6tWr5+z6AAAAisX0mZ1ly5YpPDxcP/30k95++21Vq1ZNkpSUlKQhQ4Y4vUAAAIDiMH1mJyAgQEuXLs3XPmvWLKcUBAAA4Eymw06ezMxMJScn6/r16w7trVq1KnZRAAAAzmI67Pz0008aNWqUtm/fXuB8vh8LAACUJabv2Zk8ebLS09P1ySefqFKlStq+fbteeeUVNW7cWO+//35J1AgAAFBkps/s7N69W5s2bVL79u3l5uam0NBQ9erVS35+fnruuecUFRVVEnUCAAAUiekzO1euXFFgYKAkqUqVKvrpp58kSS1bttShQ4ecWx0AAEAxmQ47TZo00bFjxyRJrVu31t///nedOXNGK1euVM2aNZ1eIAAAQHGYDjuTJk3SuXPnJEkzZ87Utm3bVLduXS1ZskTz588vciELFiyQzWbT5MmT7W3Xrl1TTEyMqlWrJl9fX0VHRys1NdVhveTkZEVFRcnb21uBgYGaOnWqbty4UeQ6AACAtZi+Z+ehhx6y/3e7du106tQpHT16VHXr1lX16tWLVMTBgwf197//Pd9j61OmTNGWLVv01ltvyd/fX+PHj9cDDzyg/fv3S/rlya+oqCgFBwfr448/1rlz5zRixAh5eHgUK3gBAADrMH1m57e8vb115513FjnoXL58WcOGDdM///lPValSxd6enp6uVatW6cUXX1TPnj3Vrl07rV69Wh9//LEOHDgg6Zfv5jpy5IjWrVunNm3aqG/fvpozZ46WLVuW7/0/AADg96lQZ3ZiY2MLvcEXX3zRVAExMTGKiopSRESE5s6da29PSkpSdna2IiIi7G1NmzZV3bp1lZiYqE6dOikxMVEtW7ZUUFCQfZnIyEiNHTtWhw8fVtu2bU3VAgAArKdQYefzzz8v1MZsNpupzt944w0dOnRIBw8ezDcvJSVFnp6eCggIcGgPCgpSSkqKfZlfB528+XnzbiYrK0tZWVn26YyMDElSdna2srOzTY3BrLztl3Q/rsDYygYvd8P8Om6Gw5+FVR4+j/K078wqL2MrzWNSKvufR3nZb0VR2mMrbD+FCjt79uwpVjEFOX36tCZNmqT4+HhVrFjR6du/leeee67A7/LauXOnvL29S6WG+Pj4UunHFRibay3sUPR157TPNbX81q1bi95ZKSsP+66oyvrYSvOYlMrPcVnW91txlNbYMjMzC7Wc6RuU09PTlZOTo6pVqzq0X7x4URUqVJCfn1+htpOUlKTz58/rzjvvtLfl5ORo3759Wrp0qXbs2KHr168rLS3N4exOamqqgoODJUnBwcH69NNPHbab97RW3jIFmT59usOluYyMDNWpU0e9e/cudP1FlZ2drfj4ePXq1UseHh4l2ldpY2xlQ4u4HabX8XIzNKd9rp75zE1ZuYU/Q/tNXKTpvkpbedp3ZpWXsZXmMSmV/eOyvOy3oijtseVdmbkd02Fn8ODBGjBggMaNG+fQ/uabb+r9998vdKK+99579fXXXzu0jR49Wk2bNtW0adNUp04deXh4KCEhQdHR0ZKkY8eOKTk5WeHh4ZKk8PBwzZs3T+fPn7e/6DA+Pl5+fn4KCwu7ad9eXl7y8vLK1+7h4VFqB15p9lXaGJtrZeWY+4fBYd1cm6n1y/pn8WvlYd8VVVkfW2kek1L5OS7L+n4rjtIaW2H7MB12PvnkkwJvQu7evbuefvrpQm+ncuXKatGihUObj4+PqlWrZm9/+OGHFRsbq6pVq8rPz08TJkxQeHi4OnXqJEnq3bu3wsLCNHz4cC1cuFApKSmaMWOGYmJiCgwzAADg98d02MnKyirwpX3Z2dm6evWqU4rKs2jRIrm5uSk6OlpZWVmKjIzU8uXL7fPd3d21efNmjR07VuHh4fLx8dHIkSM1e/Zsp9YBAADKL9Nhp0OHDvrHP/6hl19+2aF95cqVateuXbGK2bt3r8N0xYoVtWzZMi1btuym64SGhpabm9EAAEDpMx125s6dq4iICH355Ze69957JUkJCQk6ePCgdu7c6fQCAQAAisP0G5S7dOmixMRE1alTR2+++aY++OADNWrUSF999ZXuvvvukqgRAACgyEyf2ZGkNm3a6LXXXnN2LQAAAE5X7O/GAgAAKMsIOwAAwNIIOwAAwNIIOwAAwNJMhZ3s7GxVqFBB33zzTUnVAwAA4FSmwo6Hh4fq1q2rnJyckqoHAADAqUxfxnr66af11FNP6eLFiyVRDwAAgFOZfs/O0qVLdfz4cYWEhCg0NFQ+Pj4O8w8dOuS04gAAAIrLdNgZNGhQCZQBAABQMkyHnZkzZ5ZEHQAAACWiSI+ep6Wl6V//+pemT59uv3fn0KFDOnPmjFOLAwAAKC7TZ3a++uorRUREyN/fXydPntSjjz6qqlWr6p133lFycrLWrl1bEnUCAAAUiemwExsbq1GjRmnhwoWqXLmyvb1fv34aOnSoU4sr6+o9ucX0Ol7uhhZ2kFrE7VBWjs3UuicXRJnuDwCA3zvTl7EOHjyoP//5z/naa9WqpZSUFKcUBQAA4Cymw46Xl5cyMjLytX/33XeqUaOGU4oCAABwFtNhZ+DAgZo9e7ays7MlSTabTcnJyZo2bZqio6OdXiAAAEBxmA47L7zwgi5fvqzAwEBdvXpV3bp1U6NGjVS5cmXNmzevJGoEAAAoMtM3KPv7+ys+Pl4fffSRvvrqK12+fFl33nmnIiIiSqI+AACAYjEddvJ07dpVXbt2dWYtAAAATleklwomJCSof//+atiwoRo2bKj+/ftr165dzq4NAACg2EyHneXLl6tPnz6qXLmyJk2apEmTJsnPz0/9+vXTsmXLSqJGAACAIjN9GWv+/PlatGiRxo8fb2+bOHGiunTpovnz5ysmJsapBQIAABSH6TM7aWlp6tOnT7723r17Kz093SlFAQAAOEuR3rPz7rvv5mvftGmT+vfv75SiAAAAnMX0ZaywsDDNmzdPe/fuVXh4uCTpwIED2r9/vx577DEtWbLEvuzEiROdVykAAEARmA47q1atUpUqVXTkyBEdOXLE3h4QEKBVq1bZp202G2EHAAC4nOmwc+LEiZKoAwAAoEQU6T07AAAA5QVhBwAAWBphBwAAWBphBwAAWBphBwAAWJrpsLN9+3Z99NFH9ully5apTZs2Gjp0qH7++WenFgcAAFBcpsPO1KlTlZGRIUn6+uuv9dhjj6lfv346ceKEYmNjnV4gAABAcRTpPTthYWGSpLffflv9+/fX/PnzdejQIfXr18/pBQIAABSH6TM7np6eyszMlCTt2rVLvXv3liRVrVrVfsYHAACgrDB9Zqdr166KjY1Vly5d9Omnn2rDhg2SpO+++061a9d2eoEAAADFYfrMztKlS1WhQgVt3LhRK1asUK1atSRJ27ZtU58+fZxeIAAAQHGYPrNTt25dbd68OV/7okWLnFIQAACAM5k+s+Pu7q7z58/na79w4YLc3d2dUhQAAICzmA47hmEU2J6VlSVPT89iFwQAAOBMhb6MtWTJEkmSzWbTv/71L/n6+trn5eTkaN++fWratKnzKwQAACiGQoedvHtyDMPQypUrHS5ZeXp6ql69elq5cqXzKwQAACiGQoedEydOSJJ69Oihd955R1WqVCmxogAAAJzF9D07e/bscVrQWbFihVq1aiU/Pz/5+fkpPDxc27Zts8+/du2aYmJiVK1aNfn6+io6OlqpqakO20hOTlZUVJS8vb0VGBioqVOn6saNG06pDwAAlH+mHz3PycnRmjVrlJCQoPPnzys3N9dh/u7duwu9rdq1a2vBggVq3LixDMPQK6+8ovvuu0+ff/65mjdvrilTpmjLli1666235O/vr/Hjx+uBBx7Q/v377bVERUUpODhYH3/8sc6dO6cRI0bIw8ND8+fPNzs0AABgQabDzqRJk7RmzRpFRUWpRYsWstlsRe58wIABDtPz5s3TihUrdODAAdWuXVurVq3S+vXr1bNnT0nS6tWr1axZMx04cECdOnXSzp07deTIEe3atUtBQUFq06aN5syZo2nTpikuLo6nwwAAgPmw88Ybb+jNN990+pd+5uTk6K233tKVK1cUHh6upKQkZWdnKyIiwr5M06ZNVbduXSUmJqpTp05KTExUy5YtFRQUZF8mMjJSY8eO1eHDh9W2bdsC+8rKylJWVpZ9Ou87vbKzs5WdnV3omr3cC34M/5bruBkOf5phpjZXyKuvrNdZFOVpbKV5XJaHz6M87TuzysvY+F3pqLzst6Io7bEVth+bcbMX59xESEiI9u7dqzvuuKNIhf3W119/rfDwcF27dk2+vr5av369+vXrp/Xr12v06NEOoUSSOnTooB49euj555/XmDFjdOrUKe3YscM+PzMzUz4+Ptq6dav69u1bYJ9xcXGaNWtWvvb169fL29vbKeMCAAAlKzMzU0OHDlV6err8/PxuupzpMzuPPfaYXnrpJS1durRYl7DyNGnSRF988YXS09O1ceNGjRw5Uh9++GGxt3sr06dPV2xsrH06IyNDderUUe/evW/5Yf1Wi7gdt1/oN7zcDM1pn6tnPnNTVq65z++buEjT/ZWm7OxsxcfHq1evXvLw8HB1OU5VnsZWmsdlWT8mpfK178wqL2Pjd6Wj8rLfiqK0x5Z3ZeZ2TIedjz76SHv27NG2bdvUvHnzfIN55513TG3P09NTjRo1kiS1a9dOBw8e1EsvvaQ//elPun79utLS0hQQEGBfPjU1VcHBwZKk4OBgffrppw7by3taK2+Zgnh5ecnLyytfu4eHh6mdk5VT9LCXlWszvX55+Uth9nMsT8rD2ErzuCzrn8WvlYd9V1RlfWz8rixYWd9vxVFaYytsH6YfPQ8ICND999+vbt26qXr16vL393f4Ka7c3FxlZWWpXbt28vDwUEJCgn3esWPHlJycrPDwcElSeHi4vv76a4fv6oqPj5efn5/CwsKKXQsAACj/TJ/ZWb16tdM6nz59uvr27au6devq0qVLWr9+vfbu3asdO3bI399fDz/8sGJjY1W1alX5+flpwoQJCg8PV6dOnSRJvXv3VlhYmIYPH66FCxcqJSVFM2bMUExMTIFnbgAAwO+P6bDjTOfPn9eIESN07tw5+fv7q1WrVtqxY4d69eol6ZevqHBzc1N0dLSysrIUGRmp5cuX29d3d3fX5s2bNXbsWIWHh8vHx0cjR47U7NmzXTUkAABQxhQq7Nx5551KSEhQlSpV1LZt21vemHzo0KFCd75q1apbzq9YsaKWLVumZcuW3XSZ0NBQbd26tdB9AgCA35dChZ377rvPfllo0KBBJVkPypB6T24xtbyXu6GFHX558sLsDYUnF0SZWh4AgMIqVNiZOXNmgf8NAABQ1hX5np2kpCR9++23kqTmzZvf9G3FAAAArmQ67Jw/f16DBw/W3r177e+/SUtLU48ePfTGG2+oRo0azq4RAACgyEy/Z2fChAm6dOmSDh8+rIsXL+rixYv65ptvlJGRoYkTJ5ZEjQAAAEVm+szO9u3btWvXLjVr1szeFhYWpmXLlql3795OLQ4AAKC4TJ/Zyc3NLfD1zB4eHsrNzXVKUQAAAM5iOuz07NlTkyZN0tmzZ+1tZ86c0ZQpU3Tvvfc6tTgAAIDiMh12li5dqoyMDNWrV08NGzZUw4YNVb9+fWVkZOjll18uiRoBAACKzPQ9O3Xq1NGhQ4e0a9cuHT16VJLUrFkzRUREOL04AACA4irSe3ZsNpt69epl/w4rAACAssr0ZSxJSkhIUP/+/e2Xsfr3769du3Y5uzYAAIBiMx12li9frj59+qhy5cqaNGmSJk2aJD8/P/Xr1++WX9gJAADgCqYvY82fP1+LFi3S+PHj7W0TJ05Uly5dNH/+fMXExDi1QAAAgOIwfWYnLS1Nffr0ydfeu3dvpaenO6UoAAAAZzEddgYOHKh33303X/umTZvUv39/pxQFAADgLKYvY4WFhWnevHnau3evwsPDJUkHDhzQ/v379dhjj2nJkiX2ZfmuLAAA4Gqmw86qVatUpUoVHTlyREeOHLG3BwQEaNWqVfZpm81G2AEAAC5nOuycOHGiJOoAAAAoEUV6zw4AAEB5QdgBAACWRtgBAACWRtgBAACWRtgBAACWVqRvPZekzMxMJScn6/r16w7trVq1KnZRAAAAzmI67Pz0008aPXq0tm3bVuD8nJycYhcFAADgLKYvY02ePFlpaWn65JNPVKlSJW3fvl2vvPKKGjdurPfff78kagQAACgy02d2du/erU2bNql9+/Zyc3NTaGioevXqJT8/Pz333HOKiooqiToBAACKxPSZnStXrigwMFCSVKVKFf3000+SpJYtW+rQoUPOrQ4AAKCYTIedJk2a6NixY5Kk1q1b6+9//7vOnDmjlStXqmbNmk4vEAAAoDhMX8aaNGmSzp07J0maOXOm+vTpo9dee02enp5as2aNs+sDAAAoFtNh56GHHrL/d7t27XTq1CkdPXpUdevWVfXq1Z1aHAAAQHGZvow1e/ZsZWZm2qe9vb115513ysfHR7Nnz3ZqcQAAAMVlOuzMmjVLly9fzteemZmpWbNmOaUoAAAAZzEddgzDkM1my9f+5ZdfqmrVqk4pCgAAwFkKfc9OlSpVZLPZZLPZdMcddzgEnpycHF2+fFl/+ctfSqRIAACAoip02Fm8eLEMw9D/+3//T7NmzZK/v799nqenp+rVq6fw8PASKRIAAKCoCh12Ro4cKUmqX7++OnfuLA8PjxIrCgAAwFlMP3rerVs3+39fu3Yt37ee+/n5Fb8qAAAAJzF9g3JmZqbGjx+vwMBA+fj4qEqVKg4/AAAAZYnpsDN16lTt3r1bK1askJeXl/71r39p1qxZCgkJ0dq1a0uiRgAAgCIzfRnrgw8+0Nq1a9W9e3eNHj1ad999txo1aqTQ0FC99tprGjZsWEnUCQAAUCSmz+xcvHhRDRo0kPTL/TkXL16UJHXt2lX79u1zbnUAAADFZDrsNGjQQCdOnJAkNW3aVG+++aakX874BAQEOLU4AACA4jIddkaPHq0vv/xSkvTkk09q2bJlqlixoqZMmaKpU6c6vUAAAIDiMH3PzpQpU+z/HRERoaNHjyopKUmNGjVSq1atnFocAABAcZkOO78VGhqq0NBQZ9QCAADgdKYuY+Xm5urf//63+vfvrxYtWqhly5YaOHCg1q5dK8MwTHf+3HPP6a677lLlypUVGBioQYMG6dixYw7LXLt2TTExMapWrZp8fX0VHR2t1NRUh2WSk5MVFRUlb29vBQYGaurUqbpx44bpegAAgPUUOuwYhqGBAwfqkUce0ZkzZ9SyZUs1b95cp06d0qhRo3T//feb7vzDDz9UTEyMDhw4oPj4eGVnZ6t37966cuWKfZkpU6bogw8+0FtvvaUPP/xQZ8+e1QMPPGCfn5OTo6ioKF2/fl0ff/yxXnnlFa1Zs0bPPvus6XoAAID1FPoy1po1a7Rv3z4lJCSoR48eDvN2796tQYMGae3atRoxYkShO9++fXu+PgIDA5WUlKR77rlH6enpWrVqldavX6+ePXtKklavXq1mzZrpwIED6tSpk3bu3KkjR45o165dCgoKUps2bTRnzhxNmzZNcXFx8vT0LHQ9AADAegoddl5//XU99dRT+YKOJPXs2VNPPvmkXnvtNVNh57fS09MlSVWrVpUkJSUlKTs7WxEREfZlmjZtqrp16yoxMVGdOnVSYmKiWrZsqaCgIPsykZGRGjt2rA4fPqy2bdvm6ycrK0tZWVn26YyMDElSdna2srOzC12vl7v5S3debobDn2aYqc0ZzI6vPI3NrLz6ynqdUukel+Xh8yhP+86s8jI2q/+uNKu87LeiKO2xFbYfm1HIm22Cg4O1fft2tWnTpsD5n3/+ufr27auUlJRCF/lrubm5GjhwoNLS0vTRRx9JktavX6/Ro0c7BBNJ6tChg3r06KHnn39eY8aM0alTp7Rjxw77/MzMTPn4+Gjr1q3q27dvvr7i4uI0a9asfO3r16+Xt7d3keoHAAClKzMzU0OHDlV6evotv4i80Gd2Ll686HD25LeCgoL0888/m6vyV2JiYvTNN9/Yg05Jmj59umJjY+3TGRkZqlOnjnr37m3qW9tbxO24/UK/4eVmaE77XD3zmZuycm2m1v0mLtJ0f8VhdnzlaWxmZWdnKz4+Xr169ZKHh4ery7ml0jwuy/p+k8rXvjOrvIzN6r8rzSov+60oSntseVdmbqfQYScnJ0cVKtx8cXd39yI/ATV+/Hht3rxZ+/btU+3ate3twcHBun79utLS0hzezpyamqrg4GD7Mp9++qnD9vKe1spb5re8vLzk5eWVr93Dw8PUzsnKMfcX0GHdXJvp9Uv7L0VRx1cexlZUZo8RVyjN47Ksfxa/Vh72XVGV9bFZ/XdlUZX1/VYcpTW2wvZR6LBjGIZGjRpVYEiQlO9SU2G3OWHCBL377rvau3ev6tev7zC/Xbt28vDwUEJCgqKjoyVJx44dU3JyssLDwyVJ4eHhmjdvns6fP6/AwEBJUnx8vPz8/BQWFma6JgAAYC2FDjsjR4687TJmb06OiYnR+vXrtWnTJlWuXNl+v4+/v78qVaokf39/Pfzww4qNjVXVqlXl5+enCRMmKDw8XJ06dZIk9e7dW2FhYRo+fLgWLlyolJQUzZgxQzExMTcNZgAA4Pej0GFn9erVTu98xYoVkqTu3bvn62vUqFGSpEWLFsnNzU3R0dHKyspSZGSkli9fbl/W3d1dmzdv1tixYxUeHi4fHx+NHDlSs2fPdnq9AACg/Cn210UUR2EeBKtYsaKWLVumZcuW3XSZ0NBQbd261ZmlAQAAizD9recAAADlCWEHAABYGmEHAABYGmEHAABYGmEHAABYGmEHAABYGmEHAABYmkvfswMA+EW9J7eYXsfL3dDCDr980abZ7486uSDKdH9AecWZHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGmEHQAAYGl8NxaAcqM0vz+K744CrIMzOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNJcGnb27dunAQMGKCQkRDabTe+9957DfMMw9Oyzz6pmzZqqVKmSIiIi9P333zssc/HiRQ0bNkx+fn4KCAjQww8/rMuXL5fiKAAAQFnm0rBz5coVtW7dWsuWLStw/sKFC7VkyRKtXLlSn3zyiXx8fBQZGalr167Zlxk2bJgOHz6s+Ph4bd68Wfv27dOYMWNKawgAAKCMq+DKzvv27au+ffsWOM8wDC1evFgzZszQfffdJ0lau3atgoKC9N5772nw4MH69ttvtX37dh08eFDt27eXJL388svq16+f/va3vykkJKTUxgIAAMoml4adWzlx4oRSUlIUERFhb/P391fHjh2VmJiowYMHKzExUQEBAfagI0kRERFyc3PTJ598ovvvv7/AbWdlZSkrK8s+nZGRIUnKzs5WdnZ2oWv0cjfMDktebobDn2aYqc0ZzI6vPI3NrLz6ynqdUukel2X9mJQY282U5visPLaiKE+/T8wq7bEVth+bYRjmj6QSYLPZ9O6772rQoEGSpI8//lhdunTR2bNnVbNmTftyDz74oGw2mzZs2KD58+frlVde0bFjxxy2FRgYqFmzZmns2LEF9hUXF6dZs2bla1+/fr28vb2dNygAAFBiMjMzNXToUKWnp8vPz++my5XZMzslafr06YqNjbVPZ2RkqE6dOurdu/ctP6zfahG3w3TfXm6G5rTP1TOfuSkr12Zq3W/iIk33Vxxmx1eexmZWdna24uPj1atXL3l4eLi6nFsqzeOyrB+TEmO7mdIcn5XHVhTl6feJWaU9trwrM7dTZsNOcHCwJCk1NdXhzE5qaqratGljX+b8+fMO6924cUMXL160r18QLy8veXl55Wv38PAwtXOycsz9BXRYN9dmev3S/ktR1PGVh7EVldljxBVK87gsL8ekxNh+qzTHZ+WxFUd5+H1SVKU1tsL2UWbfs1O/fn0FBwcrISHB3paRkaFPPvlE4eHhkqTw8HClpaUpKSnJvszu3buVm5urjh07lnrNAACg7HHpmZ3Lly/r+PHj9ukTJ07oiy++UNWqVVW3bl1NnjxZc+fOVePGjVW/fn0988wzCgkJsd/X06xZM/Xp00ePPvqoVq5cqezsbI0fP16DBw/mSSwAACDJxWHns88+U48ePezTeffRjBw5UmvWrNETTzyhK1euaMyYMUpLS1PXrl21fft2VaxY0b7Oa6+9pvHjx+vee++Vm5uboqOjtWTJklIfCwAAKJtcGna6d++uWz0MZrPZNHv2bM2ePfumy1StWlXr168vifIAAIAFlNl7dgAAAJyBsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACyNsAMAACytgqsLAFyh3pNbTC3v5W5oYQepRdwOZeXYTK17ckGUqeUBAM7FmR0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBpFVxdAAAAKHvqPbnF9Dpe7oYWdpBaxO1QVo7N1LonF0SZ7q+wOLMDAAAsjbADAAAsjbADAAAsjbADAAAsjbADAAAsjbADAAAsjUfPAQAoIis9nm1lnNkBAACWZpmws2zZMtWrV08VK1ZUx44d9emnn7q6JAAAUAZYIuxs2LBBsbGxmjlzpg4dOqTWrVsrMjJS58+fd3VpAADAxSwRdl588UU9+uijGj16tMLCwrRy5Up5e3vr3//+t6tLAwAALlbuw87169eVlJSkiIgIe5ubm5siIiKUmJjowsoAAEBZUO6fxvrf//6nnJwcBQUFObQHBQXp6NGjBa6TlZWlrKws+3R6erok6eLFi8rOzi503xVuXDFdb4VcQ5mZuaqQ7aacXHN34V+4cMF0f8VhdnyMrWBlfWxS0cfH2JzHyr9PGNtv1mFsBSrK2C5duiRJMgzj1gsa5dyZM2cMScbHH3/s0D516lSjQ4cOBa4zc+ZMQxI//PDDDz/88GOBn9OnT98yK5T7MzvVq1eXu7u7UlNTHdpTU1MVHBxc4DrTp09XbGysfTo3N1cXL15UtWrVZLOZS6JmZWRkqE6dOjp9+rT8/PxKtK/SxtjKLyuPj7GVT4ytfCrtsRmGoUuXLikkJOSWy5X7sOPp6al27dopISFBgwYNkvRLeElISND48eMLXMfLy0teXl4ObQEBASVcqSM/Pz/LHeR5GFv5ZeXxMbbyibGVT6U5Nn9//9suU+7DjiTFxsZq5MiRat++vTp06KDFixfrypUrGj16tKtLAwAALmaJsPOnP/1JP/30k5599lmlpKSoTZs22r59e76blgEAwO+PJcKOJI0fP/6ml63KEi8vL82cOTPfZTQrYGzll5XHx9jKJ8ZWPpXVsdkM43bPawEAAJRf5f6lggAAALdC2AEAAJZG2AEAAJZG2AFw+1etA0A5RtgBIC8vL3377beuLgMASoRlHj1H6bt69aqSkpJUtWpVhYWFOcy7du2a3nzzTY0YMcJF1RXPt99+qwMHDig8PFxNmzbV0aNH9dJLLykrK0sPPfSQevbs6eoSi+TXX5Pyazk5OVqwYIGqVasmSXrxxRdLs6wSc+XKFb355ps6fvy4atasqSFDhtjHiLJjwoQJevDBB3X33Xe7uhSYdO7cOa1YsUIfffSRzp07Jzc3NzVo0ECDBg3SqFGj5O7u7uoSf+GUb+NEkSUnJxujR492dRmmHTt2zAgNDTVsNpvh5uZm3HPPPcbZs2ft81NSUgw3NzcXVlh027ZtMzw9PY2qVasaFStWNLZt22bUqFHDiIiIMHr27Gm4u7sbCQkJri6zSGw2m9GmTRuje/fuDj82m8246667jO7duxs9evRwdZlF1qxZM+PChQuGYfzyd6tevXqGv7+/cddddxlVq1Y1AgMDjR9//NHFVRZNUlKSQ+1r1641OnfubNSuXdvo0qWL8frrr7uwuuLJ+z3SuHFjY8GCBca5c+dcXZJTvfzyy8bw4cPt+2jt2rVGs2bNjCZNmhjTp083srOzXVxh0Rw8eNDw9/c32rVrZ3Tt2tVwd3c3hg8fbvzpT38yAgICjM6dOxsZGRmuLtMwDMMg7LjYF198US5DwaBBg4yoqCjjp59+Mr7//nsjKirKqF+/vnHq1CnDMMp32AkPDzeefvppwzAM4/XXXzeqVKliPPXUU/b5Tz75pNGrVy9XlVcszz33nFG/fv18Ya1ChQrG4cOHXVSV89hsNiM1NdUwDMMYNmyY0blzZyMtLc0wDMO4dOmSERERYQwZMsSVJRZZq1atjPj4eMMwDOOf//ynUalSJWPixInGihUrjMmTJxu+vr7GqlWrXFxl0dhsNmPXrl3GpEmTjOrVqxseHh7GwIEDjQ8++MDIyclxdXnFMmfOHKNy5cpGdHS0ERwcbCxYsMCoVq2aMXfuXGP+/PlGjRo1jGeffdbVZRZJly5djLi4OPv0q6++anTs2NEwDMO4ePGi0aZNG2PixImuKs8BYaeEbdq06ZY/ixYtKpehIDAw0Pjqq6/s07m5ucZf/vIXo27dusYPP/xQrsOOn5+f8f333xuGYRg5OTlGhQoVjEOHDtnnf/3110ZQUJCryiu2Tz/91LjjjjuMxx57zLh+/bphGNYMOw0aNDB27tzpMH///v1GnTp1XFFasVWqVMk4efKkYRiG0bZtW+Mf//iHw/zXXnvNCAsLc0Vpxfbr/Xb9+nVjw4YNRmRkpOHu7m6EhIQYTz31lP3vZHnTsGFD4+233zYM45f/uXV3dzfWrVtnn//OO+8YjRo1clV5xVKpUiXjhx9+sE/n5OQYHh4eRkpKimEYhrFz504jJCTEVeU54J6dEjZo0CDZbLZbPu1is9lKsSLnuHr1qipU+L/Dx2azacWKFRo/fry6deum9evXu7C64svbJ25ubqpYsaLDt+pWrlxZ6enpriqt2O666y4lJSUpJiZG7du312uvvVYuj8GbyRvLtWvXVLNmTYd5tWrV0k8//eSKsorN29tb//vf/xQaGqozZ86oQ4cODvM7duyoEydOuKg65/Hw8NCDDz6oBx98UMnJyfr3v/+tNWvWaMGCBcrJyXF1eaadPXtW7du3lyS1bt1abm5uatOmjX3+nXfeqbNnz7qouuIJDAzUuXPn1KBBA0lSamqqbty4Yf+288aNG+vixYuuLNGOp7FKWM2aNfXOO+8oNze3wJ9Dhw65usQiadq0qT777LN87UuXLtV9992ngQMHuqAq56hXr56+//57+3RiYqLq1q1rn05OTs73j2h54+vrq1deeUXTp09XREREufxH5Gbuvfde3XnnncrIyNCxY8cc5p06darc3qDct29frVixQpLUrVs3bdy40WH+m2++qUaNGrmitBJTt25dxcXF6cSJE9q+fburyymS4OBgHTlyRJL0/fffKycnxz4tSYcPH1ZgYKCryiuWQYMG6S9/+Yu2b9+uPXv2aNiwYerWrZsqVaokSTp27Jhq1arl4ip/wZmdEtauXTslJSXpvvvuK3D+7c76lFX333+/Xn/9dQ0fPjzfvKVLlyo3N1crV650QWXFN3bsWId//Fu0aOEwf9u2beX2aazfGjx4sLp27aqkpCSFhoa6upximzlzpsO0r6+vw/QHH3xQbp/4ef7559WlSxd169ZN7du31wsvvKC9e/eqWbNmOnbsmA4cOKB3333X1WUWSWho6C2f2rHZbOrVq1cpVuQ8w4YN04gRI3TfffcpISFBTzzxhB5//HFduHBBNptN8+bN0x/+8AdXl1kkc+fO1blz5zRgwADl5OQoPDxc69ats8+32Wx67rnnXFjh/+GLQEvYf/7zH125ckV9+vQpcP6VK1f02WefqVu3bqVcGYDyJi0tTQsWLNAHH3ygH3/8Ubm5uapZs6a6dOmiKVOm2C+XoOzIzc3VggULlJiYqM6dO+vJJ5/Uhg0b9MQTTygzM1MDBgzQ0qVL5ePj4+pSi+zatWu6ceNGvv+5KEsIOwAAwNK4ZwcAAFgaYQcAAFgaYQcAAFgaYQdAuWez2fTee++5ugwAZRRhB0CZl5KSogkTJqhBgwby8vJSnTp1NGDAACUkJLi6NADlAO/ZAVCmnTx5Ul26dFFAQID++te/qmXLlsrOztaOHTsUExOjo0ePurpEAGUcZ3YAlGnjxo2TzWbTp59+qujoaN1xxx1q3ry5YmNjdeDAgQLXmTZtmu644w55e3urQYMGeuaZZ5SdnW2f/+WXX6pHjx6qXLmy/Pz81K5dO/sbwU+dOqUBAwaoSpUq8vHxUfPmzbV169ZSGSuAksGZHQBl1sWLF7V9+3bNmzevwJeuBQQEFLhe5cqVtWbNGoWEhOjrr7/Wo48+qsqVK+uJJ56Q9Mtbbdu2basVK1bI3d1dX3zxhTw8PCRJMTExun79uvbt2ycfHx8dOXKkTL8sDcDtEXYAlFnHjx+XYRhq2rSpqfVmzJhh/+969erp8ccf1xtvvGEPO8nJyZo6dap9u40bN7Yvn5ycrOjoaLVs2VKS7F9yCKD84jIWgDKrqC9437Bhg7p06aLg4GD5+vpqxowZSk5Ots+PjY3VI488ooiICC1YsEA//PCDfd7EiRM1d+5cdenSRTNnztRXX31V7HEAcC3CDoAyq3HjxrLZbKZuQk5MTNSwYcPUr18/bd68WZ9//rmefvppXb9+3b5MXFycDh8+rKioKO3evVthYWH2L9F85JFH9OOPP2r48OH6+uuv1b59e7388stOHxuA0sN3YwEo0/r27auvv/5ax44dy3ffTlpamgICAmSz2fTuu+9q0KBBeuGFF7R8+XKHszWPPPKINm7cqLS0tAL7GDJkiK5cuaL3338/37zp06dry5YtnOEByjHO7AAo05YtW6acnBx16NBBb7/9tr7//nt9++23WrJkicLDw/Mt37hxYyUnJ+uNN97QDz/8oCVLltjP2kjS1atXNX78eO3du1enTp3S/v37dfDgQTVr1kySNHnyZO3YsUMnTpzQoUOHtGfPHvs8AOUTNygDKNMaNGigQ4cOad68eXrsscd07tw51ahRQ+3atdOKFSvyLT9w4EBNmTJF48ePV1ZWlqKiovTMM88oLi5OkuTu7q4LFy5oxIgRSk1NVfXq1fXAAw9o1qxZkqScnBzFxMTov//9r/z8/NSnTx8tWrSoNIcMwMm4jAUAACyNy1gAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDSCDsAAMDS/j+5SRhJziQd7wAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of data points in class 7 : 609 (28.672) %\n",
            "Number of data points in class 4 : 439 (20.669) %\n",
            "Number of data points in class 1 : 363 (17.09) %\n",
            "Number of data points in class 2 : 289 (13.606) %\n",
            "Number of data points in class 6 : 176 (8.286) %\n",
            "Number of data points in class 5 : 155 (7.298) %\n",
            "Number of data points in class 3 : 57 (2.684) %\n",
            "Number of data points in class 9 : 24 (1.13) %\n",
            "Number of data points in class 8 : 12 (0.565) %\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "my_colors = 'rgbkymc'\n",
        "test_class_distribution.plot(kind = 'bar')\n",
        "plt.xlabel('Class')\n",
        "plt.ylabel('Data points per class')\n",
        "plt.title('Distribution of yi in test data')\n",
        "plt.grid()\n",
        "plt.show()\n",
        "\n",
        "sorted_yi = np.argsort(-test_class_distribution.values)\n",
        "for i in sorted_yi:\n",
        "    print(f'Number of data points in class {i+1} : {test_class_distribution.values[i]} ({np.round((test_class_distribution.values[i]/test_df.shape[0]*100), 3)} %)')"
      ],
      "metadata": {
        "id": "9qMT6AWMauJg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "my_colors = 'rgbkymc'\n",
        "cv_class_distribution.plot(kind = 'bar')\n",
        "plt.xlabel('Class')\n",
        "plt.ylabel('Data points per class')\n",
        "plt.title('Distribution of yi in cross validation data')\n",
        "plt.grid()\n",
        "plt.show()\n",
        "\n",
        "sorted_yi = np.argsort(-cv_class_distribution.values)\n",
        "for i in sorted_yi:\n",
        "    print(f'Number of data points in class {i+1} : {cv_class_distribution.values[i]} ({np.round((cv_class_distribution.values[i]/cv_df.shape[0]*100), 3)}) %')"
      ],
      "metadata": {
        "id": "nS3B317AdU4U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prediction using a random model"
      ],
      "metadata": {
        "id": "_KEbKeNQd4DY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def plot_confusion_matrix(test_y, predict_y):\n",
        "    C = confusion_matrix(test_y, predict_y)\n",
        "\n",
        "    A = (((C.T)/(C.sum(axis = 1))).T)\n",
        "\n",
        "    B = (C/C.sum(axis = 0))\n",
        "\n",
        "    labels = [1, 2, 3, 4, 5, 6, 7, 8, 9]\n",
        "    print(\"-\"*20, \"Confusion matrix\", \"-\"*20)\n",
        "    plt.figure(figsize = (20, 7))\n",
        "    sns.heatmap(C, annot = True, cmap = 'YlGnBu', fmt = \".3f\", xticklabels = labels, yticklabels = labels)\n",
        "    plt.xlabel('Predicted class')\n",
        "    plt.ylabel('Original class')\n",
        "    plt.show()\n",
        "\n",
        "    print(\"-\"*20, \"Precision matrix (column sum = 1)\", \"-\"*20)\n",
        "    plt.figure(figsize = (20, 7))\n",
        "    sns.heatmap(B, annot = True, cmap = 'YlGnBu', fmt = \".3f\", xticklabels = labels, yticklabels = labels)\n",
        "    plt.xlabel('Predicted class')\n",
        "    plt.ylabel('Original class')\n",
        "    plt.show()\n",
        "\n",
        "    print(\"-\"*20, \"Recall matrix (Row sum = 1)\", \"-\"*20)\n",
        "    plt.figure(figsize = (20, 7))\n",
        "    sns.heatmap(A, annot = True, cmap = 'YlGnBu', fmt = \".3f\", xticklabels = labels, yticklabels = labels)\n",
        "    plt.xlabel('Predicted class')\n",
        "    plt.ylabel('Original class')\n",
        "    plt.show()"
      ],
      "metadata": {
        "id": "iC6MnV7ydITX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "random_values = np.random.rand(1,9)\n",
        "cv = (random_values/sum(sum(random_values)))[0]\n",
        "print(cv)"
      ],
      "metadata": {
        "id": "__9WaEBm25jB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sum(random_values)"
      ],
      "metadata": {
        "id": "fh8foDfs_1Rh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "test_data_len = test_df.shape[0]\n",
        "cv_data_len = cv_df.shape[0]\n",
        "\n",
        "# we create a output array that has exactly same size as the CV data\n",
        "cv_predicted_y = np.zeros((cv_data_len,9))\n",
        "for i in range(cv_data_len):\n",
        "    rand_probs = np.random.rand(1,9)\n",
        "    cv_predicted_y[i] = ((rand_probs/sum(sum(rand_probs)))[0])\n",
        "print(\"Log loss on Cross Validation Data using Random Model\",log_loss(y_cv,cv_predicted_y, eps=1e-15))\n",
        "\n",
        "\n",
        "# Test-Set error.\n",
        "#we create a output array that has exactly same as the test data\n",
        "test_predicted_y = np.zeros((test_data_len,9))\n",
        "for i in range(test_data_len):\n",
        "    rand_probs = np.random.rand(1,9)\n",
        "    test_predicted_y[i] = ((rand_probs/sum(sum(rand_probs)))[0])\n",
        "print(\"Log loss on Test Data using Random Model\",log_loss(y_test,test_predicted_y, eps=1e-15))\n",
        "\n",
        "predicted_y =np.argmax(test_predicted_y, axis=1)\n",
        "plot_confusion_matrix(y_test, predicted_y+1)\n"
      ],
      "metadata": {
        "id": "W7-Vr2z1gE0v"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Univariate Analysis"
      ],
      "metadata": {
        "id": "NLAs2HRN5j9n"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# code for response coding with Laplace smoothing.\n",
        "# alpha : used for laplace smoothing\n",
        "# feature: ['gene', 'variation']\n",
        "# df: ['train_df', 'test_df', 'cv_df']\n",
        "# algorithm\n",
        "# ----------\n",
        "# Consider all unique values and the number of occurances of given feature in train data dataframe\n",
        "# build a vector (1*9) , the first element = (number of times it occured in class1 + 10*alpha / number of time it occurred in total data+90*alpha)\n",
        "# gv_dict is like a look up table, for every gene it store a (1*9) representation of it\n",
        "# for a value of feature in df:\n",
        "# if it is in train data:\n",
        "# we add the vector that was stored in 'gv_dict' look up table to 'gv_fea'\n",
        "# if it is not there is train:\n",
        "# we add [1/9, 1/9, 1/9, 1/9,1/9, 1/9, 1/9, 1/9, 1/9] to 'gv_fea'\n",
        "# return 'gv_fea'\n",
        "# ----------------------\n",
        "\n",
        "# get_gv_fea_dict: Get Gene varaition Feature Dict\n",
        "def get_gv_fea_dict(alpha, feature, df):\n",
        "    # value_count: it contains a dict like\n",
        "    # print(train_df['Gene'].value_counts())\n",
        "    # output:\n",
        "    #        {BRCA1      174\n",
        "    #         TP53       106\n",
        "    #         EGFR        86\n",
        "    #         BRCA2       75\n",
        "    #         PTEN        69\n",
        "    #         KIT         61\n",
        "    #         BRAF        60\n",
        "    #         ERBB2       47\n",
        "    #         PDGFRA      46\n",
        "    #         ...}\n",
        "    # print(train_df['Variation'].value_counts())\n",
        "    # output:\n",
        "    # {\n",
        "    # Truncating_Mutations                     63\n",
        "    # Deletion                                 43\n",
        "    # Amplification                            43\n",
        "    # Fusions                                  22\n",
        "    # Overexpression                            3\n",
        "    # E17K                                      3\n",
        "    # Q61L                                      3\n",
        "    # S222D                                     2\n",
        "    # P130S                                     2\n",
        "    # ...\n",
        "    # }\n",
        "    value_count = train_df[feature].value_counts()\n",
        "\n",
        "    # gv_dict : Gene Variation Dict, which contains the probability array for each gene/variation\n",
        "    gv_dict = dict()\n",
        "\n",
        "    # denominator will contain the number of time that particular feature occured in whole data\n",
        "    for i, denominator in value_count.items():\n",
        "        # vec will contain (p(yi==1/Gi) probability of gene/variation belongs to perticular class\n",
        "        # vec is 9 diamensional vector\n",
        "        vec = []\n",
        "        for k in range(1,10):\n",
        "            # print(train_df.loc[(train_df['Class']==1) & (train_df['Gene']=='BRCA1')])\n",
        "            #         ID   Gene             Variation  Class\n",
        "            # 2470  2470  BRCA1                S1715C      1\n",
        "            # 2486  2486  BRCA1                S1841R      1\n",
        "            # 2614  2614  BRCA1                   M1R      1\n",
        "            # 2432  2432  BRCA1                L1657P      1\n",
        "            # 2567  2567  BRCA1                T1685A      1\n",
        "            # 2583  2583  BRCA1                E1660G      1\n",
        "            # 2634  2634  BRCA1                W1718L      1\n",
        "            # cls_cnt.shape[0] will return the number of rows\n",
        "\n",
        "            cls_cnt = train_df.loc[(train_df['Class']==k) & (train_df[feature]==i)]\n",
        "\n",
        "            # cls_cnt.shape[0](numerator) will contain the number of time that particular feature occured in whole data\n",
        "            vec.append((cls_cnt.shape[0] + alpha*10)/ (denominator + 90*alpha))\n",
        "\n",
        "        # we are adding the gene/variation to the dict as key and vec as value\n",
        "        gv_dict[i]=vec\n",
        "    return gv_dict\n",
        "\n",
        "# Get Gene variation feature\n",
        "def get_gv_feature(alpha, feature, df):\n",
        "\n",
        "    gv_dict = get_gv_fea_dict(alpha, feature, df)\n",
        "    # value_count is similar in get_gv_fea_dict\n",
        "    value_count = train_df[feature].value_counts()\n",
        "\n",
        "    # gv_fea: Gene_variation feature, it will contain the feature for each feature value in the data\n",
        "    gv_fea = []\n",
        "    # for every feature values in the given data frame we will check if it is there in the train data then we will add the feature to gv_fea\n",
        "    # if not we will add [1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9] to gv_fea\n",
        "    for index, row in df.iterrows():\n",
        "        if row[feature] in dict(value_count).keys():\n",
        "            gv_fea.append(gv_dict[row[feature]])\n",
        "        else:\n",
        "            gv_fea.append([1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9])\n",
        "#             gv_fea.append([-1,-1,-1,-1,-1,-1,-1,-1,-1])\n",
        "    return gv_fea"
      ],
      "metadata": {
        "id": "ixXblHiJuFTf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# def get_gv_fea_dict(alpha, feature, df):\n",
        "#     value_count = train_df[feature].value_counts()\n",
        "#     gv_dict =  dict()\n",
        "\n",
        "#     for i, denominator in value_count.items():\n",
        "#         vec = []\n",
        "\n",
        "#         for k in range(1, 10):\n",
        "#             cls_cnt = train_df.loc[(train_df['Class'] == k) & (train_df[feature] == i)]\n",
        "#             vec.append((cls_cnt.shape[0] + alpha*10)/ (denominator + 90*alpha))\n",
        "\n",
        "#         gv_dict[i] = vec\n",
        "#     return gv_dict"
      ],
      "metadata": {
        "id": "NU0lxOL7m32l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# def get_gv_feature(alpha, feature, df):\n",
        "#     gv_dict = get_gv_fea_dict(alpha, feature, df)\n",
        "\n",
        "#     value_count = train_df[feature].value_counts()\n",
        "\n",
        "#     gv_fea = []\n",
        "\n",
        "#     for index, row in df.iterrows():\n",
        "#         if row[feature] in dict(value_count).keys():\n",
        "#             gv_fea.append(gv_dict[row[feature]])\n",
        "\n",
        "#         else:\n",
        "#             gv_fea.append([1/9, 1/9, 1/9, 1/9,  1/9, 1/9, 1/9, 1/9, 1/9])\n",
        "\n",
        "#     return gv_fea"
      ],
      "metadata": {
        "id": "xEFItfcL469U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "unique_genes = train_df['Gene'].value_counts()\n",
        "print(\"Number of Unique Genes: \", unique_genes.shape[0])\n",
        "\n",
        "print(unique_genes.head(10))"
      ],
      "metadata": {
        "id": "Hk-6fi1K-2eP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"Ans: There are \",unique_genes.shape[0], \"different categories of genes in the train data and they are distributed as follows\",)\n",
        "s = sum(unique_genes.values)\n",
        "h = unique_genes.values/s\n",
        "\n",
        "plt.plot(h, label = \"Histogram of genes\")\n",
        "plt.xlabel('Index of gene')\n",
        "plt.ylabel('Number of occurances')\n",
        "plt.legend()\n",
        "plt.grid()\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "ZJXyWP_pA-T5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "c = np.cumsum(h)\n",
        "plt.plot(c, label = 'Cumulative distribution of genes')\n",
        "plt.grid()\n",
        "plt.legend()\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "iHvNlDgABcA1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "alpha = 1\n",
        "\n",
        "train_gene_feature_responseCoding = np.array(get_gv_feature(alpha, \"Gene\", train_df ))\n",
        "\n",
        "test_gene_feature_reponseCoding = np.array(get_gv_feature(alpha, \"Gene\", test_df))\n",
        "\n",
        "cv_gene_feature_responseCoding = np.array(get_gv_feature(alpha, \"Gene\", cv_df))"
      ],
      "metadata": {
        "id": "0GgOtwCSCPu8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"train_gene_feature_responseCoding is converted feature using coding method. The shape of gene  features: \", train_gene_feature_responseCoding.shape)"
      ],
      "metadata": {
        "id": "JOENk1IUHytG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "gene_vectorizer = CountVectorizer()\n",
        "\n",
        "train_gene_feature_onehotCoding = gene_vectorizer.fit_transform(train_df['Gene'])\n",
        "test_gene_feature_onehotCoding = gene_vectorizer.fit_transform(test_df['Gene'])\n",
        "cv_gene_feature_onehotCoding = gene_vectorizer.fit_transform(cv_df['Gene'])"
      ],
      "metadata": {
        "id": "0jrrdcSVI2YM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "cv_gene_feature_onehotCoding.shape"
      ],
      "metadata": {
        "id": "HUIY9Q3HQeMU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "train_df['Gene'].head()"
      ],
      "metadata": {
        "id": "M6GazorNJgZo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "gene_vectorizer.get_feature_names_out()"
      ],
      "metadata": {
        "id": "Ydsc1JuQKCoa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(\"train_gene_feature_onehotencoding is converted feature using one hot encoding method. the shape of gene feature: \", train_gene_feature_onehotCoding.shape)"
      ],
      "metadata": {
        "id": "fGlI8HKMKMnP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "alpha = [10**x for x in range(-5, 1)]\n",
        "\n",
        "cv_log_error_array=[]\n",
        "for i in alpha:\n",
        "    clf = SGDClassifier(alpha=i, penalty='l2', loss='log', random_state=42)\n",
        "    clf.fit(train_gene_feature_onehotCoding, y_train)\n",
        "    sig_clf = CalibratedClassifierCV(clf, method=\"sigmoid\")\n",
        "    sig_clf.fit(train_gene_feature_onehotCoding, y_train)\n",
        "    predict_y = sig_clf.predict_proba(cv_gene_feature_onehotCoding)\n",
        "    cv_log_error_array.append(log_loss(y_cv, predict_y, labels=clf.classes_, eps=1e-15))\n",
        "    print('For values of alpha = ', i, \"The log loss is: \",log_loss(y_cv, predict_y, labels=clf.classes_, eps=1e-15))\n",
        "\n"
      ],
      "metadata": {
        "id": "DBVBzWrqLAnS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "0CSMEzBHPbGQ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}